// sm-supplier-html/suppliers/www.dealsplus.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            info.store = Search.store($, product);
            //result.push(info)

            if (info.price !== null && Search.store($, product).indexOf('Amazon') < 1) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('section.info h1 a').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        let seeDeal = $('a.btn.seeIt').text()
        var isInStock = false
        if (seeDeal) {
            if (seeDeal.indexOf('See It') > -1) {
                isInStock = true
            }
        }
        return isInStock
    },
    price: function ($) {
        $('div.dealPrice del').remove();
        var price = $('div.dealPrice').first().text().trim();

        if (price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);
        }

        return price;
    },
    image: function ($) {
        var image = $('#mainImage a img').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('div.howToContent').text();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('.itemTileV5');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('.tileDealTitle').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.currentPrice').text();
        if (price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('.tileDealTitle a').attr('href');
        url = 'https://www.dealsplus.com' + url;
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src')
        return image;
    },
    store: function ($, product) {
        var store = $(product).find('div.tileStoreName').text()
        if(store){
            return store;
        }else{
            return 'nothing'
        }
    }
}

var Pagination = {

    items: function ($) {
        //
        // No items count 
        // five element in ul always is last page
        //
        var itemsCountOnPage = $('div.popularDealTile').length;

        return itemsCountOnPage;
    },
    pages: function ($) {
        //
        // No total search items count
        // five element in ul always is last page
        //
        var pages = $('div.pagination ul li:nth-child(5) a').text().trim();
        if (pages === "") {
            return null
        }
        return parseInt(pages);
    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();
        name = cleanName(name);
        return name;
    }
}
