// sm-supplier-html/suppliers/www.focalprice.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            //info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name=$('h1').text()
        name=cleanName(name)
        return name
    },
    inStock: function ($) {
        var availabilty = $('button.exclusive').text()
        var isInstock= false
        if (availabilty){
                isInstock=true            
        }
	return isInstock;

    },
    price: function ($) {
        var price=$('span#our_price_display').text()
        price=getPrice(price)
        return price
    },
    image: function ($) {
        var image=$('img#bigpic').attr('src')
        return image

    },
}

var Search = {
    container: function ($) {
        var searchproducts=$('div.product-container')
        return searchproducts
    },
    name: function ($, product) {
        var name=$(product).find('a.product-name').text();
        name = cleanName(name);
        return name
    },
    price: function ($, product) {
        var price=$(product).find('span.price.product-price').text();
        price = getPrice(price);
        return price
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.product-name');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        //('this is my url             '+ url);
        return  url;
    },

    image: function ($, product) {
        var image = $(product).find('img.replace-2x.img-responsive').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        return cleanName($("span.cat-name").text());
    }
}
