// sm-supplier-html/suppliers/www.harrods.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = replace =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
            
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.price_all').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#wrap a").attr('href');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-grid_item');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('span.product-card_brand').html();
        var name = $(product).find('span.product-card_name').html();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + ' ' + name;
    },
    price: function($, product) {
        var price = $(product).find('span.price_amount').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
            image = 'http:' + image;
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.right_column h1").text());
    }
}
var Pagination = {

     items: function($) {
        return parseInt(cleanName($("a#ctl00_ContentPlaceHolder1_ucTemplate" + 
            "_products_lnkPageProductCount18").text()));

     },
     pages: function($) {
        return parseInt(cleanName($("a#ctl00_ContentPlaceHolder1_ucTemplate_products_lklast").text()));

     }
}
