// sm-supplier-html/suppliers/abbottnutrition.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.prod-detail > h2').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {

        return null;
    },
    image: function ($) {
        var image = $('#prod-img').attr('src');
        image = 'https:' + image;

        return image;
    },
    description: function ($) {
        var des = $('div.prod-detail > p').last().text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.component');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.imgDetail > h6').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        return null;
    },
    url: function ($, product) {
        var url = $(product).find('div.imgDetail > a').attr('href');
        url = 'https://abbottnutrition.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.imgBlock > a > img').attr('src');
        image = 'https:' + image;

        return image;
    }
}

/*var Pagination = {

    items: function ($) {
        var items = ;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = ;
        var products = ;

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}*/

var Category = {
    name: function ($) {
        var name = $('section.banner-section > div > h1').text();
        name = cleanName(name)

        return name;
    }
}
