// sm-supplier-html/suppliers/www.evanscycles.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');

module.exports = evanscycles = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var availability = $('span.in-stock').text()
        var isInStock = false

        if (availability) {
            if (availability.indexOf('In Stock') > -1) {
                isInStock = true
            }
        }
        return isInStock;
    },
    price: function ($) {
        var price = $('span.current').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.title a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.current').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.evanscycles.com' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img.product-image-file').attr('src');
        return image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("h1.page-title").text());

    }
}
var Pagination = {

    items: function ($) {
        var items = cleanName($("div.showing-per-page").text())
            .match(/[0-9]+/g);

        if (items) {
            return parseInt(items[0]);
        } else {
            return null;
        }


    },
    pages: function ($) {
        var pages = cleanName($("li.end").text())

        if (pages) {
            return parseInt(pages.split(" ")[0]);
        } else {
            return null;
        }



    }
}
