// sm-supplier-html/suppliers/www.kmart.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = kmart = {
    "useJavascript": true,
    "usePhantom": "kmart.phantom.js",

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-title').html();
        (name)
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var isInStock = true;
        var checkstock = $('div.product-fulfillment-max-ships-for').html();

        if (checkstock === null) {
            isInStock = false;
        }
        return isInStock;

    },
    price: function ($) {
        var price = $('span.price-wrapper').html();
        var newprice = $('h4 span.price-wrapper').html();

        if (newprice !== null) {
            price = newprice;
        }

        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('[data-id="product-image-main"]>img').attr('src');

        if (image === undefined) {
            image = $('img.active').attr('src');
        }


        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.card-container.card-border');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('h3.card-title').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('span.card-price').text();

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.kmart.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img.card-image').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.card-container.card-border').length();
        return items;

    },
    pages: function ($) {
        var result = null;
        var pages = $('#subcat-title>span').text();

        if (pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages);
        } else {
            result = 0
            
        }
        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.ng-binding').text();
        name = cleanName(name);
        return name;
    }
}
