// sm-supplier-html/suppliers/www.interstatemusic.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.barcode = Search.barcode($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.productTitle').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var isInstock = true;
        var stockcheck = $('div.prodStock span.stockstatus').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.match(/[0-9]+/g)
            if (parseInt(stockcheck) < 8) {
                isInstock = false;
            }
        }

        return isInstock;
    },
    price: function ($) {
        var price = $('div.price').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.prodImage a img').attr('src');
        if (image.indexOf('http:') < 0) {
            image = 'http:' + image
        }
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.product-row li');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.prodListName a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.prodListName a').attr('href');
        if (url.indexOf('../..') > -1) {
            url = url.replace("../..", "");
        }
        return 'https://www.interstatemusic.com' + url
    },
    image: function ($, product) {
        var image = $(product).find('div.prodListImage img').attr('src');

        if (image.indexOf('http:') < 0) {
            image = 'https:' + image
        }
        return image;
    },
    barcode: function ($, product) {
        var barcode = $(product).find('div.skuUPC div:nth-child(3)').html()
        barcode = barcode.substring(barcode.lastIndexOf("\t") + 1, barcode.lastIndexOf(""))
        return barcode;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul.qty-item').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
    },
    pages: function ($) {
        var items = $('ul.qty-item').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('li.totalRecords').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('div.contentSubCatLvl2plus h1').text();
        name = cleanName(name);
        return name;
    }
}
