// sm-supplier-html/suppliers/www.swarovski.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    //"useJavascript" = true
    'product': function ($, next) {
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    /*
    isStock : function ($) {
        var isInStock = true;
        var availability = $('.product-info').html()
        if (availability) {
            if ((availability.indexOf('Out of stock') > -1)
                || (availability.indexOf('Out of Stock') > -1)) {
                isInStock = false
            }
        }
        return isInStock
    },
    */
    price : function ($) {
        var price = $('div.product-detail__price').text();
            price = getPrice(price);
        return price;
    },
    
    name : function($) {
        var name = $('div.product-detail__name').text();
        name = cleanName(name);

        return name;
    },
    
    image : function($) {
        var image = $('div.visible-sm.link-wrapper.c img').attr('src');
        
        return 'https://www.swarovski.com'+image;
    },
    inStock: function($) {
        return 'unchecked';

    },

}



var Search = {
    container: function ($) {
        var searchProducts = $('div.product-list-block');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('a.name').text();
        name = cleanName(name);
        return name;
        

    },
    price: function ($, product) {
        var price = $(product).find('span.product-item__price').text()
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.name');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.swarovski.com'+url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        image = 'http://www.swarovski.com' + image;
        return image;

    }
}


var Pagination = {

    items: function ($) {
        var items = $('div.listproduct');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div.listproduct');
        items = items.length;

        var products = $('em.prodcount').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h2.arrow').text();
        name = cleanName(name);
        return name;
    }
}
