// sm-supplier-html/suppliers/www.ebags.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text()
        name = cleanName(name);
        return name;

    },

    inStock: function($) {
        var isInStock = true
        var avaliability = $('div.remainingQuantity').text();
        isInStock = instockCheck(avaliability,isInStock)
        return isInStock;
    },

    price: function($) {
        var price = $('div.mainPriceCon').text();
        price = getPrice(price);
        return price;
    },

    image: function($) {
        var image = $('img#pdMainImage').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.listPageItem');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.itemProductName a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.itemProductPrice').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.ebags.com' + $(anchor).attr('href');
        }

        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.listPageImage a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.listPageItem');

        if(items.length) {
            return parseInt(items.length);
        } else {
            return null;
        }

     },
     pages: function($) {
        var numitems = $('div.listPageItem').length;

        var products = $('h3#searchResultsCount').text();
        products = products.match(/[0-9]+/g);

        if(numitems && products) {
            products = products.pop();

            var sum = products / numitems;
            return Math.ceil(sum)
        } else {
            return null;
        }
        


    }
}

var Category = {
    name: function($) {
        var name = $('h1.selectionsTitle').text();
        name = cleanName(name);
        return name;
    }
}
