// sm-supplier-html/suppliers/www.wickes.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = wickies =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true; 
        var checkstock = $('div.schoutstock').html();
        var clickandcollectonly = $('div.addToBasketButton').html();
        (clickandcollectonly);
        if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if(checkstock.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('span.isNotBulkDiscountOrWasPrice').html();
        price2 = $('span.isBulkDiscountOrWasPrice').html();

        if(price === null) {
            price = price2;
        }
        
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div#s7flyout_div_flyout div.s7staticimage img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.card.product-card');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a').attr('title')
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        let price = getPrice($(product).find('.product-card__price-value').html());
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.wickes.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https:' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('form#perpage_form1').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items)
     },
     pages: function($) {
        var items = $('form#perpage_form1').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.pagnItemInfo').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum - 1);

     }
}

var Category = {
    name: function($) {
        var name = $('h1#pcatid').text();
        name = cleanName(name);
        return name;
    }
}
