// sm-supplier-html/suppliers/www.hayneedle.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavaScript": true,
    "usePhantom": "hayneedle.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span[itemprop*="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('li.active-attr-img img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.is-one-third-desktop');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        let dollar = $(product).find('span.main-price-dollars').text()
        let cent = $(product).find('span.main-price__cents___2BnzK').text()
        let price = getPrice(dollar + '.' + cent)
        
        return price;

    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https:' + image;
        return image;

    }
}

var Pagination = {

     items: function($) {

        var items = $('div.grid-view-layout'); 
        items = items.length;
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.grid-view-layout'); 
        items = items.length;

        var result = null;
        var products = $('div.rl-info').text();
        
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.header-inner-container').text();
        name = cleanName(name);
        return name;
    }
}
