// sm-supplier-html/suppliers/www.philamfood.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = philamFood =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var checkstock = $('.title-bar h2').html();

        if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('#product_price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('#product_thumbnail').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.products tr td');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price-value span.currency').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.product-title a');
        if(anchor) {
            url = 'https://www.philamfood.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.image a img').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('.products tr td');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('div.nav-pages a').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('span.last-bread-crumb').text();
        name = cleanName(name);
        return name;
    }
}
