// sm-supplier-html/suppliers/www.swell.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {
        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span.Breadcrumb-label').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        return "UNCHECKED"

    },
    price: function($) {
        var price = $('span.ProductHeader-price ').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('span.ProductImages-mainIntrinsicWrapper img.ProductImages-mainImage').attr('src');
        return image;

    },
    description: function($) {
        var description = $('div.ProductInformation-panelContent ul li').text();
        return description;
        
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.ProductsGrid-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('span.ProductsGridItem-name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) { 
        var price = $(product).find('span.ProductsGridItem-price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.swell.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('span.ProductsGridItem-intrinsicWrapper img').attr('src');
        return image;
        
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.ProductsGrid-item');
        items = items.length;
        return items;

     },
     pages: function($) {

        var items = $('li.ProductsGrid-item');
        items = items.length;

        var result = null;
        var products = $('p.ProductsHeader-gridResults span').text();
        
        if(products !== ""){

            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;

     }
}
var Category = {
    name: function($) {
        var name = $('ul.Breadcrumb-list li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
