// sm-supplier-html/suppliers/www.quill.com/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    //"useJavascript" : true,
    'search' : function($, next) {
        var result = []

        $('.ResultsList').find('.BrowseItem').map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($, product)
            info.country = "us"
            result.push(info)
        });

        // var pages = getNumberOfPages($, result.length)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = getPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "us"
        return info
    },
      'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    var link = $(product).find('#skuName').find('a').attr('href')
    return "http://www.quill.com" + link
}


function searchName($, product) {
    return $(product).find('#skuName').text().trim()
}


function searchPrice($, product) {
    var thePrice = $(product).find('.priceupdate').text().replace("Each","").replace("$","").trim()
    return thePrice
}


function getNumberOfPages($, perpage) {
    var products = $('.ResultsSummary').text().trim()
    products = /[0-9]+/.exec(products)
    var actVal = 1
    if (products != null) {
        products = products[0]
        actVal = Math.ceil(products/perpage)
    }
    return actVal
}

function searchImage ($,product) {
    var image = $(product).find('img#SkuPageMainImg').attr('src');
    image = 'http:' + image;
    return image;
}





var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.skuAction').html()
    if(contentarea !== null) {
        if(contentarea.indexOf('out of stock') > -1) {
            isInStock = false
        }
    }
    return isInStock
}

var getPrice = function thePrice($, instock) {
    return $('#PriceLabel').text().replace("$","").replace("/Each","").trim()
}

var getName = function theName($, instock) {
    return $('h1').text().trim()

}

function image($) {
    var image = $('img#SkuPageMainImg').attr('src');
    image = 'http:' + image;
    return image; 
}
var Pagination = {

     items: function($) {
        var items = $('.BrowseItem');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.BrowseItem');
        items = items.length;

        var products = $('span#SearchCount').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.current').text();
        name = cleanName(name);
        return name;
    }
}
