// sm-supplier-html/suppliers/www.alza.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = alza = {
    'useJavascript': true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info=new CategoryInfo()
        info.name=Category.name($)
        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var availability = $('.priceDetail .avl.js-availability').text();
        isInStock=instockCheck(availability,isInStock)
        //availability = availability.match(/[0-9]+/g).pop();


        if (parseInt(availability) < 5) {
            isInStock = false;
        }

        return isInStock;
    },
    price: function ($) {
        var price
        var discountPrice = $('.price_withVat').text()
        if(discountPrice){
            price = getPrice(discountPrice+'.00')
        }else{
            var priceN = $('.pricenormal').text();
            price=getPrice(priceN)
        }
        return price
    },
    
    image: function ($) {
        var image = $('img.lazyloaded').attr('srcset');
        return image;
    },
    description: function ($) {
        var description = $('meta[name="og:description"]').attr('content');

        if (description !== null) {
            return description;
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div#boxes div.box')
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a.name').text()
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.c2').text();
        price = getPrice(price);
        if(price)
        {
            return price;
        }else{
            price=$(product).find('span.c2').text();
            price=getPrice(price+'.00')
            return price
        }
    },
    url: function ($, product) {
        var url = $(product).find('a.name').attr('href');
        return 'http://www.alza.co.uk' + url;

    },
    image: function ($, product) {
        var image = $(product).find('.pc.browsinglink>em>img').attr('data-src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div#boxes div.box')
        return parseInt(items.length)
    },
    pages: function ($) {
        var pages = $('.numberItem').text();
        return parseInt(pages)
    }
}

var Category = {
    name: function($){
        var category = $('[itemprop="name"]').text()
        return cleanName(category)
    }

}
