// sm-supplier-html/suppliers/www.sephora.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.u-fwb').text().trim();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.u-fwb.u-h3.u-lhh div span').text();
        price = getPrice(price);
		
        return price;       
    },
    image: function($) {
        var image = $('div.PdpZoom-glass.u-absolute.u-oh.ng-hide img').attr('src');
        image = 'http://www.sephora.com' + image;
		
        return image;
    },
    description: function($) {
        var description = $('#details div div').text().split('\n')[2].trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('a.u-size1of4');

        return product;
    },
    name: function($, product) {
        var brand = $(product).find('div.SkuItem-nameBrand').text();
        var itemName = $(product).find('div.SkuItem-nameDisplay').text();
        var name = brand + ' ' + itemName;
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span[data-at="sku_item_price_list"]').text();
        price = getPrice(price);

        return price
    },
    url: function($, product) {
        var url = $(product).attr('href');
        url = 'http://www.sephora.com' + url.replace(' ', '%20');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.FlexEmbed-content').attr('src');
        image = 'http://www.sephora.com' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
        
        var items = $('select option[selected="selected"]').last().attr('value');
        items = items.match(/[0-9]+/g).pop();
        items = JSON.parse(items);
		
        return items;
         
     },
     pages: function($) {
        var items = $('select option[selected="selected"]').last().attr('value');
        items = items.match(/[0-9]+/g).pop();
        items = JSON.parse(items);
        
        var products = $('div.SearchHeader h2').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / items;
        return Math.ceil(sum);
         
     }
}
var Category = {
    name: function($) {
        var name = $('li.Breadcrumb-item h1').text();
        name = cleanName(name);
        return name;
    }
}
