// sm-supplier-html/suppliers/www.ribblecycles.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        let name=$('div.product-name').text()
        name=cleanName(name)
        return name
    },
    inStock: function ($) {
        let inStock=false
        let availability=$('div.availability.in-stock span').text()
        (availability)
        if(availability.indexOf('IN STOCK')>-1 || availability.indexOf('In stock')>-1){
            inStock=true
        }
        return inStock

    },
    price: function ($) {
        let price=$('span.price').text()
        price=getPrice(price)
        return price
    },
    image: function ($) {
        let image=$('div.product-media-large img').attr('src')
        return image
    },
}

var Search = {
    container: function ($) {
        let container = $('.products-grid-secondary li.product')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('div.name a').text()
        //let name2 = $(product).find('h4').text()
            name = cleanName(name)
            return name

    },
    price: function ($, product) {
        let price = $(product).find('span.price').text()
            price = getPrice(price)
            return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('a.image').attr('href')
        return image
    }
}

var Pagination = {

    items: function ($) {},
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {}
}
