// sm-supplier-html/suppliers/www.acmetools.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var h1 = $('h1').first().html();
        
        h1 = cleanName(h1);
        return h1;
    },
    inStock: function($) {
        var info = $('.avail-info').text()
        var inStock = true
        inStock = instockCheck(info, inStock)
	return inStock
},
    price: function($) {
        var price = $('.singlePrice').first().text().replace(/\s+/g,'').replace('\r\n','');
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#productMainImage').attr('src');
        image = 'http://www.acmetools.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product_listing_container div.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product_name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.now-price td.amount').text().replace(/\s+/g,'').replace('\r\n','');
        
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.product_name a');
        if(anchor) {
            url = 'https://www.acmetools.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.product_image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('#pageSize').text();
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }
        return parseInt(items);

     },
     pages: function($) {
        var items = $('#pageSize').text(); 
        
        if(items === ""){
            return null
        }else{
            numitems = items.match(/[0-9]+/g).pop();    
        }
        

        var products = $('div.header_bar .title span.num_products').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();

        }

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.category-name').text();
        name = cleanName(name);
        return name;
    }
}
