// sm-supplier-html/suppliers/www.areyougame.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    //    'sku': (url) => {
    //        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    //    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            //          info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').text()
        name = cleanName(name)
        return name
    },
    inStock: function ($) {
        var inStock = true
        var stockcheck = $('font:contains("Out of Stock")').text()

        if (stockcheck) {
            inStock = false;
        }
        return inStock
    },
    price: function ($) {
        var price = $('strong:contains("$")').first().text()
        price = getPrice(price)
        return price
    },
    image: function($) {
        var image = $('img[name=ItemImage]').attr('src')
        image = 'http://www.areyougame.com' + image.substring(2)
        return image
    }
}

var Search = {
    container: function ($) {
        var products = $('form[name=frmCategory] table:nth-child(4) tbody td[width=175]')
        return products;
    },
    name: function ($, product) {
        var name = $(product).find('strong a').text()
        name = cleanName(name)
        return name
    },
    price: function ($, product) {
        var price = $(product).find('strong:contains("$")').text()
        price = getPrice(price)
        return price
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href')
        url = 'http://www.areyougame.com/interact/' + url
        return url
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src')
        image = 'http://www.areyougame.com' + image.substring(2)
        return image
    }
}
