// sm-supplier-html/suppliers/www.shop.bt.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.title-callouts h1').text();
        (name)
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var delivery = $('.delivery-option').text()
        inStock= inStockCheck(delivery,inStock)
	return inStock;
},
    price: function($) {
        var price = $('.bus-pricing .lprice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('#imagegallerylargeimage img').attr('src');
        return 'https://www.shop.bt.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-container');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a').attr('title');
        name = cleanName(name)
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.lprice').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.shop.bt.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://www.shop.bt.com' + image;
        return image;
    }
}

var Category = {
    name: function($) {
        return cleanName($("div.category-page-heading h1").text());
    }
}

var Pagination = {

     items: function($) {
        return parseInt($("span.page-info").text().match(/[0-9]+/g)[1]);
     },
     pages: function($) {
        return Math.ceil(this.items($) / $("span.page-info").text().match(/[0-9]+/g)[2]);

     }
}
