// sm-supplier-html/suppliers/us.memebox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = memebox =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div#productName').text();
        var brand = $('div#brandName').text();

        name = brand + " " + name;
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.specialPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#product-image-slider img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('a.brand').text();
        var name = $(product).find('a.name').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.specialPrice').text();

        if(price.indexOf('.') < 1) {
            price = price + '.00'
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://us.memebox.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.image').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.product');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('div.product');
        // items = items.length;

        var pages = $('div.pagination a').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div.titleMain').text();
        name = cleanName(name);
        return name;
    }
}
