// sm-supplier-html/suppliers/www.clarksusa.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.productName').text();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked'

    },
    price: function($) {
        var price = $('span#currentPrice').html();
        price = getPrice(price);
        return price

    },
    image: function($) {
        var image = $('div.productImage-preload img').attr('src');
        image = 'https:' + image;
        return image;

    },
    description: function($) {
        var description = $('span#Description').text().replace('®', "").replace('®', "");
        return description;  // THIS MATCHES WHAT IT IS SUPPSED TO BRING BACK BUT FAILS THE TEST //
        // if(description !== null) {
        //     return description.replace(/(\r\n|\n|\r)/gm,"");
        // }
        
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.span-6');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.itemName').html();
        var material = $(product).find('div.itemDesc').html();
        name = cleanName(name);
        return name + " " + material

    },
    price: function($, product) {
        var salePrice = $(product).find('span.salePrice').html();
        var price = $(product).find('span.retailPrice').html();

        if(salePrice === null) {
            salePrice = price;
        }

        salePrice = getPrice(salePrice);
        return salePrice;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.clarksusa.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.thumb img').attr('src');
        image = 'https:' + image;
        return image
        
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) { // ONLY LOADS ALL WITH JAVASCRIPT, NO PAGES WHEN JAVASCRIPT IS DISABLED //

//      }
// }

var Category = {
    name: function($) {
        var name = $('div.breadcrumb ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
