// sm-supplier-html/suppliers/www.camping-online.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    // 'sku': (url) => {
    //     return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    // },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            // var string = Search.url($, product)
            // if(string) {
            //     info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            // } else {
            //     info.sku = null
            // }
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h2.product_name3').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var delivery = $('.thumbnailsnspecialoffer').text()
        var isInStock = true
        isInStock = inStockCheck(delivery, isInStock)
        return isInStock;
    },
    price: function ($) {
        var price = $('span#mainprice').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('a.mygallery img').attr('src');
        return image;
    },
    url: function ($) {
        return $
    },
    description: function ($) {
        var name = $('div.promoline').text();
        name = cleanName(name);
        return name
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.row li.first_item_of_line');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h5 a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('h5 a').attr('href');
        return  url
    },

    image: function ($, product) {
        var image = $(product).find('div.product_image img').attr('src');
        return "https://www.camping-online.co.uk" + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul.row li.first_item_of_line');
        items = items.length;
        return parseInt(items);
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.main_page_products h4').text();
        name = name.split("»")
        name = name.pop()
        name = cleanName(name);
        return name;
    }
}
