// sm-supplier-html/suppliers/www.shopping.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = shopping =  {
    "useJavascript": true,
    "usePhantom": 'shopping.phantom.js',
    
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.gridBox');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2 a.productName span').html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.productPrice a').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.gridItemTop a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        url = url.substring(url.indexOf("http:") + 0);
        return url
    },

    image: function($, product) {
        var image = $(product).find('div.gridItemTop img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        return parseInt($('.gridBox').length);
     },
     pages: function($) {
        var numitems = $('.gridBox').length;

        var products = $('span.numTotalResults').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('span.preferredCategory').text();
        name = cleanName(name);
        return name;
    }
}
