// sm-supplier-html/suppliers/www.etundra.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.sku=Product.sku($)
        return info
    },
    /*

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },
    */

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.sku = Search.sku($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-title span.product-name').text();
        var brand=$('h1.product-title span.product-brandname').text()
        var brandcode=$('h1.product-title span.product-brandcode').text()
        var indexBrand=brand.lastIndexOf(" ")
        brand=brand.substring(0,indexBrand)
        name=brand+' '+brandcode+' '+name   
        if(name!=null && isNaN(name)){

        name = cleanName(name);
        }
        return name;
    },
    inStock: function ($) {
        var avilibility=$('button.btn.btn-primary.addtocart')
        var instock=false
        if(avilibility){
            instock=true
        }
	return instock;
    },
    price: function ($) {
        var price = $('div.pricing-label.standard  span.amount').text();
        return getPrice(price)
    },
    image: function ($) {
        var image = $('div.product-images a#product-zoom').attr('href');
        return 'https://www.etundra.com' + image;
    },
    sku:function($){
        var skunumber=$('h6 span').text()
        return skunumber
    }
}

var Search = {
    container: function ($) {
        var product = $('div.grid-item');
        return product;

    },
    name: function ($, product) {
        var brand = $(product).find('span.product-fullname').attr('title');
        if(brand!=null && isNaN(brand)){
            brand = cleanName(brand);
        }
        
        return brand;
    },
    price: function ($, product) {
        var price = $(product).find('div.pricing-label.standard span.amount').text();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('a.product-name.product-link').attr('href');
        url = 'https://www.etundra.com' + url;
        return url;
    },

    image: function ($, product) {
        var image = $(product).find('div.image-ctl img').attr('src');
        return 'https://www.etundra.com' + image;
    },
    sku: function($,product){
        var skunumber=$(product).find('div.product-sku span').text()
        return skunumber
    }
}

var Pagination = {

    items: function ($) {
        var items=$('ul.dropdown-menu li:first-child').text()
        if(items){
        items=items.replace(/\D/g,'')
    }
        return items;
    },
    pages: function ($) {
        var products = $('div.products-found span.badge').text().match(/[0-9]+/g).pop();
        var items = $('ul.dropdown-menu li:first-child').text().replace(/\D/g,'');
        var pages = products / Number(items);
        pages = Math.ceil(pages);
        return pages.toString();
    }
}

var Category = {
    name: function ($) {
        var name = $('div.page-title h1').text();
        name = cleanName(name);

        return name;

    }
}
