// sm-supplier-html/suppliers/www.elc.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var routine = require ("../../lib/routine.js");
var inStockCheck = require("../../lib/inStockCheck.js")
module.exports = replaceWithwebAddressChangeToCamelCase =  {
    'useJavaScript': true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.t-page-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true
        var able = $('.b-availability').text()
        inStock = inStockCheck(able,inStock)

	return inStock;
},
    price: function($) {
        var price = $('div.b-price span').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $("link[rel='image_src']").attr('href');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.b-product_tile_wrapper');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.m-title_link').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('.b-product_pricing span.b-product_price').text();
        var sale = $(product).find('.b-product_pricing .m-sales_price').text();

        if(sale !== null) {
            price = sale
        }

        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.m-title_link');
        if(anchor) {
            url = 'http://www.elc.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.b-product_image div.js-default_image a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select#grid-paging-header').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
/*        var items = $('select#grid-paging-header').text();
        numitems = items.match(/[0-9]+/g).pop();*/

        var numitems = routine.matchNumbersAndPop
        ($, 'select#grid-paging-header');

        var products = $('div.js-pagination .b-filters-box').text()
        .match(/[0-9,]+/g);

        if(numitems !== null && products !== null) {
            products = products.pop().replace(',', '');
/*            sum = products / numitems
            return Math.ceil(sum)*/
            return routine.pagesFromTotalAndProd(products, numitems);

        } else {
            return null;
        }




     }
}

var Category = {
    name: function($) {
        var name = $('h1.t-page-title').text();
        name = cleanName(name);
        return name;
    }
}
