// sm-supplier-html/suppliers/ao.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var routine = require ("../../lib/routine.js"); 
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = ao =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);
        info.nexturl=Pagination.nexturl($)

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true
        var able = $('.back-in-stock-soon').text()
        inStock = inStockCheck(able,inStock)
        return inStock;

    },
    price: function($) {
        var price = $('div.price span.now').html()
        if(price){
            price= price.replace("&#xA3;","");
            price = price.replace(",", "") + '.00';
            return price;
        }else{
            return null
        }
        
        

    },
    image: function($) {
        var image = $('li[data-media-type="image"] img').attr('src');
        image = 'http:' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productInfoArea');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h2.listerPodtitle a').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.normal-price').text();
        var nowPrice = $(product).find('div.now span').text()
        price = getPrice(price.trim()+'.00')
        if(price){
            return price
        }else{
            nowPrice = getPrice(nowPrice.trim()+'.00')
            return nowPrice
        }
        

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://ao.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img#imgProduct').attr('src');
        image = 'http:' + image;
        return image;

    }
}

var Pagination = {

     items: function($) {
        return $('div.productInfoArea').length
     },
     pages: function($) {

        // var items = routine.matchNumbersAndPop($, "#numberProductsTouch #pageSize select");
        // var prodTot = routine.matchNumbersAndPop($, "span[data-item='offerCount']");

        // return routine.pagesFromTotalAndProd(prodTot, items);
        return 0
    },
    nexturl: function($){
        var next = $('#navigationNext').attr('href')
        return next
    }
}

var Category = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    }
}
