// sm-supplier-html/suppliers/www.lumens.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = lumens =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) {
            result.push(info)
        }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.productname').text();
        var brand = $('div.brandContainer a.brand').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.salesprice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#s7placeholder img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3[itemprop="name"]').text();
        var brand = $(product).find('a[itemprop="brand"]').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.salesprice').text();


        if(price.indexOf(',') > -1) {
            price = null;
        } else if(price.indexOf('From') > -1) {
            price = null
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.name a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.lumens.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.product');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
       var items = $('div.product');
        items = items.length;

        var products = $('div.result-count-options h4').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.main-content h1').text();
        name = cleanName(name);
        return name;
    }
}
