// sm-supplier-html/suppliers/www.jarrold.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = replace =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("a#product-default").attr('href');
        return 'www.jarrold.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.col-xs-6');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('span.detail strong').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.jarrold.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('span.square img').attr('data-orig')
        var image1=$(product).find('span.square img').attr('data-src')
        var image2 =$(product).find('span.square img').attr('src')
        if(image){
            image = 'http://www.jarrold.co.uk' + image;
            return image;
        }else if(image1){
            image1 = 'http://www.jarrold.co.uk' + image1;
            return image1
        }else if(image2){
            image2 = 'http://www.jarrold.co.uk' + image2;
            return image2
        }
        //image = image.substring(0, image.indexOf('{'));

        
        

        
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.dv-con h1").text());
    }
}
var Pagination = {

     items: function($) {
        var items = Search.container($);

        var pages = cleanName($("div.dv-results strong")
            .text()).match(/[0-9]+/g);

        if(items && pages) {
            return parseInt(items.length);
        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = cleanName($("div.dv-results strong")
            .text()).match(/[0-9]+/g);



        var items = cleanName($("div.dv-results strong")
            .text()).match(/[0-9]+/g);

        if(pages && items) {
            return Math.ceil(pages.pop() / this.items($));
        } else {
            return null
        }


        

     }
}
