// sm-supplier-html/suppliers/www.tvstoybox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript": true,

    "usePhantom" : "tvstoybox.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
     'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div .product-name h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span[itemprop="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.main-proimage').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.products-grid li.item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span[itemprop="price"]').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.products-grid li.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('ul.products-grid li.item');
        items = items.length;

        var result = null;
        var products = $('p.amount').text();

        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        return result
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title').text();
        name = cleanName(name);
        return name;
    }
}
