// sm-supplier-html/suppliers/www.carealotpets.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('td#productName h1 span#Anthem_ctl00_MainContentHolder_lblName__').text();
        (name);
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span.SitePrice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#zoomimage').attr('src');
        image = 'http://www.carealotpets.com/' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.record');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.recordname a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var salePrice = $(product).find('div.SingleProductDisplayPrice').children('span').eq(1).html()
        var standardPrice = $(product).find('div.SingleProductDisplayPrice').children('span').eq(0).html();
        salePrice = getPrice(salePrice);
        standardPrice = getPrice(standardPrice);
        
  if (salePrice == null)  
	{  
		var price = standardPrice
	}    
	else  
	{  
		var price = salePrice
	} 

      return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.carealotpets.com' + $(anchor).attr('href');
            url = url.toLowerCase();
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.recordimage a img').attr('src');
        image = 'http://www.carealotpets.com' + image;
        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('table#categorysearchcriteria select').text();
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }

        return parseInt(items);
     },
     pages: function($) {
        var items = $('table#categorysearchcriteria select').text();
        if(items === ""){
            return null
        }else {
            numitems = items.match(/[0-9]+/g).pop();    
        }
        

        var products = $('span#ctl00_MainContentHolder_SearchCriteria1_lblResult').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();
        }
        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
