// sm-supplier-html/suppliers/www.tooled-up.com/logic.js
var PageInfo = require('../../lib/pageInfo.js')
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product': function ($, next) {
        var info = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = gPrice($)
        info.image = image($)
        info.country = "uk"
        return info
    },

    'sku': (url) => {
        return url.split('/')[5]
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).split('/')[5];
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var isStock = function inStock($) {
    var isInStock = true;
    var availability = $('.prodavail').html()
    if (availability) {
        if ((availability.indexOf('Out of stock') > -1)
            // || (availability.indexOf('normally dispatched') > -1)
            || (availability.indexOf('Out of Stock') > -1)) {
            isInStock = false
        }
    }
    return isInStock
}


var gPrice = function thePrice($) {
    var price = $('span.price').first().text()
    price =getPrice(price)
    return price
}

var getName = function theName($) {
    return $('h1').text().trim()
}

var image = function image($) {
    return $('.prodimage a img').attr('src');
}


var Search = {
    container: function ($) {
        var searchProducts = $('div.col-1-3');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('p.desc').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.price').text()
        price = price.substring(price.lastIndexOf("£"));
        price = price.substring(1, 7).trim();
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.tooled-up.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.tooled-up.com' + image;

    }
}


var Pagination = {

    items: function ($) {
        var items = $('div.listingboxgrid');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div.listingboxgrid');
        items = items.length;

        var products = $('div.pagewrapper p').first().text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div#breadcrumb a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
