// sm-supplier-html/suppliers/www.cduniverse.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if (info.name != '') {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('tbody tr td h1').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        var status = $('td[valign="middle"] div span').text();

        if (status != '') {
            return 'unchecked';
        } else {
            return false;
        }
    },
    price: function ($) {
        var price = $('td[valign="middle"] div span').text();

        if (price != '') {
            price = getPrice(price);
        } else {
            price = null;
        }

        return price;
    },
    image: function ($) {
        var image = $('div#igcovera1 img').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('span#description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('tbody tr');
        return product;
    },
    name: function ($, product) {
        let name = $(product).find('a.chunkytitle').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('a.childprice').first().text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a.chunkytitle').attr('href');
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('tr#rpage td span').text()
        if (items === "") {
            return null
        } else {
            items = items.replace(',', '').match(/[0-9]+/g)[1];
        }

        return parseInt(items);
    },
    pages: function ($) {
        var items = $('tr#rpage td span').text()
        if (items === "") {
            return null
        } else {
            items = items.replace(',', '').match(/[0-9]+/g)[1];
        }

        var products = $('tr#rpage td span').text()
        if (products === "") {
            return null
        } else {
            products = products.replace(',', '').match(/[0-9]+/g)[2];
        }

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.label').text();

        return name;
    }
}
