// sm-supplier-html/suppliers/www.buildabear.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require("../../lib/routine.js");
var CategoryInfo = require('../../lib/CategoryInfo.js');


module.exports = buildABear = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.hero-title').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('.product-price span').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('a.product-image img').attr('src');
        return 'http://www.buildabear.co.uk' + image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.grid-tile');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-name a').html();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.product-sales-price').html()
            price = getPrice(price)
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('.product-name a');
        if (anchor) {
            url = 'https://www.buildabear.co.uk' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('#resultsPerPageresults').text();
        items = items.match(/[0-9]+/g)
        if (items) {
            items = items.pop();
            return parseInt(items);
        } else {
            return null;
        }

    },
    pages: function ($) {
        /*    var pages = $('ul.page-list').text();
            pages = pages.match(/[0-9]+/g).pop();
            return parseInt(pages);*/

        return routine.matchNumbersAndPop($, 'ul.page-list');


    }
}

var Category = {
    name: function ($) {
        var name = $('div.product-banner h1').text();
        name = cleanName(name);
        return name;
    }
}
