// sm-supplier-html/suppliers/www.sonicdirect.co.uk/logic.js
var PageInfo = require ('../../lib/pageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = price($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('.product-detail').find('.stock-level').html()

    if(producttop !== null) {
        if((producttop.indexOf('Stock Due Soon') > -1) 
        || (producttop.indexOf('Store Exclusive') > -1)) {
            isInStock = false
        }
    }
    return isInStock
}

var price = function thePrice($) {
    return $('.product-detail').find('.current-price').text().replace("£","").replace("(inc VAT)","").trim()
    //.substring(12).replace(/\n/g,' ').replace(/\t/g,'')  
}

var getName = function theName($) {
    return $('h1').text().trim()
}

var image = function image($) {
    return $('img.product-img').attr('src');
}

var Search = {
    container: function($) {
        var searchProducts = $('div.switchIt');
        return searchProducts;

    },
    name: function($, product) {
        var model = $(product).find('span.prod-title').html();
        var name = $(product).find('span.prod-title-desc').html();
        name = model + ' ' + name
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('li.current-price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.sonicdirect.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.product-image img').eq("-1").attr('src');
        image = 'http:' + image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.prod-grid-sm');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.prod-grid-sm');
        items = items.length;

        var products = $('span.displaying').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
