// sm-supplier-html/suppliers/www.lightingdirect.com/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');


module.exports = {
    //"useJavascript" : true,
    'search' : function($, next) {
        var result = []

        $('div.js-product-grid-container ul .item').map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($,product)
            info.country = "us"
            result.push(info)
            
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = getPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "us"
        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    var url = $(product).find('a').attr('href')
    return "https://www.lightingdirect.com" + url;
}


function searchName($, product) {
    var name = $(product).find('div.product-name').text();
    var name2 = $(product).find('p.product-title').text();
    return name + ' ' + name2
}


function searchPrice($, product) {
    var price = $(product).find('strong.text-green.product-price').text()
        price =  price.substring(price.lastIndexOf("$")+1);
    return price
}

function searchImage($, product) {
    var image = $(product).find('img').attr('src');
    return image;
}


// function getNumberOfPages($) {
//     var pages = $('#bottomSearchNav').find('.pages a').toArray()
//     if (pages !== null){
//         pages.pop()
//         pages = pages.pop()
//         pages = $(pages).text()
//     } else {
//         pages = 1
//     }
//     return pages
// }





var isStock = function inStock($) {
    var isInStock = true;
    var stockcount = $('#stockCount').text()
    stockcount = /[0-9]+/.exec(stockcount)
    if (stockcount !== null)
    {
        if (stockcount[0] < 10)
        // This has been tested by increasing to 20 and swapping sufficient stock test to false
        {
            isInStock = false
        }
    }
    return isInStock
}

var getPrice = function thePrice($, instock) {
    return $('.itemPrice').text().replace("$","").trim()
}

var getName = function theName($, instock) {
    return $('h1').text().replace(/\n/g,"").replace(/\t/g,"").replace(/\r\n\r\n/g," ").replace(/\r\n/g," ").trim()

}

function image($, instock) {
    var image = $('img#productImage').attr('src');
    image = 'http:' + image;
    return image;
}

var Pagination = {

     items: function($) {
        var items = $('.product-view');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.product-view');
        items = items.length;

        var products = $('div#numberOfItemsDisplayed span').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div#breadcrumb a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
