// sm-supplier-html/suppliers/www.luckyvitamin.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = luckyvitamin =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1 span[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.new-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.MagicZoomPup img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('ul.area-description li').text();

        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC# :') + 6).substring(0, 12);
            return barcode;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.listContainer');
        return searchProducts;
    },
    name: function($, product) {
        var brand=$(product).find('span.brand').text()
        var name = $(product).find('span.name').text();
        name = cleanName(brand+' '+name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.bigPrice').first().text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('[itemprop="url"]').first();
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.luckyvitamin.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('a.imageContainer img').attr('src');
        return 'https://www.luckyvitamin.com' + image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.listContainer');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li[itemprop="itemListElement"]');
        items = items.length;

        var products = $('div.section-heading b').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.pull-left').text();
        name = cleanName(name);
        return name;
    }
}
