// sm-supplier-html/suppliers/www.ethicalsuperstore.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.product-details__title').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var check =$('.stock-unavailable')
        var inStock= true
        inStock = instockCheck(check,inStock)
        return inStock

    },
    price: function($) {
        price = $('p.product-details__price').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('a.MagicZoomPlus').attr('href');
        return 'https:'+image;

    },
    url: function ($) {
        return $
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('article.cat-list--products')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.view_product_link').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('p.cat-list__price').text();
        price = getPrice(price);
        return price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a.media__img');
        if(anchor) {
            url = 'https://www.ethicalsuperstore.com' + $(anchor).attr('href')
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.media__img img').attr('src');
        return 'https:'+image;
    }
}

var Pagination = {

    items: function($) {
        var items = $('div.product_tile')
        return parseInt(items.length);

     },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function($) {
        var name = $('div.category_detail h1.no_margin').text();
        name = cleanName(name);
        return name;
    }
}
