// sm-supplier-html/suppliers/www.clasohlson.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('span[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var stock=$('.stockInfo').text()
        var inStock= true
        inStock=instockCheck(stock,inStock)
        return inStock

    },
    price: function($) {
        var price = $('p.normalPrice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('ul.thumbnailList li a img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productListStore div.productListItems');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('span[itemprop*="name"] a').html();
        
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('strong.clearBoth').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.clasohlson.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('span.productImg a img').attr('src');
        
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div#breadCrumbs div.content ul li").last().text());
    }
}
