// sm-supplier-html/suppliers/www.williams-sonoma.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.pip-summary span > span.price-state.price-sale > span.currency.currencyUSD > span.price-amount').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('#hero').attr('src');

        return image;
    },
    description: function($) {
        var description = $('div.accordion-tab-copy').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.product-cell');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a.product-name').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span > span.price-state.price-sale > span.currency.currencyUSD > span.price-amount').text();
        if (price.indexOf(' - ') > -1) {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.product-name').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-thumb').attr('src');

        return image;
    }
}

//
// there is no items count element on page and all products count element
//
var Pagination = {

     items: function($) {
         var items = $('li.product-cell').length;
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         $('div.pagination-component ul li').last().remove();
         var pages = $('div.pagination-component ul li').last().text();
         pages = JSON.parse(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('#sub-category > div.side-nav > div.left-nav > div').text();
        name = cleanName(name);

        return name;
    }
}
