// sm-supplier-html/suppliers/www.comacotoys.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript":true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.pageTopTitle span.h1-top').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span.price-new').text();
        price = getPrice(price)
        return price

    },
    image: function($) {
        var image = $('div.ewsMainImage a img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.ewsContentItem');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h3.name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.comacotoys.com/' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.listViewOnly a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
     
     },
     pages: function($) {
        var pages = $('span.NoOfPagesBottom').html();
        pages = parseInt(pages);

        if(pages) {
            return pages
        } else {
            return null;
        }

    }
}

var Category = {
    name: function($) {
        var name = $('span.h1-top').text();
        name = cleanName(name);
        return name;
    }
}
