// sm-supplier-html/suppliers/buyithealthy.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = buyithealthy =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.productRotateImage img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('li.wsm-prod-upccode').text().replace('UPC #:', '');
        return barcode.trim();
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.wsm-cat-list-item ');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.wsm-cat-title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.wsm-cat-price-sale-value').text();
        
        if(price === "") {
            price = $(product).find('span.wsm-cat-price-price-value').text();
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.wsm-cat-title a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.wsm-cat-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div#wsm-prod-adminbar select[name="perpage"]').html();
        
        if(items) {
            items = items.match(/[0-9]+/g).pop();
            return parseInt(items);
        }
     },
     pages: function($) {
        var items = $('div#wsm-prod-adminbar select[name="perpage"]').html();
        if(items) {
            numitems = items.match(/[0-9]+/g).pop();
        }

        var products = $('span.wsm-cat-pagination-items').text();
        if(products) {
            products = products.match(/[0-9]+/g).pop();
         
        var sum = products / numitems;
        return Math.ceil(sum);
        }
     }
}

var Category = {
    name: function($) {
        var name = $('span.wsm-store-header-cat-title').text();
        name = cleanName(name);
        return name;
    }
}
