// sm-supplier-html/suppliers/www.ebuyer.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = ebuyer =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.product-hero__title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        let stock =  true
        let check = $('.purchase-info>h4').text()
        stock = instockCheck(check,stock)
        return stock
},
    price: function($) {
        var price = $('div.inc-vat>p.price').text()
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.image-gallery__hero img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div#grid-view div.grid-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3.grid-item__title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.grid-item__price div.inc-vat p.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.ebuyer.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.grid-item__img a img ').attr('src');
        if(image.indexOf('http') === -1 && image.indexOf('https') === -1) {
            image = 'https://' + image
        }
        (image)
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("h1.page-title").text());
    }
}
var Pagination = {

     items: function($) {
        return Search.container($).length;

     },
     pages: function($) {



        var pages = $("li.listing-count").text()
        .split("products")[0].match(/[0-9]+/g);

        if(pages) {
            return parseInt(pages[0]);
        } else {
            return null;
        }





     }
}
