// sm-supplier-html/suppliers/www.coggles.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')

module.exports = coggles =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('.productDescShort').html()
        isInStock=instockCheck(producttop,isInStock)
        return isInStock
    },
    price: function($) {
        var price = $('p.product-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('a.product-img-zoom-action img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productBlock');
        return searchProducts;
    },
    name: function($, product) {
        let brand = $(product).find('.productBlock_productBrand').text();
        let productTitle = $(product).find('.productBlock_productName').text();
        brand = cleanName(brand);
        productTitle = cleanName(productTitle)
        return brand + ' ' + productTitle;
    },
    price: function($, product) {
        var price = $(product).find('.productBlock_priceValue').html()
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').first().attr('href');
        return 'https://www.coggles.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        return parseInt($('div.item').length)
     },
     pages: function($) {
        var numitems = $('div.item').length;

        var products = $('p.listTitle_results').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div.listTitle h1').text();
        name = cleanName(name);
        return name;
    }
}
