// sm-supplier-html/suppliers/www.sweatband.com/logic.js
var PageInfo = require ('../../lib/pageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = getPrice($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('#frmProductDetail').html();
    var outofstock = $('.product-shop-attributes .item-availability').html();


    if(producttop !== null) {
        if((producttop.indexOf('Not in Stock') > -1) 
        || (producttop.indexOf('Not in stock') > -1)
        || (producttop.indexOf('Unavailable') > -1)) {
            isInStock = false
        }
    }

    if(outofstock !== null) {
        outofstock = outofstock.toLowerCase();
        if(outofstock.indexOf('unavailable') > -1) {
            isInStock = false;
        }
    }
    return isInStock
}

var getPrice = function thePrice($) {
    return $('.price').text().replace('£','').replace(/\n/g,' ').replace(/\r/g,'').trim()  
}

var getName = function theName($) {
    return $('h1').text().trim()
}

var image = function image($) {
    return $('.item img').attr('src');
}

var Search = {
    container: function($) {
        var searchProducts = $('div.department-products div.department-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.department-product-name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.department-product-price').text();
        price = price.split("£")[1];
        //price = getPrice(price);
        return price.trim();

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
            url = 'https://www.sweatband.com' + url;
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.department-item-image img:nth-child(2)').attr('src');
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.department-products div.department-item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
       var items = $('div.department-products div.department-item');
        items = items.length;

        var products = $('div.pager ul li').text();
        products = products.match(/[0-9]+/g).pop();
    

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h2.category-name').text();
        name = cleanName(name);
        return name;
    }
}
