// sm-supplier-html/suppliers/www.petsupplies.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = petsupplies =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h2').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('div#cannotAddToCart:not(.is-hidden)').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('temporarily unavailable for sale') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('span#lowPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.main-product-image').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.ItemPageProductSummary').text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('.search-page-image-link');
    },
    name: function($, product) {
        var name = $(product).find('.wide-search-page-item-short-description').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.wide-search-page-item-savings-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).attr('href');
        return 'https://www.petsupplies.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('span.items-per-page').text();
        products = products.match(/\d+/gm).pop();
        return parseInt(products);
     },
     pages: function($) {
        var products = $('span.items-per-page').text();
        products = products.match(/\d+/gm).pop();

        var total = $('label.search-page-label').text();
        total = total.match(/\d+/gm).pop();

        var pages = total / products;
        return Math.ceil(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div#Breadcrumb span.BreadCrumb').text();
        name = cleanName(name);
        return name;
    }
}
