// sm-supplier-html/suppliers/www.asseenontv.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': function (url) {
        var thisPattern = new RegExp(/^(\d{3,6}-\d{1,2}(?![\;\?\&\.\,])*$)/g);
        var osku = url.substring(url.lastIndexOf("p=")+2,url.lastIndexOf("&"));
        var sku = thisPattern.test(osku) ? osku : null;
        return sku;
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=")+1,Search.url($, product).lastIndexOf("&"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        name = $('h1.prodInfoTitle').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.priceDetail.salePrice').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('img#detailLargeImage').attr('src');
        image = 'http://www.asseenontv.com' + image;

        return image;

    },
    description: function($) {
        var description = $('div[itemprop="description"]').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.productItem');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('span.prodHover-name').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var salePrice = $(product).find('span.prodHover-price-sale').text();
        salePrice = getPrice(salePrice);
        var originalPrice = $(product).find('span.prodHover-price-original').text();
        originalPrice = getPrice(originalPrice)

       if (salePrice == '') {
         price = originalPrice;
       } else {
         price = salePrice;
       }
       return price;

    },
    url: function($, product) {
        //http://www.asseenontv.com/detail.php?p=xxx
        var url = $(product).find('a').attr('href');
          //  url = url.substring(0, url.indexOf('&p'));
        url = 'http://www.asseenontv.com' + url;
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
            image = 'http://www.asseenontv.com' + image;

        return image;
    }
}

//
// No pagination. Only ajax
//
/*var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}*/

var Category = {
    name: function($) {
        var name = $('h1 span.breadLink').text();
        name = cleanName(name);
        return name;
    }
}
