// sm-supplier-html/suppliers/www.shoecarnival.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        $('div#cat').remove();
        var name = $('div.pdp-name').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.item-price').text();
        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        }   

        return price;
    },
    image: function($) {
        var image = $('img.amp-main-img').attr('src');

        return image;
    },
    description: function($) {
        $('div.productMetadataDescription ul.pdp').remove();
        var description = $('div.productMetadataDescription').text().trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.cell.grid_3');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.product-name a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.current-price').text();
        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        }   

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-name a').attr('href');
        var url = 'http://www.shoecarnival.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-image img').attr('src');
        image = 'http:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = parseInt($('#ctl00_SearchBody_NavigationTop_drpPageSize option:selected').attr('value'));

         return items;
     },
     pages: function($) {
         var items = $('#ctl00_SearchBody_NavigationTop_drpPageSize option:selected').attr('value');
         //
         // There is no search result count. But we have element that displays last page number
         //
         var sum = $('span.total').text().split(' ').pop();
         //(Math.ceil(sum));

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('div.breadcrumbs span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
