// sm-supplier-html/suppliers/www.damart.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if (info.price != null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.name').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        return inStock;
    },
    price: function ($) {
        var price = $('p.price.sale').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img.photo').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.k-catalog > li');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a.name').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        let price = null;
        var salePrice = $(product).find('p.price.sale').text();
        var regPrice = $(product).find('p.price').text();

        if (salePrice != "") {
            price = salePrice
        } else {
            price = regPrice
        }

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.name');
        if (anchor) {
            url = 'https://www.damart.co.uk' + $(anchor).attr('href');
        }
        return url;
    },

    image: function ($, product) {
        var image = $(product).find('img.photo').attr('src')
        if(image){
            if(image.indexOf('http')>-1){
                return image
            }else{
                return 'https:'+image
            }
        }
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.breadcrum > span').last().text()
        name = cleanName(name);
        return name;
    }
}
