// sm-supplier-html/suppliers/www.lenox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = Lenox =  {
    "useJavascript": true,
    "usePhantom" : "lenox.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us' // JUST NEED TO ADD THE COUNTRY IN LIKE THIS //
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us' // JUST NEED TO ADD THE COUNTRY IN LIKE THIS //
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock =true
        var check=$('.stock-info-element-wrapper').text()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        //var price = $('span.lenox-red').text().replace('$', '');
        var price = $('.occ-product-title').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.occ-product-details-scene7-image img').attr('src');
        image = image.substring(0, image.indexOf('?'));
        return image;
    },
    
    description: function($) {
        
        var fullDescr = $('.occ-description div[data-bind]').text();
        if (fullDescr != null) {
            fullDescrArr = fullDescr.split('\n')[0].replace('\r', '');
            return fullDescrArr;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product');

        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span[data-bind="text: sale"]').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = 'https://www.lenox.com' + $(product).find('.product-title a').attr('href');
        return url;

    },
    image: function($, product) { 
        var image = $(product).find('.bg-product-image').css("background-image").replace('url("', '').replace('")', '');
        image = image.substring(0, image.indexOf('?'));
        return image;
    },
    
}

var Pagination = {
     items: function($) {
         var items = $('.occ-lenox-select').text();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);  // THIS IS THE MAX NUMBER IN THE DROPDOWN ON THE SEARCH PAGE //
         return items;
     },

     pages: function($) {
         var products = $('.occ-grid-items-count').text();
         products = products.match(/[0-9]+/g).pop(); // MAX NUMBER OF PRODUCTS IN THE CATEGORY (TOP OF THE PAGE) //

         var items = $('.occ-lenox-select').text(); // MAX NUMBER OF PRODUCTS ON THE SEARCH PAGE IN THE DROPDOWN//
         items = items.match(/[0-9]+/g).pop();
         
         sum = products / items; // DIVIDE BOTH //
         return Math.ceil(sum);    
     } // PLEASE DO PAGINATION LIKE THIS //
}

var Category = {
    name: function($) {
        var name = $('div.occ-side-navigation-category-title .side-navigation-title').text();
        name = cleanName(name);
        return name;
    }
}
