// sm-supplier-html/suppliers/www.macys.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
let generateSku = require('../../lib/testPattern.js')
let regex = /([^?]+)\/?$/


module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": "true",
    "usePhantom": "macys.phantom.js",
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"
        return info
    },

    'sku': (url) => {
        let sku = generateSku(url, regex)[0].match('[0-9]+')[0]
        return sku
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = generateSku(Search.url($, product), regex)[0].match('[0-9]+')[0]
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            //result.push(info)

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        let name = cleanName($('h1.p-name').html());
        return name

    },
    inStock: function ($) {

        let oos = $('.msg.inner-cont').html();

        if (oos.toLowerCase().indexOf('usually ships within') > -1) {
            return true
        } else {
            return false
        }

    },
    price: function ($) {
        var price = $('div.price').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.sortableGrid li.productThumbnailItem');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('div.productDescription a').text();
        name = cleanName(name);
        name = name.replace('more like this', '').trim()
        return name;

    },
    price: function ($, product) {

        var price = null;
        var discountPrice = $(product).find('span.discount').html();
        discountPrice = getPrice(discountPrice);
        var standardPrice = $(product).find('span.regular').html();
        standardPrice = getPrice(standardPrice);

        if (discountPrice == null) {
            price = standardPrice
        } else {
            price = discountPrice
        }
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }

        if (url.indexOf('https://') > -1) {
            return url
        } else {
            url = 'https://www.macys.com' + url;
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Pagination = {

    items: function ($) {
        var items = $('li.productThumbnail');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('li.productThumbnail');
        items = items.length;

        var products = $('span#productCount').text();
        var result = null;

        if (products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items)
        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }

        return result
    }
}

var Category = {
    name: function ($) {
        var name = $('h1#currentCatNavHeading').text();
        name = cleanName(name);
        return name;
    }
}
