// sm-supplier-html/suppliers/www.toofaced.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = tooFaced =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($("section.product-info h1").text());
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        return getPrice($("section.product-info span.price span").text());
    },
    image: function($) {
        return $("section.imagery div.hero img").not(".zoomImg").attr("src");
    },
    description: function($) {
        return $("div.description").text().trim();
    }
}

var Search = {
    container: function($) {
        return $("li.product-tile");
    },
    name: function($, product) {
        return cleanName($(product).find("div.info strong").text());
    },
    price: function($, product) {
        return getPrice($(product).find("div.info span.price").text());
    },
    url: function($, product) {
        return "https://www.toofaced.com" + $(product).find("a.product-tile-link").attr("href");
    },
    image: function($, product) {
        return $(product).find("div.image img").attr("src");
    }
}

var Category = {
    name: function($) {
        var name = $('div.products h1').text();
        name = cleanName(name);
        return name;
    }
}

var Pagination = {

     items: function($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
     }
}
