// sm-supplier-html/suppliers/search.wdrake.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('a[itemprop="image"] img').attr('src');
        image = 'https:' + image;

        return image;
    },
    description: function($) {
        var description = $('span[itemprop="description"]').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.sli_grid');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h3.ProductThumbnailCopy a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.sli_price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div a').attr('href');

        return url;
    },  
    image: function($, product) {
        var img = $(product).find('div a img').attr('src');

        return img;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.sli_sort_footer select option:selected').text();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('div.sli_sort_footer select option:selected').text();
         items = items.match(/[0-9]+/g).pop();

         var products = $('h1.BreadcrumbHeader').text();
         products = products.match(/[0-9]+/g).pop();

         var sum = products / items;

         return Math.ceil(sum);
     }
}
