// sm-supplier-html/suppliers/www.discountschoolsupply.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = discountSchoolSupply =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span#mainContent_productDetail_lblName_0').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return "UNCHECKED"

    },
    price: function($) {
        var price = $('span.detail_two').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#mainContent_productDetail_detailImage_0').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('td.dtlstProducts');

        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.BrLink').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.listing_two').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.BrLink');
        if(anchor) {
            url = 'http://www.discountschoolsupply.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
    var image = $(product).find('input[type*="image"]').attr('src');
    return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div[style="text-align:left; width:33%; margin-top:1%; margin-bottom:1%; padding-left:1%; float:left; border:0px solid black;"]').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items)
     
     },
     pages: function($) {
        var items = $('div[style="text-align:left; width:33%; margin-top:1%; margin-bottom:1%; padding-left:1%; float:left; border:0px solid black;"]').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('span#mainContent_lblItemCount').text();
        products = products.match(/[0-9]+/g).pop()

        sum = products / numitems
        return Math.ceil(sum)

    }
}

var Category = {
    name: function($) {
        var name = $('span#mainContent_lblCatTitle h1').text();
        name = cleanName(name);
        return name;
    }
}
