// sm-supplier-html/suppliers/www.next.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.Title h1').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        return inStock;
    },
    price: function ($) {
        var price = $('div.Price div.nowPrice span').text();
        price=getPrice(price)
        return price;
    },
    image: function ($) {
        var image = $('div.shotWrapper img').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('article.Item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2.Title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.Price').text();
        
        if(price){
            price = price.trim()+'.00'
        }
        price = getPrice(price)
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https:'+url;
    },

    image: function ($, product) {
        var image = $(product).find('a.Image img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.SearchedFor h1').text()
        name = cleanName(name);
        return name;
    }
}
