// sm-supplier-html/suppliers/www.thewatchery.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
      'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var brand = $('.detailsBrand').not('.detailsStyle font').text();
        var item = $('.detailsDescrp').text();
        var name = brand + '' + item;
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('p.detailsPrice span.detailsPrice').html();
        var saleprice = $('p.detailsSale').html();

        if(saleprice !== null) {
            price = saleprice;
        }
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#detailimage').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li div.box');
        return searchProducts

    },
    name: function($, product) {
        var name = $(product).find('div.box p').html();
        var brand = $(product).find('div.box h3').html();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('ul.prices li strong').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.thewatchery.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.img img').attr('src');
        return image;
    }

}


var Pagination = {

     items: function($) {
        var items = $('li div.box');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li div.box');
        items = items.length;

        var products = $('div.search-heading h1').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('').text();
        name = cleanName(name);
        return name;
    }
}
