// sm-supplier-html/suppliers/www.drsfostersmith.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.categoryname').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span.priceactive').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#productImage').attr('src');
        image = 'http://www.drsfostersmith.com' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product2014item').not('.product2014cattab');
        return searchProducts;


    },
    name: function($, product) {
        var name = $(product).find('a.product_link div').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.jason11').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.drsfostersmith.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.prodImage a img').attr('src');
        image = 'http://www.drsfostersmith.com' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.product2014item').not('.product2014cattab');

        if(items.length) {
            return parseInt(items.length);
        } else {
            return null;
        }

     },
     pages: function($) {
        var numitems = $('div.product2014item').not('.product2014cattab').length;

        var products = $('div.totalcount').text();
        products = products.match(/[0-9]+/g);

        if(numitems && products) {
            products = products.pop();

            var sum  = products / numitems;
            return Math.ceil(sum);
        } else {
            return null;
        }


     }
}

var Category = {
    name: function($) {
        var name = $('div#categoryHeading').text();
        name = cleanName(name);
        return name;
    }
}
