// sm-supplier-html/suppliers/www.surfdome.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = surfdome =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.sd-product-price-current').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.pictureDeck img:nth-child(1)').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.sd-product-features').text();
        if (description) {
            description = description.replace(/\s+/gm, ' ').trim()
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('.product-card');
    },
    name: function($, product) {
        var brand = $(product).find('.product-card__brand').text()
        var name = $(product).find('.product-card__title').text();
        name = cleanName(brand+' '+name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product-card__price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.product-card__image-container').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.product-card__image-container>img').attr('src');   
        return image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }

var Category = {
    name: function($) {
        var name = $('article.sd-category-description h1').text();
        name = cleanName(name);
        return name;
    }
}
