// sm-supplier-html/suppliers/www.ulta.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("d") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.country = 'us'
        info.image = Product.image($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            //info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("d") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.country = 'us';
            info.image = Search.image($, product);
            //result.push(info)

            if (info.price !== null) {
                result.push(info)
            }

        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var brand = $('span.ProductMainSection__brandName a').text();
        var name = $('span.ProductMainSection__productName').text();
        name = cleanName(name);
        brand = cleanName(brand);
        return brand + " " + name;

    },
    inStock: function ($) {
        return 'UNCHECKED';

    },
    price: function ($) {
        var price = $('span.ProductPricingPanel__price').text();
        price=price.replace(/\s/g,'')
        price = getPrice(price);
        return price

    },
    image: function ($) {
        var image = $('.s7staticimage img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.productQvContainer');
        return searchProducts;

    },
    name: function ($, product) {
        var brand = $(product).find('h4.prod-title a').html();
        var name = $(product).find('p.prod-desc a').html();
        name = cleanName(name);
        brand = cleanName(brand);
        return brand + " " + name;

    },
    price: function ($, product) {
        var price = $(product).find('span.regPrice').text();
        if(price){
            price = getPrice(price);
        }else{
            price=$(product).find('span.pro-new-price').text()
            price=getPrice(price)
        }
        
        return price;
    },
    image: function ($, product) {
        var image = $(product).find('div.quick-view-prod img').attr('src');
        return image;
    },
    url:function($,product){
        var url=$(product).find('a').first().attr('href')
        return 'https://www.ulta.com'+url
    }
}

var Pagination = {

    items: function ($) {
        var items = $('.productQvContainer');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('.productQvContainer');
        items = items.length;

        var products = $('span.search-res-number').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.brand-breadcrumb-ul ul li a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
