// se-app-model.old/model/matchedRecords.js
const Base = require('../lib/es/base');
const handle = require('../lib/es/handle');
const async = require('async');

class MatchedRecords extends Base {

    constructor(config, payload) {
        super(config, payload.country + 'suppliers');
        this.country = payload.country;
    }

    getMatchedByUser(payload, next) {

        //console.log('getMatchedByUser: payload: ', JSON.stringify(payload, null, 2));

        if (payload.country) {
            async.waterfall([
                (done) => {
                    this.collection = payload.country + '-*,-' + payload.country + '-www-amazon-' + (payload.country === 'uk' ? 'co-uk' : 'com') + ',-' + payload.country + '-stats*';
                    let query = {
                        size: payload.size || 10,
                        from: payload.start || 0,
                        sort: [
                            { "created": { "order": "desc" } }
                        ],
                        query: {
                            match: {
                                stateDeltaUser: payload.selectedUser
                            }
                        }
                    };
                    this.find(query, done);
                },
                (records, done) => {
                    //console.log('records: ', JSON.stringify(records, null, 2));
                    let asins = [];
                    let recordObj = {};
                    if (records.total > 0) {
                        if (!records.records || records.records.length ===0){
                            records.records = [];
                        }
                        records.records.map((thisRecord) => {
                            let asin = thisRecord.asin ? thisRecord.asin : thisRecord.stateDeltaAsin;
                            if (asin) {
                                asins.push(asin);
                            }
                            recordObj[thisRecord.id] = thisRecord;
                        });
                    }
                    done(null, asins, records.total, recordObj);
                },
                (asins, total, records, done) => {
                    this.collection = payload.country + '-www-amazon-' + (payload.country === 'uk' ? 'co-uk' : 'com');
                    let result = {
                        total: total,
                        records: records
                    };
                    if (asins.length) {
                        this.mgetNew({
                            index: this.collection,
                            type: 'amazon',
                            ids: asins
                        }, (err, amzRecords) => {
                            if (err) {
                                done(err);
                            } else {

                                //console.log('amzRecords: ', amzRecords);
                                Object.keys(result.records).map((thisKey) => {

                                    let asin = result.records[thisKey].asin ? result.records[thisKey].asin : result.records[thisKey].stateDeltaAsin ? result.records[thisKey].stateDeltaAsin : null;
                                    let thisMatch = asin && amzRecords.found ? amzRecords.found[asin] : null;

                                    result.records[thisKey].amazon = {
                                        name: thisMatch ? thisMatch._source.amazonName : 'Unknown',
                                        image: thisMatch ? thisMatch._source.amazonImage : '',
                                        url: thisMatch ? thisMatch._source.amazonUrl : ''
                                    };
                                    //console.log('Adding amazon: ', result.records[thisKey].amazon);

                                });
                                next(null, result);
                            }
                        });
                    } else {
                        next(null, result);
                    }
                }
            ], (err) => {
                next(err);
            });
        } else {
            next('NO_COUNTRY');
        }
    }

}

module.exports = MatchedRecords;


/* matchedRecords: {
    "loading": false,
        "total": 2,
            "start": 0,
                "size": 2,
                    "records" : {
        "af1fd233097d6392edde6baa2d5dc505" : {
            id: "af1fd233097d6392edde6baa2d5dc505",
                name: "GroClock The Best Child Sleep Trainer and Sleepy Farm Book",
                    url: "http://www.tesco.com/direct/groclock-the-best-child-sleep-trainer-and-sleepy-farm-book/208-8086.prd?skuId=208-8086",
                        image: "https://tesco.scene7.com/is/image/tesco/208-8086_PI_TPS946456?wid=170&hei=170&$Offers$",
                            state: "NOT_A_MATCH",
                                created: "2017-04-08T02:42:43.300Z",
                                    updated: "2018-03-13T15:42:26.110Z",
                                        supplier: "www-tesco-com",
                                            delta: [
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-08T02:42:43.300Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-08T02:42:45.950Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-12T11:02:16.448Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-12T11:02:19.300Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-15T14:25:42.930Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-15T14:25:45.973Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-16T01:12:00.045Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-16T01:12:03.235Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-16T21:39:13.113Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-16T21:39:16.302Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-25T12:14:39.106Z"
                                                },
                                                {
                                                    "price": 19.99,
                                                    "timestamp": "2017-04-25T12:14:42.162Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-11T01:05:14.828Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-16T12:50:05.465Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-16T12:50:07.885Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-17T00:35:23.027Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-17T00:35:26.627Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-18T15:45:04.821Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-18T16:03:48.281Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-19T23:10:13.553Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-19T23:27:45.859Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-20T19:53:10.715Z"
                                                },
                                                {
                                                    "price": 24.99,
                                                    "timestamp": "2017-05-20T20:00:11.186Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-05-24T15:40:55.696Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-05-25T07:15:27.527Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-05-25T09:35:34.272Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-05-25T09:39:29.283Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-01T08:31:19.326Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-01T08:48:13.790Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-06T07:51:15.381Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-06T09:28:16.107Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-06T09:29:49.891Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-08T17:45:59.473Z"
                                                },
                                                {
                                                    "price": 22.99,
                                                    "timestamp": "2017-06-08T17:49:47.761Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-13T08:41:43.124Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-13T09:00:15.014Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-13T12:14:59.137Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-13T12:18:41.034Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-17T01:57:51.504Z"
                                                },
                                                {
                                                    "price": 25.95,
                                                    "timestamp": "2017-06-17T02:01:12.675Z"
                                                },
                                                {
                                                    "price": 21.38,
                                                    "timestamp": "2017-06-19T10:27:10.425Z"
                                                },
                                                {
                                                    "price": 21.38,
                                                    "timestamp": "2017-06-19T10:43:28.546Z"
                                                },
                                                {
                                                    "price": 21.38,
                                                    "timestamp": "2017-06-19T19:18:08.976Z"
                                                },
                                                {
                                                    "price": 21.38,
                                                    "timestamp": "2017-06-19T19:21:49.057Z"
                                                },
                                                {
                                                    "price": 21.38,
                                                    "timestamp": "2017-06-20T11:21:21.727Z"
                                                }
                                            ],
                                                stringMatch: [
                                                    {
                                                        "id": 1,
                                                        "threshold": true,
                                                        "name": "Groclock The Best Child Sleep Trainer And Sleepy Farm Book",
                                                        "matched_id": "B00WPZGI9U",
                                                        "matched_name": "Groclock The Best Child Sleep Trainer And Sleepy Farm Book",
                                                        "match_distance": 0,
                                                        "lv_distance": 0,
                                                        "jaccard_distance": 0,
                                                        "price": -1,
                                                        "category": "Unknown",
                                                        "sales_rank": -1
                                                    },
                                                    {
                                                        "id": 2,
                                                        "threshold": true,
                                                        "name": "Groclock The Best Child Sleep Trainer And Sleepy Farm Book",
                                                        "matched_id": "B002APJCNE",
                                                        "matched_name": "The Gro Company Groclock Sleep Trainer",
                                                        "match_distance": 0.0736,
                                                        "lv_distance": 42,
                                                        "jaccard_distance": 0.25,
                                                        "price": -1,
                                                        "category": "Unknown",
                                                        "sales_rank": -1
                                                    }
                                                ],
                                                    asins: [
                                                        "B00WPZGI9U",
                                                        "B002APJCNE"
                                                    ],
                                                        asin: null,
                                                            price: 21,
                                                                domain: "www.tesco.com",
                                                                    sku: "208-8086",
                                                                        skuSupplier: true,
                                                                            country: "uk",
                                                                                stringMatchVersion: "3.0.0",
                                                                                    matchDistance: 0,
                                                                                        category: "kitchen_display_on_website",
                                                                                            asinMatched: {
                date: "2018-02-13T15:15:36.438Z",
                    status: "NOCHANGE"
            },
            matchStatus: "checked",
                stateDeltaAsin: null,
                    stateDeltaState: "IS_A_MATCH",
                        stateDelta: [
                            {
                                "asin": "B00WPZGI9U",
                                "state": "IS_A_MATCH",
                                "user": "mike",
                                "timestamp": "2018-03-13T15:00:13.587Z"
                            },
                            {
                                "asin": "B00WPZGI9U",
                                "state": "IS_A_MATCH",
                                "user": "mike",
                                "timestamp": "2018-03-13T15:09:23.418Z"
                            },
                            {
                                "asin": "B00WPZGI9U",
                                "state": "NOT_A_MATCH",
                                "user": "martha",
                                "timestamp": "2018-03-13T15:12:27.567Z"
                            },
                            {
                                "asin": null,
                                "state": "IS_A_MATCH",
                                "user": "mike",
                                "timestamp": "2018-03-13T15:14:51.105Z"
                            },
                            {
                                "asin": null,
                                "state": "IS_A_MATCH",
                                "user": "mike",
                                "timestamp": "2018-03-13T15:40:28.741Z"
                            },
                            {
                                "asin": null,
                                "state": "IS_A_MATCH",
                                "user": "mike",
                                "timestamp": "2018-03-13T15:42:26.110Z"
                            }
                        ],
                            stateDeltaUser: "mike",
                                notMatchedAsins: "B00WPZGI9U"
        }
    }
}, */
