// se-app-model.old/model/supplierStats.js
const handle = require('../lib/mongo/handle');
const Base = require('../lib/mongo/base');
const mongodb = require('mongodb');
const request = require('superagent');

class SupplierStats extends Base {

    constructor(config, collection) {
        super(config, collection.country + 'supplierstats');
        this.country = collection.country;
    }

    updateStats(payload, next) {
        this.update({
            query: { supplierId: payload.supplierId },
            update: {
                $set: {
                    stats: payload.stats,
                    totalProductsInSupplier: payload.totalProductsInSupplier,
                    overallAverageSalesRank: payload.overallAverageSalesRank,
                    overallProfit: payload.overallProfit,
                    overallRoi: payload.overallRoi,
                    salesRankNumber: payload.salesRankNumber,
                    totalScanned: payload.totalScanned,
                    updated: payload.updated
                }
            }
        }, (err, result) => {
            next(err, result)
        })
    }

}

module.exports = SupplierStats
