// sm-supplier-html/suppliers/www.melissaanddoug.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('#product_price').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('.main_image img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.grid-item-product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.product-standard-price').text();
        var saleprice = $(product).find('span.product-sales-price ').text();

        if (saleprice !== "") {
            price = saleprice;
        }

        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.melissaanddoug.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img.grid-item-image').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.grid-item-product');
        return parseInt(items.length);
    },
    pages: function ($) {
        var numitems = $('div.grid-item-product').length;

        var products = $('div.pagination--results-hits').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.search-results-title').text();
        name = cleanName(name);
        return name;
    }
}
