// sm-supplier-html/suppliers/www.itechdeals.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js"); 

module.exports = itechdeals =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.product-primary-column h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('.product-type-data').html()

       if(producttop !== null) {
            if ((producttop.indexOf('Out of Stock') > -1)
            || (producttop.indexOf('Pre-Order') > -1)
            || (producttop.indexOf('Unavailable') > -1)) {
                isInStock = false
            }
        }
        return isInStock
    },
    price: function($) {
        var price = $('p.special-price span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.img-box a').attr('href');
        return "https://www.itechdeals.com" + image;
    },
    description: function($) {
        var description = $('div#product-tabs div.std').text().trim();
        
        if(description !== "") {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-index');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-title').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.prod-price').text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price =$(product).find('.onsale').text()
            price = getPrice(price)
            return price
        }
        
    },
    url: function($, product) {
        var url = $(product).find('.product-info-inner>a').attr('href');
        var prefix = "https://www.itechdeals.com";
        return prefix + url;
    },
    image: function($, product) {
        var image = $(product).find('.prod-image img').attr('data-srcset');
        if (image){
            if (image.indexOf(' ') > -1){
                return  'https:'+image.split(' ')[0]
            }
        }
        return 'https://www.itechdeals.com' + image;
    }
}

var Pagination = {

     items: function($) {
/*        var items = $('div.select-list ul').html();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);*/

        return routine.matchNumbersAndPop($, 'div.select-list ul', "html");

     },
     pages: function($) {
/*        var pages = $('div.pages').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);*/

        return routine.matchNumbersAndPop($, 'div.pages');
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();
        name = cleanName(name);
        return name;
    }
}
