// sm-supplier-html/suppliers/www.officestationery.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('.productContent h1').html();
        name = cleanName(name);
        return name;

    },

    inStock: function($) {
        var isInstock = true;
        var stockcheck = $('.productContent div.discontinuedProductBox').html();
        var stocklevel = $('ul.stockCheck #stock').html();
        var availability = $('p.stock').text()
        var stockQuantity = availability.replace( /^\D+/g, '')
        
        if(stockQuantity < 1) {
            isInstock = false
        }

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('discontinued product') > -1) {
                isInstock = false;
            }
        }

        if(stocklevel !== null) {
            stocklevel = stocklevel.toLowerCase();
            if(stocklevel.indexOf('out of stock') > -1) {
                isInstock = false;
            }
        }

        return isInstock;
    },

    price: function($) {
        var price = $('.incVatPrice').html();

        if (price === null) {
            var price = $('.incVatArea').html();
        }

        if(price === null) {
            price =$('.priceExVat').text()
        }
        price = getPrice(price);
        return price;

    },

    image: function($) {
        var image = $('img.img-drop').attr('src');
        image = 'https://www.officestationery.co.uk' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.pictureList.xx li');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a strong').html();
     //   name = cleanName(name.substring(name.lastIndexOf("<strong>")+8,name.lastIndexOf("</strong>")));
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.price').html();
        price = getPrice(price);
        return price;
    },  
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        url = 'https://www.officestationery.co.uk' + url;
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.listImage').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.paganation div').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
