// sm-supplier-html/suppliers/www.fivebelow.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name= $('.product-name').text()
        name= cleanName(name)
        return name
    },
    inStock: function ($) {
        var inStock =true
        var check=$('.availability').text()
        inStock=instockCheck(check,inStock)
        return inStock
    },
    price: function ($) {
        var price=$('.price-box').text()
        price =getPrice(price+'.00')
        return price
    },
    image: function ($) {
        var image =$('.gallery-item img').attr('src')
        return image
    },
    url: function ($) {
        return $
    }
}

var Search = {
    container: function ($) {
        let container = $('#category-products-list li.item')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('h2').attr('title')

        if (name) {
            name = cleanName(name)
        } else {
            name = null
        }
        return name

    },
    price: function ($, product) {
        var price = null
        let dec = $(product).find('div.price-box').text()
        let roundedPrice = $(product).find('div.price-box span:nth-child(1)').text() + '.00'
        dec = getPrice(dec)

        if (dec == null) {
            price = roundedPrice
        } else {
            price = dec
        }
        price = getPrice(price)
        return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('img').attr('src')
        return image

    }
}

var Pagination = {

    items: function ($) {},
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {}
}
