// sm-supplier-html/suppliers/www.campusgifts.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js"); 
var inStockcheck = require('../../lib/inStockCheck')
module.exports = campusGifts =  {
    "useJavascript": true,
    "usePhantom": "campusGifts.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        isInStock = true;
        var availability = $('p.out-of-stock').html();
        isInStock=inStockcheck(availability,isInStock)
        return isInStock;

    },
    price: function($) {
        var price = $('div.price-box .regular-price .price').html();

        if (price === null) {
            price = $('div.price-box .special-price .price').html();
        }

        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.main-image img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price-box .regular-price .price').html();

        if (price === null) {
            price = $(product).find('div.price-box .special-price .price').html();
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.limiter select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
/*        var pages = $('div.pages ol').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);*/

        return routine.matchNumbersAndPop($, "div.pages ol");
     
     }
}

var Category = {
    name: function($) {
        var name = $('div.left-categorys h1').text();
        name = cleanName(name);
        return name;
    }
}
