// sm-supplier-html/suppliers/www.babyearth.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = babyearth = {

    'sku': (url) => {
        url = url.replace('.html', '')
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""))
            info.name = Search.name($, product);
            info.sku = sku.replace('.html', '');
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        var brand = $('span.product_make a').text();

        name = name + " by " + brand;
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        return isInStock;
    },
    price: function ($) {
        var price = $('span.price_extra_price').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div#left-column-main-photo img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.card');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('span.pretext').text();
        var name = $(product).find('h3').text();
        //    var brand = $(product).find('a.a_brandname').text();
        name = cleanName(brand + ' ' + name);
        return name
    },
    price: function ($, product) {
        var price = $(product).find('.price').text();
        price = getPrice(price);
        if(price){
            return price
        }else{
            price = $(product).find('[itemprop="price"]').attr('content')
            price = getPrice(price)
            return price
        }
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.babyearth.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src')
        image = image + ""
        image = image.split(' ')[0]
        return 'https:' + image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.prod_box')
        items = parseInt(items.length);

        if (items) {
            return items;
        } else {
            return null;
        }


    },
    pages: function ($) {
        var items = this.items($);

        var products = $('div#item_grid table.margin_bot20 td').not('td[align="right"]').text().match(/[0-9]+/g);

        if (products && items) {
            products = products.pop();

            return Math.ceil(products / items);
        } else if (!products && items) {
            return 1;
        } else {
            return null;
        }


    }
}

var Category = {
    name: function ($) {
        var name = $('span.bot_crumb').text();
        name = cleanName(name);
        return name;
    }
}
