// sm-supplier-html/suppliers/www.famous-smoke.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = famoussmoke = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('div.rightgap').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img.mainimg').attr('src').replace(' ', '');
        return 'www.famous-smoke.com' + image;
    },
    description: function ($) {
        var description = $('p.p-copy').text();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('a.dealitembox');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('div.dealitembrand').text();
        var name = $(product).find('div.dealitemname').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.dealpricebox .cblack').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).attr('href');
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.dealskubox img').attr('src');
        return 'https://www.famous-smoke.com' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul.filterlist').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items)
    },
    pages: function ($) {
        var items = $('ul.filterlist').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.showing-items').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name;
    }
}
