// x4-model/cache/lib/category.js
const Mongo = require('../../mongo')
const redis = require('../../redis').handle
const AmazonIdToCategoey = require('./amazonIdToCategory')

module.exports = {

    generate(config, next) {

        this.config = config
        this.redisHandle = redis(config.redis)
        this.AMAZONIDTOCAT = new AmazonIdToCategoey(this.config.redis)

        this.redisHandle.get("CATCACHEOK", (err, data) => {
            if(err) {
                next(err)
            } else {
                if(data === null || config.force) {
                    console.log("Creating Cache")
                    this.createCache(next)
                } else {
                    console.log('Category Cache OK')
                    next()
                }
            }
           
        })

    },

    getName(country, id, next) {
        this.AMAZONIDTOCAT.getName(country, id, next)
    },

    createCache(next) {
        
        this.MDB = new Mongo(this.config.mongo)

        let countries = ["uk", "us"]
        let total = countries.length

        countries.map((country) => {
            console.log("CATEGORY CACHE FOR ", country)
            this.MDB.conn(country + "categories", (err, collection) => {
                if (err) {
                    next(err)                    
                } else {
                    collection.find({}).toArray((err, data) => {
                        if (err) {
                            next(err)
                            console.error(err)
                        } else {
                            data.map((cat) => {
                                this.AMAZONIDTOCAT.setCategory(country, cat.amazonId, cat.name)
                            })
                            total--;
                            if (total === 0) {
                                this.redisHandle.set("CATCACHEOK", true)
                                next()
                            }
                        }
                    })
                }

            })
        })
    }

}
