// sm-supplier-html/suppliers/www.searsoutlet.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('section#pdp-product h1 strong').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var isInStock = true;
        // var availability =$('div#pdp--itemWrap').text();

        // if(availability.indexOf('Item Is Sold Out') > -1) {
        //     isInStock = false
        // }

        return isInStock;
    },
    price: function($) {
        //all products here have price range "$275.00 - $275.00"
        //if prices in price range are the same - we get it
        var price = $('div#pdp--itemWrap div[ng-show="so.pdpHasProduct"] strong.ng-scope').text();

        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        }  

        return price;
    },
    image: function($) {
        var image = $('div#pdp--imageWrap div a img[itemprop="image"]').attr('src');
        image = image.substring(0, image.indexOf('?'));
        return image;
    },
    description: function($) {
        $('div#overviewWrap h2').remove();
        $('div#overviewWrap ul').remove();

        var description = $('div#overviewWrap div').text().trim();;

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.plp-item--container');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a').attr('title')
        name = cleanName(name);
        if(name){
            return name;
        }else{
            name = $(product).find('a>img').attr('alt')
            name=cleanName(name)
            return name
        }
        
    },
    price: function($, product) {
        var price = $(product).find('div.plp--price').html();
            price = getPrice(price)

    ///    if(price.indexOf('-') > -1) {
    //        price = null      
    //    } else {
    //        price = getPrice(price);
    //    }   
        return price
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        url = 'http://www.searsoutlet.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src')
        if(image){
            if(image.indexOf('http')>-1){
                return image
            }else{
                return 'https:'+image
            }
        }
        return image;
    }
}

var Pagination = {

     items: function($) {
        
         var items = $('section#plp--titleInfo div p strong').html();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);
        return items;
         
     },
     pages: function($) {
         var items = $('section#plp--titleInfo div p strong').html();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);
        
        var products = $('section#plp--titleInfo').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / items;
        return Math.ceil(sum);
         
     }
}
var Category = {
    name: function($) {
        var name = $('h1#pageHeading').text();
        name =name.match( /"(.*?)"/).pop();
        //name = cleanName(name);
        return name;
    }
}
