// sm-supplier-html/suppliers/www.schoolstationery.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = schoolstationery = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        isInStock = true;
        var availbility = $('li#stock').text();

        if (availbility !== null) {
            availbility = availbility.toLowerCase();
            if (availbility.indexOf('out of stock') > -1) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function ($) {
        var price = $('strong.incVatPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("img.img-drop").attr('src');
        return 'http://www.schoolstationery.co.uk/' + image;
    },
    barcode: function ($) {
        var barcode = $('div#productPageExtraInfo span:nth-child(3)').text();
        return barcode.match(/\d+/)[0];
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('table.listings tr[style="background: rgb(255, 255, 255);"]');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('td.listingDesc a').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('td.listingPrice').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.schoolstationery.co.uk' + url;

    },
    image: function ($, product) {
        var image = $(product).find('div.panelPop img').attr('src');
        return 'https://www.schoolstationery.co.uk' + image;
    }
}
var Pagination = {

    items: function ($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

    },
    pages: function ($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.paganation a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
