// sm-supplier-html/suppliers/www.hearthsong.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('li.phw_product_details_listright h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
	return 'unchecked';
},
    price: function($) {
        var price = $('ul.phw_product_pricinglist li p.phw_product_price').text();

        if (price != '') {
            price = getPrice(price);
        } else {
            price = null;
        }

        return price;
    },
    image: function($) {
        var image = $('img#phw_product_image_main').attr('src');
        image = 'http:' + image;

        return image;
    },
    description: function($) {
        var description = $('div.ui-accordion-content').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product-item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('.name').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price').text();

        if (price.indexOf(' - ') > -1) {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.details a').attr('href');
        url = 'https://www.hearthsong.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://www.hearthsong.com' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('select.phw_search_select_record option:nth-child(4)').first().text().split(' ').shift();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('select.phw_search_select_record option:nth-child(4)').first().text().split(' ').shift();
         var products = $('span#phw_container_search_count').text().match(/[0-9]+/g).pop();

         var pages = products / items;
         pages = Math.ceil(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        $('span#phw_container_search_count').remove();
        var name = $('h1#phw_header_search').text();

        return name;
    }
}
