// sm-supplier-html/suppliers/www.booksamillion.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('#details-description-container div > span.details-title-text > strong').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var inStock = $('div.price-block-quote-text span strong').text().includes('In Stock');

        if (inStock) {
            inStock = true;
        } else {
            inStock = false;
        }
        return inStock;
    },
    price: function($) {
        var price = $('div.price-block-binding span.details-title-text strong').text();
        price = getPrice(price);
		
        return price;       
    },
    image: function($) {
        var image = $('div#details-image-container a img').attr('src');
		
        return image;
    },
    description: function($) {
        var description = $('#content > div > div:nth-child(11) > p').text().replace('\"\n', '');

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('table.feature_divider');
        return product;
    },
    name: function($, product) {
        var name = $(product).find('span.style2 a').text();
        name = name.split('$').shift()
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.style2 a').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('#result-list ol li div.meta');
        return parseInt(items.length);
         
     },
     pages: function($) {
        var numitems = $('#result-list ol li div.meta').length;
        
        var products = $('form.per-page div strong').text()
        if(products === ""){
            return null
        }else {
            products = products.split(' ')[5].replace(',', ''); 
        }

        var sum = products / numitems;
        return Math.ceil(sum);
         
     }
}

var Category = {
    name: function($) {
        var name = $('div.search span[style="color:#FFFFFF; font-size:18px; font-weight:bold; "]').text()
        name = name.replace(/[“”]+/g, '').replace(/\w\S*/g, function(txt){return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();});;
        name = cleanName(name);
        return name;
    }
}
