// sm-supplier-html/suppliers/www.outdoorgear.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        return isInStock;

    },
    price: function ($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('span.rollover-span img').attr('src');
        return image;

    },
    url: function ($) {
        return $
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.cat-prod');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('span.cat-prod-desc').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.cat-prod-price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.outdoorgear.co.uk' + url
    },

    image: function ($, product) {
        var image = $(product).find('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.cat-prod');
        items = items.length;
        return parseInt(items);
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('ol.breadcrumb li').last().text();
        name = cleanName(name);
        return name;
    }
}
