// sm-supplier-html/suppliers/www.dollartree.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('#unitPrice dt').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.image_wrap a img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.itemCell');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h6 a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('.unitPrice').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.dollartree.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.cat-thu-ima-cont a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.itemCell');

        if(items.length) {
            return parseInt(items.length);
        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = $('div.subcatDetails a.pageCount').text();
        pages = pages.match(/[0-9]+/g);


        if(pages) {
            pages = pages.pop();
            pagesVarTwo = pages.match(/\d/g)

            if(pagesVarTwo) {
                pagesVarTwo = pagesVarTwo.pop();

                return parseInt(pagesVarTwo);   
            } else {
                return null;
            }
        } else {
            return null;
        }

     }
}

var Category = {
    name: function($) {
        var name = $('td.subcatHeader h1').text();

        if(name.indexOf(':') > -1) {
            name = name.split(':').shift();
        }
        name = cleanName(name);
        return name;
    }
}
