// sm-supplier-html/suppliers/www.dollaritemdirect.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.center_block_1 h2 span').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.prod-detail-cost-value').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div div a img[title]').attr('src');
        image = 'http://www.dollaritemdirect.com' + image;

        return image;
    },
    description: function($) {
        var description = $('div.prod-detail-desc').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('#ctl00_pageContent_ctl06_productList > tbody > tr > td > div > table > tbody > tr');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('td:nth-child(1) > a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-list-cost-value').text();

        if (price.indexOf('=') > -1) {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function($, product) {
        var url = $(product).find('td:nth-child(1) > a').attr('href');
        url = 'http://www.dollaritemdirect.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-list-img').attr('src');

        image = 'http://www.dollaritemdirect.com' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('td#ctl00_pageContent_ctl06_resultCell').text().match(/[0-9]+/g);

         if(items) {
            items = parseInt(items[1]);

            return items;
         } else {
            return null;
         }
     },
     pages: function($) {
         var items = $('td#ctl00_pageContent_ctl06_resultCell').text().match(/[0-9]+/g);
         var products = $('td#ctl00_pageContent_ctl06_resultCell').text().match(/[0-9]+/g);

         if(items && products) {
            items = items[1];
            products = products.pop();


            var pages = products / items;
            pages = Math.ceil(pages);

            return pages;
         } else {
            return null;
         }

     }
}

var Category = {
    name: function($) {
        var name = $('div.page-html-content h1').text();

        return name;
    }
}
