// sm-supplier-html/suppliers/www1.macys.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript" : "true",
    "usePhantom": "macys.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
             //result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.productTitle h1.productName').html();
        var brand = $('div.productTitle a.brandNameLink').html();
        name = cleanName(name);
        brand = cleanName(brand);
        return brand + " " + name;

    },
    inStock: function($) {
        return 'unchecked'

    },
    price: function($) {
        var price = $('div.tieredPrices').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('li.imageItem img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.productThumbnail');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.shortDescription a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('.textWrapper span.colorway-price span.priceSale ').text();
        var price2 = $(product).find('.textWrapper span.colorway-price span.second-range').text();
        var price3 = $(product).find('.textWrapper span.colorway-price span.first-range').text()


        if(price === '') {
            price = price2;
        }

        if(price === '') {
            price = price3;
        }

        if (price.indexOf('-') > -1) {
            price = null;
        }    


        price = getPrice(price); 
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }

        if(url.indexOf('http://') > -1) {
            return url
        } else {
            url = 'http://www1.macys.com' + url;
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.thumbnailMainImage').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('li.productThumbnail');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.productThumbnail');
        items = items.length;

        var products = $('span#productCount').text();
        var result = null;
       
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1#currentCatNavHeading').text();
        name = cleanName(name);
        return name;
    }
}
