// sm-supplier-html/suppliers/www.microcenter.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('span[itemprop*="name"] span').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var instock = true
        var check = $('.limit').text()
        instock=instockCheck(check,instock)
        return instock

    },

    price: function($) {
        var price = $('span[itemprop*="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('li a img[itemprop*="image"]').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product_wrapper');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.normal h2 a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span[itemprop*="price"]').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h2 a');
        if(anchor) {
            url = 'http://www.microcenter.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.SearchResultProductImage').attr('src');
        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('li.product_wrapper');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
       var items = $('li.product_wrapper');
        items = items.length;

        var products = $('p.status').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div#contentResults h1:not(span)').text();
        name = cleanName(name);

        if(name.indexOf('Choose a store to see products at your local Micro Center')) {
            name = name.replace("Choose a store to see products at your local Micro Center", "")
        }
        return name.trim();
    }
}
