// sm-supplier-html/suppliers/www.zulily.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-info-name').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        $('span.product-price-box .old-price').remove();
        var price = $('span.product-price-box').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div#product-image a').attr('href');

        return image;
    },
    description: function($) {
        var description = $('div.description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.gutter_norm');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.product-name a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.special-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-name a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('a.product-image img.lazy').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         //
         // there is no element with produc count on page
         //
         var items = $('li.gutter_norm').length;

         return items;
     },
     pages: function($) {
         var items = $('li.gutter_norm').length;

         var products = $('div.results_blurb p strong').first().text();
         products = products.replace(',', '');

         var sum = products / items;

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('div.searchBanner h1').text();
        name = cleanName(name);
        return name;
    }
}
