// sm-supplier-html/suppliers/www.zappos.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript": "true",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        //var brand = $('h1.vUkNo').text();
        var name = $('h1.vUkNo').text();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        price = $('span._3r_Ou._2bQ7p').text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price = $('.mb').text()
            price=getPrice(price)
            return price
        }
        
    },
    image: function($) {
        var image = $('div._ftgE img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('article');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('[itemprop="brand"]').first().text();
        var name = $(product).find('[itemprop="name"]').last().text();
        name = cleanName(name);
        brand = cleanName(brand);
        return brand + ' ' + name;

    },
    price: function($, product) {
        price = $(product).find('[itemprop="url"]').attr('aria-label')
        price=getPrice(price)
        if(price){
            return price
        }else{
            price=$('.lb').text()
            price=getPrice(price)
            return price
        }

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product);
        if(anchor) {
            url = 'https://www.zappos.com' + $(anchor).find('[itemprop="url"]').attr('href');
        }
       return url;
    },
    image: function($, product) {
        var image = $(product).find('[itemprop="image"]').attr('content');

        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('span._3cNVs').text();
        if(!items){
            items=items.match('/\d+/g')
        }
        return parseInt(items);
     },
     pages: function($) {
        var items = $('a.product');
        items = items.length;

        var result = null;
        var products = $('h1#searchHeaderHeader').text();
        
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('h1._1e3rv').text();
        name = cleanName(name);
        return name;
    }
}
