// sm-supplier-html/suppliers/www.dsw.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.add-to-bag__product-details__product-name').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
	return "UNCHECKED"
},
    price: function($) {
        var price = $('div.product__price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.responsive-image').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-tile');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-tile__brand').text();
        var brand=$(product).find('.product-tile__detail-text').text()
        name=name+' '+brand
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var red_price=$(product).find('div.product-tile__detail-price.red').text()
        var price = $(product).find('div.product-tile__detail-price').text();
        if(red_price){
            price=getPrice(red_price)
        }else{
            price=getPrice(price)
        }
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.product-tile__link');
        if(anchor) {
            url = 'https://www.dsw.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-tile__product-image').attr('src');
        return image;
    }   
}

var Category = {
    name: function($) {
        var name = $('h1.result-list__toolbar__left__header').text();
        name = cleanName(name);
        return name;
    }
}
