// sm-supplier-html/suppliers/www.7dayshop.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            if (info.price != "") {
                result.push(info)
            }
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        let name = $('.product-name h1').text()
        name = cleanName(name)
        return name
    },
    inStock: function ($) {
        let availability = $('p.availability').text()
        var isInStock = false

        if (availability) {
            if (availability.indexOf('In stock') > -1) {
                isInStock = true
            }
        }

        return isInStock

    },
    price: function ($) {
        let price = $('p.special-price span.price').text()
        price = getPrice(price)
        return price

    },
    image: function ($) {
        let image = $('img').attr('src')
        return image

    },
    url: function ($) {}
}
var Search = {
    container: function ($) {
        let container = $('.item-area')
        return container
    },
    name: function ($, product) {
        let name = $(product).find('h2.product-name a').attr('title')
        name = cleanName(name)
        return name

    },
    price: function ($, product) {
        let price = $(product).find('li.nth-child-2np1.nth-child-3np1.nth-child-4np1.nth-child-5np1.nth-child-6np1.nth-child-7np1.nth-child-8np1').text()
        price = getPrice(price)
        return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('img').attr('src')
        return image

    }
}

var Pagination = {

    items: function ($) {},
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {}
}
