// sm-supplier-html/suppliers/uk.russellhobbs.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = russellHobbs = {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'sku': (url) => {
        let splitUrl = url.split('/');
        let last = splitUrl[splitUrl.length - 1].replace('.html', '');
        let sku = last.split('-');
        return sku[sku.length - 1];
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        //var prodno = $('div#dsPROTitle h2').text();
        //name = name + ' ' + prodno
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var availability = $('div.specs a').attr('href')

        if (availability) {
            if (availability.indexOf('amazon') > -1) {
                isInStock = false;
            }
        }

        return isInStock;
    },
    price: function ($) {

        let price = $('span.price').text()

        if(price) {
        let pence = price.slice(price.length - 2, price.length)
        let pound = price.slice(0, price.length - 2)
            price = getPrice(pound+'.'+pence)
        }


        return price
    },
    image: function ($) {
        var image = $('img').attr('src');
        return image;
    },
    description: function ($) {
        var description = $('div#tab-features .featureList ul').text().trim();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('a.element');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3').text();
        //var prodno = $(product).find('div.dsPROModel a').text();

        //name = name + ' ' + prodno;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price').text();
        //(price)
        if (price !== null) {
            var dotprice = price.substring(1, 3)
            var mainprice = price.substring(0, 2)
            price = mainprice + '.' + dotprice
        }

        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).attr('href');
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image
    }
}


var Pagination = {

    items: function ($) {


    },
    pages: function ($) {

    }
}

var Category = {
    name: function ($) {
        var name = $('div#dsContent h1').text();
        name = cleanName(name);
        return name.replace(/\w\S*/g, function (txt) {
            return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
        });;
    }
}
