// sm-supplier-html/suppliers/www.beallsflorida.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($("h1.product").text());
    },
    inStock: function($) {
	return 'unchecked'
},
    price: function($) {
        return getPrice($("div.price").text());
    },
    image: function($) {
        return $("div.s7mixedmediaviewer img").attr("src");
    },
    description: function($) {
        return $("div.tab-content p").not(".PDPwebID").text().trim();
    }
}

var Search = {
    container: function($) {
        return $(".match-height-context>li.match-height-item");
    },
    name: function($, product) {
        return cleanName($(product).find(".pr-prod-text").text());
    },
    price: function($, product) {
        return getPrice($(product).find(".price").text());
    },
    url: function($, product) {
        return $(product).find(".pr-prod-text").attr("href");
    },
    image: function($, product) {
        return $(product).find(".img-fluid").attr("src");    
    }
}

var Pagination = {

     items: function($) {
        return parseInt($(".match-height-context>li.match-height-item").length)
     },
     pages: function($) {
        var items = Pagination.items($);
        var itemTot = $("div.view-information").text()
        if(itemTot === ""){
            return null
        }else{
            itemTot = itemTot.replace(",", "").match(/[0-9]+/g)[2]; 
        }
        sum = itemTot / items
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
