// sm-supplier-html/suppliers/www.petco.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
let generateSku = require('../../lib/testPattern.js')
//let regex = /[0-9]+$/

module.exports = bestBuy = {
    "useJavascript": true,

    //  'sku': (url) => {
    //      generateSku(url, regex)[0]
    //      return url
    // },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            // info.sku = generateSku(Search.url($, product), regex)[0]
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);
        info.nexturl = Pagination.nexturl($)

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.heading-5.v-fw-regular').text();
        name = cleanName(name);

        if (name) {
            name = name.replace('Have A Question About This Product?Product Q&A', '').trim()
        }
        return name;

    },
    inStock: function ($) {
        var isInStock = true;
        let availability = $('div.col-xs-4').text()

        if (availability) {
            if (availability.indexOf('Sold Out') > -1) {
                isInStock = false
            }
            if (availability.indexOf('In Store Only') > -1) {
                isInStock = false
            }
            if (availability.indexOf('Store Pickup Only') > -1) {
                isInStock = false
            }

        }

        return isInStock
    },
    price: function ($) {
        var price = $('div.priceView-hero-price:nth-child(1)').text()
        price = getPrice(price)
        return price;

    },
    image: function ($) {
        var image = $('img').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.prod-tile');

        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('a').attr('title');
        //     name = cleanName(name);
        //     name = name.replace('New!', '').trim()
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('.product-price-promo').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return 'https:' + image;

    }
}

var Pagination = {

    items: function ($) {
        return items = $('div.list-items div.list-item').length;

    },
    pages: function ($) {
        var numitems = $('.item-count').text()
        if (numitems === "") {
            return null
        } else {
            numitems = numitems.match(/[0-9]+/g)
        }
        return numitems
    },
    nexturl: function ($) {
        var nexturl = $('.page-item>a').attr('href')
        return nexturl

    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
