// sm-supplier-html/suppliers/www.shopdisney.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = shopDisney = {
    "useJavascript": true,
    "usePhantom": "disneystore.phantom.js",
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()

            //   function formatImage(d) {
            //       if (d != null && d.startsWith("http")) {
            //           d = d.split('"')[15]
            //           return d;
            //       } else {
            //           return d
            //       }
            //   }

            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if (info.price != null) {
                result.push(info)
            }

        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.title h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var isInStock = true;
        var producttop = $('.productDescription').html();
        var availability = $('#main h1').html();
        var addButton = $('.add-container').text()
        isInStock=inStockCheck(addButton,isInStock)
        isInStock=inStockCheck(availability,isInStock)
        isInStock=inStockCheck(producttop,isInStock)
        if(addButton !==null) {
            addButton=addButton.toLowerCase()
            if(addButton.indexOf('sold out') > -1) {
                isInStock = false
            }
        }

        if (availability !== null) {
            availability = availability.toLowerCase();
            if (availability.indexOf('page not found') > -1) {
                isInStock = false;
            }
        }

        if (producttop !== null) {
            if (producttop.indexOf('This product is available in store only') > -1) {
                isInStock = false
            }
        }
        return isInStock


    },
    price: function ($) {
        var price = $('span.price-amount').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.image-wrapper img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.item-wrapper > div.item-container');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('div.title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.price').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.shopdisney.com' + $(anchor).attr('href');
        }
        return url;

    },

    image: function ($, product) {
        var image = $(product).find('div.image-container .bg-image').attr('style')
        if (image) {
            image = image.split('(')[1].replace('"', '').split('"')[0].replace(')', '')
        } else {
            image = null
        }

        return image;

    }

}

var Pagination = {

    items: function ($) {
        return parseInt(items = $('#productTiles .productTile').length)

    },
    pages: function ($) {
        var numitems = $('#productTiles .productTile').length

        var products = $('div.plNumberItems').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.pageTitle').text();
        name = cleanName(name);
        return name;
    }
}
