// sm-supplier-html/suppliers/www.lowes.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = lowes =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.h3').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').attr('content');
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    },
    description: function($) {
        var description = $('ul.disc').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }  
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul div.product-container')
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('p.product-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.js-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'www.lowes.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul div.product-container');
        return parseInt(items.length);

     },
     pages: function($) {
        var pages = $('div.v-spacing-extra-large ul.pagination').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}
var Category = {
    name: function($) {
        var name = $('div.v-spacing-large h1').text();
        name = cleanName(name);
        return name;
    }
}
