// sm-supplier-html/suppliers/electrical.coop.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var routine = require ("../../lib/routine.js"); 
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = cooperElectricShop =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.h2').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var message = $('.text-left-xs').text()
        inStock = instockCheck(message,inStock)
	return inStock;
},
    price: function($) {
        var price = $('div[itemprop = "offers"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img').attr('src');
        image = 'https://electrical.coop.co.uk' + image
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-grid-block');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.no-margin-top').text();
        if(name.indexOf('View product') > -1) {
            name = name.replace("View product", "");
        }
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://electrical.coop.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://electrical.coop.co.uk' + image;
        return image;
    }
}

//var Pagination = {

//     items: function($) {
//        var items = $('span.cs-product-per-page').text();
//        items = items.match(/[0-9]+/g).pop();
//        return parseInt(items);
//     },
//     pages: function($) {
/*        var items = $('span.cs-product-per-page').text();
        numitems = items.match(/[0-9]+/g).pop();*/

//        var numitems = routine.matchNumbersAndPop($, "span.cs-product-per-page");

/*        var products = $('div.page-count').text();
        products = products.match(/[0-9]+/g).pop();*/

//        var products = routine.matchNumbersAndPop($, "div.page-count");

/*        var sum = products / numitems;
        return Math.ceil(sum);*/

//        return routine.pagesFromTotalAndProd(products, numitems);
//     }
//}

var Category = {
    name: function($) {
        var name = $('li.active').text();
        name = cleanName(name);
        return name;
    }
}
