// sm-supplier-html/suppliers/www.rei.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = REI =  {
    "useJavascript": true,
    "usePhantom": "rei.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            
            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.product-current-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.product-primary-img').attr('src');
        return 'https://www.rei.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div#search-results>ol>li');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('article >a >h2>div').first().text();
        var name = $(product).find('article >a >h2>div').last().text();
        name = cleanName(brand+' '+name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find ('article >div >span').text();

        if (price.indexOf('-') > -1) {
            price = null
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('article >a');
        if(anchor) {
            url = 'https://www.rei.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.rei.com' + image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.product-tile');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.product-tile');
        items = items.length;

        var products = $('div.page-heading-core div.text-lead').text();
        var result = null;

        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();    
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('div.page-heading-core h1').text().replace("Results for","").replace('“',"").replace("”","");
        //name =name.match( /"(.*?)"/);
        name = cleanName(name);
        return name;
    }
}
