// sm-supplier-html/suppliers/www.monoprice.com/logic.js
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var ProductPageInfo = require('../../lib/productPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    "useJavascript": true,
    'search': function ($, next) {
        var result = [];

        $('div.row').map(function (index, product) {

            var info = new SearchPageInfo();
            info.name = searchName($, product);
            info.price = searchPrice($, product);
            info.url = searchURL($, product);
            info.image = searchImage($, product);
            info.country = "us"
            result.push(info);
        });

        // var pages = getNumberOfPages($);
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i];
        //     info.pages = pages;
        //     result[i] = info;
        // }


        return result;
    },

    'product': function ($, next) {
        var info = new ProductPageInfo();
        info.inStock = isStock($);
        info.name = getName($);
        info.price = getPrice($);
        info.image = image($);
        info.country = "us";
        return info;
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}



function searchName($, product) {
    var name = $(product).find('div.hawk-row a').first().attr('title');
    name = cleanName(name);
    return name;
}


function searchPrice($, product) {
    let price = $(product).find('.hawk-itemPrice').first().text()
    price = getPrice(price)
    return price

}


function searchURL($, product) {
    var url = $(product).find("div.hawk-row a").first().attr("href");
    return url;
}


// function getNumberOfPages($) {
//     var pages = $('.hawk-paging a').toArray();
//     pages.pop();
//     pages = $(pages.pop()).text();
//      (pages);


//     pages = /[0-9]Pages/.exec(pages)

//     if (pages != null) {
//         pages = pages[0]
//         pages = /[0-9]+/.exec(pages)
//         pages = pages[0]
//     } else {
//         pages = 1
//     }
//     return pages;
// }

function searchImage($, product) {
    return $(product).find('img').attr('src');
}

var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('#ShipDate').html();

    if (producttop !== null) {
        if ((producttop.indexOf('Out of Stock') > -1) ||
            (producttop.indexOf('Out of stock') > -1) ||
            (producttop.indexOf('Unavailable') > -1)) {
            isInStock = false
        }
    }
    return isInStock;
}

var getPrice = function thePrice($) {
    var price = $('.sale-price').text().trim().replace("$", "");
    price = /[0-9]+.[0-9][0-9]/.exec(price)
    if (price !== null) {
        price = price[0]
    }
    return price
}

var getName = function theName($) {
    return $('div.product-name').text().trim();
}

var image = function image($) {
    return $('img.img-responsive').attr('src');
}

var Pagination = {

    items: function ($) {
        var items = $('tr.b2b-product-info');

        if (items.length) {
            items = items.length;
            return parseInt(items);
        } else {
            return null;
        }


    },
    pages: function ($) {

        var pages = $('div.row div.pull-right span').text();
        var result = null;

        if (pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages);

            return result;

        } else {
            return null;
        }


    }
}

var Category = {
    name: function ($) {
        var name = $('div.hawk-top').text();
        name = cleanName(name);
        return name;
    }
}
