// sm-supplier-html/suppliers/www.gardeners.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck=require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            // result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var check= $('.availability-msg').text()
        inStock=inStockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('.price-regular').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.product-image img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-tile');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-name a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product-regular-price').text();
        price = getPrice(price);
        if(price){
            return price
        }else{
            price = $(product).find('.product-sales-price').text()
            price=getPrice(price)
            return price
        }
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.name-link');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('a.thumb-link img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('div#primary h1').text();
        name = cleanName(name);
        return name;
    }
}
