// sm-supplier-html/suppliers/www.saksoff5th.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-overview__short-description').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        //var image = $('div.s7thumb')
        var image =undefined
        //('llllllllllllllllllllllllllllllllllllllllllllll',image)
        // if(!image){
        //     return image
        // }else{
        //     image=image.css('background-image').replace('url("', '').replace('")', '').replace('wid=76', 'wid=450').replace('&hei=90', '&hei=500');
        //     return image;
        // }
        return image
        
    },
    description: function($) {
        var description = $('section.product-description').text()
        if(description !=''){
            description=description.replace(/\s+/g, " ").trim();
            return description;
        }else{
            return description
        }
        
    }
}

var Search = {
    container: function($) {
        var product = $('div.sfa-pa-product-with-swatches');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.product-text > a > p:nth-child(2)').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-sale-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-text > a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.pa-product-large').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('span.pagination-count').first().text();
         //items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('span.pagination-count').first().text();
         var products = $('span.totalRecords').first().text().replace(',', '');
         var pages = products / items;
         pages = Math.ceil(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.pa-enh-page-title span').text();
        name = cleanName(name);

        return name;
    }
}
