// sm-supplier-html/suppliers/www.christmastreeshops.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = christmasTreeShops =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info. country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
            var name = $('h1.mz-pagetitle').html();
            name = cleanName(name);
            return name;
    },
    inStock: function($) {
        var inStock = true
        
	return inStock;
},
    price: function($) {
            var price = $('span.mz-price').html();
            price = getPrice(price);
            return price;
    },
    image: function($) {
            var image = $('img#mainimage').attr('src');
            return image;
    }
}

var Search = {
    container: function($) {
            var searchProducts = $('div.mz-productlisting')
            return searchProducts;
    },
    name: function($, product) {
            var name = $(product).find('a.mz-productlisting-title').text();
            name = cleanName(name);
            return name;
    },
    price: function($, product) {
        var price1=$(product).find('span.mz-price.is-saleprice').text()
        var price = $(product).find('span.mz-price').text();
        if(price1){
            price=getPrice(price1)
        }else{
            price=getPrice(price)
        }
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.christmastreeshops.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
            var image = $(product).find('img').attr('src');
            return 'https:'+image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.ml-paging select').text();
        if(items === ""){
            return null
        }else {
           items = items.match(/[0-9]+/g).pop(); 
        }
        return parseInt(items);

     },
     pages: function($) {
        var pages = $('td.productsShown').text();
        if(pages === ""){
            return null
        }else {
            pages = pages.match(/\d/g).pop();
        }
        return parseInt(pages);

    }
}

var Category = {
    name: function($) {
        var name = $('div.directoryCategoryHeader h1').text();
        name = cleanName(name);
        return name;
    }
}
