// sm-supplier-html/suppliers/www.stage.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('section.prd-squ-details h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('#productAttributes > div.prc-det > h2').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('#s7_mixedmediaview_swatches > div > div > div:nth-child(1) > div').css('background-image').replace('url("', '').replace('")', '').replace('&wid=35&hei=50&', '&wid=500&hei=500&');

        return image;
    },
    description: function ($) {
        var des = $('#toggle-description').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('article.catalog-item');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.prc-detail > p > a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('h2.orign-prc').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.prc-detail > p > a').attr('href');
        url = 'http://www.stage.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.productImg > a > img').attr('src');

        return image;
    }
}

/*var Pagination = {

    items: function ($) {
        var items = ;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = ;
        var products = ;

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}*/

var Category = {
    name: function ($) {
        var name = $('#content > section.catalog-content > h1').text();
        name = cleanName(name)

        return name;
    }
}
