// sm-supplier-html/suppliers/www.campmor.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span[itemprop*="name"] h1').html();
        name = cleanName(name);
        return name;


    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span[itemprop*="price"]').html();
        price = getPrice(price);
        return price;


    },
    image: function($) {
        var image = $('img#productMainImage').attr('src');
        image = 'https:' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.itemthumb');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.itemthumbname h2').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.CAMProductThumbnailPrice').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        image = 'https:' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.itemthumb').length;
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.itemthumb');
        numitems = items.length;

        var products = $('span#PageResults_SummaryText').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();
        }
        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div#PageResults_Wrapper span#PageResults_SummaryText').text();
        name = name.match(/(?:"[^"]*"|^[^"]*$)/)[0].replace(/"/g, "");
        name = cleanName(name);
        return name;
    }
}
