// sm-supplier-html/suppliers/www.gooutdoors.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = goOutdoors =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            if(info.name) {
            result.push(info)
            }
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('#cphMain_pnlTitle').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var availability = $('#delOptions>div').eq(2).html() 
        isInStock = instockCheck(availability,isInStock)
        // var homeDel = $('span#countDelivery').html()
        // if (availability) {
        //     homeDel = homeDel.toLowerCase();
        //     if ((availability.indexOf('Out of Stock') > -1)
        //     || (availability.indexOf('unavailable') > -1)
        //     || (availability.indexOf('Out of stock') > -1)
        //     || (homeDel.indexOf('not available for home delivery') > -1))
        //     {
        //         isInStock = false
        //     }      
        // }
        return isInStock
    },
    price: function($) {
        var price = $('.pricing .price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.jqzoom a').attr('href');
        return 'http://www.gooutdoors.co.uk' + image;
    },
    description: function($) {
        var description = $('div#descTabSect0').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-item-holder');
        return searchProducts;
    },
    name: function($, product) {

        let brand = $(product).find('h2 .brand').text()
        let name = $(product).find('h2 span').not('span.brand').text()

        name = cleanName(brand + ' ' + name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.special').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.gooutdoors.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('h1.pSrchTitle').text();
        name = cleanName(name);
        return name;
    }
}
