// sm-supplier-html/suppliers/www.fatwallet.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.standardPageHeaderSection > div:nth-child(2) > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('span#price').text().trim();

        if (price.includes('from') || price.includes('up') || price == '') {
            price = null;
        } else {
            price = getPrice(price);
        }
        
        return price;
    },
    image: function ($) {
        var image = $('#imageContent > a > img').attr('src');
        image = 'https:' + image;

        return image;
    },
    description: function ($) {
        var des = $('#opinion').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('ul.offer-list-container li');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('a > div > h3').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('a > div > div.offer-price > span.price').text();
        if (price.includes('from') || price.includes('up') || price == '') {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'https://www.fatwallet.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a > div > div.offer-image > img').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

    items: function ($) {
        //
        // there is no elemnt with items count on page
        //
        var items = $('ul.offer-list-container li').length;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        //
        // there is no element on page with all products count
        //
        var pages = $('div.offer-paging > span').last().text().match(/[0-9]+/g).pop();
        pages = JSON.parse(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.standardPageHeaderSection > div:nth-child(5) > h1').text();
        name = cleanName(name)

        return name;
    }
}
