// sm-supplier-html/suppliers/www.notino.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop=name]').text()
        var nam1 = $('#pdHeader>p').text()

        name = cleanName(name+' '+nam1)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var check = $('#pdVariantsTile').next().text()
        inStock= instockCheck(check,inStock)
        return inStock;
    },
    price: function ($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var style = $('#pdImageGallery img').attr('src');
        //var parts = style.match(/background-image: url\("(.+)"\)/)
        //var image = parts[1]
        return style;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('span.name').text();
        var subname = $(product).find('span.subname').text();
        name = cleanName(name + ' ' + subname);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('p.price').text();

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },

    image: function ($, product) {
        var image = $(product).find('span.img-wrap img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.larger-top-margin').text()
        name = cleanName(name);
        return name;
    }
}
