// sm-supplier-html/suppliers/www.mountainsteals.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck=require('../../lib/inStockCheck')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        //var name = $('span#product_name').text().replace(/\s+/g, " ");
        var name = $('span#product_name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {  
        var inStock = true
        var check=$('#whenWillIGetIt strong').eq(1).text()
        inStock=instockCheck(check,inStock)
        return inStock
    },
    price: function($) {
        var price = $('span.price.price-inner span.price-set').text().match(/[0.-9]+/g).shift();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('input[name="pdpInfoMainImage_WithName"]').attr('value');
        image = image.split('|').shift();
        image = 'http:' + image;

        return image;
    },
    description: function($) {
        var description = $('section#product-details').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.prod-item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        $(product).find('span.price.price-inner span.pdp-strike-extra').remove();
        var price = $(product).find('span.price.price-inner').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.prod-item__image img').attr('src');

        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
         //
         // there is no element with product count on page
         //
         var items = $('div.prod-item').length;
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('div.prod-item').length;

         var products = $('span#searchTotalCount').text().trim();

         var sum = products / items;

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('div.catagory-title h1').text();
        name = cleanName(name);
        return name;
    }
}
