// sm-supplier-html/suppliers/www.lightinthebox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = lightInTheBox =  {

    "useJavascript": true,
    "usePhantom": "lightInTheBox.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'


        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('.prod-info-title h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.sale-price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.up a').attr('href');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.pagelet #item-new');

        if(searchProducts.length === 0) {
            searchProducts = $('dl.prod-item');
        }
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.prod-name-title span').text();
        
        if(name === "") {
            name = $(product).find('dd.prod-name a').text();
        }

        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.prod-item-price a').html()
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.prod-name a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.item-block img').attr('src');

        if (image === undefined) {
            image = $(product).find('.item-block img').attr('data-src');
            if(image === undefined) {
                image = $(product).find('div.img-box img').attr('src');
            }
        }
        
        return image;
    }
}
var Category = {
    name: function($) {
        var lastSpaceIndex = cleanName($("li.empty dl dt a").text())
        .lastIndexOf(" ");

        return cleanName($("li.empty dl dt a").text())
        .slice(lastSpaceIndex + 1)
    }
}
var Pagination = {

     items: function($) {
        return Search.container($).length;

     },
     pages: function($) {
        return parseInt(cleanName($("div.pagination ul.clearfix li")
            .text()).match(/[0-9]+/g).pop());

     }
}
