// sm-supplier-html/suppliers/www.everythingkitchens.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var status = true
        var message = $('.stock_availability').text()
        status = instockCheck(message,status)
	return status;
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');

        return image;
    },
    description: function($) {
        var description = $('div.product-details').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.sale-price').text();

        if (price == '') {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function($, product) {
        var url = $(product).find('h2.product-name a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-image a img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.limiter > select option[selected="selected"]').first().text().trim();
         items = parseInt(items);

         return items;
     },
     pages: function($) {
         //
         // there is no element with all products count
         //
         var pages = $('div.pages').first().text().match(/[0-9]+/g);

         if(pages) {
            pages = pages.pop();
            pages = parseInt(pages);

            return pages;
         } else {
            return null;
         }


     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();

        return name;
    }
}
