// sm-supplier-html/suppliers/www.hobbylobby.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var isInStock = true
        let availability = $('div.availability-text').text() 
        isInStock= instockCheck(availability,isInStock)
        // if(availability) {
        //     if(availability.indexOf('Out of stock') > -1) {
        //         isInStock = false
        //     }
        // }

        return isInStock



    },
    price: function ($) {
        var price = $('span.current.sale-price.sale-price-copy').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.gallery-image-content img').attr('src');

        if (!image.startsWith("http")) {
            image = 'https:' + image;
        }
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul .col-4');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('span.title').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = null;
        var salePrice = $(product).find('span.current.sale-copy.price-label').html();
        var standardPrice = $(product).find('span.current-price-copy.price-label').html();
        salePrice = getPrice(salePrice);
        standardPrice = getPrice(standardPrice);

        if (salePrice == null) {
            price = standardPrice;
        } else {
            price = salePrice;
        }
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'http://www.hobbylobby.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('span.padded-img img').attr('src');
            image = image+""

        if (!image.startsWith("http")) {
            image = 'http:' + image;
        }
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul .col-4');
        return parseInt(items.length);
    },
    pages: function ($) {
        var products = $('.number-items-buffer').text();
        if(products){
            products = products.match(/[0-9]+/g).pop();
        }else{
            products=0
        }
        return products;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.heading-info-wrapper h1').text();
        name = cleanName(name);
        return name;
    }
}
