// sm-supplier-html/suppliers/www.fredmeyerjewelers.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = fredmeyerjewelers =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('div#productDetailCol2 div#productShortDescription').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div#productListPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.mainImg').attr('src');
        return 'www.fredmeyerjewelers.com' + image;
    },
    description: function($) {
        var description = $('div#productLongDescription').text().trim();
        
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productItem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.descRating a').text();

        if(name.indexOf('-') > -1) {
            name = name.split("-").shift();
        }
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.descRating a').text();

        if(price.indexOf('-') > -1) {
            price = price.split("-").pop();
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.descRating a').attr('href').replace("..", "");
        return 'www.fredmeyerjewelers.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.imageContainer img').attr('src');
        return 'www.fredmeyerjewelers.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.m-display-options select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.m-display-options select').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.m-pager h2').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum);
     }
}
