// sm-supplier-html/suppliers/www.bonton.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = bonton =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'

            if(info.price !== null) {
                result.push(info)
            }

            

        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div .top div .widget-product-name').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span.btnSalePrice').html();

        if(price === null) {
            price = $('span.price').html();
        }

        if(price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);
        }
        return price;

    },
    image: function($) {
        var image = $('div.grid').text();

        if(image.indexOf('"ItemImage"') > -1) {
            image = image.substring(image.indexOf('"ItemImage"') + 15).substring(0, image.indexOf(','));
        }
        return image + 'arge$';
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.productListings li');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.productDesc a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.salePrice').html();

        if(price === null) {
            price = $(product).find('span.price').html();
        }

        if(price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);
        }
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.bonbon.com' + $(anchor).attr('href');
        }

        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.productImage').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.pagination-row select.refreshSelect').text();
        if(items === ""){
            return null
        }else {
           items = items.match(/[0-9]+/g).pop(); 
        }
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.pagination-row select.refreshSelect').text(); 
        if(items === ""){
            return null
        }else {
            numitems = items.match(/[0-9]+/g).pop();
        }

        var products = $('div#searchMsg').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();
        }

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('span.searchResultsString').text();
        name = cleanName(name);
        return name;
    }
}
