// sm-supplier-html/suppliers/www.tptoys.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = tpToys = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('.product-details-header h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var producttop = $('.stockcontent').html();
        var cartbutton = $('h1').text();


        if ((producttop !== null)) {
            if ((producttop.indexOf('this item is currently unavailable') > -1)
                && (!(producttop.indexOf('hidden') > -1))) {
                isInStock = false
            }
        }

        return isInStock
    },
    price: function ($) {
        var price = $('meta[itemprop="price"]').attr('content');
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('.product-main-image img').attr('src');
        return image;
    },
    description: function ($) {
        var description = $('div.description-wrapper').text().trim();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r\|\t|\t)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.prod');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.title-wrapper h3 a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.blu-price span').text();

        if (price == "") {
            price = null;
        }
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.title-wrapper h3 a').attr('href');
        return 'https://www.tptoys.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img.image1').attr('src');
        return image;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
