// sm-supplier-html/suppliers/www.brookstone.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = brookStone =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
    var isInStock = true;
    var contentarea = $('#pdpMain').text()
    if(contentarea !== null) {
        if((contentarea.indexOf('Out of Stock') > -1)
        || (contentarea.indexOf('Out of stock') > -1)
        || (contentarea.indexOf('Not in stock') > -1)) {
            isInStock = false
        }
    }
    return isInStock
    },
    price: function($) {
        var price = $('.product-price').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.item a.product-image img').attr('src');
        return image


    }
}

var Search = {
    container: function($) {
        var searchProducts = $('#search-result-items .grid-tile');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.thumb-link').attr('title');
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = null;
        var salePrice = $(product).find('span.bold').html();
        var regularPrice = $(product).find('div.product-price').html();
        salePrice = getPrice(salePrice);
        regularPrice = getPrice(regularPrice);

        if (salePrice == null) {
            price = regularPrice
        } else {
            price = salePrice
        }
        return price;

    },
    url: function($, product) {
        var url = $(product).find('.thumb-link').attr('href');
        return url
    },
    image : function($, product) {
        var image = $(product).find('div.product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = Search.container($).length;
        
        if(items) {
            return items;
        } else {
            return null;
        }

     },
     pages: function($) {
        var items = this.items($);
        var pages = $("div.pagination li").text().match(/[0-9]+/g);



        if(items && pages) {
            pages = pages.slice(pages.length / 2);
            
            return parseInt(pages.pop());
        } else if(items && !pages) {
            return 1;
        } else {
            return null;
        }        
    
     }
}

var Category = {
    name: function($) {
        var name = $('h1.category-grid-title').text();
        name = cleanName(name);
        return name;
    }
}
