// sm-supplier-html/suppliers/www.smythstoys.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = smythstoys =  {
    "useJavascript" : true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk";


            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}
var Product = {
    name: function($) {
        var name = $('h1.margn_top_10').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('.stock-status').html();
        var button = $('#main-add-to-basket').hasClass('disabled');
        var notAvailable = $('p.homeDelivery').text()
        var hiddenOutOfStock = $('div.AddToCart-PickUpInStoreAction div.padding-10 label.disabled_text').text()
        let checkPreOrder= $('.deliveryType.homeDelivery.js-stockStatus').text()
        if(checkPreOrder === null || checkPreOrder === 'undefined' || checkPreOrder === ''){
            inStock = false
        }
        // var instoreMessage= $('.instoreMessage').text()
        // if(instoreMessage !== null) {
        //     instoreMessage = instoreMessage.toLowerCase()
        //     if(instoreMessage.indexOf('pre-order') > -1 || instoreMessage.indexOf('not avaliable') > -1) {
        //         inStock = false
        //     }
        // }
        if(stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('available in store only') > -1) {
                inStock = false
            }
        }
        if(button === true) {

            inStock = false;
        }

        if(notAvailable.indexOf('Out Of Stock') > -1 ) {
            inStock = false;
        }

        return inStock;
    },
    price: function($) {
        var price = $('span[itemprop*="price"]').html();
        price = getPrice(price);
        if(price === null) {
            price = $('div.price_tag').html()
            price = price = getPrice(price)
        }
        return price;

    },
    image: function($) {
        var image = $('.item picture img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.item-panel');

        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('span.name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.price_di').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.smythstoys.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }

}
var Pagination = {

     items: function($) {
        var items = $('div.bottom .page-numbers ul.paging li.view a').text();

        if(items !== "") {
            items = items.match(/[0-9]+/g)
            items = items.pop();
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return parseInt(items)

     },
     pages: function($) {

        var items = $('div.bottom .page-numbers ul.paging li.view a').text();

        if(items !== "") {
            items = items.match(/[0-9]+/g)
            items = items.pop();
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        var pages = $('div#product-listing h1.category-name').text();
        var result = null;

        if(pages !== ""){
            pages = pages.match(/[0-9]+/g);
            pages = pages.pop();
            result = Math.ceil(pages / items);
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result
     }
}

var Category = {
    name: function($) {
        var name = $('div.container div.row div.col-md-12 h1').text();
        name = cleanName(name);
        return name;
    }
}
