// sm-supplier-html/suppliers/www.ryman.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $(".product-actions-box>h1").text();
        if(name){
            name = cleanName(name);
        }
        return name;
    },
    inStock: function($) {
        var inStock = $(".add-to-box.add-oos").text()
        var stock = true
        stock = instockCheck(inStock,stock)
        return stock
    },
    price: function($) {
        var price = $("[id^=price-including-tax]").text();
        if(price){
            price = getPrice(price);
        }
        return price;
    },
    image: function($) {
        var image = $('img#image-main').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        return $(".item");
    },
    name: function($, product) {
        var name = $(product).find(".product-name > a").text();
        if(name){
            name = cleanName(name);
        }
        return name;
    },
    price: function($, product) {
        var price = $(product).find("[id^='price-including-tax']").text();
        if(price){
            price = getPrice(price);
        }
        return price;
    },
    url: function($, product) {
        return $(product).find(".product-name > a").attr("href");
    },
    image: function($, product) {
        var image = $(product).find('img.hidden').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.products-grid li.last');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('ul.products-grid li.last');
        items = items.length;

        var products = $('p.amount').text();
        products = products.match(/[0-9]+/g).pop();


        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
