// se-app-model/model/categories.js
const Base = require('../lib/mongo/base');
const mongodb = require('mongodb');

class Categories extends Base {

    constructor(config, collection) {
        //var c = collection.country + 'categories';
        super(config, collection.country + 'categories');
        this.country = collection.country;
    }

    getCategories(next) {
        this.find({}, next);
    }

    findTotals(next) {
        this.find({}, (err, result) => {
            if (err) {
                next(err);
            } else {
                var cats = {};
                result.records.map((c) => {
                    cats[c.amazonId] = c.total;
                });
                next(null, cats);
            }
        });
    }

    generateLookup(next) {
        this.find({}, (err, result) => {
            var end = {};
            result.records.map((record) => {
                end[record.amazonId] = record.name;
            });
            next(err, end);
        });
    }

    saveCategory(payload, next) {
        this.insert(payload, next);
    }

    updateCategory(payload, next) {

        var updateCategory = {
            name: payload.name,
            total: payload.total,
            amazonId: payload.amazonId,
            iconClass: payload.iconClass
        };

        var updatePayload = {
            query: {
                _id: new mongodb.ObjectID(payload.id),
            },
            update: {

                $set: updateCategory

            }
        };

        this.update(updatePayload, (err) => {
            next(err);
        });
    }

    deleteCategory(payload, next) {
        this.delById(payload.id, next);
    }
}

module.exports = Categories;
