// se-app-model/model/searchRestructure.js

module.exports = (record) => {

   var r = {
       id: record._id,
       name: record.name,
       domain: record.domain,
       status: record.status,
       created: record.created,
       totalJobs: '10', //TODO Add total jobs to data
       completedJobs: '4', // TODO Add completed jobs to data
       analysed: '234', // TODO add analysed urs to data
       categories: [ // TODO Add Categories To Data
           {
                categoryName: 'toys and games'
           },
           {
                categoryName: 'outdoors'
           },
           {
               categoryName: 'home and garden'
           }
       ]
   }
   return r
}




/**
  "595262cca99c83585184d412" : {
    id: '595262cca99c83585184d412',
    name: 'Diy',
    supplier: 'www.diy.com',
    status: 'queued',
    created: '2017-06-27T13:51:08.356Z',
    totalJobs: 10,
    completedJobs: 4,
    analysed: 234,
    categories: [
        {
            categoryName: 'toys and games'
        },
        {
            categoryName: 'outdoors'
        },
        {
            categoryName: 'home and garden'
        }
    ]
}, **/
