// sm-supplier-html/suppliers/www.luckybrand.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('#product-content div.product-price span.price-sales').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('a.product--image.main-image.slick-slide.slick-current.slick-active > img').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('div.product--info').text().replace(/(\r\n|\n|\r)/gm, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('li.grid-tile');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('a.name-link').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price-sales').text();
        price = getPrice(price);

        return price
    },
    url: function ($, product) {
        var url = $(product).find('a.name-link').attr('href');

        return 'https://www.luckybrand.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('div.product-image > a picture source').first().attr('srcset')
        if(image != null) {
        image = "" + image.split(', ')[1].split(' ').shift();
        } else {
            image = null;
        }

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('h1.search-result-phrase-title').text().match(/[0-9]+/g).pop();
        items = JSON.parse(items);

        return items;

    },
    pages: function ($) {
        var items = $('h1.search-result-phrase-title').text().match(/[0-9]+/g).pop();
        items = JSON.parse(items);

        var products = $('h1.search-result-phrase-title').text().match(/[0-9]+/g).pop();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items;
        return Math.ceil(sum);

    }
}
var Category = {
    name: function ($) {
        var name = $('div.breadcrumb a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
