// sm-supplier-html/suppliers/www.bedbathandbeyond.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = bedbathandbeyond = {
    "useJavascript": true,
    "usePhantom": "bedbathandbeyond.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("?"))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf("?"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            if (info.name && info.price) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var isInStock = true;
        let availability = $('#atc-cta').text()
        isInStock=instockCheck(availability,isInStock)
        // if (availability !== null) {
        //     availability = availability.toLowerCase();
        //     if (availability.indexOf('usually ships in') > -1) {
        //         isInStock = true
        //     }
        // }
        return isInStock
    },
    price: function ($) {
        var price = $('div[itemprop="offers"] div').first().html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.mainProductImage').attr('src');
        image = 'http:' + image;
        return image;

    }
}

var Search = {
    container: function ($) {
        let container = $('div.grid-container>div>div>article');
        return container;

    },
    name: function ($, product) {
        var name = $(product).find('[data-locator="product_tile_title"]').text()
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('[data-locator="product_tile_price"]').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('[data-locator="product_tile_title"]>a');
        if (anchor) {
            url = 'https://www.bedbathandbeyond.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        // var image = $(product).find('img.productImage').attr('data-lazyloadsrc');
        // var img2 = $(product).find('img.productImage').attr('src');
        // if (image === undefined) {
        //     image = img2
        // }
        // image = 'https:' + image;

        let image = $(product).find('img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.listFilter select.redirPage').text();
        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g).pop();
        }

        return parseInt(items);

    },
    pages: function ($) {
        var items = $('li.listFilter select.redirPage').text();

        if (items === "") {
            return null
        } else {
            numitems = items.match(/[0-9]+/g).pop();
        }


        var products = $('li.listCount span').text();
        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }

        sum = products / numitems
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div#subHeader h1').text();
        name = cleanName(name);
        return name;
    }
}
