// sm-supplier-html/suppliers/www.novatech.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.newspec-title').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var preorder = null;
        //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('div.newspec-specprice .alert-danger').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('this product is only available to buy when in stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = $('.newspec-price').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div#newspec-imageblock img').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.col-xs-12.col-nopadding > .col-xs-12:not([id])');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('p.newspec-price-listing').text();

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.novatech.co.uk' + $(anchor).attr('href');
        }
        return url;
    },

    image: function ($, product) {
        var image = $('.product-box-image').attr('src')
        if (image) {
            return image.trim();
        }
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.search-box-liner h1').text()
        name = cleanName(name);
        return name;
    }
}
