// sm-supplier-html/suppliers/www.carid.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = carid =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.prod_name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        if ($('.status.red').html() == undefined) {
            return 'unchecked';
        } else {
            return false;
        }
    },
    price: function($) {
        var price = $('.prod_price').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('.product-images-shifter-source a').first().attr('href');
        image = 'https://www.carid.com' + image;

        return image;
    },

    description: function($) {7
        var description = $('p[itemprop="description"]').text();

        if (description != null) {
            return description;
        }
    }
}

var Search = {
    container: function($) {
        var product = $('ul.js-main-prod-list li');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('.lst_a_name').text();

        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        if ($(product).find('span[itemprop="price"]').length != 0) {
            var price = $(product).find('span[itemprop="price"]').text();
            price = getPrice(price);

            return price;
        } else {
            return null;
        }
    },
    url: function($, product) {
        var url = $(product).find('.lst_main a').attr('href');
        url = 'https://www.carid.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).attr('data-src');
        image = 'https://www.carid.com' + image;
        
        return image
    },

}

var Pagination = {

     items: function($) {
         var items = $('select[name="per-page"] option:selected').text();
         if(items === ""){
            return null
         }else {
            items = items.split(' ').shift();   
         }
         
         return parseInt(items);
     },
     pages: function($) {
         var items = $('select[name="per-page"] option:selected').text()
         if(items === ""){
            return null
         }else {
            items = items.split(' ').shift();
         }
         

         var products = $('.plab-pagination').text()
         if(products === ""){
            return null
         }else {
            products = products.match(/[0-9,]+/g).pop().replace(',', '');
         }
         
         var sum = products / items;

         //(Math.ceil(pagesCount));

         return Math.ceil(sum); 
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();
        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }

        name = cleanName(name);
        return name;
    }
}
