// sm-supplier-html/suppliers/www.therange.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = theRange = {
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        return "UNCHECKED"

    },
    price: function ($) {
        var price = $('div.price').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.rsTmb').attr('src');
        if (image.indexOf('www') > -1) {
            image = 'https://www.therange.co.uk' + image;
        } else {
            return image;
        }
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchproducts = $('.js_listing');
        return searchproducts;

    },
    name: function ($, product) {
        var name = $(product).find('h4.title').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.price').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a').first();
        if (anchor) {
            url = 'https://www.therange.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {

    },
    pages: function ($) {

    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();

        if (name.indexOf('at The Range') > -1) {
            name = name.replace("at The Range", "");
        }

        name = cleanName(name);
        return name;
    }
}
