// sm-supplier-html/suppliers/www.plowhearth.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            //result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('li.phw_product_details_listright h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return "UNCHECKED"

    },
    price: function($) {
        var price = $('p.phw_product_price').html();
        //var price2 = $('div#prod_price span.p').html();

        // if(price === null) {
        //     price = price2
        // }

        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#phw_product_image_main').attr('src');
        image = 'http:' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        // var price2 = $(product).find('div.price').html();
        var price = $(product).find('div.price').html();

        // if(price === null) {
        //     price = price2
        // }

        if(price.indexOf('-') > -1) {
            price = null
        }


        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.name');
        if(anchor) {
            url = 'http://www.plowhearth.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.image img').attr('src');
        image = 'https://www.plowhearth.com' + image;
        return image;
    }

}
var Pagination = {

     items: function($) {
        var items = $('li.phw_search_result_list_item');
        items = items.length;

        if(items) {
            return parseInt(items);
        } else {
            return null;
        }


     },
     pages: function($) {
     //    var items = $('').html(); 
     //    numitems = items.match(/[0-9]+/g).pop();
/*        var result = null; */

        var pages = $('li.phw_search_pagination_list_item ul li').eq("-2").text();
        pages = pages.match(/[0-9]+/g);

        if(pages) {
            pages = pages.pop();

            return parseInt(pages);
        } else if(Search.container($).length) {
            return 1;
        } else {
            return null;
        }

        
/*        if(pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();    
            result = parseInt(pages);
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result; */
      }
}

var Category = {
    name: function($) {
        var name = $('div#bcrumbs strong').text();
        name = cleanName(name);
        return name;
    }
}
