// sm-supplier-html/suppliers/www.onlineshoes.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = onlineShoes =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"
        
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var brand = $('.brand-name a').html();
        var item = $('.product-name span').html();
        var name = brand + '' + item;
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('#price span.current').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('picture img').attr('srcset');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div#productlist li');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.sale-price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('li a');
        if (anchor) {
            url = 'http://www.onlineshoes.com/' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.product-thumb-image img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.contextTitlelabel h1").text());
    }
}
var Pagination = {

     items: function($) {
        var items = $('div#productlist li');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('ul.ols-pagination li').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);

     }
}
