// sm-supplier-html/suppliers/www.picstop.co.uk/logic.js
var PageInfo = require ('../../lib/pageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = price($)
        info.image = image($)
        info.country = "uk"
        return info
    },
     'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('.availability').html()

    if(producttop !== null) {
        if(producttop.indexOf('Awaiting Stock') > -1) {
            isInStock = false
        }
    }
    return isInStock
}

var price = function thePrice($) {
   return $('#price_selling .selling_price').text().replace("£","").trim()
}

var getName = function theName($) {
    return $('h1').text().trim()
}

var image = function image($) {
    return 'http://www.picstop.co.uk' + $('#image img').attr('src');
}


var Search = {
    container: function($) {
        var searchProducts = $('table.gridp td.pcborder');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('span.prti a b').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.gap font.selling_price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.picstop.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.prim a img').attr('src');
        image = 'http://www.picstop.co.uk' + image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('table.gridp td.pcborder');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('table.gridp td.pcborder');
        items = items.length;

        var products = $('table.alternatet td').text();
        products = products.match(/[0-9]+/g)[2];
        
        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div#breadcrumb span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
