// sm-supplier-html/suppliers/www.buyvpc.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')
module.exports = asda = {
    "useJavascript": true,
    "usePhantom": "asda.phantom.js",

    'sku': (url) => {

        var url = url.substring(url.lastIndexOf("/") + 1)
        return url.substring(0, 9)
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
           // var sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1)
            info.name = Search.name($, product);
            //info.sku = sku.substring(0, 9);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            if (info.price != null) {
                result.push(info)
            }
        });

        return result
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product__title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        let inStock=true
        let availability = $('.pdp-oos-recommendations').not('.hidden').text()
        let check = $('.out-of-stock').text()
        inStock =inStockCheck(check,inStock)
        inStock=inStockCheck(availability,inStock)
        return inStock;
    },
    price: function ($) {
        var price = $('.buying-block .price-badge span.product__price-value').eq(1).text();
        var oldprice =$('.buying-block .price-badge span.product__price-value').eq(0).text();
        price = getPrice(price+'.00');
        oldprice = getPrice(oldprice+'.00')
        if(price){
            return price;
        }else{
            return oldprice
        }
        
    },
    image: function ($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.product-mini');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-mini__name a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.product__price-row').text()
            price = getPrice(price)
            if(!price){
                price = $(product).find('.product__price-row').text()
                price =getPrice(price+'.00')
            }

        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('.product-mini__name a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://direct.asda.com'+url;

    },

    image: function ($, product) {
        var image = $(product).find('picture source').attr('srcset');
        return image;
    }
}
