// sm-supplier-html/suppliers/www.sainsburys.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = sainsburys = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
            
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.productTitleDescriptionContainer h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var inStock = true
        var messageBox = $('div.messageBox').text()
        inStock = instockCheck(messageBox,inStock)
        return inStock;


    },
    price: function ($) {
        var price = $('p.pricePerUnit').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.s7staticimage img').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.greatOffersItem');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('.productName').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('.pricing').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a>img').attr('src');
        image = 'https://www.sainsburys.co.uk' + image;
        return image;

    }
}


var Pagination = {

    items: function ($) {
        var items = $('li.gridItem');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('li.gridItem');
        items = items.length;

        var products = $('h1#resultsHeading').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('ul#breadcrumbNavList li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
