// sm-supplier-html/suppliers/www.designtoscano.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = designToscano =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us' // PLEASE ADD COUNTRY HERE //
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'; // PLEASE ADD COUNTRY HERE //
            //result.push(info)

            if(info.price !== null) {
                result.push(info)  // GET RID OF PRODUCTS IF THEY HAVE "-" IN THE PRICE //
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $("div.detailheader:nth-of-type(1)").html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {

        var salePrice = $('span.pricesale').html();
        var price = $("span#productPricing").html();

        if(salePrice === null) {
            salePrice = price
        }

        salePrice = getPrice(salePrice);
        return salePrice;
    },
    image: function($) {
        var image = $("img#mainimage").attr("src");
        return image;
    },
    description: function($) {
        var desc = $("div.default").html().replace(/<br>.+/mig,"");
        return desc;
    }
}

var Search = {
    container: function($) {
        var products = $("div.directoryCell"); // we keep products with "-" in the name //
        return products;
    },
    name: function($, product) {
        var name = $(product).find(".thumbheader a").html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
       var salePrice = $(product).find('span.pricesale').html();
       var price = $(product).find('span#productPricing').html(); // GET RID OF PRODUCTS IF THEY HAVE "-" IN THE PRICE //

       if(salePrice === null) {
        salePrice = price
       }

       if(salePrice.indexOf("-") > -1) {
            salePrice = null;
       }

       salePrice = getPrice(salePrice);
       return salePrice
    },
    url: function($, product) {
        var url = $(product).find(".thumbheader a").attr("href");
        return "http://www.designtoscano.com"+url;
    },
    image: function($, product) {
        var img = $(product).find("div.thumbdiv img").attr("src");
        return img;
    }
}

var Pagination = {

     items: function($) {
        var items = $("div.directoryCell");
        return items.length             //  HERE WE FIND THE NUMBER OF ITEMS ON THE PAGE, SEE BOSCOVS FOR EXAMPLE //

     },
     pages: function($) {

        var items = $("div.directoryCell");
         items = items.length 

        var products = $('div.searchMessage').text().replace(",","");
        products = products.match(/[0-9]+/g).pop(); // HERE WE FIND THE TOTAL NUMBER OF ITEMS IN THE CATEGORY, THIS IS SHOWN AT THE TOP OF THE PAGE //

        sum = products / items; // WE DIVIDE THE TOTAL NUMBER OF PRODUCTS BY THE ITEMS ON THE PAGE //
        return Math.ceil(sum);
     }
}
