// sm-supplier-html/suppliers/www.hammacher.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = hammacher = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('span.productHeader').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        return isInStock;
    },
    price: function ($) {
        var price = $('span.item_price').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('a.MagicZoomPlus img').attr('src');
        return image;
    },
    description: function ($) {
        var description = $('div[itemprop="description"] p.tab_content').text();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.product_list_sub li')
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a[target="_self"]').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.subcat-prod-price span').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a[target="_self"]').attr('href');
        return 'http://www.hammacher.com' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img.sub_img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul.product_list_sub li');
        return parseInt(items.length);

    },
    pages: function ($) {
        var pages = $('span.pagination_top').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();
        if (name.indexOf('-') > -1) {
            name = name.split('-').shift();
        }
        name = cleanName(name);
        return name;
    }
}
