// sm-supplier-html/suppliers/www.lakeland.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = lakeland =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        let stock =true
        let check = $('.out-of-stock-messaging').text()
        stock = instockCheck(check,stock)
        return stock
},
    price: function($) {
        var price = $('p.price>span[itemprop="price"]').first().text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.product-gallery img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.mob-2.dts-3.product-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h4').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.lakeland.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-image img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("h1.category-lister_heading").text()).split(" (")[0];
    }
}
var Pagination = {

     items: function($) {
        return parseInt(cleanName($("select.quantity").text())
            .split("Ascending")[2].match(/[0-9]+/g)[0]);

     },
     pages: function($) {
        return parseInt(cleanName($("li a.btn")
            .text()).match(/[0-9]+/g).pop());

     }
}
