// sm-supplier-html/suppliers/www.faucetdirect.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = faucetdirect =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var titleManuf = $('span#titleManufacturer').text();
        var titleId = $('span#titleProdId').text();
        var titleName = $('span#titleName').text();

        var name = titleManuf + ' ' + titleId + ' ' + titleName;
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var price = $('.price.itemPrice').text().trim();

        if (price == '') {
            return false;
        } else {
            return true
        }       
    },
    price: function($) {
        var price = $('.price.itemPrice').text().trim();

        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        } 

        return price;          
    },
    image: function($) {
        var image = $('a#prdImageLink').attr('href');
        image = 'https:' + image;

        return image;
    },
    description: function($) {
        var description = $('.productDescriptionOver p').first().text().replace(/\u00A0/g, ' ');

        if (description != null) {
            return description;
        }
    }
}

var Search = {
    container: function($) {
        var product = $('td.product-view');

        return product;
    },
    name: function($, product) {
        var brandName = $(product).find('li.serpItemName a span.brand').text();
        var identifier = $(product).find('li.serpItemName a span.identifier').text();
        var name = brandName + ' ' + identifier;

        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.serpOurPriceAmount').text();
        if (price.includes('-')) {
            price = null;
        } else {
           price = getPrice(price);         
        }   
        
        return price;   
    },
    url: function($, product) {
        var url = $(product).find('li.serpItemName a').attr('href');
        url = 'https://www.faucetdirect.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.displayGallery.photo').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var itemsCountOnPage = parseInt($('p#resultsPerPage').text().split(' | ').pop());

         return itemsCountOnPage;
     },
     pages: function($) {
         var itemsCountAll = $('span#search_results_count').text();
         var itemsCountOnPage = $('p#resultsPerPage').text().split(' | ').pop()

         var pagesCount = itemsCountAll / itemsCountOnPage;
         //(Math.ceil(pagesCount));

         return Math.ceil(pagesCount);
     }
}

var Category = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    }
}
