// sm-supplier-html/suppliers/www.babys-mart.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = babysmart =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.descr_holder h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var availability = $('.despatch_table').html();
        var checkstock = $('#notifymediv').html();

        if (availability) {
            if ((availability.indexOf('available in') > -1)
            || (availability.indexOf('Unavailable') > -1)
            || (availability.indexOf('Out of stock') > -1))
            {
                isInStock = false
            }
        }

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if(checkstock.indexOf('this item is out of stock') > -1) {
                isInStock = false
            }
        }


        return isInStock
    },
    price: function($) {
        var price = $('.detail_prod_price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#zoomimg").attr('src');
        return 'https://www.babys-mart.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productwrapper');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3 a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.grid_PPrice').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h3 a');
        if(anchor) {
            url = 'http://www.babys-mart.co.uk' + $(anchor).attr('href');
        }
        return url;
    },

    image: function($, product) {
        var image = $(product).find('img.cthumb').attr('src');
        return 'http://www.babys-mart.co.uk' + image;
    }
}

var Pagination = {

     items: function($) {


     },
     pages: function($) {
        var pages = $('div.paging_holder').text(); 
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);

     }
}
var Category = {
    name: function($) {
        return cleanName($("h1.productH1holder").text());
    }
}

var Pagination = {

     items: function($) {
        return Search.container.length;

     },
     pages: function($) {
        var pages = $("div.paging a").text().match(/[0-9]+/g)
            .slice("3");
            
            if(pages) {
                return parseInt(pages.pop());
            } else {
                return null;
            }

            


     }
}
