// sm-supplier-html/suppliers/www.zoom.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck =require('../../lib/inStockCheck.js')
module.exports = zoom =  {
    "useJavascript": true,

 //   'sku': (url) => {
 //       return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
 //   },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
        //    info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var stock = true
        var check =$('.stockstatusmessage').text()
        stock = instockCheck(check,stock)
        return stock
},
    price: function($) {
        var price = $('span.selling-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#Default_Product_Image").attr('src');
        return 'http:' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.selling-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.zoom.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.prod');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();
        var result = null;
        var pages = $('div.pages ul li').eq("-2").text();
        if(pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages)    
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }
        return result
     }
}

var Category = {
    name: function($) {
        var name = $('div.category h1').text();
        name = cleanName(name);
        return name;
    }
}
