// sm-supplier-html/suppliers/www.hasbrotoyshop.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = hasbrotoyshop =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text()
        name = cleanName(name);
        return name;
    
    },
    inStock: function($) {
    var isInStock = true;
    var producttop = $('.productPageDivider').find('.header').text()

    if(producttop !== null) {
        if (((producttop.indexOf('Out of Stock') > -1)
        || (producttop.indexOf('out of stock') > -1))) {
            isInStock = false
        }
    }
    return isInStock

    },
    price: function($) {
        var price = $('.productPageDivider span').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#productMainImage").attr('src');
        return 'https://www.hasbrotoyshop.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product_listing_container .product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product_name a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price ').text(); 
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.product_name a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('div.image a img').attr('src');
        return 'http://www.hasbrotoyshop.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('span.num_products');

        if(items.length) {
            return parseInt(items.length);
        } else {
            return null;
        }
     },
     pages: function($) {
        var numitems = $('div.product_listing_container .product').length;

        var products = $('span.num_products').text();
        products = products.match(/[0-9]+/g);

        if(products && numitems) {
            products = products.pop();

            sum = products / numitems
            return Math.ceil(sum)
        } else {
            return null;
        }


     }
}

var Category = {
    name: function($) {
        var name = $('title').text();
        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
