// sm-supplier-html/suppliers/www.toytokyo.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = toytokyo =  {
    "useJavascript": true,
    "usePhantom": "toytokyo.phantom.js",

    'sku': (url) => {
        return url.split('/')[3]
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).split('/')[3];
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.VariationProductPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.cloudzoom').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('article.card');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h4.card-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find ('[data-test-info-type="price"]').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h4.card-title>a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.ProductList li');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('ul.ProductList li');
        items = items.length;

        var result = null;
        var pages = $('ul.PagingList li').eq("-1").text();
        
        if(pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages); 
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;

       
     }
}

var Category = {
    name: function($) {
        var name = $('div#CategoryBreadcrumb li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
