// sm-supplier-html/suppliers/www.groupon.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instochCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('#deal-title').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        var inStock = true
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('div.breakout-option-price').text();

        if (price == '') {
            price = $('div.breakout-option-value').text();
        }

        price = getPrice(price+'.00')

        return price;
    },
    image: function ($) {
        var image = $('#gallery-image-c9').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('#deal-info > section > div > div > article > div:nth-child(1)').text().replace(/\s+/g, " ").trim();

        return des;
    }
}

var Search = {
    container: function ($) {
        $('figure.card-ui.cui-c-deal.cui-c-deal-featured.v1.is-initialized').remove();
        $('figure.card-ui.cui-c-deal.v1.cui-c-deal-featured-fallback.is-initialized').remove();
        var product = $('figure.card-ui');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.cui-udc-title').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.cui-price span.cui-price-discount').text().replace(/\s+/g, " ").trim();

        if (price == '') {
            price = $(product).find('s.cui-price-original').text().replace(/\s+/g, " ").trim();
        }
        price = price.substr(1);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.cui-image').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('#pull-pagination > div > p > span > strong').text().match(/[0-9]+/g).pop();
        items = JSON.parse(items) - 1;

        return items;
    },
    pages: function ($) {
        var items = $('#pull-pagination > div > p > span > strong').text().match(/[0-9]+/g).pop();
        var products = $('#pull-pagination > div > p').text().match(/[0-9]+/g).pop();

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#pull-page-header > div:nth-child(1) > h1 > a').text();
        name = cleanName(name)

        return name;
    }
}
