// sm-supplier-html/suppliers/www.fun.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavaScript": true,
    "usePhantom": "fun.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.description = Product.description($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us";
             //result.push(info)
                result.push(info)
            

        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.prdTop__name').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var outofstock = $('#pnlOutOfStock').html();
        var quantity = $('div#pnlSizes span.ddl__val').text();

        if(quantity !== null) {
            if(quantity.indexOf('Only') > -1) {
                quantity = quantity.match(/[0-9]+/g).pop();
                if(quantity < 5) {
                    isInStock = false;
                }
            }
        }

        if(outofstock !== null) {
            isInStock = false;
        }

        return isInStock;
    },
    price: function($) {
        var price = $('span#lblPurchasePrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#imgZoom').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div#pnlDescription div[itemprop="description"]').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        let container = $("li.ps-item");
        return container;
    },
    name: function($, product) {
        var name = $(product).find('h3.name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find("span.price.cur").html();
        price = getPrice(price);
        return price
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.fun.com' + url;
    },
    image: function($, product) {
        var image = $(product).find("img").attr('src');
        return image;
    }
}


var Pagination = {

     items: function($) {
        var items = $('li.ps-item');
        return items.length
    
     },
     pages: function($) {
        var items = $('li.ps-item');
        numitems = items.length

        var result = null;
        var products = $('span#lblCountTop').text();
            
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / numitems)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;

     }
}

var Category = {
    name: function($) {
        var name = $('span#lblHeaderTitle').text();
        name = cleanName(name);
        return name;
    }
}
