// sm-supplier-html/suppliers/www.staples.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = Staples =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("_")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("_")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) {
            result.push(info)
            }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.skuName').first().text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
    return 'UNCHECKED'
},
    price: function($) {
        var price = $('#SkuPriceUpdate').first().text();
        var anotherPrice = $('.delivery-info__price_section').text()
        anotherPrice = getPrice(anotherPrice)
        price = getPrice(price);
        if(price){
            return price;
        }else{
            return anotherPrice
        }
        
    },
    image: function($) {
        var image = $("div#STP--Product-Image img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.standard-type__large_product_tile');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.standard-type__product_title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.standard-type__price').text();

        if(price !== null) {
            price = price.toLowerCase();
        
        if(price.indexOf('as low as') > -1) {
            price = null
            }
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'www.staples.com' + $(anchor).attr('href');
        }
        return 'https://' + url;

    },

    image: function($, product) {
        var image = $(product).find('.standard-type__product_image').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.stp--product-list div.tile-container');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.stp--product-list div.tile-container');
        items = items.length;

        var result = null;
        var products = $('a#all-products-tab span.count').text();
        
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else{
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        
        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('div.results-title h1').text();
        name = cleanName(name);
        return name;
    }
}
