// sm-supplier-html/suppliers/www.pishposhbaby.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.eyItemTitle').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('#Price0 > div > font').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('a#zoom1').attr('href');

        return image;
    },
    description: function ($) {
        var description = $('div.captionText').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('table.listTable');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.contentsName a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.salePriceColor').text();

        if (price == '') {
            price = null;
        } else {
            price = getPrice(price);
        }

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.contentsName a').attr('href');
        url = 'http://www.pishposhbaby.com/' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.itemImg').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.SortPagingLinksCount').first().text().match(/[0-9]+/g)[1];
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('div.SortPagingLinksCount').first().text().match(/[0-9]+/g)[1];
        var products = $('div.SortPagingLinksCount').first().text().match(/[0-9]+/g).pop();

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.eyTitle').text();

        return name;
    }
}
