// sm-supplier-html/suppliers/www.costco.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')


module.exports = costco =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    }, 

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').first().text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock = true
        var able =$('.price-note-wrapper').text()
        inStock = inStockCheck(able,inStock)
	return inStock;
},
    price: function($) {
        var price = $('span.notranslate').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div img").attr('src');
        return 'https://www.costco.co.uk'+image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-name-container').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-price-amount').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.costco.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.costco.co.uk' + image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("ul.breadcrumbs li").last().text());
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.productList_item').length;

        if(items) {
            return parseInt(items);
        } else {
            return null;
        }

     },
     pages: function($) {
        var numitems = $('div.productList_item').length;

        var products = $('span.x_y_string').text();
        products = products.match(/[0-9]+/g);

        if(products) {
            products = products.shift();
            return Math.ceil(products / numitems)
        } else {
            return null;
        }




     }
}
