// sm-supplier-html/suppliers/www.rakuten.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)


        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var inStock = true;
        var checkstock = $('div.sold-out').text();

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('sold out') > -1) {
                inStock = false;
            }
        }
        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('this product is currently not available.') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('div.main-price span.text-primary').html();
        price = getPrice(price);
        if (price === null) {
            price = 'sold out'
        }
        (price);

        return price;

    },
    image: function ($) {
        var image = $('img.product-img-main').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.r-product__wrapper');
        return searchProducts;


    },
    name: function ($, product) {
        var name = $(product).find('.r-product__name').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        let price = null;
        let dollarPrice = $(product).find('span.r-product__price-text').text();
        //if(dollarPrice) {
        //dollarPrice = dollarPrice.split('<')[1].replace(',','').split('>')[1]
        //let centPrice = $(product).find('span.r-product__price-text small').html();
        price = getPrice(dollarPrice);
        //} else {
        //    price = null
        //}
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Pagination = {

    items: function ($) {
        var items = $('div.FabContainer');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('div.pagination ul li').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
    }
}

var Category = {
    name: function ($) {
        var name = $('div.info_title').text();
        name = cleanName(name);
        return name;
    }
}
