// sm-supplier-html/suppliers/www.buybuybaby.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = buyBuyBaby =  {
    "useJavaScript": "true",
    "usePhantom": "buybuybaby.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name =$('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock=true
        
        return 'UNCHECKED'
    },
    price: function($) {
        var price = $('[itemprop="offers"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image=null
        image=$('a#mainProductImgZoom').attr('href')
        if(image){
            image='https:' + image
        }
        return image
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('article')
        return searchProducts;
    },
    name: function($, product) {

        var name = $(product).find ('.tealium-product-title').text();
        if(name){
            name = cleanName(name)
        }else{
            mame=null
        }
        
        return name;
    },
    price: function($, product) {
        var price = $(product).find ('div[data-locator="product_tile_price"]').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.tealium-product-title a');
        if(anchor) {
            url = 'https://www.buybuybaby.com' + $(anchor).attr('href');
        }
   //     var url = $(product).find('div.prodName a').attr('href')
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.tealium-product-tile img').attr('src')
    
        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('li.listFilter select').text();
        
        if(items !== ""){
            items = items.match(/[0-9]+/g).pop();
        }else {
          if(Search.container($).length > 0) {
                items = 0;
            }  
        }

        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.listFilter select').text();
        
        if(items !== ""){
            items = items.match(/[0-9]+/g).pop();
        }else {
          if(Search.container($).length > 0) {
                items = 1;
            }  
        }

        var products = $('li.listCount ').text();
        var result = null;

        if(products !== ""){
           products = products.match(/[0-9]+/g).pop();
           result = Math.ceil(products / items);
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1.catHeader.isL2Category').text();
        name = cleanName(name);
        return name;
    }
}
