// sm-supplier-html/suppliers/www.potterybarn.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        let name=$('div.pip-summary h1').text()
        name=cleanName(name)
        return name
    },
    inStock: function ($) {
        var inStock = true
        var check=$('.receiving-method-label.radio-box-label').text()
        inStock=instockCheck(check,inStock)
        return inStock
    },
    price: function ($) {
        let price=$('div.pip-summary span.product-price span').last().text()
        
        price=price.trim()
        var checkprice=getPrice(price)
        if(checkprice){
            price=checkprice
        }else{
            price=getPrice(price+'.00')
        }
        return price
    },
    image: function ($) {
        let image=$('div.hero-image.toolbox-for-pip img').attr('src')
        return image
    },
}

var Search = {
    container: function ($) {
        let container = $('li.product-cell')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('a.product-name').text()
            name = cleanName(name)
        return name

    },
    price: function ($, product) {
        let price = $(product).find('.price-amount').first().text()
        let roundedPrice = $(product).find('.price-amount').first().text() + '.00'
        price = getPrice(price)

        if (price == null || price =='') {
            price = getPrice(roundedPrice)
        }
        return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('img.product-thumb').attr('src')
        return image

    }
}

var Pagination = {

    items: function ($) {

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        let name=$('h1.breadcrumb-list-heading').text()
        return name
    }
}
