// sm-supplier-html/suppliers/www.webstaurantstore.com/logic.js
var SearchPageInfo = require('../../lib/searchPageInfo.js')
var ProductPageInfo = require('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck= require('../../lib/inStockCheck.js')

module.exports = {
    //"useJavascript" : true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'search': function ($, next) {
        var result = []

        $('#product_listing').find('div.ag-item').map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = searchName($, product)
            info.sku = url($, product).substring(url($, product).lastIndexOf("/")+1,url($, product).lastIndexOf("."));
            info.price = searchPrice($, product)
            info.url = url($, product)
            info.image = searchImage($, product)
            info.country = "us"
            if(info.price) {
                if(info.price.indexOf("XX") === -1) {
                    result.push(info)
                }
            } 
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    },

    'product': function ($, next) {
        var info = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = Price($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.image)
        //info.barcode = barcode($, info.barcode)
        info.country = "us"
        return info
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    var link = $(product).find('.description').attr('href')
    return "https://www.webstaurantstore.com" + link
}


function searchName($, product) {
    return $(product).find('a.description').text().trim()
}


function searchPrice($, product) {
    var thePrice = $(product).find('p.price').first().text();//.text().replace("From","").replace("$","").replace("/Case","").replace("/Each","").trim()
    //price = price.replace(',', '');
    thePrice=getPrice(thePrice)
    return thePrice;
}


function getNumberOfPages($) {
    return 1
}

function searchImage($, product) {
    let image = $(product).find('.productImage').attr('src')
    // if(image.indexOf('https') !== 0) {
    //     image = 'https:' + image
    // }   
    return image;
}





var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.notifyHeadingText').text()
    isInStock=instockCheck(contentarea,isInStock)
    return isInStock
}

var Price = function thePrice($, instock) {
    let price=$('p.price').first().text()
    price=getPrice(price)
    return price
}

var getName = function theName($, instock) {
    return $('h1').text().trim()

}

var image = function theImage($, instock) {
    return $('#zoom img').attr('src');
}

// var barcode = function theImage($, instock) {
//     var gtin14 = $('span.side-description').html()
//         gtin14 = gtin14.match(/\d+/)[0] 
//     return gtin14
// }


var Pagination = {

    items: function ($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

    },
    pages: function ($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.page-header').text();
        name = cleanName(name);
        return name;
    }
}
