// sm-supplier-html/suppliers/www.zzounds.com/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');


module.exports = {
    "useJavascript" : true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("-")+1,url.lastIndexOf(""))
    },

    'search' : function($, next) {
        var result = []

        $('#leaf-products').find('.product-card').map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.sku = url($,product).substring(url($,product).lastIndexOf("-")+1,url($,product).lastIndexOf(""));
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($, product)
            info.country = "us"
            result.push(info)
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = getPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.image)
        info.country = "us"
        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    var link = $(product).find('.search-image').attr('href')
    return /*"http://www.webstaurantstore.com" +*/ link
}


function searchName($, product) {
    return $(product).find('div.span-21 a.prod-name').html().replace("&quot;",'"');
}


function searchPrice($, product) {
    var thePrice = $(product).find('span.price').text().replace("$","").replace(",","").trim()
    return thePrice
}


function getNumberOfPages($) {
    var pages = $('.pag-control a').toArray()
    if (pages !== null){
        pages = pages.pop()
        pages = $(pages).text()
    } else {
        pages = 1
    }
    return pages
}

function searchImage ($,product) {
    return $(product).find('img.product-img').attr('src');
}


var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.stock-txt').text()
    if(contentarea !== null) {
        if((contentarea.indexOf('Left!') > -1)
        || (contentarea.indexOf('Preorder Today') > -1)
        || (contentarea.indexOf('More Expected Soon!') > -1)) {
            isInStock = false
        }
    }
    return isInStock
}

var getPrice = function thePrice($, instock) {
    return $('.pricing-financing').find('.price').text().replace("$","").trim()
}

var getName = function theName($, instock) {
    return $('#product-title').text().trim()

}

var image = function Image($, instock) {
    return $('img#product-image').attr('src');
}

var Pagination = {

     items: function($) {
        var items = $('div.search-result-product-wrapper');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.search-result-product-wrapper');
        items = items.length;
        
       
        var products = $('div.span-8').text();
        products = products.match(/[0-9]+/g);
        
        (products);
        var result = null;

        if(products !== null) {
            products = products.pop();
            result = Math.ceil(products / items)
        }else{
             if($('#leaf-products').find('.product-card').length > 0) {
                result = 1;
            }
        }

        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1#cat-page-title').text();
        name = cleanName(name);
        return name;
    }
}
