// sm-supplier-html/suppliers/www.graingergames.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = graingergames =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.first').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#packshot img").attr('src');
        return 'http://www.graingergames.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.result');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.first').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.graingergames.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.resultImg img').attr('src');
        return 'www.graingergames.co.uk' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.viewCount select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.viewCount select').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.resultsCount').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div#resultsTop h1').text();
        name = cleanName(name);
        return name;
    }
}
