// sm-supplier-html/suppliers/uk.bodybuilding.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = bodyBuilding =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("=")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
            
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.disco-product__name').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var outofstock = $('.disco-product__discontinued-msg').text();
        isInStock = instockCheck(outofstock,isInStock)
        if(outofstock.indexOf('not available') > -1) {
            isInStock = false;
        }

        return isInStock

    },
    price: function($) {
        var price = $('.SkuGroup__sale-price').first().text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('p.summary').text();
        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product');
        return product;
    },
    name: function($, product) {
        var name = $(product).find('a.product__name').text();
        let brand = $(product).find('.product__brand').text()
        name = cleanName(brand + ' ' + name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product__price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
            url = url.replace('www.bodybuilding.com', 'uk.bodybuilding.com')
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div#CategoryPageProductListing article');
         return parseInt(items.length)
     },
     pages: function($) {
         var numitems = $('div#CategoryPageProductListing article').length;

         var products = $('div.content span.results-field span.count').html();
         products = products.match(/[0-9]+/g).pop();

         var sum = products / numitems;
         return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div#search-results-product h2').text();
        name = cleanName(name);
        return name;
    }
}
