// sm-supplier-html/suppliers/www.superbiiz.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var inStock = true
        var check=$('[itemprop="availability"]').text()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('strong font[itemprop="price"]').text();
        price = getPrice(price);
		
        return price;       
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        image = 'http://www.superbiiz.com' + image;
		
        return image;
    },
    description: function($) {
        return null;
    }
}

var Search = {
    container: function($) {
        var product = $('tr[valign="middle"] td[align="center"]');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a img').attr('alt');
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).next().find('a font').text();
        price = getPrice(price);

        return price
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        url = 'http://www.superbiiz.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image  = 'http://www.superbiiz.com' +  image;

        return image;
    }
}

var Pagination = {

     items: function($) {
        
        var items = $('td[valign="top"] select option[selected="selected"]').attr('value');
        items = JSON.parse(items);
		
        return items;
         
     },
     pages: function($) {
        var items = $('td[valign="top"] select option[selected="selected"]').attr('value');
        items = JSON.parse(items);
        
        var products = $('tbody tr:nth-child(2) td table tbody tr:nth-child(2) td:nth-child(1)').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / items;
        return Math.ceil(sum);
         
     }
}
