// sm-supplier-html/suppliers/www.wholesalehalloweencostumes.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.product-name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var instock = true
        var check =$('.availability').text()
        inStock=instockCheck(check,instock)
        return inStock
},
    price: function($) {
        var price = $('.regular-price>span>sup').first().text();
        price = getPrice(price.trim()+'.00');
        return price;
    },
    image: function($) {
        var image = $('a > div > img').attr('src');

        return image;
    },
    description: function($) {
        var des = $('div.product-description__content-panel').text().replace(/\s+/g, " ").trim();

        return des;
    }
}

var Search = {
    container: function($) {
        var product = $('li.product-grid__cell');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a[itemprop="url"]').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('[itemprop="price"]').text();
        price = getPrice(price);

        if ($(product).find('p.product-prices__price strong').length > 1) {
            price = null;
        }

        return price;
        
    },
    url: function($, product) {
        var url = $(product).find('a[itemprop="url"]').attr('href');
        url = 'https://www.wholesalehalloweencostumes.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img[itemprop="image"]').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         //
         // there is no items count element
         //

         var items = $('li.product-grid__cell').length;
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         //
         // there is no all products count element
         //
         $('ul.pagination li').last().remove()
         var pages = $('ul.pagination li a').last().text();
         pages = JSON.parse(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('h2.secondary-nav__title').text();
        name = cleanName(name);

        return name;
    }
}
