// sm-supplier-html/suppliers/www.oneida.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.price-box').text();
        price = price.match(/[0-9.,]+/gm).pop();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('p.product-image img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.simple-details').text();

        if (description) {
            description = description.replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('ul.products-grid li.item');
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = price.match(/[0-9.,]+/gm).pop();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('h2.product-name a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('a.product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.products-grid li.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('ul.products-grid li.item');
        items = items.length;

        var products = $('div.amount').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}


var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
