// sm-supplier-html/suppliers/www.sportsdirect.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("=")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var brand = $('span[itemprop="brand"]').text();
        var name = $('span[itemprop="name"]').text()
        name = cleanName(brand+' '+name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.pdpPriceRating .pdpPrice').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#imgProduct').attr('src');
        return image;

    }


}

var Search = {
    container: function($) {
        var searchProducts = $('div#productlistcontainer li');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('span.productdescriptionbrand').html();
        var name = $(product).find('span.productdescriptionname').html();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + " " + name;

    },
    price: function($, product) {
        var price = $(product).find('span.curprice').html();
        price = getPrice(price);
        return price;


    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.sportsdirect.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.MainImage').attr('src');
        return image;
    }


}

var Pagination = {

     items: function($) {
     
     },
     pages: function($) {
        var pages = $('div.paginationWrapper .PageNumber').text();   

        var numb = pages.match(/\d/g);
        page = numb.pop();
        page = JSON.parse(page);
        return page;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.s-white span').text();
        name = cleanName(name);
        return name;
    }
}
