// sm-supplier-html/suppliers/www.the-house.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockcheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var stock = true
        var check = $('[itemprop="availability"]').text()
        stock = instockcheck(check,stock)
        return stock
    },
    price: function($) {
        var price = $('#lowest').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.product-image>a>img').attr('src');
        
        return image;
    },
    description: function($) {
        var description = $('div#product-description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.products-grid div.category-product');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('strong.displayname').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.saleprice').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.category-product-title a').attr('href');
        url = 'http://www.the-house.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('.category-product-mainimage img').attr('data-original')
        return 'https:'+image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.items-per-page div button').first().text();
         items = parseInt(items);

         return items;
     },
     pages: function($) {
        var items = $('div.items-per-page div button').first().text();
         
        var result = null;
        var products = $('h5.pull-left strong').text();
         
        if(products !== "") {
            products = products.match(/[0-9,]+/g).shift();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
         
        return result;
     }
}

var Category = {
    name: function($) {
        var name = $(' div.col-xs-12.collapse-group.category-title.editorial > h1 > strong').text();

        return name;
    }   
}
