// sm-supplier-html/suppliers/www.calendars.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = calendars =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.fontsize19.semibold').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('form#productDetailForm').text();
        if (stockMessage) {
            if (stockMessage.toLowerCase().indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('span#prodctprice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#largeProductImageandGallery img:nth-child(1)').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div#tabProductDetails').text();
        if (description) {
            description = description.replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('div.floatLeft.categoryProduct1, div.floatLeft.categoryProduct1Last');
    },
    name: function($, product) {
        var name = $(product).find('a.searchResultText').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.productDetailBottomBox').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.calendars.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return 'http://www.calendars.com' + image;
    }
}

var Pagination = {

     items: function($) {
        return parseInt(items = $('div#productList div#productListLineProduct').length);
     },
     pages: function($) {
        var pages = $('div#pagination').text();
        if(pages === ""){
            return null
        }else {
            if(pages.indexOf('of') > -1) {
            pages = pages.substring(pages.indexOf('of'));
            pages = pages.match(/[0-9]+/g).shift();
            return parseInt(pages)
            }
        }
       
     }
}

var Category = {
    name: function($) {
        var name = $('h1.categoryH1').text();
        name = cleanName(name);
        return name;
    }
}
