// sm-supplier-html/suppliers/video-games.products.half.ebay.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = '';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        // return cleanName($("h1.pdppagetitle").text());
        var name = $("table h1.pdppagetitle").text();
        ("........." + name);

    },
    inStock: function($) {

    },
    price: function($) {

    },
    image: function($) {

    },
    description: function($) {
        
    }
}

var Search = {
    container: function($) {
        return $("table tbody tr td table tbody tr");
    },
    name: function($, product) {
        // return cleanName($(product).find("td div b a").text());
    },
    price: function($, product) {
        return getPrice($(product).find("a.RentalPriceLinks").text());

    },
    url: function($, product) {

    },
    image: function($, product) {  
        return $("a.imageborder img.imageborder").attr("src");
    }
}

var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}
