// sm-supplier-html/suppliers/www.aardstore.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.prod-details h2').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var checkstock = $('li.dispatch').html();

        if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('sorry sold out') > -1) {
                inStock = false
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('span.price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.prod-img a img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('dl.product');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('dt.title a').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('dd.price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.aardstore.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('dt.image img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        return cleanName($("div#col2 h2").text());;
    }
}
