// sm-supplier-html/suppliers/www.physioroom.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = physioRoom =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($("div.producttitle h1").text());
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        return getPrice($("span.price").text());
    },
    image: function($) {
        return "http://www.physioroom.com" + $("img.mainproductimage").attr("src");
    },
    description: function($) {
        return ""
    }
}

var Search = {
    container: function($) {
        return $("div.productrow");
    },
    name: function($, product) {
        return cleanName($(product).find("h4").text())
    },
    price: function($, product) {
        return getPrice($(product).find("div.description p.pricing span.price").text());
    },
    url: function($, product) {
        return $(product).find("a").attr("href");
    },
    image: function($, product) {
        return "http://www.physioroom.com" + $(product).find("img").attr("src");
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }

var Category = {
    name: function($) {
        return cleanName($("div#two-col-left div h1").text());
    }
}
