// sm-supplier-html/suppliers/www.hamleys.com/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var getPrice = require('../../lib/getPrice.js');
module.exports = {
    //"useJavascript" : true,
    'search' : function($, next) {
        var result = []

        $('.productsLayoutModeThumb').find('.stockThumb').map(function(index, product) {

            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($,product)
            info.country = "uk"
            result.push(info)
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = Price($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "uk"
        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}


function url($, product) {
    var link = $(product).find('.productThumbImage').find('a').attr('href')
    return "http://www.hamleys.com/" + link
}


function searchName($, product) {
    return $(product).find('.productThumbName').find('a').text().trim()
}


function searchPrice($, product) {
    var thePrice = $(product).find('.stockthumbpriceNow').text()
    if (thePrice == "" || thePrice == null)
    {
        thePrice = $(product).find('.stockthumbprice').text()
    }
    thePrice = /([0-9]+.[0-9][0-9])|([0-9]+)/.exec(thePrice)
    if (thePrice !== null){
        thePrice = thePrice[0].replace("£","").trim()
    }else{
        thePrice = ""
    }
    return thePrice
}

function searchImage($, product) {
     var image= $(product).find('img.lazy').attr('src');
      image = "https://www.hamleys.com/"+image
      return image
}


// function getNumberOfPages($) {
//     var matching = $('#pagenumber a').toArray()
//     matching.pop()
//     matching = (($(matching[matching.length-1]).text()))
//     matching = /[0-9]+/.exec(matching)
//     return matching[0]
// }





var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.detailInfo').html()
    if(contentarea !== null) {
        if((contentarea.indexOf('Out of Stock') > -1)
        || (contentarea.indexOf('Unavailable') > -1))
        {
            isInStock = false
        }
    }
    return isInStock
}

var Price = function thePrice($, instock) {
    var otherPrice = $('.detailOurPrice').text()
    var price = $('.productprice').text()
    if(otherPrice !=''){
        price = getPrice(otherPrice)
    }else{
        price= getPrice(price)
    }
    //price = /[0-9]+.[0-9][0-9]/.exec(price)
    return price//[0]
}

var getName = function theName($, instock) {
    var name = $('h1').text()
    name = cleanName(name)
    return name
    

}

var image = function image($, instock) {
    var image = $('img.productMain').attr('src');
    return image;
}
var Category = {
    name: function($) {
        return cleanName($("div.detailPageTitle h1").text());
    }
}
var Pagination = {

     items: function($) {
        return Search.container($).length;

     },
     pages: function($) {
        return parseInt(cleanName($("div.detailPageTitle")
            .text()).match(/[0-9]+/g)[0] / this.items($));

     }
}
