// sm-supplier-html/suppliers/www.opentip.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = opentip = {
    "useJavascript": true,

    'sku': function (url) {
        var thisPattern = new RegExp(/^(\d{5,9}(?![\;\?\&\.\,])*$)/g);
        var osku = url.substring(url.lastIndexOf("=") + 1, url.lastIndexOf(""));
        var sku = thisPattern.test(osku) ? osku : null;
        return sku;
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {

        var name = $('div.products_name').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        let availability = $('#stock_quantity').text()
        var isInStock = true
        isInStock = instockCheck(availability,isInStock)
        // if (availability) {
        //     if (availability.indexOf('In Stock') > -1) {
        //         isInStock = true
        //     }
        // }
        return isInStock


    },
    price: function ($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.sp-large img').attr('src');
        return image;
    },
    barcode: function ($) {
        var barcode = $('div#products_upc').text()
        if (barcode) {
            barcode = barcode.match(/[0-9]+/g)[0]
            return barcode
        } else {
            return null
        }

        //      if (barcode) {
        //          if (barcode.indexOf('UPC') > -1) {
        //              barcode = barcode.substring(barcode.indexOf('UPC') + 3).substring(0, 12);
        //              return barcode
        //          } else {
        //              return null
        //          }
        //      }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.block_itemlist div.item-detail');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h5 a').attr('title')
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.products_price').text()

        //    if(price === "") {
        //        price = $(product).find('span[style="font-family: arial,helvetica,sans-serif;font-size: 12px;color: #C71515;font-weight: bold;"]').text();
        //     }
        price = getPrice(price);
        if(price){
        return price;
    }else{
        price =$(product).find('.data>div>span').text()
        price = getPrice(price)
        return price
    }
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div#products_nav_header b:nth-child(2)').text();
        if (items) {
            items = items.match(/[0-9]+/g).pop();
            return parseInt(items);
        }

    },
    pages: function ($) {
        var items = $('div#products_nav_header b:nth-child(2)').text();
        if (items) {
            numitems = items.match(/[0-9]+/g).pop();
        }

        var products = $('div#products_nav_header').text();
        if (products) {
            products = products.match(/[0-9]+/g).pop();
            sum = products / numitems
            return Math.ceil(sum)
        }

    }
}

var Category = {
    name: function ($) {
        var name = $('div.headerNavigation').text();

        if (name.indexOf('»') > -1) {
            name = name.split("»").pop();
        }
        name = cleanName(name);
        return name;
    }
}
