// sm-supplier-html/suppliers/www.giggle.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('div.top-price > div > div > div > div > span').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('#wrap > a > img').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('div.shortDescription').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.productlisting div.product.producttile');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.name > a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.pricing > div > div > div.salesprice').text();

        if (price == '') {
            price = $(product).find('div.pricing > div > div').text();
        }
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.name > a').attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.image > div.thumbnail > p > a > img').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        //
        // there is no element with items count on page
        //
        var items = $('div.productlisting div.product.producttile').length;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('div.productlisting div.product.producttile').length;
        var products = $('#search > div.searchresultsheader > div.pagination > label').first().text().match(/[0-9]+/g).shift();;

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#content > div.producthits > h1').text().trim();
        name = cleanName(name)

        return name;
    }
}
