// sm-supplier-html/suppliers/www.hifigear.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = replace =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }



}

var Product = {
    name: function($) {
        var name = $('div.product-name h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-image img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        return cleanName($("div.category-title h1").text());
    }
}
var Pagination = {
     items: function($) {

     var items = cleanName($("p.amount").text()).split("total")[0]
            .match(/[0-9]+/g);

            if(items) {
                return parseInt(items[1]);
            } else {
                return null;
            }

/*        return parseInt(cleanName($("p.amount").text()).split("total")[0]
            .match(/[0-9]+/g)[1]);*/

     },
     pages: function($) {
        var pages = cleanName($("p.amount").text()).split("total")[0]
            .match(/[0-9]+/g)

            if(pages) {
                return Math.ceil(parseInt(pages[2]) / this.items($));
            } else {
                return null;
            }

     }
}
