// sm-supplier-html/suppliers/www.bigbadtoystore.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = bigBadToyStore = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h3').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var instock =true
        var status = $('div.product-order-section div').first().text();
        instock = instockCheck(status,instock)

        return instock;
    },
    price: function ($) {
        var price = $('div.price').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('img.rsImg.rsMainSlideImage').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('div.col-md-12').text()

        return description;
    }
}

var Search = {
    container: function ($) {
        //$('td:nth-child(2) > table > tbody > tr > td > table').first().remove();
        var product = $('div.panel.panel-search-result.instock-search-result');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('span.product-name').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var priceInt = $(product).find('span.price-integer').html()
        priceInt = priceInt.replace('$', '')
        var priceDec = $(product).find('span.price-decimal').html()

        var price = getPrice(priceInt + '.' + priceDec)

        return price
    },
    url: function ($, product) {
        var url = $(product).find('a').last().attr('href');
        url = 'https://www.bigbadtoystore.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('td > table > tbody > tr > td > table > tbody > tr > td > table > tbody > tr > td:nth-child(2) > font').text().trim();
        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g)[1];
            items = JSON.parse(items);
        }
        return items;
    },
    pages: function ($) {
        var items = $('td > table > tbody > tr > td > table > tbody > tr > td > table > tbody > tr > td:nth-child(2) > font').text().trim();

        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g)[1];
        }


        var products = $('td > table > tbody > tr > td > table > tbody > tr > td > table > tbody > tr > td:nth-child(2) > font').text().trim();
        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }
        var sum = products / items;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('a#rptMenu_ctl00_lnkMenuName').text();

        if (name.indexOf(':') > -1) {
            name = name.split(':').shift()
        }

        name = cleanName(name);
        return name;
    }
}
