// sm-supplier-html/suppliers/www.elfcosmetics.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = elfcosmetics =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        return isInStock;
    },
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div[itemprop="description"]').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }   
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.col-xs-12.col-md-4.product-listing-holder')
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-img-description h4').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        var saleprice = $(product).find('span.price .new-price').text();

        if(saleprice) {
            price = saleprice
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-img-description a').attr('href');
        return 'http://www.elfcosmetics.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.category-image-wrap img').attr('src')
        return image;
    }
}

var Pagination = {

     items: function($) {
        var pages = $('div.pagination').text();
        pages = pages.match(/[0-9]+/g)

        var items = $('div.col-sm-6');

        if(items.length && pages) {
          return parseInt(items.length);  
      } else {
        return null;
      }


     },
     pages: function($) {
        var pages = $('div.pagination').text();
        pages = pages.match(/[0-9]+/g)

        if(pages) {
            pages = pages.pop();

            return parseInt(pages);
        } else {

            return null;
        }



     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb-container li.active').text();
        name = cleanName(name);
        return name.replace(/\w\S*/g, function(txt){return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();});;
    }
}
