// sm-supplier-html/suppliers/www.tjhughes.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var isntockCheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            if(info.price != "") {
            result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($("h1.product-content__title").text());
    },
    inStock: function($) {
        var availability = $('span#js-product-stock-message-out-of-stock').not('[style="display:none"]').text()
        var isInStock = true
        isInStock = isntockCheck(availability,isInStock)
        // if (availability) {
        //     if (availability.indexOf('Item in Stock') > -1) {
        //         isInStock = true
        //     }
        // }
	return isInStock;
},
    price: function($) {
        return getPrice($("span.GBP").text());
    },
    image: function($) {
        return $("img.cloudzoom").attr("src");
    },
    description: function($) {
        return $("div.product-tabs__content__cms p").first().text();
    }
}

var Search = {
    container: function($) {
        return $("li.col").not(".infinitescroll-load-more");
    },
    name: function($, product) {
        return cleanName($(product).find("div.product__details__title a").text()).replace("\n", " ");
    },
    price: function($, product) {
        return getPrice($(product).find("span.product-content__price--inc").text());
    },
    url: function($, product) {
        return "https://www.tjhughes.co.uk" + $(product).find("div.product__details__title a").attr("href");
    },
    image: function($, product) {
        return $(product).find("div.product__image a img").attr("src");
    }
}

var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}

var Category = {
    name: function($) {
        var name = $('div.product-listings-description h1#js-category-h1').text();
        name = cleanName(name);
        return name;
    }
}
