// sm-supplier-html/suppliers/www.adorama.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": true,
    "usePhantom": "adorama.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)

        if (info.inStock == false) {
            info.price = null;
        } else {
            info.price = Product.price($)
        }


        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $("[itemprop='name'] > span").text();
        var name2 = $('div.item-details h1 a').html();

        if (name === "") {
            name = name2;
        }

        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var inStock = true;
        var checkstock = $('div.item-details').html();


        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();

            if (checkstock.indexOf('we are sorry, this item is no longer available.') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('strong.your-price').html().replace("$", "");
        //price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.productImage').attr('src');
        if (image.indexOf('https:') > -1) {
            return image;
        } else {
            image = 'http:' + image;

        }

        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div div .item').not('.hide').not('.next-page');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('h2 a').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('strong.your-price').text()
            price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img.productImage').attr('src');
        image = 'https://www.adorama.com' + image
        return image;

    }
}

var Pagination = {

    items: function ($) {
        var items = $('span.index-count b').text();

        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g).pop();
        }
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('span.index-count-showing ').text();

        if (items === "") {
            return null
        } else {
            numitems = items.match(/[0-9]+/g).pop();
        }


        var products = $('span.index-count-total').text();

        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }

        sum = products / numitems
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('header.list-title h1').text();
        name = cleanName(name);
        return name;
    }
}
