// sm-supplier-html/suppliers/www.puritan.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = puritan =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.container-header h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.container-outofstock');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock:') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('p.offer-price').first().text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.container-product-img img').attr('src');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('div#tab div#tab-1').text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('ul.cat-item-grid li:not([class])');
    },
    name: function($, product) {
        var name = $(product).find('div.cat-item-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.cat-item-price').first().text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.cat-item-name a').attr('href');
        return 'http://www.puritan.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');
        return 'http:' + image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }
var Category = {
    name: function($) {
        var name = $('div#mainContent ul#breadcrumbs li').eq("-2").text();
        name = cleanName(name);
        return name;
    }
}
