// sm-supplier-html/suppliers/www.gear4music.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js')
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replace = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var check = $('.tooltip-source').first().text()
        var instock = true
        instock = instockCheck(check,instock)
        return instock;
    },
    price: function ($) {
        var price = $('span.c-val').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("img.main-image").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.result-list li');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.gear4music.com' + url;

    },
    image: function ($, product) {
        var image = $(product).find('span.plt-rw-img img').attr('src');
        return image;
    }
}

var Category = {
    name: function ($) {
        var name = $('h2.section-title').text();
        name = cleanName(name);
        return name;
    }
}
