// sm-supplier-html/suppliers/www.zavvi.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = zavvi =  {
    "useJavascript": "true",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".html"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".html"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.productName_title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
         var isInStock = true;
         var producttop = $('.availability').html();
         var preorder = $('.productbuy').html();
        var check =$('.productStockInformation_prefix').text()
        isInStock = instockCheck(check,isInStock)
         if(producttop !== null) {
             if(producttop.indexOf('Out of Stock') > -1) {
                 isInStock = false;
             }
         }

         if(preorder !== null) {
             preorder = preorder.toLowerCase();
             if (preorder.indexOf('pre-order now') > -1) {
                 isInStock = false;
             }
         }
         return isInStock;
    },
    price: function($) {
        var price = $('p.productPrice_price').text();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('.productImageCarousel_image').attr('src');
        return 'https:'+image;
    }
}

var Search = {
    container: function($) {
        var searchproducts = $('div.productBlock');
        return searchproducts;

    },
    name: function($, product) {
        var name = $(product).find('h2.productBlock_productName').html().replace("<br>","");
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.productBlock_priceValue').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
    var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.zavvi.com'+url;

    },
    image: function($, product) {
        var image = $(product).find('img.productBlock_image').attr('src');
        if(image.indexOf('https://') === -1 ) {
            return "http:" + image;
        } else {
            return image
        }

    }
}

var Pagination = {

     items: function($) {
        var items = $('div.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var result = null;
        var pages = $('div.pagination_pageNumbers').text();
        if(pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages)
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
    }
}

var Category = {
    name: function($) {
        var name = $('div.listTitle h1.listTitle_text').text();
        name = cleanName(name);
        return name;
    }
}
