// sm-supplier-html/suppliers/www.pet360.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = pet360 =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = ''
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = '';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $("h1").html();
        name =cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $(".price-box-container span.price").text().replace(/[^\d.]/mig,"");
        return price;
    },
    image: function($) {
        var image = $("p.image img").attr("src");
        return image;
    },
    description: function($) {
        var desc = '';
        return desc;
    }
}

var Search = {
    container: function($) {
        var products = $("span.equalizer-inner:not(:has(span:contains( - )))");
        return products;
    },
    name: function($, product) {
        var name = $(product).find("h2 a").html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find("span.price").text().replace(/[^\d.]/mig,"");
        return price;
    },
    url: function($, product) {
        var url = $(product).find("h2 a").attr("href");
        return url;
    },
    image: function($, product) {
        var img = $(product).find("a.image img").attr("src");
        return img;
    }
}
/**
var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}
**/

var Category = {
    name: function($) {
        var name = $('div#breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
