// sm-supplier-html/suppliers/www.pamperedchef.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        // info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
       'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($("h1.m-title-none").text());
    },
    inStock: function($) {
        var inStock = true
    	return inStock;
    },
    price: function($) {
        var price = null
        var regularPrice = $('span.our_price_display').text()
        var oldPrice = $('span.old-price').text()
        var newPrice = $('span.new-price').text()
        regularPrice = getPrice(regularPrice)
        oldPrice = getPrice(oldPrice)
        newPrice = getPrice(newPrice)

        if (newPrice) {
            price = newPrice
        } else {
            price = oldPrice
        }
        if (!price) {
            price = regularPrice
        }

        return price
    },
    image: function($) {
        var image = $("a.fancybox-imageOnly img").attr("src");
        image = 'https://www.pamperedchef.com' + image
        return image
    },
    description: function($) {
        return $("div.desc_blk_inn p").text().trim();
    }
}

var Search = {
    container: function($) {
        return $(".item-collection>div");
    },
    name: function($, product) {
        return cleanName($(product).find("h4").text());
    },
    price: function($, product) {
        // return getPrice($(product).find("p.collection-product-price").text());
        var price = null
        var regularPrice = $(product).find('p.collection-product-price').text()
        var newPrice = $(product).find('p.collection-product-price.now-price').text()
        regularPrice = getPrice(regularPrice)
        newPrice = getPrice(newPrice)

        if (newPrice) {
            price = newPrice
        } else {
            price = regularPrice
        }

        return price
    },
    url: function($, product) {
        var url = $(product).find("h4>a").attr("href");
        url = 'https://www.pamperedchef.com' + url
        return url
    },
    image: function($, product) {
        var image = $(product).find("div.collection-item-image a img").attr("src");
        image = 'https://www.pamperedchef.com' + image
        return image
    }
}

var Pagination = {

     items: function($) {
        return parseInt($("select.recordsPerPageOptionsSelector1 option").text().match(/[0-9]+/g)[4]);
     },
     pages: function($) {
        items = Pagination.items($);
        pages = $("div.tag span.tag-name").text().match(/[0-9]+/g)[0];

        return (Math.ceil(pages / items));
     }
}
var Category = {
    name: function($) {
        var name = $('h1.category-name').text();
        name = cleanName(name);
        return name;
    }
}
