// sm-supplier-html/suppliers/www.fragrancedirect.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = fraganceDirect =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('.prod-detail-content').html();
        var availability = $('#content h1').html();
        var able = $('.availability-msg').text()
        isInStock = inStockCheck(able, isInStock)

        if(producttop !== null) {
            if ((producttop.indexOf('Out of Stock') > -1) 
            || (producttop.indexOf('Not in Stock') > -1)
            || (producttop.indexOf('Unavailable') > -1)) {  
                isInStock = false
            }
        } 

        if(availability !== null) {
            availability = availability.toLowerCase();
            if(availability.indexOf('product unavailable') > -1) {
                isInStock = false
            }
        }

        return isInStock   
    },
    price: function($) {
        var price = $('span.price-sales').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#productdetail-image img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('p[itemprop="description"]').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.grid-tile');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.product-pricing').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
            url = 'http://www.fragrancedirect.co.uk' + url;
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('srcset');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select[name="perpage"]').text();
        items = items.match(/[0-9]+/g)

        if(items) {

            items = items.pop();

            return parseInt(items);

        } else {
            return null;
        }


     },
     pages: function($) {
        var pages = $('span.pagnNumbers').text();
        pages = pages.match(/[0-9]+/g)

        if(pages) {
            pages = pages.pop();

            return parseInt(pages);
        } else {
            return null;
        }


     }
}

var Category = {
    name: function($) {
        var name = $('h1.hide-for-small').text();
        name = cleanName(name);
        return name;
    }
}
