// sm-supplier-html/suppliers/www.repairclinic.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.productDetailWrap > div.nameNumberLogoWrapper > div.nameNumberWrapper > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('div.productDetailWrap > div.helpColumn > h3').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('a.viewLargerImg.slick-slide.slick-current.slick-active > img').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('p.productDescription').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.productWrap');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.productImage > h3 > a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.productPriceWrap > h3').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.productImage > h3 > a').attr('href');
        url = 'https://www.repairclinic.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.imgSkillWrap > a > img').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        //
        // there is no elemnt with items count on page
        //
        var items = $('div.productWrap').length;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        //
        // there is no element on page with all products count
        //
        $('div.paginationControls a.nextArrow').remove()
        var pages = $('div.paginationControls a').last().text();
        pages = JSON.parse(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#siteContent > div.siteLayoutFrame > div.contentTop > h1').text();
        name = cleanName(name)

        return name;
    }
}
