// sm-supplier-html/suppliers/www.mandmdirect.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

}

var Product = {
    name: function($) {
        var name = $('h1.st-product-title').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var avaibility=$('div.carousel-inner').text()
        var instock=false
        if(avaibility.indexOf('In stock')>-1){
            instock=true
        }
	return instock;
},
    price: function($) {
        var price = $('h2.st-selling-price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.slider-for__item img').attr('src');
        return image; 
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.span3.item.plitem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h4').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.itemimage a');
        if(anchor) {
            url = 'http://www.mandmdirect.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        var name = $('div#breadcrumbContainer p.mbreadcrumb span').eq("-1").text().replace("\>\n","");
        name = cleanName(name);
        return name;
    }
}

var Pagination = {

     items: function($) {
        var items = $('.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.item');
        items = items.length;

        var products = $('span#xProductsLabel').text();
        products = products.match(/[0-9]+/g).pop();


        sum = products / items
        return Math.ceil(sum)
     }
}
