// sm-supplier-html/suppliers/www.solidsignal.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = solidsignal =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').attr('content');
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img[itemprop='image']").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('span[itemprop="sku"]').text();
        return barcode.trim();
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div[style="position:relative; width:200px; font-size:12px;  display: inline-block; border: 1px solid #CFCFCF; height: 320px; margin: 10px; text-align: center; cursor: pointer;  padding: 4px; vertical-align: top;"]');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div[style="height: 65px; border: 1px solid white; display: inline-block;overflow:hidden; width: 100%; cursor: pointer; font-weight: bold; font-size: 12px;"]').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('b[style="font-size: 19px; color:navy;  text-decoration: none;"]').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'www.solidsignal.com/' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div[style="position:relative; width:200px; font-size:12px;  display: inline-block; border: 1px solid #CFCFCF; height: 320px; margin: 10px; text-align: center; cursor: pointer;  padding: 4px; vertical-align: top;"]');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div[style="position:relative; width:200px; font-size:12px;  display: inline-block; border: 1px solid #CFCFCF; height: 320px; margin: 10px; text-align: center; cursor: pointer;  padding: 4px; vertical-align: top;"]');
        items = items.length;

        var result = null;
        var products = $('span[style="font-size: .66em;"]').text();
        
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
