// sm-supplier-html/suppliers/search.bodybuilding.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('h1.fn').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('span.price').first().text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('a.bb-image-viewer').attr('href');

        return image;
    },
    description: function($) {
        //
        // there is no structure for description
        //
        return null;
    }
}

var Search = {
    container: function($) {
        var product = $('div.bb-prod-listing__col');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a.bb-prod-listing__col-header').text().split(',').shift();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.bb-prod-listing__price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.bb-prod-listing__col-header').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img[itemprop="image"]').attr('src');
        image = 'http:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         //
         // there is no element with search items count on page
         //
         var items = $('div.bb-prod-listing__col').length;

         return items;
     },
     pages: function($) {
         var items = $('div.bb-prod-listing__col').length;

         var products = $('div.search-details__detail-label').text();
         products = products.match(/[0-9]+/g).shift();

        var sum = products / items;
        return Math.ceil(sum)
     }
}
