// sm-supplier-html/suppliers/www.beautyflash.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.split('/')[3]
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($);
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).split('/')[3];
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockMessage = $('span.stock-spacer').text();
        if (stockMessage && stockMessage.toLowerCase().indexOf('out of stock') != -1) {
            inStock = false;
        }
        return inStock;
    },
    price: function($) {
        var price = $('div.prod-price span[itemprop=price]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#main_image img').attr('src');
        return 'https://www.beautyflash.co.uk' + image;
    },
    description: function($) {
        var description = $('span[itemprop=description]:nth-child(1)').text();
        if (description) {
            description = description.replace(/(\r|\n|\r\n)/gm, ' ').replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('div.brand-prod-box');
    },
    name: function($, product) {
        var name = $(product).find('h5.prod-box-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product-box-price div').text()
        if(price !== ''){
            price= getPrice(price.text())
            return price
        }else{
            price = $(product).find('.product-box-price').text()
            price = getPrice(price)
            return price
        }
    },
    url: function($, product) {
        var url = $(product).find('a:nth-child(1)').attr('href');
        return 'https://www.beautyflash.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.beautyflash.co.uk' + image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }

var Category = {
    name: function($) {
        var name = $('ol.breadcrumb li.active').text();
        name = cleanName(name);
        return name;
    }
}
