// sm-supplier-html/suppliers/www.lambda-tek.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replace = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.ProductInfo h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('span.PriceSmall').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img[alt="Product Image"]').attr('src');
        var item = $('td#prodTabOverview a').attr('href');

        // item = item.split("/")
        // item.pop()
        // item.shift()
        // item.shift()

        // url = item.shift();


       // return url + '/' + item + '/' + image;
        return image;
    },
    barcode: function ($) {
        var barcode = $('div#ProductDetails').html();
        //return barcode.match(/[0-9]+/g).pop();
        return barcode
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.productlist');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.col-sm-7 h3').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.col-sm-5 h5:nth-child(2) b:nth-child(4)').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a[itemprop="url"]');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img.product-image').attr('src');
        return 'https://www.lambda-tek.com/shop/' + image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("td#MainContent div h1").text());
    }
}
