// sm-supplier-html/suppliers/www.xtra-vision.co.uk/logic.js
var PageInfo = require ('../../lib/pageInfo.js')
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');


module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = getPrice($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
} 

var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('.buyDetails').html()

    if(producttop !== null) {
        if(producttop.indexOf('Awaiting Stock') > -1) {
            isInStock = false
        }
    }

    return isInStock
}

var getPrice = function thePrice($) {
    return $('.buyDetails').find('.price').text().replace("£","")  
}

var getName = function theName($) {
   var name = $('h1').html();
   name = cleanName(name);
   return name;
}

function image($) {
   var image = $('.image img').attr('src');
   image = 'http:' + image;
   return image;   
}


var Search = {
    container: function($) {
        var searchProducts = $('section.grid > ul');
        searchProducts = $(searchProducts).children('li').not(".advert");
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('section.info h3').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('section.pricing span.price').text().trim().replace("£","");
        //price = getPrice(price);//** BREAKS IT FOR SOME REASON **//
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.xtra-vision.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('section.image a img').attr('src');
        image = 'http:' + image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('section.grid > ul');
        items = $(items).children('li').not(".advert");
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('section.pagination ul li').eq("-1").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div#products h1').text();
        name = cleanName(name);
        return name;
    }
}
