// sm-supplier-html/suppliers/www.famousfootwear.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var brand = $('div.style-brandLogo a img').attr('alt');
        var name = $('h1.PD_BrandStyle span[itemprop*="name"]').html();

        brand = cleanName(brand);
        name = cleanName(name);

        return brand + " " + name; 


    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span[itemprop*="price"]').html();


        price = getPrice(price);

        return price

    },
    image: function($) {
        var image = $('div.prodImg img').attr('src');
        image = 'http://www.famousfootwear.com' + image;
        return image;

    },
    description: function($) {
        var description = $('div#ProductDescription span[itemprop*="description"]').text().trim();
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
        
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productCell');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('span.brand').text();
        var name = $(product).find('span.name').text();

        
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + " " + name;

    },
    price: function($, product) {
        var salePrice = $(product).find('span.salePrice').html();
        var price = $(product).find('span.variantPrice').html();

        if(salePrice === null) {
            salePrice = price;
        }

        salePrice = getPrice(salePrice);

        return salePrice
    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.famousfootwear.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.productImage a img').attr('src');
        image = 'http://www.famousfootwear.com' + image;
        return image;
        
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.view-links').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
        

     },
     pages: function($) {

        var items = $('div.view-links').text();
        items = items.match(/[0-9]+/g).pop();

        var products = $('div.searchRecordCount').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / items;
        return Math.ceil(sum);

     }
}

var Category = {
    name: function($) {
        var name = $('div.breadCrumbs h2 span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
