// sm-supplier-html/suppliers/us.sportsdirect.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = sportsdirect =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("-")+1,url.lastIndexOf("?"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("-")+1,Search.url($, product).lastIndexOf("?"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
        var brand =$('span[itemprop="brand"]').text()
        name = cleanName(brand+' '+name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('div.addToBasketContainer a').html();
        inStock=instockCheck(stockcheck,inStock)
        if(stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('pre-order now') > -1) {
                inStock = false
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('div.pdpPrice').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#imgProduct').attr('src');
        return image;

    },
    description: function($) {
        var description = $('span[itemprop="description"]').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }

}

var Search = {
    container: function($) {
        var searchProducts = $('ul.s-productscontainer2 li');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('span.productdescriptionbrand').html();
        var name = $(product).find('span.productdescriptionname').html();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + " " + name;

    },
    price: function($, product) {
        var price = $(product).find('span.curprice').html();
        price = getPrice(price);
        return price;


    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://us.sportsdirect.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.MainImage').attr('src');
        return image;
    }


}

var Pagination = {

     items: function($) {
        var items = $('ul.s-productscontainer2 li');
        return parseInt(items.length);
     },
     pages: function($) {
        var items = $('ul.s-productscontainer2 li');
        numitems = items.length;

        var result = null;
        var products = $('span.pppt').text();
           
        if(products !== ""){
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / numitems);    
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }    

        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1.s-white').text();
        name = cleanName(name);
        return name;
    }
}
