// sm-supplier-html/suppliers/www.sunandski.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product__name').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockMessage = $('span#product-detail p.text-center').text();
        var check = $('.btn-group').text()
        inStock = instockCheck(check,inStock)

        if (stockMessage) {
            if (stockMessage.toLowerCase().indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('span.product__price.product__price--lrg').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.product__image').attr('src');
        return 'https:' + image;
    },
    description: function($) {
        var description = $('article.product-details__item p').text();
        return description;
    }
}

var Search = {
    container: function($) {
        return $('div.products__wrapper li.product-list__item');
    },
    name: function($, product) {
        var name = $(product).find('span.product__name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product__price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.sunandski.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('span.product__image img').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('select[data-mz-value]').text();
        products = products.match(/\d+/gm);

        if(products) {
            products = products.pop()

            return parseInt(products);
        } else {
            return null;
        }

     },
     pages: function($) {
        var products = $('select[data-mz-value]').text();
        products = products.match(/\d+/gm);

        var total = $('div.products__header span:contains("Results")').text();
        total = total.match(/\d+/gm);

        if(products && total) {

            products = products.pop();
            total = total.pop();


            var sum = total / products;
            return Math.ceil(sum);

        } else {
            return null;
        }

     }
}

var Category = {
    name: function($) {
        return cleanName($("h1.page__title").text());
    }
}
