// sm-supplier-html/suppliers/www.thatpetplace.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },
/*
    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },
*/
    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            //info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-details-full-content-header-title').text();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var avilibility=$('p.product-line-stock-msg-in').text()
        var instock=false
        if(avilibility.indexOf('In Stock')>-1){
            instock=true
        }
	return instock;

    },
    price: function ($) {
        var price = $('span.product-views-price-lead').text();
        price=getPrice(price)
        return price

    },
    image: function ($) {
        var image = $('div.item-details-image-gallery-detailed-image img').attr('src');
        if(!image){
            image=$('div.bx-viewport img').attr('src')
        }
        return image;

    },
}

var Search = {
    container: function ($) {
        var product = $('div.facets-items-collection-view-cell-span3');
        return product;

    },
    name: function ($, product) {
        var brand = $(product).find('a.facets-item-cell-grid-title span').text();
        brand = cleanName(brand);
        return brand;

    },
    price: function ($, product) {
        var price = $(product).find('div.product-views-price').text();
        price=getPrice(price)
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a.facets-item-cell-grid-title').attr('href');
        return 'https://www.thatpetplace.com'+url;
    },

    image: function ($, product) {
        var image = $(product).find('img.facets-item-cell-grid-image').attr('src');
        return image;

    }
}

var Pagination = {

    items: function ($) {
        var items=$('h1.facets-facet-browse-title').attr('data-quantity')
        return items;
    },
    pages: function ($) {
        var products = $('h1.facets-facet-browse-title').attr('data-quantity');
        var items=$('select.facets-item-list-show-selector option:first-child').last().attr('class');
        var pages = Number(products) / Number(items);
        pages = Math.ceil(pages);
        return pages.toString();
    }
}

var Category = {
    name: function ($) {
        var name = $('div.facets-browse-category-heading-main-description h1').text();
        name = cleanName(name);
        return name
    }
}
