// sm-supplier-html/suppliers/www.footaction.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('#pdp_title > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('#pdp_price > span').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('#s7MixedMediaViewer_swatches > div > div > div > div').attr('style').split('url("').pop().replace('");', '').replace('wid=56', 'wid=500').replace('hei=56', 'hei=500');

        return image;
    },
    description: function($) {
        var description = $('span[itemprop="description"]').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div#endeca_search_results ul li').not('.clearRow');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a').next().attr('title');
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.product_price em').text();

        if (price == '') {
            price = $(product).find('p.product_price b').text();
        }

        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').next().attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('span > a:nth-child(2) > img').attr('src');
        image = 'http:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.itemsPerPage > span').first().text();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('div.itemsPerPage > span').first().text();
         var products = $('div#searchResultsInfo span').first().text().match(/[0-9]+/g).shift();

         var pages = products / items;
         pages = Math.ceil(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('div.header_text h1').text();
        name = cleanName(name);

        return name;
    }
}
