// sm-supplier-html/suppliers/www.cyclestore.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js"); 

module.exports = cycleStore =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1 div').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.pricetag').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('meta[property="og:image"]').attr('content');
        return image;
    },
    description: function($) {
        var description = $('form#productDetail strong').text();
        return description;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ParentDivClass');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.productlisttitle span.ddmx').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.PriceDivHolder span').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.cyclestore.co.uk/' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.LargeImage').attr('src');
        return 'https://www.cyclestore.co.uk/' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div#DropDiv2').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
     },
     pages: function($) {
/*        var pages = $('span.PageNofilter').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);*/

        return routine.matchNumbersAndPop($, 'span.PageNofilter');
    }
}

var Category = {
    name: function($) {
        var name = $('span.SpanTitleChild').text();
        name = cleanName(name);
        return name;
    }
}
