// sm-supplier-html/suppliers/www.matalan.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }/*,
    //Logic File
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }*/

}

var Product = {
    name: function ($) {
        var name = $('h1.product__hero__sidebar__title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        return 'unchecked';

    },
    price: function ($) {
        var price = $('span.product__hero__sidebar__price__price').html().replace(/\n/g, "").replace("&#xA3", "").replace(";", "").replace(/\s/g, "");
        price = price + '.00';
        return price;

    },
    image: function ($) {
        var image = $('picture img').attr('src');
        (image);
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProduct = $('div.search-results__item');
        return searchProduct;

    },
    name: function ($, product) {
        var name = $(product).find('div.results__detail__title').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.results__detail__price').html()
        price = getPrice(price)
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.matalan.co.uk' + $(anchor).attr('href');
        }
        return url;


    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;

    }
}
var Category = {
    name: function ($) {
        return cleanName($("h1.collection__intro__title").text());
    }
}
/*var Pagination = {

     items: function($) {
        return "TestItems";

     },
     pages: function($) {
        return "TestPage";

     }
}*/
