// sm-supplier-html/suppliers/www.hottopic.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = hottopic = {
    "useJavascript": true,
    "usePhantom": "hottopic.phantom.js",
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            //result.push(info)

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-name').html();
        name = cleanName(name)
        return name;

    },
    inStock: function ($) {
        var inStock = true
        var able = $('.availability').text()
        inStock = instockCheck(able,inStock)
        return inStock;

    },
    price: function ($) {
        var price = $('span[title="Sale Price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.product-primary-image a.main-image img').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.grid-tile')
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('div .product-name h2 a').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.product-price span.price-standard').text();
        var price2 = $(product).find('span.price-sales').text();
        var price3 = $(product).find('div.product-discounted-price').text();


        if (price === '') {
            price = price2;
        }

        if (price === '') {
            price = price3;
        }

        if (price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);
        }
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('.product-name h2 a').attr('href');
        if (url.startsWith("http") != true) {
            url = 'https://www.hottopic.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.product-image a img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.grid-tile');
        return parseInt(items.length);
    },
    pages: function ($) {
        var items = $('li.grid-tile');
        numitems = items.length;

        var result = null
        products = $('div.pagination .results-hits').text();

        if (products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / numitems);
        } else {
            if (Search.container($).length > 0) {
                result = 1
            }
        }


        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.html-slot-container span[style="font-family: roboto; font-size:30px; color:#FFF;"]').text();
        name = name.replace(/\w\S*/g, function (txt) { return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(); });
        name = cleanName(name);
        return name;
    }
}
