// sm-supplier-html/suppliers/www.americasbest.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
let generateSku = require('../../lib/testPattern.js')
let regex = /[0-9]+$/

module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'sku': (url) => {
        let sku = generateSku(url, regex)[0]
          return sku
      },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = generateSku(Search.url($, product), regex)[0]
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        price = $('div.price-starting   span span').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.similar-products__thumb').attr('src');
        return image;

    },
    url: function ($) {
        return $
    }   
}

var Search = {
    container: function($) {
        var searchProducts = $('div.similar-products__item')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.productDescription').find('a').first().attr('title')
        name = cleanName(name).replace(' more like this','');
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.similar-products__price').text();
        price = getPrice(price);
        return price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a.similar-products__link');
        if(anchor) {
            url = 'https://www.americasbest.com' + $(anchor).attr('href')
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.similar-products__thumb').attr('src');
        return 'https://www.americasbest.com'+image
    }
}

var Pagination = {

    items: function($) {
        var items = $('div.similar-products__item')
        return parseInt(items.length);

     },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function($) {
        
        return null
    }
}
