// sm-supplier-html/suppliers/www.tesco.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = Tesco =  {
    'useJavaScript': true,
    'usePhantom': 'tesco.phantom.js',

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
     'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.product-details-tile__title-wrapper h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var preorder = null;
     //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('div.stock-messaging-wrapper').text();

        if(stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if(stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        if(preorder !== null) {
            preorder = preorder.toLowerCase();
            if(preorder.indexOf('pre-order') > -1); 
            {
                inStock = false
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.price-control-wrapper span.value').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.product-image__container img').attr('src');
        return image;
    },
    url: function($) {
        return $;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-list--list-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('img.product-image').attr('alt');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price-details--wrapper span.value').first().text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.product-tile--title');
        if(anchor) {
            url = 'https://www.tesco.com' + $(anchor).attr('href');
        }
        return url;
    },
    
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);
     },
     pages: function($) {
        return 1;
     }
}

var Category = {
    name: function($) {
        var name = $('div#breadcrumb ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
