// sm-supplier-html/suppliers/www.morphyrichards.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = morphyrichards =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.page_title').text();
        var itemnum = $('dd[itemprop="sku"]').text();
        name = name + ' - ' + itemnum;
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var isInStock = true;
        var stock = $('span.outofstock').text()

        if(stock !== "") {
            isInStock = false;
        }
        return isInStock;
    },
    price: function($) {
        var price = $('div.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img.img").attr('src');
        return 'www.morphyrichards.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.col-xs-6');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-name a').text();
        name = cleanName(name);
        name = 'Morphy Richards' + ' ' + name;
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price').text();
        
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.morphyrichards.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.product-image img').attr('src');
        return 'https://www.morphyrichards.co.uk' + image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("h1.page_title").text());
    }
}
