// sm-supplier-html/suppliers/www.allsaints.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if(info.name != "" && info.price != "") {
            result.push(info)
            }
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product__title.js-product-title').text()
        name=cleanName(name)
        return name;
    },
    inStock: function ($) {
        var availabilty = $('div.pdisp_addtocart')
        var isInstock= false
        if (availabilty){
                isInstock=true
        }
	return isInstock;
    },
    price: function ($) {
        var price = $('h2.product__price').text();
    //    if(price.match(/\\./gi).length!=1){
    //        price=null
        if(price.indexOf('-') > -1) {
        price = null
        }
        else{
            price = getPrice(price);
        }
        return price;
    },
    image: function ($) {
        var image = $("img.rsImg.rsMainSlideImage.js-product-image").attr('src');
        return image;
    },
}

var Search = {
    container: function ($) {
        var searchProducts = $('.product-item.category-product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-item__name__text').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var pricenow = $(product).find('.product-item__price-new').text();
        var price=$(product).find('.product-item__price').text()
        pricenow=getPrice(pricenow)
        if(pricenow){
            return pricenow
        }else{
            price=getPrice(price)
            return price
        }

    },
    url: function ($, product) {
        var url = null
        var anchor = $(product).find('a.mainImg');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.allsaints.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('picture img').first().attr('src')
        image = 'https:' + image
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
}

var Category = {
    name: function ($) {
        return cleanName($('h1.category-title').text())
    }
}
