// sm-supplier-html/suppliers/www.ocado.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = ocado =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('strong[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var stockcheck = $('p.stock-status.oos').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock
    },
    price: function($) {
        var price = $('h5').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("li.zoomable img").attr('src');
        return 'https://www.ocado.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.main-column div.fop-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.fop-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.fop-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.ocado.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.fop-img').attr('src');
        return 'https://www.ocado.com' + image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("li span.crumb-trail-simple-text").text());
    }
}
