// sm-supplier-html/suppliers/www.wilko.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck.js')
module.exports = wilko =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.name').last().html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('.productDescShort').html();
        var outofstock = $('.out-of-stock ').html();
        var availability = $('h1').html();
        var check =$('.stock-wrapper').text()
        var check1 = $('.unavailable-msg').text()
        isInStock=instockCheck(check1,isInStock)
        isInStock = instockCheck(check,isInStock)
        if(producttop !== null) {
        if(producttop.indexOf('This product is available in store only') > -1) {
            isInStock = false
            }
        }

        if(outofstock !== null) {
        outofstock = outofstock.toLowerCase();
        if(outofstock.indexOf('out of stock') > -1) {
            isInStock = false;
            }
        }

        if(availability !== null) {
        availability = availability.toLowerCase();
        if(availability.indexOf('product unavailable') > -1) {
            isInStock = false;
            }
        }

        return isInStock;
    },
    price: function($) {
        var price = $('.prod-price').first().text();
        var price1=$('.pdp-price').first().text()
        if(price){
            price = getPrice(price);
        }else{
            price = getPrice(price1);
        }
        
        
        return price;

    },
    image: function($) {
        var image = $('div#productdetail-image img.activeZoom').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.product-link.name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
            if(url.indexOf('https://www.wilko.com') === -1) {
                url = 'https://www.wilko.com' + url;
            }
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://www.wilko.com' + image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.catResults').html();

        items = items.match(/[0-9]+/g).shift();
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.itemperpagebox select#itemperpage').html();
        numitems = items.match(/[0-9]+/g).shift();

        var products = $('div.catResults').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.prodList h1').html();
        name = cleanName(name);
        return name;
    }
}
