// sm-supplier-html/suppliers/www.coolshop.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = coolshop = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = "" + Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').last().text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var inStock = true
        var able = $('.thing-stock-stock-status').text()
        var second = $('.thing-buy-box-whitebox').text()
        inStock = instockCheck(second,inStock)
        inStock = instockCheck(able,inStock)
        return inStock;
    },
    price: function ($) {
        var smallNum = $('.after-decimal').first().text()
        var price = $('div.price-with-decimal-large').clone().children().remove().end().text();
        
        price = getPrice(price+'.'+smallNum);
        return price;
    },
    image: function ($) {
        var image = $("img#primary-thumbnail").attr('src');
        return image;
    },
    barcode: function ($) {
        var barcode = $('').text();
        return barcode;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product-card').not('div.product-card.nomore-products-card.hide');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.product-cards__title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var d = $(product).find('div.product-cards__price').not('span').html();

        let price = null

        if (d != null) {
            var int = d.substring(d.lastIndexOf(";") + 1, d.lastIndexOf("<s"))
            var dec = d.substring(d.lastIndexOf("l") + 3, d.lastIndexOf("</"))
            price = getPrice(int + '.' + dec)
        }

        // if(price.indexOf('£') > 1) {
        //     price = $(product).find('div.price').not('div.price--guide').text();
        // // }

        // if(price.indexOf('£') > 1) {
        //      price = $(product).find('div.price').not('div.price--guide').text();
        // }

        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.coolshop.co.uk' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("section.search-description h1").text());
    }
}
