// sm-supplier-html/suppliers/tjmaxx.tjx.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = tjmaxx =  {
    "useJavascript": true,
    "usePhantom": "tjmaxx.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("?"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("?"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) { 
            result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var brand = $('h1.product-brand').text();
        var name = $('h2.product-title').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('p.price span.product-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.zoom-lens img').attr('src').replace('//', '');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a.product-link span.product-title').text();
        var brand = $(product).find('a.product-link span.product-brand').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find ('.product-details span.product-price').text();
        
        if (price.indexOf('–') > -1) {
            price = null;
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.product-link');
        if(anchor) {
            url = 'http://tjmaxx.tjx.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-link img').attr('src').replace('//', '');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('.product');
        items = items.length;
        return parseInt(items);
     
     },
     pages: function($) {
        var items = $('.product');
        items = items.length;

        var result = null;
        var products = $('div.item-count').text();

        if(products !== ""){
            products = products.match(/[0-9,]+/g).pop();
                
                if(products.indexOf(",") > -1) {
                    products = products.replace(",","");
                }

            result = Math.ceil(products / items);
        
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;
     }
}


var Category = {
    name: function($) {
        var name = $('h2#category-name').text();
        name = cleanName(name);
        return name;
    }
}
