// sm-supplier-html/suppliers/www.directgolf.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = directgolf =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }


}

var Product = {
    name: function($) {
        var name = $('span#ProductName').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {

	return 'unchecked';
},
    price: function($) {
        var price = $('div.pdpPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#imgProduct").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul#navlist li');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.s-producttext-top-wrapper a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.curprice').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.directgolf.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.img-responsive').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("span#dnn_ctr185761_ViewTemplate_ctl00_CategoryHeaderTop_lblCategoryHeader").text());
    }
}
