// se-app-model/lib/saveJobSupplierPayload.js
const formatDate = require('./formatDate');
const createJobIdName = require('./createJobIdName.js');
const makeSafe = require('./makeSafe.js');
module.exports = (data) => {


    let name = 'Supplier Search' ;
    let jobType = name.toUpperCase(name).replace(' ','_');
    let typePayload = data[data.type];
    let jobInfo;

    let fullSearch = data.fastApi ? true : false
    let fullSearchName = data.fastApi ? ': FULLSEARCH' : ""

    if(data.stagingMatchedSearch) {
        fullSearchName = data.stagingMatchedSearch ? ': STAGING_MATCHED ' : ""
    }


    let jobPayload = {
        status: "QUEUED",
        created: new Date(),
        userName: data.userName,
        userId: data.userId,
        jobType: jobType,
        fastApi: data.fastApi,
        country: data.country,
        name: data.supplierName + fullSearchName,
        size: data.size,
        categorySearchAge: data.categorySearchAge,
        options: {
            supplierShippingCost: data.options.supplierShippingCost,
            supplierDiscount: data.options.supplierDiscount
        },
        globalFilters: data.globalFilters,
        supplierId: data.supplierId,
        domain: data.domain.toLowerCase(),
        selectedPeriod: data.selectedPeriod,
        fullSearch: fullSearch,
        hasPIF: data.hasPIF,
        totalScanned: data.totalScanned,
        stagingMatchedSearch: data.stagingMatchedSearch,


    }

    let jobIdPayload = {
        country: data.country,
        jobType: jobPayload.jobType.replace('- ','').toUpperCase().replace(/ /g,'_'),
        jobName: data.supplierName.trim().toUpperCase().replace(/ /g,'_'),
        userName: data.userName,
        date: formatDate(jobPayload.created.toISOString())
    }


    var qName = createJobIdName(jobIdPayload)
    qName = makeSafe(qName)
    jobPayload.idName = qName

    console.log("JOB PAYLOAD:", jobPayload.idName)

    return jobPayload;

}

function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}
