// sm-supplier-html/suppliers/www.offbroadwayshoes.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us' //<j/>
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'; //<j/>
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.main_info_box h1').text(); //<j/>
        name = cleanName(name);

        return name; //<j/>
    },
    inStock: function($) {
        return 'unchecked'; //<j/>
    },
    price: function($) {
    //<j>
        var price = $('div.main_info_box h3').text();
        price = getPrice(price);

        return price;
    //</j>

    },
    image: function($) {
        var image = $('div.main_big_img a img').attr('src'); //<j/>

        return image; //<j/>

    },
    description: function($) {
        description = $('div.product_desc p').text();
        (description);

        return description;

        
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.box_prodcut_min');

        return searchProducts;

    },
    name: function($, product) {
        var productName = $(product).find('div.product_brand_name h2').text();
        productName = cleanName(productName);

        return productName;

    },
    price: function($, product) {
        var productPrice = $(product).find('h3').text();

        productPrice = getPrice(productPrice);


        return productPrice;

    },
    url: function($, product) {
        var url = null;

        var anchor = $(product).find('a');

        if(anchor) {
            url = 'https://www.offbroadwayshoes.com' + (anchor).attr('href');
        }

        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.box_prodcut_min_img > img').attr('src'); 
        

        return image;
        
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.visible-xs label.selectedViewOption').text();
        items = items.match(/[0-9]+/g);

        return parseInt(items[0]);

     },
     pages: function($) {
        var items = $('div.visible-xs label.selectedViewOption').text();
        items = items.match(/[0-9]+/g);
        items = parseInt(items[0]);

        productCount = $('div.breadcrumb_paging_container_bottom label.resultCount').text();
        productCount = productCount.match(/[0-9]+/g);

        var sum = productCount / items;

        sum = Math.ceil(sum);

        return sum;


     }
}

var Category = {
    name: function($) {
        var name = $('td.breadcrumbs ul li.active').text();
        name = cleanName(name);
        return name;
    }
}
