// sm-supplier-html/suppliers/www.crateandbarrel.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
           // 
        });

        return result
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        let oss = $('.secondary-message').text().includes('This item is currently on backorder.')

        if (oss) {
            return false;
        } else {
            return 'unchecked';
        }
    },
    price: function ($) {
        var price = $('span.regPrice').html();
        if (price.indexOf('-') > -1) {
            price = null;
        }
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.hwProductImage').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.jsItemRow');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a.product-miniset-title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var salePrice = $(product).find('span.spanSale').text();
        var price = $(product).find('span.regPrice').text();
        salePrice = getPrice(salePrice);
        price = getPrice(price);

        if(salePrice != "") {
            price = salePrice;
        } else {
            price = price;
        }
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.crateandbarrel.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.spillImage').attr('src');
        return 'https:' + image;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name;
    }
}
