// sm-supplier-html/suppliers/www.miniinthebox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'//Please fill in country here //
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'; //Please fill in country here //
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.prod-info-title h1').text();
        name = cleanName(name);
        name = name.replace(/\s+#\d+$/, '');
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.current-price strong[itemprop="price"]').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.normal');
        return image ? image.attr('src') : '';
    },
    description: function($) {
        var description = $('div.prod-description-feature-list').text();
        
        if(description !== null) {
            
            return description.replace(/(\r\n|\n|\r|)/gm,"").replace(/(\t)/gm," ");
        }
        

    }
}

var Search = {
    container: function($) {
        return $('div.pagelet.product-list dl.item-block');
    },
    name: function($, product) {
        var name = $(product).find('span.prod-name-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('dd.prod-item-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('dd.prod-name a').attr('href');
        return url;
    },
    image: function($, product) {
        //Somehow this doesn't work for the last product//
        var image = $(product).find('div.img-box img.prod_img').attr('src');
       
        if(image === undefined) {
            image = $(product).find('div.img-box img.prod_img').attr('data-src');
        }
        return image; // the problem was that due to javascript it wasnt loading all the images so the last products had a slightly different source, () helps to debug//
    }
}

var Pagination = {

     items: function($) {
        return 60; //This is the number of items on the page, if you come across a dropdown where you can select the items per page let me know//
     },
     pages: function($) {
        var products = $('h2.flash-banner-title span').text(); // get the total number of products iun this category, this was at the top of the page //
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / 60; // Divided the total number of products by the number of products on the page //
        return Math.ceil(sum); // using ceil to round the number up //

     }
}


var Category = {
    name: function($) {
        var name = $('title').text();
        name = cleanName(name);
        name = name.split('|');
        name = name[1];

        if(name.indexOf('for 2016')) {
            name = name.replace("for 2016", "");
        }
        if(name.indexOf('for 2017')) {
            name = name.replace("for 2017", "");
        }
        return name.trim();
    }
}
