// sm-supplier-html/suppliers/www.shopbop.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text().trim();

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.priceBlock').text().trim();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('#productZoomImage').attr('href');

        return image;
    },
    description: function($) {
        var description = $('div[itemprop="description"]').html().split('<br>')[0].trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('.hproduct');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('.title').text().trim();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.retail-price').text()
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.url').attr('href');
        url = 'https://www.shopbop.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('span.productBrowseMainImage img').attr('src').trim();

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = parseInt($('#pagination-container-top div.more span.selected').text().split(' ').pop());

         return items;
     },
     pages: function($) {
         var products = $('span#product-count').text().split(' ').shift();
         var items = $('#pagination-container-top div.more span.selected').text().split(' ').pop();


         sum = products / items;

         //(Math.ceil(pagesCount));

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('h1.pageHeadingSpan').text();
        name = cleanName(name);
        return name;
    }
}
