// sm-supplier-html/suppliers/www.saksfifthavenue.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-overview__short-description').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var status = $('strong.product__label-message-display-label').text();

        if (status.indexOf('Pre-Order') > -1) {
            status = false;
        } else {
            status = true;
        }

        return status;
    },
    price: function($) {
        var status = $('strong.product__label-message-display-label').text();
        var price;

        if (status.indexOf('Pre-Order') > -1) {
            price = null;
        } else {
            price = $('section.product__sku-selection div div.product-pricing dl dd span[itemprop="price"]').text();
            price = getPrice(price);
        }

        return price;
    },
    image: function($) {
        var image = $('#video-player video').attr('poster');
        image = 'http:' + image;

        return image;
    },
    description: function($) {
        var description = $('section.product-description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.sfa-pa-product-with-swatches');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('p.product-description').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-text a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.pa-product-large').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('span.currentView.pagination-count').first().text();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('span.currentView.pagination-count').first().text();

         var products = $('span.totalRecords').first().text().replace(',', '');

         var sum = products / items;

         return Math.ceil(sum);
     }
}


var Category = {
    name: function($) {
        var name = $('div.search-cat span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
