// sm-supplier-html/suppliers/www.ashford.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = ashford =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('a.qvPrdURL').text();
        name = name.replace(/\s+/gm, ' ');
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.outOfStockMsg');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('table#pricing').text();
        price = price.match(/[0-9.,]+/gm).pop();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.productDetailImage').last().attr('src');
        return 'http://www.ashford.com' + image;
    },
    description: function($) {
        return 'xxx';
        /*
        var description = $('h3.prod_desc').text();
        if (description) {
            description = description.replace(/\s+/gm, ' ').trim();
            return description;
        }
        */
    }
}

var Search = {
    container: function($) {
        return $('div.watch');
    },
    name: function($, product) {
        var brand = $(product).find('a.brand_name').text();
        var name = $(product).find('a.prod_name').text();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + ' ' + name;
    },
    price: function($, product) {
        var price = $(product).find('div.product-price').text();
        price = price.replace(/\s+/gm, '').replace(/\$/, '').trim();
        price = price.replace('Clearance:','');
        price = price.replace(',','');
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-desc a').attr('href');
        return 'http://www.ashford.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http://www.ashford.com' + image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }

var Category = {
    name: function($) {
        var name = $('title').text();
        name = cleanName(name);
        return name;
    }   
}
