// sm-supplier-html/suppliers/www.officeplayground.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = officeplayground =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('p.availability');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('p.special-price, span.regular-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#image-main').attr('src');
        return image;
    },
    description: function($) {
        var description = $('dl#collateral-tabs dd:nth-child(2) div.std div');
        if (description) {
            return description.text().replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('li.item');
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.special-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.item'); 
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
       var items = $('li.item'); 
       items = items.length;

       
       var result = null;
       var products = $('li.amshopby-advanced ol li.active a span.count').text();


        if(products !== "") {
            products = products.match(/[0-9,]+/g).shift();   
            result = Math.ceil(products / items);
        } else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
       
        return result;
     }
}
var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
