// sm-supplier-html/suppliers/www.heb.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = heb =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.pdp-product-location');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('location: ') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.s7staticimage img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('p[itemprop=description]').text();
        return description.trim();
    }
}

var Search = {
    container: function($) {
        return $('li.responsivegriditem');
    },
    name: function($, product) {
        var name = $(product).find('span.responsivegriditem__title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.cat-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.heb.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.gridproductview');
        return parseInt(items.length);
     },
     pages: function($) {
        var numitems = $('li.gridproductview').length;

        var products = $('div.pagenofn').text();
        products = products.match(/\d+/gm).pop();

        var pages = products / numitems;
        return Math.ceil(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();
        name = cleanName(name);
        return name;

    }
}
