// sm-supplier-html/suppliers/mbeans.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        $('span#IPINManuName').remove();
        var name = $('h1#IPItemName').text().toLowerCase();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div#IPPricePrice div').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div#IPMainImageOuter img').attr('src');
        image = 'http://mbeans.com' + image;

        return image;
    },
    description: function($) {
        var description = $('div#IPDescription').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.SSRLGV-Item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.SSRLGVI-Name').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.SSRLGVIP-Price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.SSRLGVI-Name a').attr('href');

        return url;
    },
    image: function($, product) {
      //  var image = $(product).find('div.SSRLGVI-Image');
      //  Image embedded on page in a way that makes it complex to get out setting to null for now
          var image = null;
        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.SSDBV-XToY').text();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('div.SSDBV-XToY').text();
         items = items.match(/[0-9]+/g).pop();

         var products = $('div.PHCHeadline').text();
         products = products.match(/[0-9]+/g).pop();

         var sum = products / items;

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('ul#PageTrail li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
