// sm-supplier-html/suppliers/www.tfaw.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = tfaw =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span.blackheader').first().text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.profile-desc').text();
        if (stockMessage) {
            if (stockMessage.toLowerCase().trim().indexOf('this item is currently not available') != -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('td[itemprop=price] span.redheader').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.cover-image').attr('src');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('span[itemprop=description]').text();
        description = description.replace(/\s+/gm, ' ').trim();
        return description;
    }
}

var Search = {
    container: function($) {
        return $('table.results_mod td[bgcolor=#FFFFFF]');
    },
    name: function($, product) {
        var name = $(product).find('a.product-profile-link').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.smallbodytextbold font').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.product-profile-link').attr('href');
        return 'http://www.tfaw.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.cover-image').attr('src');
        return 'http:' + image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('select[name=_results_limit_search] option').last().text();
        
        if(products !== "") {
           products = products.match(/\d+/gm).pop();
           products = parseInt(products); 
        }else {
            if(Search.container($).length > 0) {
                products = 1;
            }
        }
        return products;
     },
     pages: function($) {
        var products = $('select[name=_results_limit_search] option').last().text();
        
        if(products !== "") {
           products = products.match(/\d+/gm).pop();
           products = parseInt(products); 
        }else {
            if(Search.container($).length > 0) {
                products = 1;
            }
        }

        var result = null;
        var total = $('div.small-corners-light table tr:nth-child(1)').text();
        
        if(total !== ""){
            total = total.match(/\d+/gm)[0];    
            result = Math.ceil(total / products);
        }else {
             if(Search.container($).length > 0) {
                result = 1;
            }
        }
        return result
     }
}

var Category = {
    name: function($) {
        var name = $('a.sectionheader').text();
        name = cleanName(name);
        return name;
    }
}
