// sm-supplier-html/suppliers/www.paulaschoice.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)
        return info
    },

    // 'sku': (url) => {
    //     return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    // },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-name').html()
        name = cleanName(name)
        return name

    },
    inStock: function ($) {
        return 'UNCHECKED'

    },
    price: function ($) {
        var price = $('div.product-pricing span').html()
        price = getPrice(price)
        return price

    },
    image: function ($) {
        var image = $('div.product-primary-image a').attr('href')

        image = 'https://www.paulaschoice.co.uk' + image
        return image
    },
    description: function($) {
        var description = $('div.product-info-container section.product-short-description p').html();
        description = cleanName(description)
        return description
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product-tile');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.product-name a').html();
        name = cleanName(name);
        return name
    },
    price: function ($, product) {
        var price = $(product).find('div.product-pricing span').last().html();
        price = getPrice(price)
        return price

    },
    url: function ($, product) {
        var url = $(product).find('a.name-link').attr('href')
        if(url){
            if(url.indexOf('http')>-1){
                return url
            }else{
                return 'https://www.paulaschoice.co.uk' +url
            }
        }
        return url

    },

    image: function ($, product) {
        var image = $(product).find('div.product-image a.thumb-link img').attr('src')
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul#search-result-items li.grid-tile').length;
        return items
    },
    pages: function ($) {
        var items = $('ul#search-result-items li.grid-tile').length;

        var products = $('div.results-hits').text();
        products = products.match(/[0-9]+/g).pop()
        return Math.ceil(products / items);
    }
}

var Category = {
    name: function ($) {
        var name = $('div.cat-banner h1').text()
        name = cleanName(name)
        return name;
    }
}
