// sm-supplier-html/suppliers/www.roadrunnersports.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = roadrunnersports =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        // info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span.prod_title h1').text();
        //name = name.replace(/\s+/gm, ' ').trim();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockMessage = $('span.prod_nostock_title');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = null
        var regPrice = $('span.prod_detail_reg_price').first().text();
        var salePrice = $('span.prod_select_title3 > span.prod_detail_sale_price').text()  
        if (salePrice) {
            if (salePrice.indexOf('-') === -1) {
                price = getPrice(price);
            }
        } else {
            price = getPrice(regPrice)
        }
        return price
    },
    image: function($) {
        // var image = $('div.s7flyoutStaticImage img:nth-child(1)').attr('src');
        // return image;
        var style = $('div.s7thumb[state=selected]').attr('style') 
        // if(style != undefined){
        //     var imageUrl = style.match(/url\("(.+)"\)/)[1]
        //     imageUrl = imageUrl.replace(/wid=\d+/, 'wid=445')
        //     imageUrl = imageUrl.replace(/hei=\d+/, 'hei=445')
        // }
        
        return style
    },
    description: function($) {
        var description = $('div#grp_1 p').text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        var container = $('figure.product_colorways');
        return container;
    },
    name: function($, product) {
        var name = $(product).find('div.shoe-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = null
        var listPrice = $(product).find('span.celebros_list_price.price').text();
        var salePrice = $(product).find('span.celebros_sale_price.price').text();
        if (salePrice) {
            if (salePrice.indexOf('-') !== -1) {
                price = getPrice(listPrice);
            }
        } else {
            price = getPrice(listPrice)
        }
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.roadrunnersports.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('data-yo-src');
        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('select[name=p]').text();
        if(products != ''){
            products = products.match(/\d+/gm).pop();
        }
        return parseInt(products);
     },
     pages: function($) {
        var products = $('select[name=p]').text();
        if(products != ''){
        products = products.match(/\d+/gm).pop();
        }
        var total = $('td#product_count').text();
        if(products != ''){
        total = total.match(/\d+/gm).pop();
        }
        var pages = total / products;
        return Math.ceil(pages);
     }
}
var Category = {
    name: function($) {
        var name = $('div.top-banner-section h1').text();
        name = cleanName(name);
        return name;
    }
}
