// sm-supplier-html/suppliers/www.regentproducts.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('td.pageHeading:nth-child(1)').text();
        name = cleanName(name.replace(/\[\d+\]/gm, ''));
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        return inStock;
    },
    price: function ($) {
        var price = $('td.pageHeading:nth-child(2)').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('.main_table_heading_inner img[name]').attr('src');
        return 'http://www.regentproducts.com/' + image;
    },
    description: function ($) {
        return '';
    }
}

var Search = {
    container: function ($) {
        return $('table.productListing tr.productListing-odd,tr.productListing-even');
    },
    name: function ($, product) {
        var name = $(product).find('td:nth-child(3) a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('td:nth-child(4)').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('td:nth-child(3) a').attr('href');
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('td:nth-child(1) img').attr('src');
        return 'http://www.regentproducts.com/' + image;
    }
}

var Pagination = {

    items: function ($) {
        return 25;
    },
    pages: function ($) {
        var products = $('td.smallText:contains("Displaying")').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / 25;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('td.pageHeading').text();
        name = cleanName(name);
        return name;
    }
}
