// sm-supplier-html/suppliers/www.entertainmentearth.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = enterainmentearth = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()

            var name = Search.name($, product);
            var price = Search.price($, product);
            var url = Search.url($, product);
            var image = Search.image($, product);


            info.name = name;
            info.price = price;
            info.url = url;
            info.image = image;
            info.country = 'us';
            result.push(info)

        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-title').text();

        if (name.indexOf('- Free Shipping') > -1) {
            name = name.replace(" - Free Shipping", "");
        }
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var isInStock = true;
        var checkstock = $('h5.product-status').html();

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if ((checkstock.indexOf('out of stock') > -1)
                || (checkstock.indexOf('coming in') > -1)
                || (checkstock.indexOf('sold out') > -1)) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function ($) {
        var price = $('span.product-price').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.product-image img').attr('src');
        image = 'https:' + image
        return image;
    },
    description: function ($) {
        var description = $('div#Description').text();

        if (description !== null) {
            return description.replace(/\s+/gm, ' ').trim();
        }

    }

}

var Search = {
    container: function ($) {
        var products = $('div.product-results div.item.grid-view')
        return products;
    },
    name: function ($, product) {
        var name = $(product).find('h4.item-name').text();
        name = cleanName(name)
        return name;
    },
    price: function ($, product) {
        var price = null;
        var price = $(product).find('b.item-price').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('h4 a').attr('href');
        return 'https://www.entertainmentearth.com' + url;

    },
    image: function ($, product) {
        var image = null;
        var image = $(product).find('div.image img').attr('src');
        return 'https:' + image;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.hitlist-title').text()
        name = cleanName(name);
        return name;
    }
}
