// sm-supplier-html/suppliers/www.hobbycraft.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = hobbycraft =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var availability = $('.pull-right.hidden-xs').text()
        var isInStock = true
        isInStock = instockCheck(availability,isInStock)
	return isInStock;
},
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.mobile-image-blocker img").attr('src');
        return 'www.hobbycraft.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.list_element');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('span.productName').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.hobbycraft.co.uk/' + url.replace('../../', '');

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src').replace('../../', '');
        return 'http://www.hobbycraft.co.uk/' + image;

    }
}

var Pagination = {

     items: function($) {
        var items = ('div.list_element');

        var pages = $('div.pagination span.pagin-tablet').text();
        pages = pages.match(/[0-9]+/g)

        if(items && pages) {
            return parseInt(items.length);
        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = $('div.pagination span.pagin-tablet').text();
        pages = pages.match(/[0-9]+/g)

        var items = ('div.list_element');

        if(pages && items) {
            return parseInt(pages.pop());
        } else {
            return null;
        }
     }
}

var Category = {
    name: function($) {
        var name = $('h1.category_heading').text();
        name = cleanName(name);
        return name;
    }
}
