// sm-supplier-html/suppliers/www.ewatches.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var brand = $('div.productAttributes h2.productBrand').text();
        var name = $('div.productAttributes h1').text();
        name = brand + " " + name;
        return name;

    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('ul.productInfoSecondary li.priceFinal').text();
        price = getPrice(price)
        return price;
    },
    image: function($) {
        var image = $('img.img-responsive').attr('src');
        return image; 
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.results li.col-xs-6');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('div.box h3').text();
        var name = $(product).find('div.box p').text();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var normPrice = $(product).find('ul.prices li b').text();
        var price = $(product).find('ul.prices li strong').text();

        if(normPrice === "") {
            normPrice = price;
        }

        normPrice = getPrice(normPrice);
        return normPrice;


    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.ewatches.com' + (anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image; 

    }
}

var Pagination = {

      items: function($) {
        var items = $('div.search-heading h1').text();
        items = items.match(/[0-9]+/g);
        items = items[1];
        return parseInt(items);


     },
        pages: function($) {
        var items = $('div.search-heading h1').text();
        items = items.match(/[0-9]+/g);
        items = items[1];

        var products = $('div.search-heading h1').text();
        products = products.match(/[0-9]+/g);
        products = products[2];

        sum = products / items;

        return Math.ceil(sum); 

     }
}

var Category = {
    name: function($) {

    }
}
