// sm-supplier-html/suppliers/www.chewy.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = chewy = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);
        info.nexturl=Pagination.nexturl($)
        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div#product-title h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        let check = true
        let availiablity = $('.btn-addtocart.btn-pdp-addtocart.btn.btn--commerce.mr-xs').val()
        if (availiablity) {
            availiablity = availiablity.toLowerCase()
            if (availiablity.indexOf('unavailable') > -1 || availiablity.indexOf('out of stock') > -1 || availiablity.indexOf('pre-order') > -1) {
                check = false
            }
        }
        return check;
    },
    price: function ($) {
        var price = $('span.ga-eec__price').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('a#Zoomer').attr('href');
        image = 'http:' + image;

        return image;
    },
    description: function ($) {
        var description = $('section.descriptions__content.cw-tabs__content--left').text().replace(/\s+/g, " ").trim();;

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('article.product-holder');

        return product;
    },
    name: function ($, product) {
        var name = $(product).attr('data-name')
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('p.autoship>strong').text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price=$(product).attr('data-price')
            price=getPrice(price)
            return price
        }
        
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'https://www.chewy.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.image-holder img').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

    items: function ($) {
        return parseInt($('article.product-holder').length)
    },
    pages: function ($) {
        var products = $('p.results-count').text();
        if (product) {
            products = products.match(/of\s[0-9]+/g).pop();
            products = products.match(/[0-9]+/g).pop();
        } else {
            products = 0
        }
        return products
    },
    nexturl: function ($){
        var next= $('.cw-pagination__next').attr('href')
        return next
    }
    
}

var Category = {
    name: function ($) {
        var name = $('h1.results-cat-title').text();
        name = cleanName(name);
        return name;
    }
}
