// sm-supplier-html/suppliers/www.currys.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var routine = require ("../../lib/routine.js");
var getPrice = require('../../lib/getPrice.js');

module.exports = currys =  {
    "useJavascript": true,

    'sku': (url) => {
        url = url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."))
        url = url.replace('-pdt','').split('-').pop()
        return url
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf("."));
            info.name = Search.name($, product)
            info.sku = sku.replace('-pdt','').split('-').pop()
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        
        var brand = $('.page-title').find('span').html();
        var name = $('.page-title span').eq(1).html();
        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;

        var proChannels = $('.prd-channels').html();
        if (proChannels) {
            if(proChannels.indexOf('Sorry this item is out of stock') > -1) {
                isInStock = false
            }
            if(proChannels.indexOf('Not available for delivery') > -1) {
                isInStock = false
            }
        }

        var checkstock = $('.oos').html();
        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if(checkstock.indexOf('out of stock') > -1) {
                isInStock = false;
            }
        }

        return isInStock
    },
    price: function($) {
        var price = $('div.prd-amounts strong.current').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.swiper-slide a img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('article.result-prd');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('header.productTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find ('strong.price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.desc a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.product-images img.image').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select#sViewCount').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
     },
     pages: function($) {
/*        var items = $('select#sViewCount').text(); 
        numitems = items.match(/[0-9]+/g).pop();*/

        var numitems = routine.matchNumbersAndPop($, "select#sViewCount");

        if(numitems !== null) {
           var products = routine.matchNumbersAndPop($, 'section[role="main"] .nosp strong');
           return routine.pagesFromTotalAndProd(products, numitems);
        } else {
            return null;
        }
/*        var products = $('section[role="main"] .nosp strong').text();
        products = products.match(/[0-9]+/g).pop()

        sum = products / numitems
        return Math.ceil(sum)
      */


     }
}


var Category = {
    name: function($) {
        var name = $('h1.pageTitle').text();
        name = cleanName(name);
        return name;
    }
}
