// sm-supplier-html/suppliers/www.cartridgemonkey.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = cartridgeMonkey =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.prd_productImage').attr('src');
        return image;
    },
}

var Search = {
    container: function($) {
        var searchProducts = $('div.clearfix.g-bc-s.prd-summ-entry');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.g-fw-b.g-m-t-10.productName').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('strong').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.cartridgemonkey.com' + url 
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        
        if(image.startsWith("http")) {
            image = image
        } else {
            image = 'https:' + image
        }

        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('div.g-1col-Main h1').text();
        name = cleanName(name);
        return name;
    }
}
