// sm-supplier-html/suppliers/www.dickies.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('#pdpMain > div.product-col-2.product-detail > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('#product-content > div.product-price > span[itemprop="offers"] span.price-sales').text();
        price = getPrice(price);

        if (price == '') {
            price = null;
        }

        return price;
    },
    image: function ($) {
        var image = $('div.product-primary-image > a > img.zoomImg').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('#pdpMain > div.product-col-2.product-detail > div.product-description-info > div > p > span').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('li.grid-tile');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('a').attr('title')
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.product-info > div.product-pricing > span').text();
        let promoPrice = $(product).find('.product-sales-price').text()

        if (promoPrice != null) {
            price = promoPrice
        }

        if (price.indexOf(' - ') > -1) {
            price = null;
        } else {
            price = getPrice(price);
        }
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.pagination > div').first().text().match(/[0-9]+/g)[1];
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('div.pagination > div').first().text().match(/[0-9]+/g)[1];
        var products = $('div.pagination > div').first().text().match(/[0-9]+/g).pop();

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#secondary > h1').text();
        name = cleanName(name)

        return name;
    }
}
