// sm-supplier-html/suppliers/www.simplygames.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = simplygames =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


var Product = {
    name: function($) {
        var name = $('.product_content h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
    var isInStock = true;
    var checkstock = $('.out_of_stock').html();

    if (checkstock !== null) {
        checkstock = checkstock.toLowerCase();
        if(checkstock.indexOf('out of stock') > -1) {
            isInStock = false;
        }
    }

    return isInStock

},
    price: function($) {
        var price = $('.price_point').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('.product_image img').attr('src');
        image = 'http://www.simplygames.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.list_item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.button').attr('title');
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('a span').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.simplygames.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return 'http://www.simplygames.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.list_item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var pages = $('div.content_3 ul.paging').html();

        if(pages.indexOf('&#xBB') > -1){
            pages = pages.match(/[0-9]+/g);
            pages.pop();
            pages = pages.pop();
        } else {
            pages = pages.match(/[0-9]+/g).pop();
        }
        return parseInt(pages)
     }
}


var Category = {
    name: function($) {
        var name = $('div.content_wrapper h2.list_header').text();
        name = cleanName(name);
        return name;
    }
}
