// sm-supplier-html/suppliers/www.wayfair.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = wayfair = {
    "useJavaScript": true,
    "usePhantom": "wayfair.phantom",

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.pl-PageTitle-header').html();
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div.ProductDetailMainImage-outOfStockOverlay').text();
        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('h2.ProductPricing-amount').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.ProductDetailMainImage-mainImageResponsive img').first().attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.BrowseProductGrid div.Grid-item.Grid-item--row.Grid-item--wrap');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2.ProductCard-name').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.ProductCard-price').text();
        price = getPrice(price)
        return price
    },
    url: function ($, product) {
        var url = ''
        var link = $(product).find('a.ProductCard')
        if (link.length) {
            url = link.attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.ImageComponent img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('select.pagination_select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
    },
    pages: function ($) {
        var items = $('select.pagination_select').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('span.filter_remove_wrap').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1#browse-page-title').text();
        name = cleanName(name);
        return name;
    }
}
