// sm-supplier-html/suppliers/www.bitsandpieces.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.title_bfont').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        price = $('b.clear').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.product_img a img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product_tile').not('.MB_PRODUCTSLOT');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('.prodtilename').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var red_price=$(product).find('.colorred').text()
        var price = $(product).find('.colorblack').text();
        if(price){
            price=getPrice(price)
        }else{
            price = getPrice(red_price);
        }
        return price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.bitsandpieces.com' + $(anchor).attr('href').replace("?quickview=true","");
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.prodtileimage>img').attr('src');
        return image
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.product_tile').not('.MB_PRODUCTSLOT');
        return parseInt(items.length);

     },
     pages: function($) {
        var pages = $('ul.pagination').text();
        if(pages === ""){
            return null
        }else {
           pages = pages.match(/[0-9]+/g).pop(); 
        }
        if(parseInt(pages)===null){
            return 0
        }else{
            return parseInt(pages)
        }
        
     }
}

var Category = {
    name: function($) {
        var name = $('div.category_detail h1.no_margin').text();
        name = cleanName(name);
        return name;
    }
}
