// sm-supplier-html/suppliers/www.overstock.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'useJavascript': true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            if (Search.url($, product).startsWith("https://www.overstock.com") == true) {
                var info = new SearchPageInfo()
                info.name = Search.name($, product)
                info.price = Search.price($, product);
                info.url = Search.url($, product);
                info.image = Search.image($, product);
                info.country = "us"
            result.push(info)
        }
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('.product-title h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div.out-of-stock-label').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock
    },
    price: function ($) {
        var price = $('span.price-value-container').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('.hero img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product-tile');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.product-title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        //    var salePrice = $(product).find('div.product-price.sale').text()
        var price = $(product).find('span.from-price').text()

        //      if (salePrice.indexOf("-") > 1) {
        //          salePrice = null
        //      } else {
        //          salePrice = salePrice
        //      }

        //    var priceDollar = $(product).find('span.price-dollar').text();
        //    var priceCent = $(product).find('span.price-cent').text();
        //    var price = priceDollar + '.' + priceCent
        price = getPrice(price)

        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('.img-middle img').attr('src');
        return image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("h1.search-nav-h1").text());
    }
}
