// sm-supplier-html/suppliers/www.drinkstuff.com/logic.js

var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithwebAddressChangeToCamelCase = {

    'sku': (url) => {
        url = url.split('#')[0]
        url = url.split('&')[0]
        return url.substring(url.lastIndexOf("=") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).split('#')[0]
            var sku = Search.url($, product).split('&')[0]
            var name = Search.name($, product).split('£')[0].trim()
            info.name = name;
            info.sku = sku.substring(sku.lastIndexOf("=") + 1, sku.lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1#name').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var producttop = $('.availability').html();
        var checkstock = $('div#productversionscontainer div#footer strong').html();
        var preorder = $('div#productversionscontainer div#header span').text();
        var availability = $('#not_available').html();
        isInStock=instockCheck(preorder,isInStock)
        if (producttop !== null) {
            if ((producttop.indexOf('Out of Stock') > -1)
                || (producttop.indexOf('Not in Stock') > -1)
                || (producttop.indexOf('Unavailable') > -1)) {
                isInStock = false
            }
        }

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('out of stock') > -1) {
                isInStock = false;
            }
        }

        if (preorder !== null) {
            preorder = preorder.toLowerCase();
            if (preorder.indexOf('pre-order') > -1) {
                isInStock = false;
            }
        }

        if (availability !== null) {
            availability = availability.toLowerCase();
            if (availability.indexOf('discontinued') > -1) {
                isInStock = false;
            }
        }



        return isInStock
    },
    price: function ($) {
        var price = $('div#pvprice').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img#productimage').attr('src');
        if(image !== null){
            if (image.indexOf('https:') < 0) {
                image = 'https:' + image;
            }
        }
        return image;
    },
    description: function ($) {
        var description = $('div#desc_description').text();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a').attr('title');
        var quantity = $(product).find('p.price span:nth-child(2)').text();
        name = name.replace("/", "");
        name = cleanName(name);
        quantity = cleanName(quantity);
        return name + ' ' + quantity;
    },
    price: function ($, product) {
        var price = $(product).find('p.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.drinkstuff.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src');


        if (image.indexOf('https:') < 0) {
            image = 'https:' + image;
        }
        return image.trim();
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.product')
        (items.length);


        if (items) {
            return parseInt(items.length)
        } else {
            return null;
        }
    },
    pages: function ($) {
        var pages = $('span#control').text();
        pages = pages.match(/[0-9]+/g)

        if (pages) {
            pages.pop();
            return parseInt(pages);
        } else {
            return null;
        }


    }
}

var Category = {
    name: function ($) {
        var name = $('h1#categoryheader').text();
        name = cleanName(name);
        return name;

    }
}
