// sm-supplier-html/suppliers/www.nothingbutsavings.com/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var logger = require('../../lib/logger.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = {
    //"useJavascript" : true,
    'search' : function($, next) {
        var result = []

        $('.product-holder').find('li').map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($,product)
            info.country = "us"
            result.push(info)
        });

        // var pages = getNumberOfPages($, result.length)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = pPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "us"
        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    var link = $(product).find('.product-tile').find('a').attr('href')
    return "http://www.nothingbutsavings.com" + link
}


function searchName($, product) {
    return $(product).find('.title').text()
}


function searchPrice($, product) {
    var thePrice = $(product).find('b').text()
    thePrice = /[0-9]+.[0-9][0-9]/.exec(thePrice)
    if (thePrice !== null) {
        thePrice = thePrice[0]
    } else {
        thePrice = ""
    }

    return thePrice
}


function getNumberOfPages($, products) {
    var pages = $('.items-row').find('span').html()
    var retn = 1
    if ((pages !== null) && (pages !== "")) {
        pages = pages.substring(pages.indexOf('of'))
        pages = /[0-9]+/.exec(pages)
        if (pages !== null) {
            pages = pages[0]
            retn = Math.ceil(pages / products)
        }
    }
    return retn
}

function searchImage($, product) {
    return $(product).find('.product-tile img').attr('src');
}




var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.product-dlist2').eq(2).text()
    isInStock =instockCheck(contentarea,isInStock)
    return isInStock
}

var pPrice = function thePrice($, instock) {
    var pirce= $('.product-dlist2-price').text()
    pirce=getPrice(pirce)
    return pirce
}

var getName = function theName($, instock) {
    var name =$('#desktop-title').text()
    name=cleanName(name)
    return name

}

var image = function image($, instock) {
    return $('#pika-large').attr('src');
}
var Pagination = {

     items: function($) {
        var items = $('.product-holder li');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.product-holder li');
        items = items.length;

        var products = $('div.items-row span').text();
        products = products.match(/[0-9]+/g);
        products = products[2];
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('ul.breadcrumbs li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
