// sm-supplier-html/suppliers/www.fabric.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js"); 

module.exports = fabric =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.PageHeader').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var outOfStockMessage = $('div#ctlProductDetail_divAddForBackorder').text();
        if (outOfStockMessage && outOfStockMessage.toLowerCase().indexOf('item is on backorder') > -1) {
            inStock = false;
        }
        return inStock;
    },
    price: function($) {
        var price = $('div.DetailPriceReduced').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        return $('a#ctlProductDetail_hrefEnlargeImage img').attr('src');
    },
    description: function($) {
        var description = $('span#ctlProductDetail_lblDescription').text();
        if (description) {
            description = description.replace(/(\r|\n|\r\n)/gm, ' ').replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.catalog-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-item-desc').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.product-item-name div.bold').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        return $(product).find('div.product-item-desc a').attr('href');
    },
    image: function($, product) {
        return $(product).find('div.product-item-image img').attr('src');      
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.catalog-item').length;

        if(items) {
            return parseInt(items)   
        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = $("ul.pagination").text().match(/[0-9]+/g);

        if(pages) {
            return parseInt(pages.pop());
        } else {
            return null;
        }

     }
}

var Category = {
    name: function($) {
        var name = $('meta[name="com.silverpop.pagename"]').attr('content');
        name = cleanName(name)
        return name;
    }
}
