// sm-supplier-html/suppliers/www.petsmart.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.ws-product-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var inStock = true;
        var checkstock = $('div.pet-product-shopoption-field-disabled div.pet-product-shopoption-heading').html();
        (checkstock);

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('not sold online') > -1) {
                inStock = false;
            }
        }
        return inStock;

    },
    price: function ($) {
        var price = $('span.kor-product-sale-price-value').html();

        if (price.indexOf('to') > -1) {
            price = null;
        }

        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.ws-product-image').attr('src');
        return image;

    }
}

let Search = {
    container: ($) => {
        var container = $('ul.search-result-items li')
        return container;

    },
    name: function ($, product) {
        var name = $(product).find('div.product-name h3').html();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
       // ($(product).html())

        var price = $(product).find('div .product-price .price-sales').html();
        if(price === null) {
            price = $(product).find('div .product-price .price-regular').html();
        }
        
        if(price.indexOf('-') > -1) {
            price = null;
        }
        
        return getPrice(price);

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.petsmart.com' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;

    }
}
var Pagination = {

    items: function ($) {
        var items = $('li.grid-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('li.grid-tile');
        items = items.length;

        var products = $('div.results-hits').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.product-browse-hero-title h1').text();
        name = cleanName(name);
        return name;
    }
}
