// sm-supplier-html/suppliers/www.bbtoystore.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript":true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("-")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("-")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('meta[property="og:title"]').attr('content');
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var stockcount = $('div#js-inventory-message b').text();

        if(stockcount !== "") {
            stockcount = stockcount.match(/[0-9]+/g);

            if(stockcount < 5) {
                isInStock = false;
            }
        }

        return isInStock;
    },
    price: function($) {
        var price = $('div#js-price-value').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('span#js-main-image-zoom img').attr('src');
        return 'http://www.bbtoystore.com/mm5/' + image;
    },
    description: function($) {
        var description = $('div.product-information--description').text();

        if(description !== "") {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div#js-product-list .category-product')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('a strong').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http://www.bbtoystore.com/mm5/' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div#js-product-list .category-product').length;
        return parseInt(items);
     },
     pages: function($) {
        var pages = $('span.page-links-container').text();
        
        if(pages === ""){
            return null
        }else {
            pages = pages.match(/[0-9]+/g).pop();
        }
    
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name;
    }
}
