// sm-supplier-html/suppliers/www.watches2u.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        //info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.page_products_details5_top_h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var status = $('div.stockin').text();

        if (status.indexOf('In Stock') > -1) {
            status = true;
        } else {
            status = false;
        }

        return status;
    },
    price: function($) {
        var price;
        var status = $('div.stockin').text();

        if (status.indexOf('In Stock') > -1) {
            price = $('span.text_sale').text();
        }
            price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        
        return image;
    },
    /*
    description: function($) {
        var description = $('div.page_products_details5_top_description').text().replace(/\s+/g, " ").trim();

        return description;
    }
    */
}

var Search = {
    container: function($) {
        var product = $('div.xcomponent_products_medium');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('img.xcomponent_products_medium_image').attr('title');
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price;
        if ($(product).find('div.xcomponent_products_medium_price span.price').text()) {
            price = $(product).find('div.xcomponent_products_medium_price span.price').text();
        }
        if ($(product).find('div.xcomponent_products_medium_price span.sale').text()) {
            price = $(product).find('div.xcomponent_products_medium_price span.sale').text();
        }
        if ($(product).find('div.xcomponent_products_medium_price span.alternatives span').text()) {
            price = $(product).find('div.xcomponent_products_medium_price span.alternatives span').text();
        }

        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.xcomponent_products_medium_link').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.xcomponent_products_medium_image').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.page_search_results_subbar_right div').first().text();
         items = items.match(/[0-9]+/g)[1];
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('div.page_search_results_subbar_right div').first().text();
         items = items.match(/[0-9]+/g)[1];

         var products = $('div.page_search_results_subbar_right div').first().text();
         products = products.match(/[0-9]+/g).pop();

         var sum = products / items;

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div[itemprop*="breadcrumb"] span').text();
        name = cleanName(name);
        return name;
    }
}
