// sm-supplier-html/suppliers/www.wag.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('span#normalSinglePrice').text();
        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        }   

        return price;
    },
    image: function($) {
        var image = $('a.MagicZoomPlus').attr('href');
        image = 'http:' + image;

        return image;
    },
    description: function($) {
        //
        // May be many <p>
        //
        var description = $('#GreenAttributesWLP_026 > p:nth-child(3)').text().replace(/\u00A0/g, ' ');
        description = cleanName(description);

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product-box');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('span.item-name').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.base-price').text();
        if(price.indexOf('-') > -1) {
            price = null      
        } else {
            price = getPrice(price);
        }   

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.product-box-link').attr('href').replace(':443', '');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-image-inner img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = parseInt($('span.searched-stats').text().split(' ')[0].split('-')[1]);

         return items;
     },
     pages: function($) {
         var items = $('span.searched-stats').text().split(' ')[0].split('-')[1];
         var  products = $('span.searched-stats').text().split(' ').pop().replace(',', '');

         var sum = products / items;
         //(Math.ceil(sum));

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        return cleanName($("span.searched-term").text());
    }
}
