// sm-supplier-html/suppliers/www.shopko.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = shopko =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    }

}

var Product = {
    name: function($) {
        var name = $('h1.item-name').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isinstock = true;
        // var availablility = $('.shipInlineMessage').html();
        var stockcheck = $('.shipInlineMessage').html();
        isinstock = instockCheck (stockcheck,isinstock)
        // if (availablility !== null) {
        //     availablility = availablility.toLowerCase();
        //     if (availablility.indexOf('not available') > -1){
        //         isinstock = false;
        //     }
        // }

        // if (stockcheck !== null) {
        //     stockcheck = stockcheck.toLowerCase();
        //     if (stockcheck.indexOf('out of stock') > -1) {
        //         isinstock = false;
        //     }
        // }
        return isinstock;
    },





    price: function($) {
        var price = $('div.product-price-value').html();
        price = getPrice(price);
          return price;
    },
    image: function($) {
        var image = $('div.zoomProduct img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.catalog-entity-thumbnail article.ProductBody');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h1.product-title').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.product-price-value').text();
        price = getPrice(price)
        return price


    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        url = 'http://www.shopko.com' + url
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.product-thumb a img ').attr('src');
        return image;
    }
}
