// sm-supplier-html/suppliers/www.parts-express.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var inStockCheck=require('../../lib/inStockCheck')
module.exports = partsexpress =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').first().text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock = true
        var check=$('[itemprop="offers"]').text()
        inStock=inStockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('[itemprop="image"]').attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div.ProducDetailsNote').text();

        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC') + 3).substring(0, 12);
            return barcode;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.CatProdBox');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.CatItemName').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.CatPrice').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.parts-express.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.CatItemImage img').attr('src');
        return 'https://www.parts-express.com' + image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.CatProdBox');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.CatProdBox');
        items = items.length;

        var products = $('div.PaginationFooter div.AllResultBox').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.BreadCrumbLabel span span').text();
        name = cleanName(name);
        return name;
    }
}
