// sm-supplier-html/suppliers/www.ffx.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = ffx =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }


}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock =true
        return inStock
},
    price: function($) {
        var price = $('div.ToolPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.ImageArea img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div#option_top_1 p').html();
        barcode = barcode.split("<br>");
        return barcode.pop().match(/\d+/)[0]
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ToolPromotions a');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span').clone().children().remove().end().text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).attr('href');
        return 'http://www.ffx.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.section-top-offers div h1").text());
    }
}
var Pagination = {

     items: function($) {
        var items = cleanName($("div.ListingResults").text()).split("for")
            [0].match(/[0-9]+/g);

            if(items) {
                return parseInt(items[1]);
            } else {
                return null;
            }

     },
     pages: function($) {
/*        return Math.ceil(parseInt(cleanName($("div.ListingResults").text())
            .split("for")[0].match(/[0-9]+/g)[2]) / this.items($));*/


        var pages = cleanName($("div.ListingResults").text())
            .split("for")[0].match(/[0-9]+/g);

            if(pages) {

                pages = parseInt(pages[2]);

                return Math.ceil(pages / this.items($));
            } else {
                return null;
            }

     }
}
