// sm-supplier-html/suppliers/cpc.farnell.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js')
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require("../../lib/routine.js");

module.exports = cpcfarnell = {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('div.productPrice span[data-type="EXCL_INCL"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("img#productMainImage").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('tr.altRow');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('td.description a').attr('title')
        //    var partnum = $(product).find('p.wordBreak a').text();
        //    var item = $(product).find('td.description p:nth-child(3) a').attr('title');

        //    name = brand + ' ' + partnum + ' ' + item;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.qty_price_range').text();
        price = getPrice(price)
        price = price * 1.2

        //    if(price !== "" || price !== null) {
        //        price = price.match(/\(([^)]+)\)/)[0].replace("(", "").replace(")", "");
        //        price = getPrice(price);
        //    }
        return price.toString();
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('div.productThumbnailWrap img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('select#mySelectTop').text().match(/[0-9]+/g);
        if (items) {
            items = items.pop();
            return parseInt(items);
        } else {
            return null;
        }
    },
    pages: function ($) {
        var items = $('select#mySelectTop').text().match(/[0-9]+/g);

        if (items) {
            numitems = items.pop();
        } else {
            return null;
        }


        var products = $('span#titleProdCount').text();

        if (products) {
            products = products.match(/[0-9]+/g).pop();
            var sum = products / numitems;
            return Math.ceil(sum);
        } else {
            return null;
        }
    }
}

var Category = {
    name: function ($) {
        var name = $('div#prodResultHeader h1').text();
        name = cleanName(name);
        return name;
    }
}
