// sm-supplier-html/suppliers/www.fleetfarm.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = fleetfarm =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'

            if(info.price !== null) {
                result.push(info)
            }

        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('[itemprop="name"]').text();

        // if (name === null) {
        //     name = $('h1.item_title').html();
        // }
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true
        var checkstock = $('div.message').html();
        var outofstock = $('#inventory_out_message').attr('style');
        var check = $('[for="shipping-order"]').text()
        inStock =instockCheck(check,inStock)

        if(outofstock !== undefined) {
            if(outofstock.indexOf('display: none') < 0) {
                inStock = true;
            }
        } else {
            inStock = false
        }

        if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('available in select stores only') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('[itemprop="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('[itemprop="image"]').attr('src');
        image = 'https://www.fleetfarm.com' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-tile');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.product-name').text()

     //   if(name === undefined) {
    //        name = $(product).find('ul li.twoPaddingTop a.productListLink').text();
     //   }
        
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = null
        var salePrice = $(product).find('div.sale-price').text();
        salePrice = getPrice(salePrice);

        var regularPrice = $(product).find('div.regular-price').text();
        regularPrice = getPrice(regularPrice);

        if(salePrice == "") {
            price = regularPrice
        } else {
            price = salePrice
        }
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.fleetfarm.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://www.fleetfarm.com' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.category-items-per-page.dropdown ').text();
        
        items = items.match(/[0-9]+/g);
        items = items.shift()
        items = parseInt(items)

     //   if(items) {
      //      items = parseInt(items.pop());
     //   }
        
        return items;

     },
     pages: function($) {
        var pages = $('a.page-num').last().text();
        pages = pages.match(/[0-9]+/g);

        if(pages) {
            pages = parseInt(pages.pop());
        } else {
            pages = 1;
        }
        
        return pages;
     }
}


var Category = {
    name: function($) {
        var main = $('li.leftNavTitle').text();
        var sub = $('li.leftNavCategorySelected a').text();

        if(sub !== "") {
        name = main + ' - ' + sub
        } else {
            name = main
        }

        name = cleanName(name);
        return name;
    }
}
