// sm-supplier-html/suppliers/www.box.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = box =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }


}

var Product = {
    name: function($) {
        var name = $('.p-title span.p-title-desc').first().text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock = true
        var able = $('.p-stock').text()
        inStock = inStockCheck(able,inStock)
	return inStock;
},
    price: function($) {
        var price = $('.p-price p').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("td.p-image img").attr('src');
        return 'http://www.box.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-list div.product-list-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.p-list-sell').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.lazyimage').attr('src');
        return 'http://www.box.co.uk' + image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.override-wrapper h1.Meta-Title").text());
    }
}
var Pagination = {

     items: function($) {
        return parseInt($("div.pq-pagination-numbers p").text().match(/[0-9]+/g)[1]);

     },
     pages: function($) {
         var pages = Math.ceil($("div.pq-pagination-numbers p").text().match(/[0-9]+/g)[4] / this.items($));
         return pages


     }
}
