// sm-supplier-html/suppliers/www.worldstores.co.uk/logic.js
var PageInfo = require ('../../lib/pageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = getPrice($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
} 

var isStock = function inStock($) {
    var isInStock = true;
    var availability = $('.ws-availability').html() 
    if (availability) {
        if ((availability.indexOf('Weeks') > -1)
        || (availability.indexOf('Discontinued') > -1)
        || (availability.indexOf('Out of stock') > -1) 
        || (availability.indexOf('NOT AVAILABLE') > -1))

        {
            isInStock = false
        }       
    }
    return isInStock
}


var getPrice = function thePrice($) {
    var price = $('.ws_product_details_item_price_price').text()
    price = /£[0-9]+.[0-9][0-9]/.exec(price)
     if (price !== null){
        price = price[0].replace("£","").trim()
    }else{
        price = "N/A"
    }  
    return price
    (price)
}

var getName = function theName($) {   
    return $('.ws_item_name').text().trim()
}

var image = function image($) {
    return $('#imageMain').attr('src');
}



var Search = {
    container: function($) {
        var searchProducts = $('div.products-list-item-container');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h2.product_description a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.item-price').text().replace("From","").replace("£","").trim();
        //price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.worldstores.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.image a img').attr('src');
        image =  image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.products-list-item-container');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.products-list-item-container');
        items = items.length;

        var products = $('div.showing').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
