// sm-supplier-html/suppliers/www.bergners.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js')
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = bergners = {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("-") + 1, url.lastIndexOf("."))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("-") + 1, Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            // result.push(info)

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('a.product-name').first().attr('data-name')
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        let stock = true
        let check = $('.alertBox-message').text()
        stock = instockCheck(check,stock)
        return stock
    },
    price: function ($) {
        var price = $('.price--main').text()
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("li.productView-imageCarousel-main-item slick-slide a").attr('href');
        return 'https:' + image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h4.card-title a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.price--main').first().text();
        price = getPrice(price)

        if (price) {
            return price
            
        }else{
            price = $(product).find('#price').text();
            price = getPrice(price)
            return price
        }
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('h4.card-title a');

        url = 'https://www.bergners.com' + $(anchor).attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.card-img-container img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.pagination-row select').text();

        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g).shift();
        }

        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div.controls-module').html();

        return items
    }
}

var Category = {
    name: function ($) {
        var name = $('span.searchResultsString li.last').text();
        name = cleanName(name);
        return name;
    }
}
