// sm-supplier-html/suppliers/www.thetoyshop.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')
module.exports = beauty =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },
     'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
   
}

var Product = {
    name: function($) {
        var name = $('div.name').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
         var isInStock = true;
         var availability = $('.actions').html();
         var checkstock = $('div.prod_add_to_cart').html();
         //var deliveryAvailable = $('.deliveryOptionTop').first().text()
         let outofStock = $('.deliveryNotAvailable').text()
         let discontinued = $('.yCmsComponent').text()
         let outof = $('.outofStockText').text()
         isInStock = inStockCheck(outof,isInStock)
        //isInStock =inStockCheck(deliveryAvailable,isInStock)
         isInStock=inStockCheck(outofStock,isInStock)

        //  if (deliveryAvailable) {
        //     if (deliveryAvailable.indexOf('In Stock') > -1) {
        //         isInStock = true
        //     } else {
        //         isInStock = false
        //     }
        // }

        if (discontinued !== null) {
            if(discontinued.indexOf('Discontinued') > -1) {
                isInStock = false
            }
        }
    
        if (availability !== null) {
            if(availability.indexOf('Out of Stock') > -1) {
                isInStock = false
            }
        }

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if(checkstock.indexOf('discontinued') > -1) {
                isInStock = false
            }
        }

        return isInStock
    },
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.zoomImg img').attr('src');
        image = 'http://www.thetoyshop.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.thetoyshop.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https://www.thetoyshop.com' + image;
        return image;

    }
}




var Pagination = {

     items: function($) {
        var items = $('li.prod_list');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.prod_list');
        items = items.length;


        var products = $('ul.pagination li').eq("-2").text();
        products = products.match(/[0-9]+/g).pop();
       
        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.seoTag').text();
        name = cleanName(name);
        return name;
    }
}
