// sm-supplier-html/suppliers/www.artsupplywarehouse.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
let generateSku = require('../../lib/testPattern.js')
let regex = /([^=]+)\/?$/


module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'sku': (url) => {
        let sku = generateSku(url, regex)[0].match('[0-9]+')[0]
        return sku
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = generateSku(Search.url($, product), regex)[0].match('[0-9]+')[0]
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div#prodDetailTitle').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var preorder = null;
        //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('div#prodDetailSpecs').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('this item is currently out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = null
        var description = $('div#prodDetailDescription').text()
        var listPrice = description.match(/List\s*Price:\s*([\$0-9.]+)/)
        var ourPrice = description.match(/Our\s*Price:\s*([\$0-9.]+)/)
        if (ourPrice) {
            price = getPrice(ourPrice[1])
        } else if (listPrice) {
            price = getPrice(listPrice[1])
        }
        return price
    },
    image: function ($) {
        var image = $('img#prodDetailImg').attr('src');
        image = 'http://www.artsupplywarehouse.com/' + image
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('tr.productsListDisplay');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('td:nth-child(3)').text()
        name = cleanName(name)
        return name
    },
    price: function ($, product) {
        var price = $(product).find('td:nth-child(4)').find('span').remove().end().text()
        price = getPrice(price)
        return price
    },
    url: function ($, product) {
        var url = $(product).find('td:nth-child(3) a').attr('href')
        url = 'http://www.artsupplywarehouse.com/' + url
        return url
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src')
        image = 'http://www.artsupplywarehouse.com/' + image
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('.page-title.category-title').text()
        name = cleanName(name);
        return name;
    }
}
