// sm-supplier-html/suppliers/www.aliveanddirty.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = aliveanddirty =  {
    'useJavascript' : true,

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
   },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('header.product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.prices span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#product-featured-image').attr('src');
        return 'https:' + image;
    },
    description: function($) {
        var description = $('div#myTabContent  div#collapse-tab1');
        if (description) {
            return description.text().replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('li.list-product');
    },
    name: function($, product) {
        var brand = $(product).find('h3.list-product__vendor').text()
        var name = $(product).find('h4.list-product__title').text();
        name = cleanName(brand + ' ' + name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.money').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a:nth-child(4)').attr('href');
        return 'https://www.aliveanddirty.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.grid-item-border.no_crop_image.grid-item.product-item');
        return parseInt(items.length)
     },
     pages: function($) {
        var pages = $('ul.pagination-page').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb').html();
        name = name.split("<span>").pop();
        name = cleanName(name).replace("</span>", "");
        return name;
    }
}
