// sm-supplier-html/suppliers/www.yoyo.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript":"true",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            //result.push(info)
            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div .productTitle h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
       var inStock = true;
       var checkstock = $('div.discontinuedBanner').html();

       if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('shop for other ') > -1) {
                inStock = false
            }
       }
       return inStock;
    },
    price: function($) {
        var price = $('span.singlePrice').html();
        price = getPrice(price);
        return price

    },
    image: function($) {
        var image = $('img#pdpMainImageImg').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.listing div.product-box');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div .product-name span.item-name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find("span.q-product-price__current-price").text();

        return getPrice(price);


    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href').replace(':443','');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.q-product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.listing div.product-box');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.listing div.product-box');
        items = items.length;

        var result = null;
        var products = $('span.searched-stats').text();
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();    
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        

        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('div.left-nav-title span').text();
        name = cleanName(name);
        return name;
    }
}
