// sm-supplier-html/suppliers/www.guitarcenter.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('#headingLeft > div > h1 > div').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var coins = $('span.topAlignedPrice sup').text().match(/[0-9]+/g).pop();
        $('span.topAlignedPrice sup').remove()

        var pounds = $('span.topAlignedPrice').text().trim();

        var price = pounds + '.' + coins;
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('#productWidgets > div > div.span17 > div.hidden > a').attr('href');

        return image;
    },
    description: function($) {
        var description = $('section#product-overview').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.product-container');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.productTitle a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var coins = $(product).find('span.productPrice sup.decimals').text();
        $(product).find('span.productPrice sup.decimals').remove()
        var pounds = $(product).find('span.productPrice').text();

        var price = pounds + '.' + coins;
        price = getPrice(price);

        return price;

    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        url = 'http://www.guitarcenter.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('data-original')

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('select.displayNumber').first().text().match(/[0-9]+/g)

         if(items) {
            return parseInt(items.pop());
         } else {
            return null;
         }



     },
     pages: function($) {
         var items = $('select.displayNumber').first().text().match(/[0-9]+/g);
         var products = $('p.searchResultsUpdateContainer').first().text().match(/[0-9]+/g);

         if(items && products) {
            items = items.pop();
            products = products.pop();
            var pages = products / items;
            pages = Math.ceil(pages);

            return pages;

         } else {
            return null;
         }


     }
}

var Category = {
    name: function($) {
        var name = $('div#mainPageHeader h1').text();
        name = cleanName(name);

        return name;
    }
}
