// sm-supplier-html/suppliers/www.robertdyas.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.product-name').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true
        var availability = $('p.availability').text()
        inStock = instockCheck(availability,inStock)
        return inStock;

    },
    price: function($) {
        var price = $('span.special-price span.price').html();
        var price2 = $('span.regular-price').html();

        if(price === null) {
            price = price2;
        }    
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.product-image-gallery img#image-main').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.special-price span.price').html();
        var price2 = $(product).find('span.price').html();

        if(price === null) {
            price = price2;
        }

        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('li.item a img').attr('src');
        return image;
        
    }
}
var Pagination = {

     items: function($) {
        var items = $('li.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.item');
        items = items.length;

        var products = $('p.amount').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
