// sm-supplier-html/suppliers/www.cassart.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = cassart =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }/*,
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }*/


}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var stock = $('#product-stock').text()
        var inStock =true
        inStock =instockCheck(stock,inStock)
        return inStock
},
    price: function($) {
        var price = $("meta[property='og:price:amount']").attr('content');
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#product-image img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.prod');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h5 a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.blu-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h5 a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.cassart.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.im img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.content h1").text());
    }
}
/*var Pagination = {

     items: function($) {
        return "TestItems";

     },
     pages: function($) {
        return "TestPage";

     }
}
*/
