// sm-supplier-html/suppliers/www.carolwrightgifts.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = carolwrightgifts =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) {
            result.push(info)
             }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div[itemprop="name"] h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span#PRICE').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.prdimg').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div[itemprop="description"] p:nth-child(2)').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.galleryul_cat_container ul.galleryul_cat li').not('li.separator')
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('p.title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.price').not('p strike').html();

        if((price.indexOf('to') > -1)
        || (price.indexOf('Additionals') > -1)) {
            price = null;
        } else if (price.indexOf('strike') > -1) {
            price = price.substring(price.indexOf('....') + 6).substring(0, 4);
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('p.title a').attr('href');
        return url 
    },
    image: function($, product) {
        var image = $(product).find('div.prdctimg img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.galleryul_cat_container ul.galleryul_cat li').not('li.separator');
        return parseInt(items.length)
     },
     pages: function($) {
        var items = $('div.galleryul_cat_container ul.galleryul_cat li').not('li.separator');
        numitems = items.length;

        var products = $('div#breadcrumbs').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/\(([^)]+)\)/)[1];
            products = products.match(/[0-9]+/g).pop();

        }

        sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div#itemHeader h1').text();
        name = cleanName(name);
        return name;
    }
}
