// sm-supplier-html/suppliers/www.6pm.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockcheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var brand = $('#overview h1').text();


        brand = cleanName(brand);
        return brand;
    },
    inStock: function ($) {
        let stock = true
        let check = $('#buyBox>div>form>div>div').last().text()
        stock = instockcheck(check, stock)
        return stock
    },
    price: function ($) {
        var price = $('#productRecap>div>aside>div>div>div>span').first().text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('#detailImage img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.searchPage div article');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('p[itemprop="brand"]').text()
        var item = $(product).find('p[itemprop="name"]').text();
        name = brand + ' ' + item;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        let price = $(product).find('a').attr('aria-label')
        price = price.split('$')[1].split(' ')[0]
        //var price = $(product).find('a[itemprop="url"] div:last-child p:last-child span:first-child').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a[itemprop="url"]')
        if (anchor) {
            url = 'https://www.6pm.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {

        var items = $('#searchResults a');
        return parseInt(items.length);
    },
    pages: function ($) {
        var pages = $('div.bottom .pagination span.last').text();
        if (pages === "") {
            return null
        } else {
            pages = pages.match(/[0-9]+/g).pop();
            return parseInt(pages)
        }
    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();

        if (name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }

        if (name.indexOf(',') > -1) {
            name = name.split(',').shift();
        }

        name = cleanName(name);
        return name;
    }
}
