// sm-supplier-html/suppliers/www.sonicelectronix.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = sonicelectronix =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var isInStock = true;
        var avaliability = $('span.outOfStock').text();

        if(avaliability !== "") {
            isInStock = false;
        } 

        return isInStock;

    },
    price: function($) {
        var price = $('div.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#mainProductImage").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div.summaryContent').text();

        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC') + 8).substring(0, 12);
            return barcode.trim();
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productRow');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.productName').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.addProductToCart .price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.sonicelectronix.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.imageContainer img').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.productRow');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.productRow');
        items = items.length;

        var pages = $('div.bottomPaginationControl div.pagenav a.paginationItem').text();
        pages = pages.match(/\d/g).pop();
        return parseInt(pages) 
     }
}

var Category = {
    name: function($) {
        var name = $('div.categoryBlueBar h1').text();
        name = cleanName(name);
        return name;
    }
}
