// sm-supplier-html/suppliers/www.pharmaca.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = pharmaca =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        //info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.price-wrapper').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.fotorama__img').attr('src');
        return image;
    }
    /*
    description: function($) {
        var description = $('div.tab-content .std p').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
    */
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-item-info');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a.product-item-link').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.special-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.product-item-link').attr('href');
        return url
    },
    image: function($, product) {
        var image = $(product).find('img.product-image-photo').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.product-item-info');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.item');
        items = items.length;

        var products = $('p.amount').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('span.base').text();
        name = cleanName(name);
        return name;
    }
}
