// sm-supplier-html/suppliers/www.thebrilliantgiftshop.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("=")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop*="name"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('h3.productNowPrice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $("meta[property='og:image']").attr('content').replace('\r\n','').replace('\n','');
        (image);
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productPreviewWrapper');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('p.productTitle a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('p.productPrice').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.thebrilliantgiftshop.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.productPreviewImage img').attr('src');
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div.productPreviewWrapper');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.productPreviewWrapper');
        items = items.length;

        var products = $('span#numItems').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('nav#breadcrumbNav div span.breadcrumbcuelink h1').text();
        name = cleanName(name);
        return name;
    }
}
