// sm-supplier-html/suppliers/www.whsmith.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = whsmith =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var isInStock = true;
        var contentarea = $('#stock_status').html()
        if(contentarea !== null) {
            contentarea = contentarea.toLowerCase();
            if((contentarea.indexOf('weeks') > -1)
            || (contentarea.indexOf('only') > -1))
            {
                isInStock = false
            }
        }
        return isInStock
    },
    price: function($) {
        var price = $('.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#main_image").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('li.barcode').text();
        return barcode.match(/\d+/)[0];
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product_title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.product_image').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.products_per_page').html();
        
        if(items !== null) {
            items = items.match(/[0-9]+/g).pop()
            items = parseInt(items)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        
        return items;

     },
     pages: function($) {
         var items = $('div.products_per_page').html();
        
        if(items !== null) {
            items = items.match(/[0-9]+/g).pop()
            items = parseInt(items)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        var result = null;
        var products = $('div#category_summary p').text();

        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
 
        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
