// sm-supplier-html/suppliers/www.smiggle.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Smiggle = {
    'useJavaScript': true,
    'usePhantom': 'tesco.phantom.js',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var preorder = null;
        //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('div.stock-messaging-wrapper').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        if (preorder !== null) {
            preorder = preorder.toLowerCase();
            if (preorder.indexOf('pre-order') > -1);
            {
                inStock = false
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = $('div.product-price-view').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('.s7staticimage img').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.product__list>div.product.product__item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.product__name').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = null;
        var salePrice = $(product).find('span.product__price.product__price--sale').text()
        var standardPrice = $(product).find('.product__price').text();

        salePrice = getPrice(salePrice);
        standardPrice = getPrice(standardPrice);

        if (salePrice != "") {
            price = salePrice
        } else {
            price = standardPrice
        }

        return price;

    },
    url: function ($, product) {
        var url = $(product).find('a.product__item--description').attr('href');
        return url;
    },

    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.smiggle.co.uk' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div#breadcrumb ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
