// sm-supplier-html/suppliers/www.worldmarket.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var routine = require("../../lib/routine.js");


module.exports = worldMarket = {

    'sku': (url) => {
        url = url.split("?")[0]
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).split("?")[0]
            info.name = Search.name($, product);
            info.sku = sku.substring(sku.lastIndexOf("/") + 1, sku.lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    price: function ($) {
        var price = $('span.ml-item-price-now').html();
        price = getPrice(price);
        return price;
    },
    inStock: function ($) {
        var inStock = true;
        return inStock;
    },
    image: function ($) {
        var image = $('img.mainimage').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.ml-grid-view-item').not('div.ml-grid-view-item.ml-view-more-link-container');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.ml-thumb-name').text()
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.ml-item-price').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.worldmarket.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}


var Pagination = {

    items: function ($) {
        var items = $('select.results-per-page').text();

        if (items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        } else {
            if (Search.container($).length > 0) {
                items = 1;
            }
        }

        return items;

    },
    pages: function ($) {
        var items = $('select.results-per-page').text();

        if (items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        } else {
            if (Search.container($).length > 0) {
                items = 1;
            }
        }

        var result = null;
        var products = $('div.showing').text();

        if (products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }


        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.title-big-text').text();
        name = cleanName(name);
        return name;
    }
}
