// sm-supplier-html/suppliers/www.toysrus.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var routine = require ("../../lib/routine.js");


module.exports = toysrus =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('#deliveryBox').text()
        var lookForDelivery = $('li#deliveryBox div.container p').text()

        // toysrus remove out of stock items and replace them with empty pages
        // This checks if this is an example of those pages
        // "<h2>There are no items found</h2>"
        if(producttop !== null) {
            if (producttop.indexOf('Sorry, this item is out of stock for delivery') > -1) {
                isInStock = false
            }
        }

        if(lookForDelivery.indexOf('Not available for Delivery') > -1) {
            isInStock = false
        }


        return isInStock
    },
    price: function($) {
        var price = $('.sale').html();
        price = getPrice(price);
        if(price === null) {
            price = $('.price h3.price span[itemprop="price"]').attr('content')
        }
        return price;
    },
    image: function($) {
        var image = $('a#mainImage img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-card');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.title a').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.toysrus.co.uk' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}


var Pagination = {

     items: function($) {
        var items = $('select.results-per-page').text();

        if(items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        }else {
             if(Search.container($).length > 0) {
                items = 1;
            }
        }

        return items;

     },
     pages: function($) {
        var items = $('select.results-per-page').text();

        if(items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        }else {
             if(Search.container($).length > 0) {
                items = 1;
            }
        }

        var result = null;
        var products = $('div.showing').text();

        if(products !== ""){
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        }else{
            if(Search.container($).length > 0) {
                result = 1;
            }
        }


        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.title-big-text').text();
        name = cleanName(name);
        return name;
    }
}
