// sm-supplier-html/suppliers/www.staples.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = staples = {

    'sku': (url) => {
        url = url.split('?')[0]
        url = url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."))
        return url
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).split('?')[0]
            info.name = Search.name($, product)
            info.sku = sku.substring(sku.lastIndexOf("/") + 1, sku.lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var outofstock = $('span.OOSTextWrap').text();

        if (outofstock !== null) {
            if (outofstock.indexOf('out of stock') > -1) {
                isInStock = false
            }
        }
        return isInStock
    },
    price: function ($) {
        var price = $('span.priceupdate').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('#SkuPageMainImg').attr('src');
        image = 'https:' + image;
        return image;
    },
    description: function ($) {
        var description = $('div.productHighlights ul').text().trim();
        description = description.replace(/\s+/g, ' ');
        if (description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.detailsBlk');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3#skuName a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('h3#skuName a').attr('href');
        return 'https://www.staples.co.uk' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img#SkuPageMainImg').attr('src');
        return 'https:' + image;
    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();

        if (name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }

        name = cleanName(name);
        return name;
    }
}
