// sm-supplier-html/suppliers/www.ahalife.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {

    },
    inStock: function ($) {

    },
    price: function ($) {

    },
    image: function ($) {

    },
    url: function ($) {

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul#products li');  
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('p.productName.ng-binding').text();
        name = cleanName(name);
        return name
    },
    price: function ($, product) {

        var price = $(product).find('p.productPrice.ng-binding.ng-scope.sale-product-price').text();
        var standardPrice = $(product).find('p.productPrice').text();
        price = getPrice(price);
        standardPrice = getPrice(standardPrice);

        if(price == '') {
            price = standardPrice
        }



        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.ahalife.com' + $(anchor).attr('href');
        }
        return url;

    },

    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;

    }
}

var Pagination = {

    items: function ($) {
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
    }
}
