// sm-supplier-html/suppliers/www.amerimark.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1#item_name').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div#sold_out').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('sold out') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = null
        var label = $('p#item_price').text()
        if (label.indexOf(' to ') === -1) {
            price = getPrice(label)
        }
        return price;
    },
    image: function ($) {
        var image = $('img#main_image').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.grid_3');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2.itemTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = null
        var itemPrice1 = $(product).find('p.itemPrice').text()
        
        var itemPrice2 = $(product).find('div.itemWrapper p:not([class])').text()
        if (itemPrice2) {
            price = getPrice(itemPrice2)
        } else if (itemPrice1 && itemPrice1.indexOf(' to ') === -1) {
            price = getPrice(itemPrice1)
        }
        return price
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.itemLink');
        if (anchor) {
            url = 'https://www.amerimark.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.itemImage').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.thumbHeaderLarge').text()
        name = cleanName(name);
        return name;
    }
}
