// sm-supplier-html/suppliers/www.cultbeauty.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = cultbeauty =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var brand = $('div.productBrandTitle').text();
        var name = $('div.productTitle').text()

        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var availabilty = $('div.header')

        var isInstock= true
        isInstock=inStockCheck(availabilty,isInstock)
        if (availabilty){
            
                isInstock=false
            
        }
	return isInstock;
},
    price: function($) {
        var price = $('span.productPrice.js-product-price ').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img.js-big-image").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productGridItem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        var brand = $(product).find('div.productGridBrandTitle span').text();

        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.productGridPrice').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.productGridItemInner');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        //('this is my url             '+ url);
        return 'http://www.cultbeauty.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.productGridImg').attr('src');

        if(image.indexOf('https://') < 0) {
            image = $(product).find('img.productGridImg').attr('data-src');
        }
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div.heading h1").text());
    }
}
