// sm-supplier-html/suppliers/www.rcwilley.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        // info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#prodName').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('div.grid-full.box.warn').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price1 = $('div#price span.price strong').text();
        var price2 = $('div#price span.price span').text();
        var price = price1 + '.' + price2;
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#mainImage').attr('src');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('#description-tab h2').text();
        description = description.replace(/(\r|\n|\r\n)/gm, ' ').replace(/\s+/, ' ').trim();
        return description;
    }
}

var Search = {
    container: function($) {
        return $('div#resultsContainer a.product');
    },
    name: function($, product) {
        var name = $(product).find('span.name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price1 = $(product).find('span.price strong').text();
        var price2 = $(product).find('span.price span').text();
        var price = price1 + '.' + price2;
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).attr('href');
        return 'http://www.rcwilley.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.image').attr('src');
        return 'http:' + image;
    }
}

var Category = {
    name: function($) {
        var name = $('div#breadCrumbs a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
var Pagination = {

     items: function($) {
        var items = $('div#resultsContainer a.product');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
    

        var pages = $('div.resultPages a').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);

     
     }
}
