// sm-supplier-html/suppliers/www.payless.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text()
		name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var check =$('.availability-msg').text()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('.NONRangePriceSale').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#primary-image').attr('src');
        return image;
    },
    description: function($) {
        $('div.product-tabs #tab1 a.print-page').remove();
        var description = cleanName($('div.product-tabs #tab1').text().trim());
        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product-tile');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a.name-link').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product-price').eq(10).text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.name-link').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-image img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = parseInt($('h1#results-products').text().replace(')', '').split('(').pop());

         return items;
     },
     pages: function($) {
         var items = $('div.results-hits').text().split('\n')[3].split(' ').shift();
         var products = $('h1#results-products').text().replace(')', '').split('(').pop();

         var sum = products / items;
         //(Math.ceil(pagesCount));

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
