// sm-supplier-html/suppliers/www.crutchfield.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var contentarea = $('.prodBuyBox').html()
        if(contentarea !== null) {
            if((contentarea.indexOf('Out of stock') > -1)
            || (contentarea.indexOf('Not in Stock') > -1)
            || (contentarea.indexOf('Low Stock') > -1)
            || (contentarea.indexOf('Email me when available') > -1)
            || (contentarea.indexOf('Ship date unknown') > -1)
            || (contentarea.indexOf('Where\'s the price?') > -1)
            || (contentarea.indexOf('This item is no longer available.') > -1)) {
                isInStock = false
            }
        }
        return isInStock
    },
    price: function($) {
        var price = $('.priceInfo').find('.priceNow').text().replace("$","").replace(",","").trim()
        price = /([0-9]+.[0-9][0-9])/.exec(price)
        if (price !== null){
            price = price[0]
        } else {
            price = ""
        }
        return price
    },
    image: function($) {
        var image2 = $('a.productMedium img.mainProductPhoto').attr('src');

        if(image2.indexOf('http:') > -1) {
            return image2;
        } else {
            image2 = 'http:' + image2; 
        }

        
        return image2;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.list-unstyled li.list');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.prodTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('.prodImageContainer a').attr('href');
        return 'http://www.crutchfield.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.prodImage').attr('src') ;
        return 'http:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.list-unstyled li.list');
        return parseInt(items.length);
     },
     pages: function($) {
        var items = $('ul.list-unstyled li.list');
        numitems = items.length;

        var result = null;
        var products = $('span.showingSome').text();
        if(products !== ""){
             products = products.match(/[0-9]+/g);
             result = Math.ceil(products / numitems); 
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        return result
     }
}

var Category = {
    name: function($) {
        var name = $('h1.headline').text();
        name = cleanName(name);
        return name;
    }
}
