// sm-supplier-html/suppliers/www.biglots.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = biglots =  {
    "useJavascript" : true,
    "usePhantom" : "biglots.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("?"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("?"))
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.product-name h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockMessage = $('.out-of-stock').html();
        if(stockMessage) {
            stockMessage  = stockMessage.trim()
        }

        if(stockMessage === 'This item is currently out of stock.') {
            inStock  = false;
        }

        return inStock;
    },
    price: function($) {
        var price =  $('.regular-price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.viewer-thumb-image').attr('src');
        image = 'http:' + image;
        return  image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.product-grid>li');

        if(!searchProducts) {
            searchProducts = [];
        }

        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-name').text()
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.regular-price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.biglots.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');
        image = 'http:' + image;
        return  image;
    }
}

var Pagination = {

     items: function($) {
        return $('.product-grid>li').length
     },
     pages: function($) {
        var numitems = $('.product-grid li').length;

        var products = $('.item-range-top').frist().text();
        var result = null;
        
        // if(products !== ""){
        //     products = products.match(/[0-9]+/g).pop()  
        //     result = Math.ceil(products / numitems)
        // }else {
        //    if(Search.container($).length > 0) {
        //         result = 1;
        //     }
        // }
        return products
     }
}


var Category = {
    name: function($) {
        var name = $('div.title-bar h1.title').text();
        name = cleanName(name);
        return name;
    }
}
