// sm-supplier-html/suppliers/www.marksandspencer.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = marksAndSpencer =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.product-title').text();
        name = cleanName(name);
        return name
    },
    inStock: function($) {
        var isInStock = true;
        var outofstock = $('div.information span.off-sale-msg').text();

        if(outofstock !== null) {
            if(outofstock.indexOf('no longer available') > -1) {
                isInStock = false;
            }
        }
        return isInStock
    },
    price: function($) {
         var price = $('.price1').text();
         price = getPrice(price);
         return price;
    },
    image: function($) {
        var image = $('meta[property="og:image"]').attr('content');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('p[itemprop="description"]').text();

        if(description !== "") {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-wrap');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.body2').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price1').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('.detail a').attr('href');
        return 'http://www.marksandspencer.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.model-view').attr('src');
        return 'http:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.product-wrap');
        return parseInt(items.length);
     },
     pages: function($) {
        var items = $('li.product-wrap');
        numitems = items.length;

        var products = $('div#topcontrol .item-count span.count').text();
        products = products.match(/[0-9]+/g).pop();
        
        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1 span#landingPage').text();
        name = cleanName(name);

        if(name !== "") {
            name = name.replace(/\w\S*/g, function(txt){return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();});
        }
        return name;
    }
}
