// x4-model/influx/index.js
const Influx = require('influx');

class SaveToInflux {

    constructor(payload) {
        this.host = payload.host
        this.database = payload.database
        this.measurement = payload.measurement
        this.data = payload.data ? payload.data : []
        this.schema = payload.schema
    
        this.influx = new Influx.InfluxDB({
            host: payload.host,
            database: payload.database,
            schema: payload.schema,
            measurement: payload.measurement 
        })
    }

    add(data) {        
        data.measurement = this.measurement
        this.data.push(data)
    }

    save(next) {
        let measurement = this.measurement
        this.influx.writePoints(this.data, {
            measurement: measurement        
        }).then(() => {
            this.data = []
            next()
        }).catch(err => {
            console.error('Error saving data to InfluxDB! ', err)
            this.data = []
            next(err)
        })
    }

} 

module.exports = SaveToInflux
