// sm-supplier-html/suppliers/www.abt.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    //    'sku': (url) => {
    //        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    //    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            //            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            if (info.name != "") {
                result.push(info)
            }
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        let name = $('h1').text()
        name = cleanName(name)
        return name
    },
    inStock: function ($) {

        let availability = $('#product_top_container').html()

        var isInStock = false

        if (availability) {
            if (availability.indexOf('In Stock') > -1) {
                isInStock = true
            }
        }
        return isInStock

    },
    price: function ($) {
        let price = $('#price').attr('content')
            price = getPrice(price)
            return price

    },
    image: function ($) {
        let image = $('#prod_thumbs img').attr('src')
        return image

    },
    url: function ($) {

    }
}

var Search = {
    container: function ($) {
        let container = $('li')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('div.cl_title').text()
        name = cleanName(name)
        return name

    },
    price: function ($, product) {
        let price = $(product).find('div.deal_price:not(.regular_price)').text()

        if (price) {
            price = price.substring(price.lastIndexOf("$") + 1, price.lastIndexOf(""))
        }
        price = getPrice(price)
        return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('img').attr('src')
        return image

    }
}

var Pagination = {

    items: function ($) {},
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {}
}
