// sm-supplier-html/suppliers/www.merrell.com/logic.js
 var ProductPageInfo = require ('../../lib/productPageInfo.js');
 var SearchPageInfo = require ('../../lib/searchPageInfo.js');
 var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
 var CategoryInfo = require ('../../lib/CategoryInfo.js');
 var cleanName = require('../../lib/cleanName.js');
 var getPrice = require('../../lib/getPrice.js');
 
module.exports = meller =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) {
            result.push(info)
        }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}
 
 var Product = {
     name: function($) {
        var name = $('h1.product-name-v2').text();
         name = cleanName(name);
         return name;
     },
     inStock: function($) {
        var isInStock = true;
        return isInStock;
     },
     price: function($) {
         var price = $('span.price-sales').text();
         price = getPrice(price);
         return price;
     },
     image: function($) {
        var image = $('meta[property="og:image"]').attr('content');
         return image;
     },
     description: function($) {
        var description = $('meta[property="og:description"]').attr('content');
        
        if(description !== "") {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
     }
 }
 
 var Search = {
     container: function($) {
       var searchProducts = $('li.grid-tile');
        return searchProducts;
     },
     name: function($, product) {
          var name = $(product).find('a.name-link').text();
         name = cleanName(name);
         return name;
     },
     price: function($, product) {
         var price = $(product).find('span.product-sales-price').text();
 
        if(price == "") {
            price = null
        }
        price = getPrice(price);
         return price;
     },
     url: function($, product) {
         var url = $(product).find('a.name-link').attr('href');
         return url;
     },
     image: function($, product) {
         var image = $(product).find('a.thumb-link img').attr('src');
         return image;
     }

}

var Category = {
    name: function($) {
        var name = $('div.wrap-top-search h1').text();
        name = cleanName(name);
        return name;
    }
}
