// sm-supplier-html/suppliers/untitled.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockcheck = require('../../lib/inStockCheck.js')

module.exports = Boscovs =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    }

}

var Product = {
    name: function($) {
        var name = $('').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock = true
        var availabilty = $('').text()
        inStock = instockcheck(availabilty,inStock)
	return inStock
},
    price: function($) {
        var price = $('').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find ('').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = '' + $(anchor).attr('href');
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}
