// sm-supplier-html/suppliers/www.fila.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = fila =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.product-add-to-cart');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div#product-content-inner div.product-price').text();
        price = price.match(/[0-9.,]+/gm).pop();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.product-primary-image img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.product-bottom-details');
        if (description) {
            return description.text().replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('li.grid-tile');
    },
    name: function($, product) {
        var name = $(product).find('a.name-link').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.product-pricing').text();
        price = price.match(/[0-9.,]+/gm).pop();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');   
        return image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }
