// sm-supplier-html/suppliers/shop.mercola.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockcheck = require('../../lib/inStockCheck.js')
module.exports = mercola =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("ct/")+3,url.lastIndexOf("/")-2)
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("ct/")+3,Search.url($, product).lastIndexOf("/")-2);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#productname').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        let stock=true
        let check= $('a.notifyme').text()
        stock = instockcheck(check,stock)
	return stock;
},
    price: function($) {
        var price = $('li.yourprice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#prodmainimage img').attr('src');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('div.description').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.productitem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.TruncatedPrdName').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('li.prodprice .price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('h3.prodname a').attr('href');
        return url;

    },
    image: function($, product) {
        var image = $(product).find('li.prodimage img').attr('src');
        return 'http:' + image;   
    }
}

var Pagination = {

     items: function($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
