// sm-supplier-html/suppliers/www.tkmaxx.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockcheck = require('../../lib/inStockCheck.js')
module.exports = TkMaxx = {
    'useJavaScript': true,
    'usePhantom': 'tkmaxx.phantom.js',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = null
        var brand = $('div.product-details p').text();
        var title = $('div.product-details h1').text();
        name = cleanName(brand + ' ' + title);
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var preorder = null;
        
        //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('div.stock-messaging-wrapper').text();
        var check = $('.tjx-body').text()
        inStock = instockcheck(check,inStock)

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        if (preorder !== null) {
            preorder = preorder.toLowerCase();
            if (preorder.indexOf('pre-order') > -1);
            {
                inStock = false
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = $('div.product-price h2').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.thumb_ img').attr('src');
        return 'https://www.tkmaxx.com' + image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.c-product-grid__item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = null;
        var brand = $(product).find('strong.u-spaced-2').text();
        var title = $(product).find('span.c-product-card__label').text();
        name = cleanName(brand + ' ' + title);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('strong').last().text();
        var clearPrice = $(product).find('.u-colour-xi').text()
        
        price = getPrice(price);
        if(price){
            return price;
        }else{
            clearPrice = getPrice(clearPrice)
            return clearPrice
        }
        

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('.c-product-card');
        if (anchor) {
            url = 'https://www.tkmaxx.com' + $(anchor).attr('href');
        }
        return url;
    },

    image: function ($, product) {
        var image = $(product).find('img.c-product-card__thumbnail').attr('src');
        return 'https://www.tkmaxx.com' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div#breadcrumb ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
