// sm-supplier-html/suppliers/www.harborfreight.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = harborfreight =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.product-information-title h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('div.sale').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.thumbnails-inner a img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.product-name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.sale').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-image img').attr('src');
        return image;

    }
}

var Pagination = {
    items: function($) {
        var items = $('div.limiter select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
    },
    pages: function($) {
        var items = $('div.limiter select').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.pager p.amount').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function($) {
        var name = $('h1.category-title').text();
        name = cleanName(name);
        return name;
    }
}
