// sm-supplier-html/suppliers/www.marisota.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.stockPriceWrap').text();
        if (stockMessage) {
            if (stockMessage.toLowerCase().indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('h3.productNowPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.rsMainSlideImage').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div#productDescription').text();
        if (description) {
            description = description.replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('div.productPreviewWrapper');
    },
    name: function($, product) {
        var name = $(product).find('p.productTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.productPrice').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('p.productTitle a').attr('href');
        return 'http://www.marisota.co.uk' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.productPreviewImage img').attr('src'); 
        return image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('select.rppSelect').text();
        products = products.match(/\d+/gm).pop();
        return parseInt(products);
     },
     pages: function($) {
        var products = $('select.rppSelect').text();
        products = products.match(/\d+/gm).pop();

        var total = $('span#numItems').text();
        var sum = total / products;

        return Math.ceil(sum);
     }
}


var Category = {
    name: function($) {
        var name = $('div#copy h2').text();
        name = cleanName(name);
        return name;
    }
}
