// sm-supplier-html/suppliers/www.viatrading.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.responsive-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('.details_info_box .unit_price:nth-child(6) .price_value').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('.item_image').attr('src');
        image = 'http://www.viatrading.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProduct = $('div .product_wrapper ');
        return searchProduct;

    },
    name: function($, product) {
        var name = $(product).find('div .product_name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price =$(product).find('div .unit_price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.viatrading.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.center-block').attr('src');
        image = 'http://www.viatrading.com' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div .product_wrapper');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div .product_wrapper');
        items = items.length;

        var products = $('div.pageShowing').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('a.collapsed').text();
        name = cleanName(name);
        return name;
    }
}
