// sm-supplier-html/suppliers/shop.1sale.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck=require('../../lib/inStockCheck')
var cleanUrl =require('../../lib/cleanUrl')
module.exports = sale =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock =true
        var check=$('[itemprop="availability"]').html()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('#ProductPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $(".product-single__image").attr('srcset');
        image= cleanUrl(image)
        return image;
    }
}

var Search = {
    container: function($) {
        return $('.on-sale');

    },
    name: function($, product) {
        var name= $(product).find(".product-grid-title").text();
        name=cleanName(name)
        return name
    },
    price: function($, product) {
        var price = $(product).find('.grid-link__title span').last().text();
        price = getPrice(price);
       // (price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.grid-link');
        if(anchor) {
            url = "https://shop.1sale.com" + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.product__img').attr('srcset');
        image = cleanUrl(image)
        return 'https:'+image;
    }
}

var Pagination = {

     items: function($) {
        var items = Search.container($).length;
        if(items) {
            return items;
        } else {
            return null;
        }
     },
     pages: function($) {
        if(this.items($)) {
            return 1;
        } else {
            return null;
        }
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('|') > -1) {
            name = name.split('|').pop();
        }

        name = cleanName(name);

        //(name);
        return name;
    }
}
