// sm-supplier-html/suppliers/www.fatbraintoys.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = fatbraintoys = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.product-label').first().text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var inStock = true
        var stockcheck = $('div.iparProdRow').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock
    },
    price: function ($) {
        var price = $('span.product-price').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img[name=main_image]').attr('src');
        return 'https:' + image;
    },
    description: function ($) {
        var description = $('div.tab-content div.col-xs-12.col-sm-9 span').text();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r|)/gm, '').replace(/\s+/gm, ' ');
        }
    }
}

var Search = {
    container: function ($) {
        var container = $('ul.product-list-row-fbt li.item')
        return container;
    },
    name: function ($, product) {
        var name = $(product).find('h3.product-catalog-label').html();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('h3.iparPrice').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find("a:contains('More Info')").attr('href');
        url = 'https://www.fatbraintoys.com' + url;
        return url
    },
    image: function ($, product) {
        var image = ''
        var imageElem = $(product).find('div.image-product-list img');
        if (imageElem.length) {
            image = imageElem.attr('data-original')
        }
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('table.iparProdList');
        return parseInt(items.length);
    },
    pages: function ($) {

        var items = $('table.iparProdList');
        numitems = items.length;

        var products = $('div.listing_header').text();
        products = products.match(/[0-9]+/g);
        products = products[2];

        sum = products / numitems;


        return Math.ceil(sum);

    }
}

var Category = {
    name: function ($) {
        var name = $('div.product-list-head h1').text();
        name = cleanName(name);
        return name;
    }
}
