// sm-supplier-html/suppliers/global.oup.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = OxfordUniversityPress =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("?"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.barcode = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("?"));  
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("?"));            
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    // 'pagination' : function ($, next) {

    //     var info = new PaginationNumberInfo()
    //     info.items = Pagination.items($);
    //     info.pages = Pagination.pages($);

    //     return info
    // },
}

var Product = {
    name: function($) {
        var name = $('h1.product_title').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('.content_block .left_column .text_colour_orange').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.content_block .product_image_shadow img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('.content_block.full_width.enclosed .inset_text_block').text();
        
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|)/gm,"").replace(/(\s+)/gm," ").trim();
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.content_block.search_results table tr');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3 a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p').last().text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('h3 a').attr('href');
        return 'https://global.oup.com' + url;
    },
    image: function($, product) {
        var image =  $(product).find('img').attr('src');
        return image;
    }
}

// var Pagination = {
//      items: function($) {
//         return 10;
//      },
//      pages: function($) {
//         var numberOfPages = '';
//         var text = $('div.content_block.search_results select#start_top option:nth-child(1)').text();
//         var matches = /page\s+\d+\s+of\s+(\d+)/i.exec(text);
//         if (matches) {
//             numberOfPages = parseInt(matches[1]);
//         }
//         (numberOfPages);
//         return numberOfPages;
//      }
// }
