// sm-supplier-html/suppliers/www.flexxmemory.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = flexxmemory = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var isInStock = true;
        var stockcheck = $('span.VariationProductInventory').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                isInStock = false
            }
        }
        return isInStock;
    },
    price: function ($) {
        var price = $('span.ProductDetailsPriceIncTax').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("img.slide-image").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h4.card-title').text();
        if (name.indexOf('...') > -1) {
            name = name.replace("...", "");
        }
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.price-section.price-section--withTax span.price').not('span.price.price--rrp').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul.product-list li');
        var pages = $('ul.paging-list').html();

        if (items && pages) {
            return parseInt(items.length);
        } else {
            return null;
        }




    },
    pages: function ($) {
        var pages = $('ul.paging-list').html();
        var items = $('ul.product-list li');
        if (pages && items) {

            pages = pages.match(/[0-9]+/g);

            if (pages) {
                pages = pages.pop();

                return parseInt(pages);
            } else {
                return null;
            }

        } else {
            return null;
        }

    }
}

var Category = {
    name: function ($) {
        var name = $('h1.category-heading').text();
        name = cleanName(name);
        return name;
    }
}
