// sm-supplier-html/suppliers/www.superdrug.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"
        //(info)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $(".pdp__productName").text();
        if(name){
            name = cleanName(name);
        }
        return name;
    },
    inStock: function($) {    
        var inStock = $('#productDetailAddToBasket').attr('value');
        if(inStock === 'Add to basket'){
            inStock = true;
        } else {
            inStock = false;
        }
        return inStock;
    },
    price: function($) {
        var price = $(".pricing__now").text();
        if(price){
            price = getPrice(price);
        }
        return price;
    },
    image: function($) {
        var image = $('.pdp-gallery__zoom img').attr('src');
        image = 'https://www.superdrug.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        return $(".item__content");
    },
    name: function($, product) {
        var name = $(product).find(".item__productName").text();
        if(name){
            name = cleanName(name);
        }
        return name;
    },
    price: function($, product) {
        var price = $(product).find(".item__price--now-only").text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price=$(product).find('.item__price--now').text()
            price=getPrice(price)
            return price
        }
        
    },
    url: function($, product) {
        return "https://www.superdrug.com" + $(product).find(".item__text a").attr("href");
    },
    image: function($, product) {
        var image = $(product).find('.item__image img').attr('src');
        image = 'https://www.superdrug.com' + image;
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('.productgroup > li');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.productgroup > li');
        items = items.length;

        var products = $('div.data-display-view-all span').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name;
    }
}
