// sm-supplier-html/suppliers/www.gamestop.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": true,
    "usePhantom": 'gamestop.phantom.js',
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.country = 'us';
            info.image = Search.image($, product);
            info.condition = Search.condition($, product);

            if (info.price !== null && info.condition.indexOf('NEW') > -1) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.ats-prod-title').text();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var inStock = true;
        var checkstock = $('div.buttonna a').text();
        var preorder = $('li.ats-prodBuy-inventory').text();
        var preowned = $('strong.ats-prodBuy-condition').html();

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('not available') > -1) {
                inStock = false;
            }
        }

        if (preorder !== null) {
            preorder = preorder.toLowerCase();
            if (preorder.indexOf('release date') > -1) {
                inStock = false;
            }
        }

        if (preowned !== null) {
            preowned = preowned.toLowerCase();
            if (preowned.indexOf('pre-owned') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('h3.ats-prodBuy-price').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.boxart img').attr('src');
        image = 'www.gamestop.com' + image;
        return image;
    }

}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('.ats-product-title a').text();
        
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        // jQuery.fn.justtext = function() {

        //     return $(this).clone()
        //             .children()
        //             .remove()
        //             .end()
        //             .text();
        
        // };
        
        // var price = $(product).find('.pricing.ats-product-price').justtext();
        // var condition = $(product).find('h4.ats-product-condition').html();

        // if (condition !== null) {
        //     condition = condition.toLowerCase()
        //     if (condition.indexOf('pre-owned') > -1) {
        //         price = null
        //     }
        // }
        var price = $(product).find('.ats-product-price').clone().children().remove().end().text()

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('.ats-product-title a');

        if (anchor) {
            anchor = 'https://www.gamestop.com' + $(anchor).attr('href');
        }
        return anchor;

    },
    condition: function ($, product) {

        let condition = $(product).find('h4.ats-product-condition').text()
        return condition;
    },

    image: function ($, product) {

        var image = $(product).find('img').first().attr('src');
        return 'https://www.gamestop.com' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.result_count_first .result_set').text();

        if (items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items)
        } else {
            if (Search.container($).length > 0) {
                items = 1;
            }
        }

        return items;
    },
    pages: function ($) {
        var items = $('div.result_count_first .result_set').text();

        if (items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items)
        } else {
            if (Search.container($).length > 0) {
                items = 1;
            }
        }

        var products = $('h3[role="presentation"]').text();
        var result = null;

        if (products !== "") {
            products = products.split('of').pop();

            if (products.indexOf(",") > -1) {
                products = products.replace(",", "");

            }
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items);
        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('title').text();
        if (name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
