// sm-supplier-html/suppliers/www.natchezss.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = natchezss =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        info.description = Product.description($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
     'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#image").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $("table.data-table").text();

        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC') + 16).substring(0, 12);
            return barcode;
        }
    },
    description: function($) {
        var description = $('div.std').text().trim();
            if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
        
        
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select.styled-select').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
        var items = $('select.styled-select').text(); 
        numitems = items.match(/[0-9]+/g).pop();


        var products = $('p.amount').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems
        return Math.ceil(sum)
     }
}
var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
