// sm-supplier-html/suppliers/www.all-battery.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var availbility=$('span.prod-detail-cost-value').text()
        var inStock=true
        if(availbility.indexOf('Soon') > -1){
            inStock=false
        }
        return inStock;

    },
    price: function($) {
        price = $('span.prod-detail-cost-value').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div#gallery-main a img').attr('src');
        return 'http://www.all-battery.com'+image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-list-item');
        return searchProducts;
    },

    name: function($, product) {
        var name = $(product).find('h5').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var red_price=$(product).find('span.product-list-cost-value').text()
        red_price = getPrice(red_price);

        return red_price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('h5 a');
        if(anchor) {
            url = 'http://www.all-battery.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.product-list-img').attr('src');
        return 'http://www.all-battery.com'+image
    }
}

var Pagination = {

    items: function($) {
        var items = $('div.product-list-item')
        return parseInt(items.length);

     },
     pages: function($) {
        var pages = $('ul.pagination').text();
        if(pages === ""){
            return null
        }else {
           pages = pages.match(/[0-9]+/g).pop(); 
        }
        if(parseInt(pages)===null){
            return 0
        }else{
            return parseInt(pages)
        }
        
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
