// sm-supplier-html/suppliers/www.base.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var routine = require ("../../lib/routine.js"); 
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = base =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    "usePhantom": "base.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }


}

var Product = {
    name: function($) {
        var name = $('.product-main-title h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var isInStock = true;

        var availability = $('.stock-message').html() 
        var availability2 = $('.price-info').last().text()
        isInStock = instockCheck(availability2,isInStock)

        if (availability) {
            if ((availability.indexOf('Pre-order') > -1)
            || (availability.indexOf('Unavailable') > -1)
            || (availability.indexOf('Out of stock') > -1))
            {
                isInStock = false
            }      
        }

        return isInStock
    },
    price: function($) {
        var price = $('.price-info .price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#mainImage").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.cell');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price a').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.title a');
        if(anchor) {
            url = 'http://www.base.com' + $(anchor).attr('href');
        }

        return url;

    },

    image: function($, product) {
        var image = $(product).find('div.image img').attr('src');

        return 'https://www.base.com' + image;
    }
}

var Category = {
    name: function($) {
        return cleanName($("div.page-header h1").text());
    }
}
var Pagination = {

     items: function($) {
        return Search.container($).length;
     },
     pages: function($) {
        return routine.matchNumbersAndPop($, "div.plist-centered div[style]");
     }
}
