// sm-supplier-html/suppliers/www.shopgadgetsandgizmos.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    }

}

var Product = {
    name: function($) {

    },
    inStock: function($) {

    },
    price: function($) {

    },
    image: function($) {

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('table:nth-child(3) td:nth-child(1) div');
        //var searchProducts = $('table:nth-child(3) div');
        (searchProducts.html());
        return searchProducts;

    },
    name: function($, product) {

    },
    price: function($, product) {

    },
    url: function($, product) {

    },
    image: function($, product) {

    }
}
