// sm-supplier-html/suppliers/www.herbergers.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div[itemprop="name"]').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('span#salePrice_displayPriceString').first().text();

        if (price == '') {
            price = $('#price_priceString').text();
        }
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div.s7thumb').css('background-image').replace('url("', '').replace('")', '').replace('wid=75', 'wid=500').replace('hei=75', 'hei=500');

        return image;
    },
    description: function ($) {
        var des = $('div.widget-content-desc').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('li.productListingsItem');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.productDesc.productDesc-breakWordWrap > a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('#salePrice').text();

        if (price == '') {
            price = $(product).find('#price').text();
        }
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.productDesc.productDesc-breakWordWrap > a').attr('href');
        url = 'http://www.herbergers.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.productImage').attr('src');
        image = 'https:' + image

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.countDiv.pagination-module > select > option').last().text();
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('div.countDiv.pagination-module > select > option').last().text();
        var products = $('#searchMsg').text().trim().match(/[0-9]+/g).pop();

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name)

        return name;
    }
}
