// sm-supplier-html/suppliers/www.gandermountain.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = gandermountain =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var saleprice = $('div.saleprice').text();
        var price = $('div.price span.regPrice').text();


        if(saleprice !== "") {
            price = saleprice;
        }

        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#mainProductImg").attr('src');
        return 'http:' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.browse-products');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('p#short').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price-point').text();
        var saleprice = $(product).find('span.red-message').text();

        if(saleprice !== "") {
            price = saleprice;
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'www.gandermountain.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img.product-image').attr('src');

        if(image.indexOf('|') > -1) {
            image = image.substring(0, image.indexOf('|'))
        }

        return 'http:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('p.per-page').html();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
        var items = $('p.per-page').html(); 
        numitems = items.match(/[0-9]+/g).pop();


        var products = $('p.page-numbers').text();
        products = products.substring(products.indexOf('of') + 3)
        products = products.match(/[0-9]+/g).shift();

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();
        
        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }

        if(name.indexOf('for Sale') > -1) {
            name = name.replace("for Sale", "");
        }

        name = cleanName(name);
        return name;
    }
}
