// sm-supplier-html/suppliers/www.cpooutlets.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.sku=Product.sku($)
        return info
    },
/*
    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },
*/
    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.sku = Search.sku($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.productname.pnameText').text();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var avilibility=$('div.addtocart button.addtocartbutton.imagebutton')
        var instock=false
        if(avilibility){
            instock=true
        }
	return instock;

    },
    price: function ($) {
        var price = $('div.price div.salesprice.value').text();
        if(price){
            price=price.replace(/\s/g,'')
            price=getPrice(price)
        }else{
            price = $('div.price div.standardprice.retail.value').text()
            price=price.replace(/\s/g,'')
            price=getPrice(price)
        }
        return price
    },
    image: function ($) {
        var image = $('div.productimage ul:first-child img').attr('src');
        return 'https://www.cpooutlets.com' + image;

    },
    sku: function ($) {
        var skunumber=$('span#product-id').text()
        skunumber=skunumber.replace(/\s/g,'')
        return skunumber

    }
}

var Search = {
    container: function ($) {
        var product = $('div.product.producttile.cmpr_listing');
        return product;

    },
    name: function ($, product) {
        var brand = $(product).find('div.name a').attr('title');
        brand = cleanName(brand);
        return brand;
    },
    price: function ($, product) {
        var price = $(product).find('div.price div.salesprice').text();
        if(price){
            price=getPrice(price)
        }else{
            price=$(product).find('div.price div.standardprice').text()
            price = getPrice(price);
        }
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('div.name a').attr('href');
        return url;

    },

    image: function ($, product) {
        var image = $(product).find('div.image img').attr('src');
        return 'https://www.cpooutlets.com' + image;

    },
    sku: function($,product){
        var skunumber=$(product).find('div.prodid').text()
        skunumber=skunumber.replace(/\s/g,'')
        return skunumber
    }
}

var Pagination = {

    items: function ($) {
        var items=$('ul.dropdown-menu li:first-child').text()
        items=items.replace(/\D/g,'')
        return items;

    },
    pages: function ($) {        
    var products = $('div.products-found span.badge').text().match(/[0-9]+/g).pop();
    var items = $('ul.dropdown-menu li:first-child').text().replace(/\D/g,'');
    var pages = products / Number(items);
    pages = Math.ceil(pages);
    return pages.toString();

    }
}

var Category = {
    name: function ($) {
        var name = $('div#catHeading h1').text();
        ('this is my name'+name)
        name = cleanName(name);
        return name
    }
}
