// sm-supplier-html/suppliers/www.activerideshop.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop=name]').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div.product-info__availability').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = $('meta[itemprop=price]').attr('content');
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = null
        var container = $('div.zoomWindowContainer div.zoomWindow').first().attr('style')
        if(container){
            var parts = container.match(/url\("(.+?)"\)/) 
            if (parts) {
                image = 'https:' + parts[1]
            }
            return image
        }else{
            return null
        }
        
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product.product--zoom');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2.product_title').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.money').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('h2.product_title a');
        if (anchor) {
            url = 'https://www.activerideshop.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.product__inside__image img').attr('src')
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.title-under').first().text()
        name = cleanName(name);
        return name;
    }
}
