// sm-supplier-html/suppliers/www.monsterpetsupplies.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = pcworldbusiness = {
    "useJavascript": true,
    "usePhantom": "pcworldBusiness",

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
            
        });

        return result
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop="name"]').text()
        name = cleanName(name);
        
        return name;

    },
    inStock: function ($) {
        var inStock= true
        var stock =$('.monster-product-desc-details-inline').text()
        inStock = instockCheck(stock,inStock)
        return inStock;

    },
    price: function ($) {
        var price = $('p.varient-price').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var img = $('img').attr('src');
        return img;

    },
    description: function ($) {
        var description = $('meta[name="description"]').attr('content');

        if (description !== "") {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.col-md-4.col-sm-6.product-list-item');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('a.product-title').text();
        name = cleanName(name)

        return name;

    },
    price: function ($, product) {
        var price = $(product).find('a.product-price').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.monsterpetsupplies.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var img = $(product).find('img').attr('src');
        return 'https:' +  img;

    }
}

var Category = {
    name: function ($) {
        var name = $('div.categoryGray h1').text();
        name = cleanName(name);
        return name;
    }
}
