// sm-supplier-html/suppliers/www.ajmadison.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'sku': function (url) {
        var thisPattern = new RegExp(/^([0-9A-Z]{6,15}(?![\;\?\&\.\,])*$)/g);
        var osku = url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."));
        var sku = thisPattern.test(osku) ? osku : null;
        return sku;
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('.table-cell .red').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('.mz-figure img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.py1.mb1.divided.border-gray-4');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('div.inline-block.bold').text();
        brand = cleanName(brand);
        var name = $(product).find('div.block.mr2').text();
        name = cleanName(name);
        return brand + ' ' + name;
    },
    price: function ($, product) {
        var price = $(product).find('td.py05.nowrap.right-align').text()
        price = getPrice(price)
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href')
        url = 'https://www.ajmadison.com' + url;
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div#layoutWrapperContent td.t9').text();
        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g).pop();
        }


        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div#layoutWrapperContent td.t9').text();

        if (items === "") {
            return null
        } else {
            numitems = items.match(/[0-9]+/g).pop();
        }


        var products = $('td.t8').text();

        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }



        sum = products / numitems
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name
    }
}
