// sm-supplier-html/suppliers/www.waterstones.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.barcode = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('#scope_book_title').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        var status = $('div.book-actions > div > div > div > div.perk-container > ul > li.feature-bad').text();

        if (status != '') {
            status = false;
        } else {
            status = true;
        }
        return status;
    },
    price: function ($) {
        var price;
        var status = $('div.book-actions > div > div > div > div.perk-container > ul > li.feature-bad').text();

        if (status != '') {
            price = null;
        } else {
            price = $('div.price > div > b').last().text();
        }
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('#scope_book_image').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('#scope_book_description').first().text().replace(/\s+/g, " ").trim();

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.book-preview');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('a>img').attr('alt')
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'https://www.waterstones.com' + url

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        //
        // there is no elemnt with items count on page
        //

        var items = $('div.book-preview').length;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        //
        // there is no element on page with all products count
        //

        var pages = $('div.pagination.span12 > div > span').last().text().match(/[0-9]+/g).pop();
        pages = JSON.parse(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.heading-pagination > div.input-wrapper > h1').text().trim();
        name = cleanName(name)

        return name;
    }
}
