// sm-supplier-html/suppliers/jet.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = jet = {
    "useJavascript": true,
    "usePhantom": "jet.phantom.js",


    // 'sku': function (url) {
    //     var thisPattern = new RegExp(/^([0-9a-z]{32}(?![\;\?\&\.\,])*$)/g);
    //     var osku = url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""));
    //     var sku = thisPattern.test(osku) ? osku : null;
    //     return sku;
    // },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($);
        info.inStock = Product.inStock($);
        info.price = Product.price($);
        info.image = Product.image($);
        info.country = 'us'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)

            // if (info.price !== null) {
            //     result.push(info);
            // }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1 span.krhuHS').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var isInStock = true;
        var availability = $('span.eBcoCg').text();

        if (availability !== "") {
            if (availability.indexOf('This item is unavailable right now') > -1) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function ($) {
        var price = $('span.bcoGcE').text();
        price = getPrice(price);
        return price ? price : null;
    },
    image: function ($) {
        var image = $("img.fMzbJG").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('#__next>div[display="flex"]>div>div>div>div>div[display="flex"]>div>a[aria-label]')
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product)
        var brand
        if (name) {
            name = $(name).attr('aria-label')
            brand = name.split(',')[0]
            brand = brand.split(':')[1]
            name = name.split(',')[1]
            name = name.split(":")[1]
            name = cleanName(brand) + ' ' + CleanName(name)
        }
        name = cleanName(name);
        if (name) {
            return name;
        } else {
            name = $(product).find('span[font-size="2"]').first().text()
            name = cleanName(name)
            return name
        }

    },
    price: function ($, product) {
        var price = $(product)
        var tmpPrice
        if (price) {
            price = $(price).attr('aria-label')
            tmpPrice = price.split(',')[3]
            tmpPrice = getPrice(tmpPrice)
            if (tmpPrice) {
                return tmpPrice
            } else {
                var scondtem
                price.split(',').map((element, index) => {
                    if (element.indexOf('effective price') > -1) {
                        scondtem = price[index]
                    }
                })
                tmpPrice = getPrice(scondtem)
                if (tmpPrice) {
                    return tmpPrice
                } else {
                    tmpPrice = getPrice(scondtem + '.00')
                    if (tmpPrice) {
                        return tmpPrice
                    } else {
                        tmpPrice = $(product).find('[font-weight="bold"]').last().text()
                        tmpPrice=getPrice(tmpPrice)
                        return tmpPrice
                    }
                }

            }
        }

        // price = getPrice(price+'.00')
        // if(price){
        //     return price
        // }else{
        //     price = $(product).find('[font-weight="bold"]').last().text()
        //     price=getPrice(price)

        //     return price
        // }

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product);
        if (anchor) {
            url = $(anchor).attr('href');
        }
        if (url.indexOf('undefined') > -1) {

        }
        return 'https://jet.com' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}


var Pagination = {

    items: function ($) {
        var items = $('ul.list-grid li');
        items = items.length;

        if (!items) {
            return null;
        } else {
            return parseInt(items);
        }


    },
    pages: function ($) {
        var pages = $('div.pagination ul').text();
        var result = null;

        if (pages !== "") {
            pages = pages.match(/[0-9,]+/g).pop();
            result = parseInt(pages);
        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.module-filters h1').text();
        name = cleanName(name);
        return name;
    }
}
