// sm-supplier-html/suppliers/www.wexphotovideo.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = wexphotographic =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("-")+1,url.lastIndexOf("/"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("-")+1,Search.url($, product).lastIndexOf("/"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.producttitle').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var stock = $('.stock').text()
        var inStock =true
        inStock=instockCheck(stock,inStock)
        return inStock
},
    price: function($) {
        var price = $('div#productsummary #ctl00_pnlMainContent_price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#mainimage_large img').attr('src').replace("..", "");
        return 'www.wexphotoandvideo.com'+ image;
    },
    description: function($) {
        var description = $('p.pshort').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.listing-product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-excerpt h3 a').text().replace('Find out more about the ','');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.wexphotovideo.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http://www.wexphotovideo.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul#productList li');
        return parseInt(items.length)
     },
     pages: function($) {
        var items = $('ul#productList li');
        numitems = items.length;

        var products = $('div#matching h3').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / numitems;
        return Math.ceil(sum);

     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('-') > -1) {
            name = name.split("-");
            name = name.shift();
        }

        name = cleanName(name);
        return name;
    }
}
