// sm-supplier-html/suppliers/www.vitacost.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript": "true",
    "usePhantom": "vitacost.phantom.js",

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $("h1[itemprop='name']").text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.pOurPriceM').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('#wrap img').attr('src');
        image = 'http://www.vitacost.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.productWrapper li.product-block');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.description a').html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.pOurPriceM').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.description a');
        if(anchor) {
            url = 'http://www.vitacost.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.pb-image img').attr('src');
        image = 'http://www.vitacost.com' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('.productWrapper li.product-block'); 
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
       return 1;
     }
}

var Category = {
    name: function($) {
        var name = $('h3.bcs span.bcNoArw').text();
        name = cleanName(name);
        return name;
    }
}
