// sm-supplier-html/suppliers/www.maplin.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = maplin =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.description = Product.description($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div .product-summary h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var checkstock = $('ul.stock-status').html();
        var discontinued = $('.add_to_cart_form').html();

        if(checkstock !== null) {
            if(checkstock.indexOf('icon-remove-sign') > -1)
             {
                inStock = false;
            }

        }

        if (discontinued !== null) {
            discontinued = discontinued.toLowerCase()
            if (discontinued.indexOf('discontinued') > -1) {
                inStock = false
            }
        }

    return inStock;
    },
    price: function($) {
        var price = $('p.new-price ').html();
            if(price !== null) {
            price = price.replace('<span>','').replace('</span>','')
            }
        price = getPrice(price)
        return price;

    },
    image: function($) {
        var image = $('div.product-images img').attr('src');
        return image
        
    },
    description: function($) {
        var description = $('div.productDescription').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.lister-details.mb.row');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h4.product-description').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.price-tag').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.maplin.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http:' + image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('select[name="productsPerPage"]').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
     },
     pages: function($) {
        var items = $('select[name="productsPerPage"]').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div.list-summary p').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum - 1);
     }
}

var Category = {
    name: function($) {
        var name = $('div.rightcol-3-4 h1').text();
        name = cleanName(name);
        return name;
    }
}
