// sm-supplier-html/suppliers/www.origins.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = origins =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if(name !== null) {
            result.push(info)
        }
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.product-full__info .product-heading').first().text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.product-sku-price__value').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.product-full__carousel__slide img").attr('src');
        return 'www.origins.co.uk' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-grid__item');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('h2').text();
        var item = $(product).find('h3').text();
        name = brand + ' ' + item;
        name = cleanName(name);

        if(name === "") {
            name = null
        }
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.product-sku-price__value').last().text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.origins.co.uk' + url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-brief__img-link img').attr('src');
        return 'https://www.origins.co.uk' + image;
    }
}

var Pagination = {

     items: function($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h2.content-header__title').text();
        name = cleanName(name);
        return name;
    }
}
