// sm-supplier-html/suppliers/www.garden.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {

    },
    inStock: function ($) {

    },
    price: function ($) {

    },
    image: function ($) {

    },
    url: function ($) {

    }
}

var Search = {
    container: function ($) {
        let container = $('div#product-list-pro')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('.product-list-name').text()
        name = cleanName(name)
        return name

    },
    price: function ($, product) {
        let price = $(product).find('.special-price').text()
        let standardPrice = $(product).find('.price').text()
        price = getPrice(price)
        standardPrice = getPrice(standardPrice)
        
        if(price == "") {
            price = standardPrice
        }
        return price

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return url

    },

    image: function ($, product) {
        let image = $(product).find('img').attr('src')
        return image

    }
}

var Pagination = {

    items: function ($) {},
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {}
}
