// sm-supplier-html/suppliers/www.boxed.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = boxed = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
            
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        // var name1 = $('h2').first().text();
        var name=$('h1').first().text()
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        var inStock= true
        var check=$('h2').text()
        inStock=instockCheck(check,inStock)
        return inStock
    },
    price: function ($) {
        var price = $('[data-bx="pdp-details-price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div.product-gallery-wrapper img').attr('src');
        image = 'https:' + image;

        return image
    },
    description: function ($) {
        var description = $('div.details p').text().replace(/(\r\n|\n|\r)/gm, "");

        return description;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.g-product-list li > div');

        return searchProducts;
    },
    name: function ($, product) {
        //  var name = $(product).find('.g-brand-text').text()
        let name = $(product).find('a').attr('href');
            name = name.split('/')[3].replace(/-/g,' ')
            name = cleanName(name)

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'https://www.boxed.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('figure img').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('span.results-count').text();
        if (items === "") {
            return null
        } else {
            items = items.match(/[0-9]+/g).pop();
            items = JSON.parse(items);
        }
        return items;
    },
    pages: function ($) {
        //
        // one page. products loads by ajax
        //
        var items = $('span.results-count').text();
        if (items === "") {
            return null
        } else {
            numitems = items.match(/[0-9]+/g).pop();
            numitems = JSON.parse(numitems);

        }


        var products = $('span.results-count').text();
        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }
        sum = products / numitems;
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
