// sm-supplier-html/suppliers/www.skinstore.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },
/*
    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },
*/
    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            //info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.product-title-wrap h1.product-title').text();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var avilibility=$('p.availability span.product-stock-message').text()
        var instock=false
        if(avilibility.indexOf('In stock')>-1){
            instock=true
        }
	return instock;

    },
    price: function ($) {
        var price = $('section.productoptions.product-detail-block p.product-price').text();
        price=getPrice(price)
        return price
    },
    image: function ($) {
        var image = $('div.product-image.main-product-image img').attr('src');
        return image;

    },
}

var Search = {
    container: function ($) {
        var product = $('.productListProducts_product');
        return product;

    },
    name: function ($, product) {
        var brand = $(product).find('h2').text();
        brand = cleanName(brand);
        return brand;

    },
    price: function ($, product) {
        var price = $(product).find('.productBlock_priceValue').text();
        price=getPrice(price)
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.skinstore.com' + url;

    },

    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;

    }
}

var Pagination = {

    items: function ($) {
        var items=$('span.js-enhanced-ecommerce-data').last().attr('data-product-position')
        return items;

    },
    pages: function ($) {
        var products = $('div.listTitle p.listTitle_results').text().replace(/\D/g,'');
        var items=$('span.js-enhanced-ecommerce-data').last().attr('data-product-position');
        var pages = Number(products) / Number(items);
        pages = Math.ceil(pages);
        return pages.toString();
        }
}

var Category = {
    name: function ($) {
        var name = $('div.listTitle h1.listTitle_text').text();
        name = cleanName(name);
        return name

    }
}
