// sm-supplier-html/suppliers/www.lookfantastic.com/logic.js
var PageInfo = require ('../../lib/pageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product' : function($, next) {
        var info  = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = price($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var isStock = function inStock($) {
    var isInStock = true;
    var producttop = $('.availability').html()

    if(producttop !== null) {
        if ((producttop.indexOf('Sold out') > -1)) {  
            isInStock = false
        }
    }

    return isInStock
}


var price = function thePrice($) {
    var price=$('.productPrice_price').text()
    price=getPrice(price)
    return price
}

var getName = function theName($) {
    return $('h1').text().trim()
}

var image = function image($) {
    return $('img.product-img').attr('src');
}

var Search = {
    container: function($) {
        var searchProducts = $('.productBlock');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('.productBlock_productName').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('.productBlock_priceValue').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.productBlock_link').first();
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.lookfantastic.com'+url;

    },
    image: function($, product) {
        var image = $(product).find('.productBlock_imageContainer>img').attr('src');
        image = image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.item_listRow div.item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.item_listRow div.item');
        items = items.length; 

        var products = $('p.listTitle_results').text();
        products = products.match(/[0-9]+/g).pop();
        

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
