// sm-supplier-html/suppliers/bluemercury.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var inStockCheck =require('../../lib/inStockCheck')
module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var brand = $('.brand-reviews>h2').text()
        var name =$('.h1').last().text()
        name = cleanName(brand+' '+name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('span#AddToCartText').first().text();
        inStock=inStockCheck(stockcheck,inStock)
        return inStock;
    },
    price: function ($) {
        var price = $('[itemprop="price"]').last().text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('div.product-image.slick-current.slick-active img.small--hide').attr('src');
        image = 'https:' + image
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.ajax-product-list.product-wrap div.product-container');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h6').text();
        var name1 =$(product).find('h5').text()
        name = cleanName(name+' '+name1);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('p.price').text();
        price= getPrice(price)
        if(price){
            return price
        }else{
            price = $(product).find('.price').text()
            price= getPrice(price)
            return price
        }
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://bluemercury.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('span.product-image-wrap img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1#main-title').text()
        name = cleanName(name);
        return name;
    }
}
