// sm-supplier-html/suppliers/www.walgreens.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var addCheckDigit = require('../../lib/barcode.js').addCheckDigit;
var isValid = require('../../lib/barcode.js').isValid;
var instockCheck = require('../../lib/inStockCheck.js')


module.exports = walgreens = {
    "useJavascript": true,
    "usePhantom": "walgreens.phantom.js",

    'sku': (url) => {
        var sku = "" + url.substring(url.lastIndexOf("=") + 1, url.lastIndexOf("-"))
        sku = "" + sku.replace('prod', '').trim()
        return sku
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {

            var info = new SearchPageInfo()
            var sku = "" + Search.url($, product).substring(Search.url($, product).lastIndexOf("=") + 1, Search.url($, product).lastIndexOf("-"));
            info.barcode = null; //formatBarcode(Search.barcode($, product));
            info.name = Search.name($, product);
            info.sku = "" + sku.replace('prod', '').trim()
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)

        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1#productName').text().replace("1.0ea", "");
        name = cleanName(name);


        return name;

    },
    inStock: function ($) {
        var inStock = true;
        //var checkstock = $('article.wag-rec-outofstock').html();
        var checknotOnline = $('.product-add-list-cont').text()
        inStock = instockCheck(checknotOnline, inStock)
        // if(checknotOnline !== null) {
        //     checknotOnline = checknotOnline.toLowerCase()
        //     if(checknotOnline.indexOf('not sold online' > -1)) {
        //         inStock= false
        //     }
        // }

        // if (checkstock !== null) {
        //     checkstock = checkstock.toLowerCase();
        //     if (checkstock.indexOf('temporarily out of stock') > -1) {
        //         inStock = false
        //     }
        // }
        return inStock;
    },
    price: function ($) {
        var price = $('span.product__price span').text();
        var doc = $('span.product__price sup:last-child').text()
        price = price + '.' + doc
        price = getPrice(price);
        if (price) {
            return price;
        } else {
            price = $('h2#regular-price span').text()
            doc = $('h2#regular-price sup:last-child').text()
            price = price + '.' + doc
            price = getPrice(price);
            return price
        }

    },
    image: function ($) {
        var image = $('img#productImage-gettab').attr('src');
        if (image) {
            if (image.indexOf('http:') > -1) {
                return image;

            } else {
                image = 'http:' + image;
            }
        } else {
            image = $('img#productImg').attr('src')
            if (image) {
                if (image.indexOf('http:') > -1) {
                    return image;

                } else {
                    image = 'http:' + image;
                }
            }

        }



        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.wag-product-cards');
        return searchProducts;

    },
    name: function ($, product) {
        var name = "" + $(product).find('a[name="product-title"]').attr('title') //.replace("(1 ea)","");
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        let dotprice = $(product).find('.wag-price-black sup').last().text().trim()
        let price = $(product).find('.wag-price-black span').text().trim()
        price = getPrice(price + '.' + dotprice)
        if(price){
            return price
        }else{
            dotprice = $(product).find('.product__price>sup').last().text()
            price = $(product).find('.product__price').last().clone().children().remove().end().text()
            if(price){
                price = getPrice(price +'.'+dotprice)
                return price
            }else{
                dotprice=$(product).find('.product__price>div>sup').last().text()
                price=$(product).find('.product__price>div').last().clone().children().remove().end().text()
                price=getPrice(price+'.'+dotprice)
                return price
            }
        }
        // let salePrice = $(product).find('.wag-price-black sup').text()
        return price

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.walgreens.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        //    if(image.indexOf('http:') > -1) {
        //        return image;

        //    } else {
        //       image = 'http:' + image;
        //    }
        return 'https:' + image;
    },
    barcode: function ($, product) {

        //    var barcode = $(product).find('span.sr-only.ng-binding').text()//.replace("(1 ea)","");
        //    barcode = cleanName(barcode).split(' ').pop().trim();
        //    barcode = barcode.slice(-12)
        //     return null;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('.wag-product-cards');
        if (items.length) {
            items = items.length;
            return parseInt(items);
        } else {
            return items;
        }


    },
    pages: function ($) {
        var items = $('article.wag-product-card-width_b');
        items = items.length;

        var products = $('#resultcount').text();
        if (products) {
            products = products.match(/[0-9]+/g);
            if (products) {
                products = products[0];
                return products
            } else {
                return 0;
            }
        } else {
            return 0;
        }
    }
}

var Category = {
    name: function ($) {
        var name = $('.h1__page-title').text();
        name = cleanName(name);
        return name;
    }
}
