// sm-supplier-html/suppliers/www.overclockers.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = overclockers =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var preorder = null;
        //   var preorder = $('form.add-to-basket-view').html();
        var stockcheck = $('span.delivery_container').first().text();
        inStock=instockCheck(stockcheck,inStock)
        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.article_details_price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("table.zoom1_container img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('meta[itemprop="gtin13"]').attr('content');
        barcode = barcode.split(":")[0];
        return barcode;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ck_listing div.artbox');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('span.ProductTitle').text()
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product_img img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.artbox');
        return parseInt(items.length);
     },
     pages: function($) {
        var pages = $('div.pagination .display_sites').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div#breadcrumb b').text();
        name = cleanName(name);
        return name;
    }
}
