// sm-supplier-html/suppliers/www.beauty.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = beauty =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("?"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("?"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    }

}

var Product = {
    name: function($) {
        var name = $('div.prod h1:nth-child(3)').text().replace('NEW!','');
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('.prodprice p span').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("meta[property='og:image']").attr('content');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.itemGridbty');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.info a.oesLink').attr('title');
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find ('#priceTd span').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.beauty.com' + $(anchor).attr('href');
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('.prodImg img').attr('src');
        return image;
    }
}
