// sm-supplier-html/suppliers/www.steinmart.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
     'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.ml-product-name').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span[itemprop*="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.mainimage').attr('src')
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ml-grid-view-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.ml-thumb-name a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.ml-item-price').html();
        var salePrice = $(product).find('span.ml-item-price.ml-message-price-now').html();
        price = getPrice(price);
        salePrice = getPrice(salePrice)

         if(salePrice == "") {
             price = price
         } else {
             price = salePrice
         }

         if(price){
             return price
         }else{
             price = $(product).find('.ml-item-price').text()
             price = getPrice(price)
             return price
         }

        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.steinmart.com' + $(anchor).attr('href');
        }
        return url;


    },
    image: function($, product) {
        var image = $(product).find('div.ml-thumb-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.ml-dir-header-category-head-text h1').text();
        name = cleanName(name);
        return name;
    }
}
