// sm-supplier-html/suppliers/www.c21stores.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.product-details__info.hover-zoom__container > hgroup > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('div.product-details__price-group > div.product-details__price-row > p > span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div > div.product-details__primary-image.slick-slide.slick-current.slick-active > a > img').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('div.product-details__description-body').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('.tile-box.tile-element')

        return product;
    },
    name: function ($, product) {
        var brand = $(product).find('[itemprop="brand"]').text();
        var name=$(product).find('[itemprop="name"]').text()
        name = cleanName(brand+' '+name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.price>span.list').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('[itemprop="url"]').attr('href');

        return 'https://www.c21stores.com'+url;
    },
    image: function ($, product) {
        var image = $(product).find('[itemprop="image"]').attr('src');

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul > li.pagination__node.pagination__node--showing').first().text()
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g)[1];
            items = JSON.parse(items);
        }
    
        return items;
    },
    pages: function ($) {
        var items = $('ul > li.pagination__node.pagination__node--showing').first().text()
        if(items === ""){
            return null
        }else {
          items = items.match(/[0-9]+/g)[1];  
        }

        
        var products = $('ul > li.pagination__node.pagination__node--showing').first().text()
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();
        }

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#main > div.view > h1').text();
        name = cleanName(name)

        return name;
    }
}
