// sm-supplier-html/suppliers/www.katom.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".html"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".html"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock=true
        var check=$('p.days').text()
        inStock=instockCheck(check,inStock)
	return inStock
},
    price: function($) {
        var price = $('.price-display').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.zoomPad img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.row .product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product a').attr('title');
        name = cleanName(name);
        return name;
    },  
    price: function($, product) {
        var price = $(product).find('.kPrice').html();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price = $(product).find('.aPrice').text()
            price =getPrice(price)
            return price
        }
        
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.img img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('.row .product').length;
        return parseInt(items);
     },
     pages: function($) {

        var products = $('span.results').first().text();
        if(products){
            products = products.match(/[0-9]+/g).pop();
        }else{
            products=0
        }
        
        return products;

        
     },
}

var Category = {
    name: function($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name;
    }
}
