// sm-supplier-html/suppliers/www.worldmusicsupply.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = thinkgeek =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
            
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        return cleanName($('h1').html())
    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('span.out_stock').html();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        return getPrice($('strong.redtext').html());
    },
    image: function($) {
        var image = $('img').attr('src');
        return image
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.caption');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('strong').html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = getPrice($(product).find('span.pricenew').html());
      
        return price;
    },
    url: function($, product) {
        return $(product).find('a').attr('href');
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = parseInt($('div .contentblob .product').length);

        if(items) {
            return items
        } else {
            return null;
        }
        
     },
     pages: function($) {
        var pages = $('div.pagenav ul.list-clean').text();
        var result = null
        
        if(pages !== ""){
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages);
        }else {
            if(Search.container($).length > 0){
                result = 1
            }
        }
        return result

     }
}

var Category = {
    name: function($) {
        var name = $('h1.cat_title').text();
        name = cleanName(name);
        return name;
    }
}
