// sm-supplier-html/suppliers/www.northerntool.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = northerntool =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.title').text();
        name = name.replace(/\s+/gm, ' ').trim();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div#description-text');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('item is on backorder with the manuafacturer') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.col div.sale-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#main-image').attr('src');
        return 'http://www.northerntool.com' + image;
    },
    description: function($) {
        var description = $('div#description-text');
        if (description) {
            return description.text().replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('div.prod-listing');
    },
    name: function($, product) {
        var name = $(product).find('div.title').text();
        name = cleanName(name);
        name = name.replace('FREE SHIPPING — ','');
        return name;
    },
    price: function($, product) {
        let price;

        let salePrice = $(product).find('div.sale-price-med').text()
            salePrice = getPrice(salePrice)

        let onlyPrice = $(product).find('div.only-price-med').text()
            onlyPrice = getPrice(onlyPrice)

        if(salePrice == '') {
            price = onlyPrice 
        } else {
            price = salePrice
        }
        
       return price;
    },
    url: function($, product) {
        var url = $(product).find('div.title a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.thumb').attr('src');
        return 'http://www.northerntool.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('div#items-per-page').text();
        products = products.match(/\d+/gm).pop();
        return parseInt(products);
     },
     pages: function($) {
        var products = $('div#items-per-page').text();
        products = products.match(/\d+/gm).pop();

        var total = $('div.nt-header div.total').text();
        total = total.match(/\d+/gm).pop();

        var pages = total / products;
        return Math.ceil(pages);
     }
}
var Category = {
    name: function($) {
        var name = $('div.heading h1').text();
        name = cleanName(name);
        return name;
    }
}
