// sm-supplier-html/suppliers/www.rockymountaintrail.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.detailsH1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span#lblOurPriceRange').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#Thumbnails .imgProduct a').attr('data-large');
        return 'www.rockymountaintrail.com' + image;
    },
    description: function($) {
        var description = $('p.DetailsDescription').text().trim();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.tdWrapper');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('p.productBrandName a.productsNameLink').attr('title');
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.productsPrice').text();
        var sale = $(product).find('span.productsPrice .productsSale').text();

        if(sale !== "") {
            price = sale;
        } 

        if(price === "") {
            price = null;
        }

        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = $(product).find('p.productBrandName a.productsNameLink').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('td.prodImage img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select#ddlProductsPerPage2').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);

     },
     pages: function($) {
        var items = $('select#ddlProductsPerPage2').text(); 
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('div#ProductsTable div:nth-child(1)').html();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('h1.Header1').text();
        name = cleanName(name);
        return name;
    }
}
