// sm-supplier-html/suppliers/lelandfly.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("product/")+8,url.lastIndexOf("/"))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("product/")+8,Search.url($, product).lastIndexOf("/"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.pro-slider-title h1').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.pro-slider-price-detail p').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('.images a[itemprop="image"]').attr('href');

        return image;
    },
    description: function($) {
        var description = $('div.z-content-inner').text();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.cat-list-detail');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.cat-list-detail-title a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.cat-list-detail-regprice').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.cat-list-detail-title a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.cat-list-detail-thumb img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.cat-list-detail').length;
         return items;
     },
     pages: function($) {
         var pages = parseInt($('ul.page-numbers li').last().prev().text());
         return pages;

     }
}

var Category = {
    name: function($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name
    }
}
