// sm-supplier-html/suppliers/www.meijer.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
		name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('li#product-image img[itemprop="image"]').attr('src');
        image = 'https:' + image;

        return image;
    },
    description: function($) {
        var description = $('div[itemprop="description"]').text();
        description = cleanName(description);

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.thumbnail-wrapper');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h3.mjr-product-name a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.product-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('h3.mjr-product-name a').attr('href');
        url = 'https://www.meijer.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.prod-img').attr('src');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = parseInt($('button.dropdown-toggle.results').text());

         return items;
     },
     pages: function($) {
         var items = $('button.dropdown-toggle.results').text();
         var  products = $('span#noOfResults').text();

         var sum = products / items;

         return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('ol li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
