// sm-supplier-html/suppliers/www.katespade.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = kateSpadecom =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = ''
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = '';
            result.push(info)
        });

        return result 
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').text().trim().replace(/\n|\t|\r/mig,"");
        return name;
    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('div#product-content span.price-sales').text().replace("$","");
        return price;
    },
    image: function($) {
        var img = $('img[itemprop="image"]').attr('src');
        return img;
    },
    description: function($) {
        var desc = $('div#small-details').text().trim();
        return desc;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-tile');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a.name-link').text().trim();
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price-sales').text().replace("$","");
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.name-link').attr("href");
        return url;
    },
    image: function($, product) {
        var img = $(product).find('img.beta').attr("src");
        return img;
    }
}
