// sm-supplier-html/suppliers/www.gardenersedge.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('p.prod-det-price').first().text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('a#Zoomer').attr('href');

        return image;
    },
    description: function($) {
        var description = $('#prod-det-info > div').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.cat-product');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('p.cat-product-title a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.cat-product p.fc-orange').text();
        
        if(price === "") {
            price = $(product).find('div.cat-product p.fs-20').text();
            
        }

        price = getPrice(price);

        return price;
    },

    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href').replace("\n", "");
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('div.cat-product a img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.cat-product'); 
        items = items.length;
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.cat-product'); 
        items = items.length;

        var results = null;
        var products = $('div.paging').text();

        if(products !== "") {
            products = products.match(/[0-9,]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        

        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('div.child-cat-non-featured-content h1').text();

        return name;
    }
}
