// sm-supplier-html/suppliers/www.discountwatchstore.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#itemnamefield').html();
        (name);
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        var price = $('span.ProductPrice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('a.main-image-link img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h3').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.price').html();
         price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        } 
        return 'https://www.discountwatchstore.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.product-grid  ul li');
        return parseInt(items.length);
     },
     pages: function($) {
        var pages = $('ul.pagination').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages.match(/\d/g).pop());
     }
}

var Category = {
    name: function($) {
        var name = $('section.CategoryHeading h1').text();
        name = cleanName(name);
        return name;
    }
}
