// sm-supplier-html/suppliers/www.theworks.co.uk/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');


module.exports = {


    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    
     'search' : function($, next) {
        var result = []

        $('#product-list').find('div.product').map(function(index, product) {

            var info = new SearchPageInfo()
            info.barcode = url($, product).substring(url($, product).lastIndexOf("/")+1,url($, product).lastIndexOf(""));
            info.name = searchName($,product)
            info.sku = url($, product).substring(url($, product).lastIndexOf("/")+1,url($, product).lastIndexOf(""));
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($, product)
            info.country = "uk"
            result.push(info)
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = gPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "uk"
        return info
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}


function url($, product) {
    return $(product).find('.product-image').find('a').attr('href')
}

function searchName($, product) {
    return $(product).find('h2').text().trim()
}

function searchPrice($, product) {
    return $(product).find('.attr-price-now').text().replace("£","").trim()
}


function getNumberOfPages($) {
    var pages = $('.pagination-wrapper').find('.controls').find('.pages').find('.last-child').find('a').attr('data-page-number')
    return pages
}

function searchImage ($,product) {
    return $(product).find('.product-image img').attr('src');
}





var isStock = function inStock($) {
    var isInStock = true;
    var contentarea = $('.product-stock-status').html()
    if(contentarea !== null) {
        if(contentarea.indexOf('Out of Stock') > -1) {
            isInStock = false
        }
    }
    return isInStock
}

var gPrice = function thePrice($, instock) {
    var price = $('span.amount').last()
    price =getPrice(price)
    if (price !== null){
        price = price[0]
    }else{
        price = ""
    }

    if (price.indexOf('.') == -1)
    {
        price = price + ".00"
    }

    return price

}

var getName = function theName($, instock) {
    return $('h1').text().trim()

}

var image = function image($) {
    return $('#im1 img').attr('src');
}

var Pagination = {

     items: function($) {
        var items = $('.hproduct');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('.hproduct');
        items = items.length;

        var products = $('div.message-area h2').text();
        products = products.match(/[0-9]+/g).pop();
       
        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb ol li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
