// sm-supplier-html/suppliers/www.americangolf.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = americangolf = {
    "useJavascript": true,

    'sku': (url) => {
        url = url.split('?')[0]
        return url.substring(url.lastIndexOf("-") + 1, url.lastIndexOf("."))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).split('?')[0]
            info.name = Search.name($, product);
            info.sku = sku.substring(sku.lastIndexOf("-") + 1, sku.lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }


}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var inStock = true
        var able =$('.delivery-option-availability').text()
        inStock= inStockCheck(able,inStock)
        return inStock;
    },
    price: function ($) {
        var price = $('div[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("a.product-image img").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.grid-tile');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.product-sales-price').text();
        price = getPrice(price);

        if (price === "") {
            price = null;
        }
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
            url = url+""

        if (url.startsWith("https://www.americangolf.co.uk") != true) {
            url = "https://www.americangolf.co.uk" + url
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Category = {
    name: function ($) {
        return cleanName($('h3').first().contents().filter(function () {
            return this.type === 'text';
        }).text());
    }
}
var Pagination = {

    items: function ($) {
        var items = Search.container($).length;

        var pages = pagesCheck($, items);

        if (items && pages) {
            return items;
        } else {
            return null;
        }

    },
    pages: function ($) {

        var itemsOnPage = this.items($);

        return pagesCheck($, itemsOnPage);

    }
}

function pagesCheck($, itemsOnPage) {
    var totalItems = parseInt(cleanName($("div.pagination div.results-hits"
    ).text().split("Products found")[0]))

    if (itemsOnPage && totalItems) {
        return Math.ceil(totalItems / itemsOnPage);
    } else {
        return null;
    }
}
