// sm-supplier-html/suppliers/www.electricshopping.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js");

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.ProductNm h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.ProductPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('a[rel="prodImage"] img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.ProductDescriptionContainer').text().trim();
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.ProductList li');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.ProductDetails a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('em.p-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.pname').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.ProductImage img').attr('src');
        return image;   
    }
}

var Pagination = {

     items: function($) {
        var pages = this.pages($);

        var items = parseInt($('ul.ProductList li').length);

        if(pages && items) {
            return items;
        } else {
            return null;
        }


     },
     pages: function($) {
        var pages = $('ul.PagingList').html();

        if(pages) {
          pages = pages.match(/[0-9]+/g);  
        } else {
            pages = null; 
            
        }
        if(pages) {
            //(pages);
            return parseInt(pages.pop());
        } else {
            return null;
        }

     }
}

var Category = {
    name: function($) {
        var name = $('h1.TitleHeading').text();
        name = cleanName(name);
        return name;
    }
}
