// sm-supplier-html/suppliers/www.findmeagift.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = findMeAGift =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span#actionPanelPricesThePrice').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.main-image img').attr('src');
        return 'http:' + image;
    },
    description: function($) {
        var description = $('div.product-description').text();       
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.col-xs-6.col-md-4.col-lg-4');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-thumbnail__title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.product-thumbnail__price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.product-thumbnail__title a').attr('href');
        return 'http://www.findmeagift.co.uk' + url;
    },
    image: function($, product) {
    //    var image = null
        var image = $(product).find('img').attr('data-src')
        return 'http:' + image
    }
}

var Pagination = {

     items: function($) {
        return parseInt($('div.thumbnail-container').length);
     },
     pages: function($) {
        var pages = $('div.pagination ul li').eq('-3').text();
        pages = pages.match(/[0-9]+/g);
        return parseInt(pages)
     }  
}

var Category = {
    name: function($) {
        var name = $('title').text();
        if(name.indexOf('|') > -1){
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
