// sm-supplier-html/suppliers/www.surlatable.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = surlatable =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('div.product-priceMain span.hide').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#productLeft img').attr('src');
        return 'http://www.surlatable.com' + image;
    },
    description: function($) {
        var description = $('div#product-description')
                            .find('h2').text('').end()
                            .find('a').text('').end()
                            .text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('dl.item.ec_listing');
    },
    name: function($, product) {
        var name = $(product).find('a.name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.name').attr('href');
        return 'http://www.surlatable.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http://www.surlatable.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('dl.item.ec_listing');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('ul.pages li').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}
var Category = {
    name: function($) {
        var name = $('div#items h2.header').text();
        name = cleanName(name);
        return name;
    }
}
