// sm-supplier-html/suppliers/www.freenetelectrical.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = freenetelectrical = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var inStock = true
        var delivery = $('.product-summary-bottom p').text()
        inStock = instockCheck(delivery,inStock)
        return inStock;
    },
    price: function ($) {
        var price = $('ins span.woocommerce-Price-amount').text()
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("div.woocommerce-product-gallery__image img").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ul.products li');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price ins').text();

        if (price === "") {
            price = $(product).find('span.price').text();
        }
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.freenetelectrical.co.uk' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return 'https://www.freenetelectrical.co.uk' + image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("h1.page-title").text());
    }
}
var Pagination = {

    items: function ($) {
        var items = cleanName($("p.woocommerce-result-count").text())
            .match(/[0-9]+/g);

        if (items) {
            return parseInt(items[1]);
        } else {
            return null;
        }

    },
    pages: function ($) {
        /*        return Math.ceil(parseInt(cleanName($("p.woocommerce-result-count")
                    .text()).match(/[0-9]+/g)[2]) / this.items($));*/

        var pages = cleanName($("p.woocommerce-result-count")
            .text()).match(/[0-9]+/g);

        if (pages) {
            pages = parseInt(pages[2]);

            return Math.ceil(pages / this.items($));
        } else {
            return null;
        }

    }
}
