// sm-supplier-html/suppliers/www.buxomcosmetics.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('.product-name').text();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return 'unchecked';

    },
    price: function($) {
        price = $('h3').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('.s7staticimage img').attr('src');
        return image;

    },
    url: function ($) {

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-tile:not([data-recommendername])')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.product-name-link').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.product-sales-price').first().text();
        var otherPrice = $(product).find('.product-pricing').text()
        otherPrice = getPrice(otherPrice);
        price = getPrice(price)
        if(otherPrice){
            return otherPrice
        }
        if(price){
            return price
        }
        return null
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a.product-name-link');
        if(anchor) {
            url = $(anchor).attr('onclick');
        }
        if(url.indexOf(',') > -1){
            url=url.split(",")[1]
            //url = url.split("'")[2]
            if(url.indexOf('https') === -1){
                url = 'https://www.buxomcosmetics.com'+ url
            }
        }
        
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').first().attr('data-src');
        return image
    }
}

var Pagination = {

    items: function($) {
        var items = $('div.product_tile')
        return parseInt(items.length);

     },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function($) {
        var name = $('div.category_detail h1.no_margin').text();
        name = cleanName(name);
        return name;
    }
}
