// sm-supplier-html/suppliers/www.jomashop.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var brand = $('h1 .brand-name').text();
        var title = $('h1 .product-name').text();
        name = brand + ' ' + title
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('#final-price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('a.MagicZoomPlus img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h2.product-name a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.product-image-wrapper img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('-') > -1) {
            name = name.split('-').shift();
        }

        name = cleanName(name);
        return name;
    }
}
