// sm-supplier-html/suppliers/www.joke.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('span.h1').text();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var stock = true
        var check = $('#stocklevel').text()
        stock = instockCheck(check,stock)
        return stock;
    },
    price: function($) {
        price = $('li.yourprice').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('a.thumbnail:first-child img').attr('src');
        return 'https:'+image;

    },
    url: function ($) {

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.thumbnail')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('a.h4').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).attr('data-product-price')
        price = getPrice(price);
        return price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.joke.co.uk' + $(anchor).attr('href')
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'https:'+image
    }
}

var Pagination = {

    items: function($) {
        var items = $('li.thumbnail')
        return parseInt(items.length);

     },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function($) {
        var name = $('h1.h2').text();
        name = cleanName(name);
        return name;
    }
}
