// se-app-model.old/lib/saveJobReversePayload.js
const formatDate = require('./formatDate');
const createJobIdName = require('./createJobIdName.js');
const makeSafe = require('./makeSafe.js');

module.exports = (data) => {

    let name = 'Reverse Search';
    let jobType = name.toUpperCase(name).replace(' ','_');
    let typePayload = data[data.type];
    let jobInfo;

    if(Array.isArray(data[data.type])) {
        jobInfo = data[data.type].length + ' Asins';
    } else {
        typePayload = [data[data.type]];
        jobInfo = 'Single Asin ' + data[data.type];
    }

    let jobPayload = {
        status: "QUEUED",
        created: new Date(),
        userName: data.userName,
        userId: data.userId,
        jobType: jobType,
        subType: 'ASIN',
        data: typePayload,
        country: data.country,
        name: jobInfo + ' - ' + name
    }

    let jobIdPayload = {
        country: data.country,
        jobType: 'reverseSearch',
        jobName: jobPayload.name.replace('- ','').toUpperCase().replace(/ /g,'_'),
        userName: data.userName,
        date: formatDate(jobPayload.created.toISOString())
    }



    var qName = createJobIdName(jobIdPayload)
    qName = makeSafe(qName)
    jobPayload.idName = qName

    return jobPayload;

}

function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}
