// se-app-model.old/lib/saveJobCategoryPayload.js
const formatDate = require('./formatDate');
const createJobIdName = require('./createJobIdName.js');
const makeSafe = require('./makeSafe.js');

module.exports = (data) => {

    let name = 'Category Search' ;
    let jobType = name.toUpperCase(name).replace(' ','_');
    let typePayload = data[data.type];
    let jobInfo;

    let jobPayload = {
        status: "QUEUED",
        created: new Date(),
        userName: data.userName,
        userId: data.userId,
        jobType: jobType,
        subType: data.category,
        category: data.category,
        fastApi: data.fastApi,
        data: typePayload,
        country: data.country,
        name: data.originalCategory,
        size: data.size,
        scroll: data.scroll,
        categorySearchAge: data.categorySearchAge
    }

    let jobIdPayload = {
        country: data.country,
        jobType: jobPayload.jobType.replace('- ','').toUpperCase().replace(/ /g,'_'),
        jobName: jobPayload.subType.replace('_display_on_website','').toUpperCase(),
        userName: data.userName,
        date: formatDate(jobPayload.created.toISOString())
    }


    var qName = createJobIdName(jobIdPayload)
    qName = makeSafe(qName)
    jobPayload.idName = qName

    return jobPayload;

}

function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}
