// sm-supplier-html/suppliers/www.farmandfleet.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript":"true",
    "usePhantom":"farmandfleet.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us";
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div#pdpName h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var availability = $('.stock-msg').text()
        var isInStock = true
        isInStock=instockCheck(availability,isInStock)

        // if (availability) {
        //     if (availability.indexOf('In Stock') > -1) {
        //         isInStock = true
        //     }
        // }


	return isInStock;
},
    price: function($) {
        var price = $('span[itemprop="price"]').text();
        var saleprice = $('div.active-price .price').text();

        if(saleprice !== "") {
            price = saleprice;
        }
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.list-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.list-item-name a').text();
        name = cleanName(name);
       return name;
    },
    price: function($, product) {
        var price = $(product).find('div.list-item-price .reg-price').text();
        var price2 = $(product).find('div.list-item-price .sale-price').text(); 
        
        if (price2 !== "") {
            price = price2
        }
        
        price = getPrice(price); 
        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.list-item-name a').attr('href');
        return 'https://www.farmandfleet.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('data-src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.list-item')
        return parseInt(items.length);
     },
     pages: function($) {
        var numitems = $('div.list-item').length
        var products = $('div.list-header-text span').text();
        var result = null;

        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / numitems);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        
        return result;

     }
}

var Category = {
    name: function($) {
        var name = $('div.list-header-text h1').text();
        name = cleanName(name);
        return name;
    }
}
