// sm-supplier-html/suppliers/www.shop4world.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = shop4world = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var availability = $('.product_stock_level').html();
        var status = $('div.mobilewrap div.producttop p span').html();
        var oss = $('div.col_xs_12').text();

        if (availability !== null) {
            if ((availability.indexOf('Out of stock') > -1)
                || (availability.indexOf('Pre-order') > -1)
                || (availability.indexOf('Last one') > -1)
                || (availability.indexOf('Limited stock') > -1)) {
                isInStock = false
            }
        }

        if (status !== null) {
            status = status.toLowerCase();
            if (status.indexOf('currently unavailable') > -1) {
                isInStock = false
            }
        }

        if (oss.indexOf('Want to know when this item is back in stock?') > -1) {
            isInStock = false

        }

        return isInStock
    },
    price: function ($) {
        var price = $('.priceblock .product_price span').text();
        price = getPrice(price);
        if(price)
        {
            return price
        }else{
            price =$('div.text_primary').first().text()
            price = getPrice(price)
            return price;
        }
        
    },
    image: function ($) {
        var image = $('a.fancybox').attr('href');
        return image;
    },
    description: function ($) {
        var description = $('div.descriptionwrapper').text();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product_box');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('b.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src')
        //   return image.match(/\(([^)]+)\)/)[1].replace("'", "").replace("'", "");
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('select[name="pagesize"]').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
    },
    pages: function ($) {
        var items = $('select[name="pagesize"]').text();
        numitems = items.match(/[0-9]+/g).pop();

        products = $('div#main .productcount').text();
        if (products.indexOf(',') > -1) {
            products = products.replace(",", "");
            products = products.match(/[0-9]+/g).shift();
        } else {
            products = products.match(/[0-9]+/g).pop();
        }

        sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.catpageh1').text();
        name = cleanName(name);
        return name;
    }
}
