// sm-supplier-html/suppliers/www.decathlon.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = decathlon = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('span#productName').text();
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('span#real_price_value').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("div#viewerImage img").attr('src');
        return 'http://www.decathlon.co.uk/' + image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.new-product-thumbnail');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-label').text();
        var brand = $(product).find('.product-brand').text();

        name = brand + ' ' + name;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.decathlon.co.uk' + url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

      //  if (image.indexOf('imgLoader') > -1) {
      //      image = $(product).find('img.product_visuel').attr('data-original');
      //  }
        return 'https://www.decathlon.co.uk' + image;
    }
}
var Category = {
    name: function ($) {
        return cleanName($("div.title-pre-listing h1").text());
    }
}
