// sm-supplier-html/suppliers/www.chasing-fireflies.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = chasingFireFlies =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },

}

var Product = {
    name: function($) {
        var name = $('h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('.ml-product-instock-msg.ml-collection-directory-stock-display').first().text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.detailImage img.mainimage').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ml-grid-view-item');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find("div.ml-thumb-name").text();
        
        return cleanName(name);
    },
    price: function($, product) {
        var price = $(product).find("span.ml-item-price").text();
        var price1 =$(product).find('.ml-item-price-now').text()
        price1=getPrice(price1)
        price=getPrice(price);
        if(price){
            return price
        }else{
            return price1
        }
    },
    url: function($, product) {
        var url = $(product).find("div.ml-thumb-name a").attr("href");

        url = "https://www.chasing-fireflies.com" + url;

        return url;
    },
    image: function($, product) {
        var imageUrl = $(product).find("div.ml-thumb-image img").attr("data-deferred-src");

        return imageUrl;

    }
}

var Pagination = {

     items: function($) {
        var items = $('span.ml-products-per-page-item-1').text();

        if(items) {
            items = items.slice(items.length / 2);

            return parseInt(items);
        } else {
            return null;
        }

     },
     pages: function($) {
       var totalItems = $("a.ml-products-per-page-view-all").text();

       totalItems = totalItems.match(/[0-9]+/g);

       if(totalItems) {
        totalItems = totalItems.pop();
       } else {
        return null;
       }

       var maxItems = this.items($);


       var pages = totalItems / maxItems;
       pages = Math.ceil(pages);

       return pages;
    }
}

var Category = {
    name: function($) {
        var name = $('ol.breadcrumb li').last().text();
        name = cleanName(name);
        return name;
    }
}
