// sm-supplier-html/suppliers/www.mothercare.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = motherCare = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        var contentarea = $('.availability').html();
        var checkstock = $('div.b-availability').html();

        if (contentarea !== null) {
            if ((contentarea.indexOf('Out of Stock') > -1)
                || (contentarea.indexOf('Out of stock') > -1)) {
                isInStock = false
            }

        }

        if (checkstock !== null) {
            if (checkstock.indexOf('Out of Stock') > -1) {
                isInStock = false;
            }
        }
        return isInStock
    },
    price: function ($) {
        var price = $('div.m-sales span.m-price_floor').html();
        var price2 = $('div.m-sales span.m-price_dec').html();
        price = price + '.' + price2;

        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('meta[property="og:image"]').attr('content');
        return image;
    },
    description: function ($) {
        var description = $('div.b-product_info-wrap').text().trim();

        if (description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.b-product_tile');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.b-product_title a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price= $(product).find('.b-product_price').text();
        // var big = $(product).find('.m-sales_price .m-price_floor').text();
        // var small = $(product).find('.m-sales_price .m-price_dec').text();
        // price = big+'.'+small
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.b-product_image a').attr('href');

        if (url.startsWith("http") != true) {
            url = 'https://www.mothercare.com' + url
        }

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('.b-product_tile');
        return parseInt(items.length)
    },
    pages: function ($) {
        var items = $('.b-product_tile');
        numitems = items.length;

        products = $('div.m-pagestat').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('h1.m-search-title').text();

        if (name.indexOf('(') > -1) {
            name = name.split('(').shift();
        }

        name = cleanName(name);
        return name.replace(/\w\S*/g, function (txt) { return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase(); });;
    }
}
