// sm-supplier-html/suppliers/www.audiovisualonline.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var inStockCheck =require('../../lib/inStockCheck.js')
module.exports = replaceWithSupplierName = {

    'product': ($) => {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    },

    'search': ($, next) => {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map((index, product) => {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': ($, next) => {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': ($) =>  {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.main').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        let check =$('.product-single-stock-status').text()
        let stock = true
        stock = inStockCheck(check,stock)
        return stock;
    },
    price: function($) {
        price = $('span.js-product-single-price').text();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.product-single-image-hero-image').attr('src');
        return image;

    },
    url: function ($) {
        return $
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-container')
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h3.product-title').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var red_price=$(product).find('span.product-cost-now-price').text()
        red_price = getPrice(red_price);
        return red_price;
    },

    url: function($, product) {
         var url = null;
        var anchor = $(product).find('h3.product-title a');
        if(anchor) {
            url = $(anchor).attr('href').replace("?quickview=true","");
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.product-thumb-image').attr('src');
        return image
    }
}

var Pagination = {

    items: function($) {
        var items = $('div.product-container')
        return parseInt(items.length);

     },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function($) {
        var name = $('h1.page-primary-title').text();
        name = cleanName(name);
        return name;
    }
}
