// sm-supplier-html/suppliers/www.organizeit.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        // info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            if(info.url) {
            result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = cleanName($("div.hproduct h1.fn").text());
        name = cleanName(name)
        return name
    },
    inStock: function ($) {
        var inStock = true
        var stockcheck = $('div#Stock_Status').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock
    },
    price: function ($) {
        var price = getPrice($("div.pa span.price").text());
        price = getPrice(price)
        return price
    },
    image: function ($) {
        var image = $("img#prodImg").attr("src");
        image = 'https://www.organizeit.com' + image
        return image
    },
    description: function ($) {
        return '';
    }
}

var Search = {
    container: function ($) {
        var container = $('div#targetDiv table tr:nth-child(even) td');
        return container
    },
    name: function ($, product) {
        var name = $(product).find("a.itemname").text();
        name = cleanName(name)
        return name
    },
    price: function ($, product) {
        var price = $(product).find("span.regprice").text();
        price = getPrice(price)
        return price
    },
    url: function ($, product) {
        var url = $(product).find("a.itemname").attr("href");

        if (url) {
            if (url.startsWith('/')) {
                url = 'https://www.organizeit.com' + url
            } else {
                url = 'https://www.organizeit.com' + '/' + url
            }
            return url
        } else {
            return null
        }

    },
    image: function ($, product) {
        var index = $(product).index()
        var image = $(product).parent('tr').prev().find('td').eq(index).find('img').attr('src')
        image = 'https://www.organizeit.com' + image
        return image
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }
var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
