// sm-supplier-html/suppliers/www.toolbarn.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = toolbarn = {
    "useJavascript": true,

    'sku': (url) => {
         url = url.replace(/\/$/, "");
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product) ? Search.url($, product).replace(/\/$/, "") : "";
            info.name = Search.name($, product)
            info.sku = sku.substring(sku.lastIndexOf("/") + 1, sku.lastIndexOf("."))
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('.product-name>h1').first().text();
        if (name != null) {
            name = cleanName(name)
        } else {
            name = null
        }
      //  name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var instock=true
        var check =$('.brand-primary').first().text()
        instock =instockCheck(check,instock)
        return instock
    },
    price: function ($) {
        var price = $('span[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("div.item img").attr('src');
        return image;
    },
    barcode: function ($) {
        var barcode = $('table#product-attribute-specs-table').text();

        if (barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC') + 27).substring(0, 12);
            return barcode;
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3.product-name a').text();
        name = cleanName(name);
        //name = name.replace(/\s+/g,' ').trim()
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.price-box').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('h3 a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('div.details img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.item');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var result = null;
        var pages = $('ol.pagination li').eq("-2").text();

        if (pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
            result = parseInt(pages);
        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }

        return result;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.toolbar-upper-container div.brand-category-name').text();
        name = cleanName(name);
        return name;
    }
}
