// sm-supplier-html/suppliers/www.justmylook.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)
        return info
    },

    // 'sku': (url) => {
    //     return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    // },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1#js-product-content__title').text()
        name = cleanName(name)
        return name
    },
    inStock: function ($) {
        let stock =true
        let check =$('.js-product-stock-messages>span').not('[style="display:none"]').text()
        stock = instockCheck(check,stock)
        return stock

    },
    price: function ($) {
        var price = $('span.product-content__price--inc').text()
        price = getPrice(price)
        return price
    },
    image: function ($) {
        var image = $('div.product__image__main a img').attr('src')
        image = 'https://www.justmylook.com' + image
        return image
    },
    description: function($) {
        var description = $('div#product__description').text();
        description = cleanName(description)
        return description
    }


}

var Search = {
    container: function ($) {
        var searchProducts = $('ul#js-search-results-products__list li');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product__details').text()
        name = cleanName(name)
        return name
    },
    price: function ($, product) {
        var price = $(product).find('span.product-content__price--inc').text()
        price = getPrice(price)
        return price
    },
    url: function ($, product) {
        var url = $(product).find('.product__image a').attr('href')
        url = 'https://www.justmylook.com' + url
        return url
    },

    image: function ($, product) {
        var image = $(product).find('div.product__image a img').attr('data-src')
        image = 'https://www.justmylook.com' + image
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('ul#js-search-results-products__list li').length;
        return items
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.breadcrumb ul li').last().text()
        name = cleanName(name)
        return name
    }
}
