// sm-supplier-html/suppliers/www.richersounds.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = richersounds = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h2.product-title-attr').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var isInStock = true;
        return isInStock;
    },
    price: function ($) {
        var price = $('span.price-wrapper').attr('data-price-amount')
        return price;
    },
    image: function ($) {
        var image = $("img.fotorama__img").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('ol.products li.product-item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('span.sku-product').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.price-wrapper').attr('data-price-amount');
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('h3 a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img.product-image-photo').attr('src');
        return image;
    }
}
var Pagination = {

    items: function ($) {
        var items = $('li.product');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('li.product');
        items = items.length;

        var products = $('div.category-info span.black').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.breadcrumbs li').last().text();
        name = cleanName(name);
        return name;
    }
}
