// sm-supplier-html/suppliers/www.dillards.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var checkstock = $('div.noLongerAvailable').html();

        if(checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if(checkstock.indexOf('this item is no longer available for purchase on dillards.com') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('span.price').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.mainImage').attr('src');
        image = 'http:' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.wrapper-item')

        return searchProducts;

    },
    name: function($, product) {
        let brand = $(product).find('span.product-name strong').text();
        var name = $(product).find('span.product-name .product-name').text();
        name = cleanName(brand + ' ' + name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.price-number').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.dillards.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img.product-img').attr('src');
        return 'https:' + image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('li.wrapper-item').not(':last-child');
        return parseInt(items.length);
     },
     pages: function($) {
        var numitems = $('li.wrapper-item').not(':last-child').length;  

        var products = $('ul.pagin span[data-twist="items-text"]').text();
        products = products.match(/[0-9]+/g).pop();
        
        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('section#product-results h1').text();
        name = cleanName(name);
        return name;
    }
}
