// sm-supplier-html/suppliers/www.shoptv.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us";
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.product-name').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var stockcheck = $('span.soldOut').html();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('currently sold out') > -1) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function($) {
        var price = $('span.product-price-sale-value').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.zoom-image').attr('src');
        image = 'http://www.shoptv.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchproducts = $('li.product');
        return searchproducts

    },
    name: function($, product) {
        var name = $(product).find('span.product-title').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.sale-price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'http://www.shoptv.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('span.product-image img').attr('src');
        image = 'http://www.shoptv.com' + image;
        return image;

    }

}


var Pagination = {

     items: function($) {
        var items = $('li.product');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('div#category-control-foot div.pagination a.last').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);

    
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs h1').text();
        name = cleanName(name);
        return name;
    }
}
