// sm-supplier-html/suppliers/www.beautybrands.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div[itemprop=name]').first().text() 
        if (name) {
            name = cleanName(name)
        } else {
            name = $('a#Headline').text()
            name = cleanName(name)
        }
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div.productLongDesc span.title').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = null
        var defaultPrice = $('span.ml-item-price').text()
        var oldPrice = $('span.ml-item-price-was').text()
        var newPrice = $('span.ml-item-price-sale').text()
        if (newPrice) {
            price = getPrice(newPrice)
        } else if (oldPrice) {
            price = getPrice(oldPrice)
        } else if (defaultPrice) {
            price = getPrice(defaultPrice)
        }
        return price
    },
    image: function ($) {
        var image = $('img#mainimage, img.detailImage').attr('src');
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.ml-grid-view-item');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.ml-thumb-name').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = null
        var oldPrice = $(product).find('span.ml-item-price-was').text()
        var nowPrice = $(product).find('span.ml-item-price').text()
        var salePrice = $(product).find('span.ml-item-price-sale').text()
        if (salePrice && salePrice.indexOf('-') === -1) {
            price = getPrice(salePrice)
        } else if (nowPrice && nowPrice.indexOf('-') === -1) {
            price = getPrice(nowPrice)
        } else if (oldPrice && oldPrice.indexOf('-') === -1) {
            price = getPrice(oldPrice)
        }
        return price
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.beautybrands.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.ml-thumb-image img').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('ol.breadcrumb li').last().text()
        name = cleanName(name);
        return name;
    }
}
