// sm-supplier-html/suppliers/www.samsclub.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = samsclub =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.description = Product.description($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            //result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1 span').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var outofstock = $('div.biggraybtn').text();

        if(outofstock !== "") {
            if(outofstock.indexOf('Out of stock online') > -1) {
                isInStock = false;
            }
        }

        return isInStock;
    },
    price: function($) {
        var price = $('ul.lgFont span.hidden').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.imgHolder2 img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div[itemprop="description"] p:nth-child(2)').text().trim();
        
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r|\t)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.sc-product-card');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.sc-product-card-title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.Price-group span.visuallyhidden').html();
        //var price2 = $(product).find('footer div.sc-cents').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.sc-product-card>a');
        if(anchor) {
            url = 'https://www.samsclub.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.sc-product-card-image').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.sc-product-card');
        items = items.length;
        return parseInt(items);
     },
     pages: function($) {
        var items = $('div.sc-product-card');
        numitems = items.length;

        var products = $('ul.filters-01 span[ng-bind="plpCtrl.totalCount"]').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1.catLeftTitle').text();
        name = cleanName(name);
        return name;
    }
}
