// sm-supplier-html/suppliers/groceries.morrisons.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var base =$('.productTitle').first().not('.offer').text()
        base = cleanName(base)
        return base
    },
    inStock: function ($) {
        var inStock = true
        var available = $('.stock-status').text()
        inStock = inStockCheck(available,inStock)
        return inStock
    },
    price: function ($) {
        var price = $('h5').text()
        price = getPrice(price)
        return price
    },
    image: function ($) {
        var img = $('.galleryImages a').attr('href')
        return 'https://groceries.morrisons.com' + img
    },
    url: function ($) {
        return $
    }
}

var Search = {
    container: function ($) {
        var container = $('ul.fops li.fops-item')
        return container

    },
    name: function ($, product) {
        let name = $(product).find('h4').text()
        name = cleanName(name)
        return name

    },
    price: function ($, product) {
        let price = $(product).find('h5').text()
        price = getPrice(price)

        let roundedPrice = $(product).find('h5').html()

        let price2= $(product).find('.price-group-wrapper').text()
        if(roundedPrice != null) {
        roundedPrice = ""+roundedPrice.split(';')[1]
        roundedPrice = ""+roundedPrice.split('\n')[0]
        }

        if (price == null && roundedPrice != null) {
            price = roundedPrice + '.00'
        } else {
            price = price
        }

        price2 = getPrice(price2)

        return price2

    },
    url: function ($, product) {
        let url = $(product).find('a').attr('href')
        return 'https://groceries.morrisons.com' + url

    },

    image: function ($, product) {
        let image = $(product).find('img').attr('src')
        return 'https://groceries.morrisons.com' + image


    }
}

var Pagination = {

    items: function ($) {
    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
    }
}
