// sm-supplier-html/suppliers/www.pcrichard.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var text = $('div#text-pdp-stock').text();
        if (text && text.trim() == 'This product is only available for In-Store Pickup') {
            inStock = false;
        }
        return inStock;
    },
    price: function($) {
        var price = $('div.price-row').text();
        price = getPrice(price); // if a price has a comma take the comma out see supplier file//
        return price;
    },
    image: function($) {
        var image = $('img#imagePlus').attr('src');
        return 'http://www.pcrichard.com' + image;
    },
    description: function($) {
        return '';
    }
}

var Search = {
    container: function($) {
        return $('div.product-tile');
    },
    name: function($, product) {
        var name = $(product).find('.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.salePrice').text();
        price = getPrice(price); // use getPrice to clean the price //
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.pcrichard.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('a.prodImage img').attr('src');
        return 'http://www.pcrichard.com' + image;
    }
}

var Pagination = {
     items: function($) {
        var items = $('ul.dropdown-menu').text();
        
        if(items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        }else {
            if(Search.container($).length > 0) {
                items = 1;
            }
        }

        
        return items;

     },
     pages: function($) {
        var items = $('ul.dropdown-menu').text();
        
        if(items !== "") {
            items = items.match(/[0-9]+/g).pop();
            items = parseInt(items);
        }else {
            if(Search.container($).length > 0) {
                items = 1;
            }
        }

        var products = $('span.summary').text();
        var result = null;
        if(products !== "") {
            products = products.match(/[0-9]+/g).pop();
            result = Math.ceil(products / items)
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }

        
        return result;
        
     }
}
var Category = {
    name: function($) {
        var name = $('div.page-title h1').text();
        name = cleanName(name);
        return name;
    }
}
