// sm-supplier-html/suppliers/www.sunglasses-shop.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = sunglassesshop =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('meta[itemprop="name"]').attr('content');
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('meta[itemprop="price"]').attr('content');

        if(price.indexOf('.') < 1) {
            price = price + '.00';
        }

        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.item img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('meta[itemprop="sku"]').attr("content");
        return barcode
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.productContainerBox');
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('span.text-bold.brandName').html();
        var name = $(product).find('span.brandProductName').html();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + ' ' + name;
    },
    price: function($, product) {
        var price = $(product).find('span.productPrice').text();

        if(price.indexOf('.') < 1) {
            price = price + '.00';
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.sunglasses-shop.co.uk'+url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.PVBox');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('ul.pagination li').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div#breadcrumb div').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
