// sm-supplier-html/suppliers/www.bonanza.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replace =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('span[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('div.item_price_text').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img.main_image").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div.upc .item_detail_data').text();
        return barcode;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.gridblock');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('li.item_title').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('li.item_price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.bonanza.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('span#the_search_term').text();
        name = cleanName(name);
        return name;
    }
}
