// sm-supplier-html/suppliers/www.365games.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')


module.exports = {
    "useJavascript": true,

    /*
        'sku': (url) => {
            return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
        },
    */

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1 a').html();
        name = cleanName(name);
        if (!name) {
            name = $('h1').html()
            name = cleanName(name)
        }
        return name;

    },
    inStock: function ($) {
        var isInStock = true;
        var producttop = $('div.col_xs_8.col_lg_7.col_xl_8 span').html();
        var aviable = $('.block_md.basket_button_container').text()
        isInStock = instockCheck(aviable, isInStock)
        if (producttop === null) {
            isInStock = false
        }
        return isInStock
        /*
 
        if(producttop !== null) {
            if(producttop.indexOf('Out of stock') > -1) {
                isInStock = false
            }
        }
        if(producttop !== null) {
            if(producttop.indexOf('Currently unavailable') > -1) {
                isInStock = false
            }
        }
        if(producttop !== null) {
            if(producttop.indexOf('Limited stock') > -1) {
                isInStock = false
            }
        }
        if(producttop !== null) {
            if(producttop.indexOf('Last one - hurry!') > -1) {
                isInStock = false
            }
        }
        if(producttop !== null) {
            if(producttop.indexOf('Pre-order') > -1) {
                isInStock = false
            }
        }
*/

    },
    price: function ($) {
        var price = $('div.kb_grid_2').text();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('a.product_zoom img').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product_box');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h3 > a').text();
        name = cleanName(name);
        if (name) {
            return name;
        } else {
            name = $(product).find('h4 > a').text();
            name = cleanName(name)
            return name
        }


    },
    price: function ($, product) {
        var price = $(product).find('.price').text();

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        //    var regExp = /\(([^)]+)\)/;
        //    var matches = regExp.exec(image);
        //    image = matches[1].replace("'", "");
        //    image = image.substring(0, image.length - 1);
        return image;

    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.product_grid div.col_xs_6')
        return parseInt(items.length);

    },
    pages: function ($) {
        var pages = $('.collapse_menu_header.text_sm').last().text();
        if (pages !== "") {
            pages = pages.match(/[0-9]+/g).pop();
        }
        return parseInt(pages);
    }
}


var Category = {
    name: function ($) {
        var name = $('h1.text_xl').text();

        name = cleanName(name);
        return name;
    }
}
