// sm-supplier-html/suppliers/www.theshoemart.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.product-info-main div.panel-a h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.panel-b div.price-box span.price').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div.product-image a').attr('href');

        return image;
    },
    description: function($) {
        var description = $('div.std').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('span.show a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price-box').text();

        if (price.indexOf('Special') > -1) {
            price = $(product).find('p.special-price span.price').text();
        } else {
            price = $(product).find('span.regular-price span.price').text();
        }

        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('span.show a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('p.amount').first().text().trim().split(' ').shift();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('p.amount').first().text().trim().split(' ').shift();
         var products = $('p.amount').first().text().trim().split(' ')[2];

         var sum = products / items;
         sum = Math.ceil(sum);

         return sum;
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();

        return name;
    }
}
