// sm-supplier-html/suppliers/www.menkind.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
//Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = menKind =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.page-title').text();
        name = cleanName(name);
        return name;


    },
    inStock: function($) {
        var isInStock = true;
        var availability = $('div.unavailable').html();
        var page = $('h1').html();

        if (availability !== null) {
            availability = availability.toLowerCase();
            if(availability.indexOf('out of stock') > -1) {
                isInStock = false;
            }
        }
        if(page !== null) {
            if (page.indexOf('404') > -1) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function($) {
        var price = $('div.product-info-main span.special-price .price').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('a.MagicZoom').attr('href');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('strong.name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price
    },
    url: function($, product) {
        var url = $(product).find('strong.name a').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img.product-image-photo').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('h1#page-title-heading span').text();
        name = cleanName(name);
        return name;
    }
}
