// sm-supplier-html/suppliers/www.thenorthface.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = theNorthFace =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-content-info-name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.product-content-info-offer-price.offer-price.offer-price-js.product-price-amount-js').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = undefined
        // var image = $('div.s7thumb').css('background-image');
        // image = image.replace('url("', '').replace('")', '');
        // image = image.substring(0, image.indexOf('?'));

        return image;
    },
    description: function($) {
        var description = $('div.pdp-details-desc-container').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product-block');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('span.product-block-name-wrapper').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-block-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        url = $(product).find('a').attr('href');

        return url;
    },
    image: function($, product) {
        image = $(product).find('figure > a > picture source').first().attr('srcset');
        image = 'https:' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
       return parseInt(items = $('div.product-block').length)
     },
     pages: function($) {

     }
}

var Category = {
    name: function($) {
        var name = $('h1.page-breadcrumb-token').text();
        name = cleanName(name);
        return name;
    }
}
