// sm-supplier-html/suppliers/www.play-asia.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        var brand = $('[itemprop="brand"]').text()
        name = cleanName( brand+' '+name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var check=$('[itemprop="availability"]').text()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $("span[itemprop='price']").html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img[itemprop='image']").attr('src');
        image = 'http:' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.p_prev');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).attr('title')
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.p_prev_p').html();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price = $(product).find('.price_val').text();
            price=getPrice(price)
            return price
        }
        
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.p_prev a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find("img[data-hdpi='1']").attr('src');
        image = 'http:' + image;
        return image;
    }
}
var Pagination = {

     items: function($) {
        var items = $('.p_prev');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('div#n_pf_filter_cat_paging span#n_pf_filter_cat_pb_pages span#n_pf_filter_cat_pb_mp').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div#navlangbar h2').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
