// sm-supplier-html/suppliers/www.babymonitorsdirect.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = babymonitorsdirect =  {
/*
    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },
*/
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            //info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var isInStock = true;

        var availability = $('p.stock').html()
        if (availability) {
            if (availability.indexOf('Out of stock') > -1) {
                isInStock = false
            }
        }

        return isInStock
    },
    price: function($) {
        var price = $('p.product-page-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div.woocommerce-product-gallery__image a").attr('href');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-small.box');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('p.name.product-title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price_old=$(product).find('del span').text()
        var price = $(product).find('ins span').text();
        if(!price_old){
            price=getPrice(price_old)
        }else{
            price = getPrice(price);
        }
        return price;
    },
    url: function($, product) {
        var anchor = $(product).find('p.name.product-title a').attr('href');
        
        return anchor;
    },

    image: function($, product) {
        var image = $(product).find('div.image-fade_in_back a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.limiter select').text();
        items = items.match(/[0-9]+/g);

        if(items) {
            return parseInt(items.pop());
        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = $('div.toolbar-bottom .pages').text();
        pages = pages.match(/[0-9]+/g)

        var items = this.items($);

        if(pages && items) {
            return parseInt(pages.pop());
        } else if(items) {
            return 1;
        } else {
            return null;
        }

     }
}

var Category = {
    name: function($) {
        var name = $('span.last-crumb').text().replace("View All", "");
        name = cleanName(name);
        return name;
    }
}
