// sm-supplier-html/suppliers/www.aerosoles.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = aerosoles = {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            // result.push(info)

            if (info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $("h1.product-name span[itemprop='name']").text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('span.sale-price').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img.zoom-main-image').attr('src');
        image = 'http:' + image;
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.category-product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-tile-name').html();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('.price-markdown').text();
        var price2 = $(product).find('.product-tile-price').text();


        if (price === "") {
            price = price2;
        }

        if (price.indexOf('-') > -1) {
            price = null
        } else {
            price = getPrice(price);

        }
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a#dynAnchor');
        if (anchor) {
            url = 'http://www.aerosoles.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.product-tile-image a img').attr('src');
        image = 'http:' + image;
        return image;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.category-header-title h1').text();
        name = cleanName(name);
        return name;
    }
}
