// sm-supplier-html/suppliers/www.wdrake.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck=require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
            
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('.product-main__title').first().text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var inStock= true
        var check=$('.product-variants__note').text()
        inStock=instockCheck(check,inStock)
    return inStock
    },
    price: function($) {
        var price = $('.price__amount--normal').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div.js-product-image.slick-current.slick-active').data('img-url')
        return image
    },
    description: function($) {
        // var description = $('span[itemprop="description"]').text().replace(/\s+/g, " ").trim();

        // return description;
        var description = $('section.product-description dd').text().replace(/\s+/g, ' ').trim();
        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.products-layout__item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.row.title-row a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
       var price = $(product).find('div.pricing.col-xs-12').html();
       price = getPrice(price);
       return price;
    },
    url: function($, product) {
        let url = $(product).find('div.title-row a').attr('ng-href');
            url = ""+url.substring(url.lastIndexOf("url=")+4,url.lastIndexOf(""))

        return decodeURIComponent(url);
    },  
    image: function($, product) {
        var img = $(product).find('figure.product__image img').attr('src');
        return img;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.sli_sort_footer select option:selected').text();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);

         return parseInt(items);
     },
     pages: function($) {
         var items = $('div.sli_sort_footer select option:selected').text();
         items = items.match(/[0-9]+/g).pop();

         var products = $('h1.BreadcrumbHeader').text();
         products = products.match(/[0-9]+/g).pop();

         var sum = products / items;

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('ul.breadcrumb li').last().text();
        name = cleanName(name);
        return name;
    }
}
