// sm-supplier-html/suppliers/www.competitivecyclist.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = competitivecyclist =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').find('div').html('').end().text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('h2.toos-label');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('span.price-sale').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#main_product_image').attr('src');
        if (image.indexOf('data:image/png') == -1) {
            image = 'http:' + image;
        }
        return image;
    },
    description: function($) {
        var description = $('div.prod-desc').text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('div#products div.product');
    },
    name: function($, product) {
        var brand = $(product).find('span.ui-pl-name-brand').text()
        var name = $(product).find('span.ui-pl-name-title').text();
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + ' ' + name;
    },
    price: function($, product) {
        var price = $(product).find('span.price-retail, span.price-sale').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.competitivecyclist.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.ui-pl-img').html();
        image = image.substring(image.lastIndexOf("//"),image.lastIndexOf("jpg")) 
        return 'http:' + image + 'jpg';
    }
}

var Pagination = {

     items: function($) {
        var items = $('div#products div.product').length;
        return parseInt(items);
     },
     pages: function($) {
        var pages = $('section.main div.pag-wrap ul li a').text();
        if(pages === ""){
            return null
        }else {
            pages = pages.match(/[0-9]+/g).pop();
            pages = pages.match(/\d/g).pop();
        }

        
        return parseInt(pages);
     }
}
