// sm-supplier-html/suppliers/www.udans.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').find('strong').html('').end().text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        return inStock;
    },
    price: function($) {
        var price = $('span.lead-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.item-image-gallery img:nth-child(1)').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div.item-detailed-info div.well-body:nth-child(3)').text();
        if (description) {
            description = description.replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('div#item-list div.span4');
    },
    name: function($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.lead-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('h2 a').attr('href');
        return 'http://www.udans.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('div.item-cell-thumbnail img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var products = $('div.item-listing-items-limit select').text();
        products = products.match(/\d+/gm).pop();
        return parseInt(products);
     },
     pages: function($) {
        var products = $('div.item-listing-items-limit select').text();
        products = products.match(/\d+/gm).pop();

        var total = $('b.total-items').text();
        var sum = total / products;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        return cleanName($("div.view-header ul.breadcrumb li.active").text());
    }
}
