// sm-supplier-html/suppliers/www.improvementscatalog.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.gwt-product-detail-product-name-label h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var inStock=true
        var check=$('[itemprop="availability"]').attr('content')
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price;

        if ($('div.gwt-product-detail-top-price').text() != '') {
            price = $('div.gwt-product-detail-top-price').text();

            if (price.indexOf(' - ') > -1) {
                price == null;
            } else {
                price = getPrice(price);
            }
        } else {
            price = $('div.gwt-pdp-main-stacked-price-now-label').text();

            if (price.indexOf(' - ') > -1) {
                price == null;
            } else {
                price = getPrice(price);
            }
        }

        return price;
    },
    image: function($) {
        var image = $('img.iwc-main-img').attr('src');

        return image;
    },
    description: function($) {
        var description = $('div.pdp-single-tab-description-content').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.gridItem');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price;
        if ($(product).find('span.priceNow').text() != '') {
            price = $(product).find('span.priceNow').text();

            if (price.indexOf(' - ') > -1) {
                price = null;
            } else {
                price = getPrice(price);
            }
        } else {
            price = $(product).find('span.price-grid').text();

            if (price.indexOf(' - ') > -1) {
                price = null;
            } else {
                price = getPrice(price);
            }
        }

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');

        return 'https:'+url;
    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src')

        return 'https:'+image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('span.sli_bct_num_results').text().split(' ').pop();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         var items = $('span.sli_bct_num_results').text().split(' ').pop();
         var products = $('span.sli_bct_total_records').text();

         var pages = products / items;
         pages = Math.ceil(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('div#browseHead h1').text();

        return name;
    }
}
