// sm-supplier-html/suppliers/salonskincare.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = salonskincare =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
     'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div.product-name h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
    return 'unchecked'
},
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("span.image_url").text();
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.productitem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.productitem--title').text();
        var brand =$(product).find('.productitem--vendor').text()
        brand = cleanName(brand)
        name = cleanName(name);
        return brand+' '+name;
    },
    price: function($, product) {
        var price = $(product).find('.price--main').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.productitem--title>a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://salonskincare.co.uk'+url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.item-inner');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        var pages = $('div.pages ol li').eq("-2").text();
        if(pages){
            pages = pages.match(/[0-9]+/g).pop();
            return parseInt(pages);
        }else{
            return null
        }
        
        
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
