// sm-supplier-html/suppliers/www.dcshoes.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'// can you add country here //
        info.description = Product.description($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'; // can you add country here //
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.productname-container h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.salesprice ').html();
		price = getPrice(price); // use getPrice() to clean the price //
		return price;
    },
    image: function($) {
        var image = $('.productthumbnail').attr('src');
		
		return image;
    },
    description: function($) {
        var description = $('div.pdp-desc-long').text();
        
        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.producttileinner');
		
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.name').text();
        name = cleanName(name);
		
        return name;
    },  
    price: function($, product) {
        var price = $(product).find('div.price div.salesprice').text(); // this is how we get the price  //
		price = getPrice(price);
		
		return price;
    },
    url: function($, product) {
        var url = $(product).find('.name a').attr('href'); // dont need 2 .find //
		return url;
    },
    image: function($, product) {
        var image = $(product).find('.jsThumbnailReplace img').attr('src'); // dont need 2 .find //
		return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $("div.refinement-box-values ul li.selected").text().match(/[0-9]+/g);
        if(items) {
            return parseInt(items[0]);
        }
     },
     pages: function($) {
        var totalItems = $("div.producthits-categoryslots").attr("data-result-count");

        (totalItems, "................");
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('|') > -1) {
            name = name.split('|').pop();
        }
        name = cleanName(name);
        return name;
    }   
}
