// sm-supplier-html/suppliers/www.dickssportinggoods.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = dicksSportingGoods = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if (info.name != "") {
                result.push(info)
            }

        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        let inStock=true
        let available=$('.ship-To-Me-Msg').text()
        if(available !== null){
            available=available.toLowerCase()
            if(available.indexOf('not available') > -1) {
                inStock = false
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('span.final-price').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('#main-image img').attr('src');
        //image = 'http://www.dickssportinggoods.com' + image;
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product-grid').not('pdp-spec-cont');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.fplpTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.item-price').html();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.dickssportinggoods.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        return 'https:' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('span.page-select').text();
        items = items.match(/[0-9]+/g)

        if (items) {
            items = items.pop();
            return parseInt(items);
        } else {
            return null;
        }

    },
    pages: function ($) {
        var items = $('span.page-select').text();
        numitems = items.match(/[0-9]+/g);

        if (numitems) {
            numitems = numitems.pop();
        } else {
            return null;
        }



        var products = $('span.product-count').text();
        products = products.match(/[0-9]+/g)

        if (products) {
            products = products.pop();

        } else {
            return null;
        }


        var sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('h1#category-name').text();
        name = cleanName(name);
        return name;
    }

}
