// sm-supplier-html/suppliers/www.lovelula.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('#inside-product-cart-form > div > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('#product-price > div.product__price-amount').text();
        var price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div#main-image img').attr('src');
        image = 'https://www.lovelula.com' + image;

        return image;
    },
    description: function($) {
        var des = $('div.description').text().replace(/\s+/g, " ").trim();

        return des;
    }
}

var Search = {
    container: function($) {
        var product = $('li[data-hook="products_list_item"]');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('a.info').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-box-price').text();
        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('a.info').attr('href');
        url = 'https://www.lovelula.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.product-box-image a img').attr('src');
        image = 'https://www.lovelula.com' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('div.layout-number-results a.selected').first().text();
         items = JSON.parse(items);

         return items;
     },
     pages: function($) {
         //
         // there is no all products count element
         //

         var pages = $('span.last a').attr('href').split('=').pop();
         pages = JSON.parse(pages);

         return pages;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.title-line span').text();
        name = cleanName(name);

        return name;
    }
}
