// sm-supplier-html/suppliers/www.acehardware.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = acehardware = {

 //   'sku': (url) => {
 //       return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
 //   },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        // info.barcode = Product.barcode($)
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
      //      info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'

            if (info.price && info.name) {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1.mz-pagetitle').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var stock = $('.mz-productdetail-shipping-text').text()
        var inStock = true
        inStock = inStockCheck(stock,inStock)
        return inStock;

    },
    price: function ($) {
        var price = $('div.mz-pricestack.mz-pricestack-range:last-child').html();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('img.mz-productimages-thumbimage').attr('src');
        return 'https:'+image;

    },
    // barcode: function ($) {
    //     var barcode = $('div.prodC3').html();
    //     barcode = barcode.substring(barcode.lastIndexOf("|") + 2).substring(0, 12)
    //     return barcode;
    // }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li');
        return searchProducts;

    },
    name: function ($, product) {
        let name = $(product).find('a').text()
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('.mz-price').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.prod').length;
        return parseInt(items)

    },
    pages: function ($) {
        var items = $('li.prod').length;

        var products = $('div.show_products').text();
        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }


        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
