// sm-supplier-html/suppliers/www.pebblegrey.co.uk/logic.js
var PageInfo = require('../../lib/pageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = {
    'product': function ($, next) {
        var info = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = price($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var isStock = function inStock($) {
    var isInStock = true;
    var availability = $('.prod-info').html()
    if (availability) {
        if ((availability.indexOf('Out of Stock') > -1)
            || (availability.indexOf('unavailable') > -1)
            || (availability.indexOf('Out of stock') > -1)) {
            isInStock = false
        }
    }
    return isInStock
}


var price = function thePrice($) {
    $('.prod-price').find('span').remove()
    var price = /[0-9]+\.[0-9][0-9]/.exec($('.prod-price').html())
    if (price !== null) {
        price = price[0]
    } else {
        price = ""
    }
    return price
}

var getName = function theName($) {
    return $('h1.prod-title').text().trim()
}

var image = function image($) {
    return 'http://www.pebblegrey.co.uk' + $('.slides img').attr('src');
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product-item-info');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('a.product-item-link').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('span.price').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('img').eq("-1").attr('src');
        return image;

    }
}


var Pagination = {

    items: function ($) {
        // var items = $('div.product-items');
        // items = items.length;
        // return parseInt(items);

    },
    pages: function ($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('ul.breadcrum li').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
