// sm-supplier-html/suppliers/www.sunnysports.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = replaceWithwebAddressChangeToCamelCase =  {
    "useJavascript":true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1[itemprop*="name"] span[itemprop*="title"]').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        let stock = true
        let check = $('p>strong').first()
        stock = instockCheck(check,stock)
        return stock;

    },
    price: function($) {
        var price = $('strong[itemprop*="price"]').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.large-img img').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.item-list div.item').not('div.item-list .next-page').not('div.item-list .hide');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.item-details h2 a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('strong.your-price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'https:' + image;
        return image;

    }
}
var Pagination = {

     items: function($) {
        var items = $('div.item-list div.item').not('div.item-list .next-page').not('div.item-list .hide');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.item-list div.item').not('div.item-list .next-page').not('div.item-list .hide');
        items = items.length;

        var products = $('span.index-count-total').text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('header.list-title h1').text();
        name = cleanName(name);
        return name;
    }
}
