// sm-supplier-html/suppliers/ecosaveledlights.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var routine = require ("../../lib/routine.js"); 

module.exports = ecosaveledlights =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($) 
        info.image = Product.image($)
        info.barcode = Product.barcode($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf("."));            
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span.h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#image-main").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div.bullet_codes li').html();
        if(barcode) {
         return barcode.match(/\d+/)[0];
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.price-box a span.price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select[title="Results per page"]').text().match(/[0-9]+/g);
        if(items){
            items = items.pop();
            return parseInt(items)
        } else {
            return null;
        }
     },
     pages: function($) {
/*        var items = $('select[title="Results per page"]').text();
        
        if(items) {
            numitems = items.match(/[0-9]+/g).pop();
        }*/

        var numitems = routine.matchNumbersAndPop($, 'select[title="Results per page"]');



/*        var products = $('p.amount').text();
       
        if(products) {
            products = products.match(/[0-9]+/g).pop();
            

        }*/

        var products = routine.matchNumbersAndPop($, "p.amount");

/*        var sum = products / numitems;
        return Math.ceil(sum);*/

        return routine.pagesFromTotalAndProd(products, numitems);
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();
        name = cleanName(name);
        return name;
    }
}
