// sm-supplier-html/suppliers/www.sallybeauty.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = sallyBeauty = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);
        info.nexturl=Pagination.nexturl($)

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-name').html();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('span.price-holder').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img.primary-image').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.product-tile');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('div.tile-body.tile-body--desktop p.pdp-link__brand').text()
        var name = $(product).find('div.tile-body.tile-body--desktop a[itemprop="url"]').text();
        name = cleanName(brand+' '+name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('[itemprop="price"]').first().text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price=$(product).find('[itemprop="lowprice"]').first().text()
            price= getPrice(price)
            return price
        }
        
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.pdp-link__name');
        if (anchor) {
            url = 'https://www.sallybeauty.com'+$(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('.tile-image').attr('src');
        return image;
    }
}
var Pagination = {

    items: function ($) {
        return parseInt($('div.product-tile').length);
    },
    pages: function ($) {
        var products = $('.result-count').text();
        if(products){
            products = products.match(/[0-9]+/g).pop();
        }else{
            products=0
        }
        
        return products
    },
    nexturl: function ($){
        var nexturl = $('.text-center>button')
        if(nexturl){
            nexturl=nexturl.attr('data-url')
        }else{
            nexturl=null
        }
        return nexturl
    }
}

var Category = {
    name: function ($) {
        var name = $('.header.page-title').text();
        name = cleanName(name);
        return name;
    }
}
