// sm-supplier-html/suppliers/www.carphonewarehouse.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require("../../lib/routine.js");

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1#productTitle').text();
        var features = $('div.fConfiguration li.active').text();

        if (features !== "") {
            name = name + ' ' + features;
        }
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('span#upfrontCostDodoFull').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('meta[property="og:image"]').attr('content');
        return image;
    },
    description: function ($) {
        var description = $('meta[property="og:description"]').attr('content');
        if (description !== "") {
            return description.replace(/(\r\n|\n|\r)/gm, "");
        }
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('label.name').html();
        name = name.substring(name.lastIndexOf("=") + 6, name.lastIndexOf(""));
        name = name.replace('\n</a>', '')
        name = cleanName(name)

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.priceprprty').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('p.handset a').attr('href');
        return 'https://www.carphonewarehouse.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        return parseInt($('div.product').length);
    },
    pages: function ($) {
        var numitems = $('div.product').length;
        /*
                var products = $('div[ng-if="totalResults >0"]').text();
                products = products.match(/[0-9]+/g).pop();
        
                var sum = products / numitems;
                return Math.ceil(sum);*/


        var products = routine.matchNumbersAndPop($, 'div[ng-if="totalResults >0"]');

        return routine.pagesFromTotalAndProd(products, numitems);



    }
}

var Category = {
    name: function ($) {
        var name = $('div.left-side-of-banner h1').text();
        name = cleanName(name);
        return name;
    }
}
