// sm-supplier-html/suppliers/www.echemist.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var routine = require ("../../lib/routine.js"); 
var inStockCheck = require("../../lib/inStockCheck.js")
module.exports = echemist =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.right-side h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true
        var availbility = $('.stock-status').text()
        inStock=inStockCheck(availbility,inStock)
        
	return inStock;
},
    price: function($) {
        var price = $('div.price-info').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.image-holder img').attr('src');
        return 'http://www.echemist.co.uk' + image;
    },
    description: function($) {
        var description = $('p[itemprop="description"]').text();

        if(description !== null) {
            return description.replace(/(\r\n|\n|\r)/gm,"");
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-box');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h4[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('h5').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a[itemprop="url"]').attr('href');
        return url;
    },
    image: function($, product) {
        var image = $(product).find('img[itemprop="image"]').attr('data-original');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.dropdown ul.dropdown-menu').text();
        items = items.match(/[0-9]+/g);

        if(items) {

            items = items.pop();
            return parseInt(items);
        } else {
            return null;
        }

     }, 
     pages: function($) {
/*        var items = $('li.dropdown ul.dropdown-menu').text();
        numitems = items.match(/[0-9]+/g).pop();*/

        var numitems = routine.matchNumbersAndPop($, 'li.dropdown ul.dropdown-menu');
/*
        var products = $('div.pagination-header span.showing').text();
        products = products.match(/[0-9]+/g).pop();*/

        var products = routine.matchNumbersAndPop($, 'div.pagination-header span.showing');
        
/*        sum = products / numitems;
        return Math.ceil(sum);*/
        
        return routine.pagesFromTotalAndProd(products, numitems);
     }  
}

var Category = {
    name: function($) {
        var name = $('div.category-desc h2').text();
        name = cleanName(name);
        return name;
    }
}
