// sm-supplier-html/suppliers/www.bhphotovideo.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck =require('../../lib/inStockCheck')
module.exports = bhPhoto =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('[itemprop="name"]').text();
        var brand =$('[itemprop="brand"]').text()
        name = cleanName(brand+' '+name);
        return name;

    },
    inStock: function($) {
        var check= $('.salesComments span').first().text()
        var instock = true
        instock = instockCheck(check,instock)
        return instock

    },
    price: function($) {
        var price = $('.itc-you-pay-price').last().text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img#mainImage').attr('src');
        return image


    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.js-item');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('[itemprop*="brand"]').html();
        var name = $(product).find('[itemprop*="name"]').html();
        var name = brand + " " + name;
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.itc-you-pay-price').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;


    },
    image : function($, product) {
       var image = $(product).find('div.img-zone a img[itemprop*="image"]').attr('data-src');
       //var imageSplit = image.split("/").pop()
       //var image = "https://static.bhphoto.com/images/images250x250/" + imageSplit;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.top .selectbox p').text();
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }
      
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.top .selectbox p').text(); 
            if(items === ""){
                return null
            }else {
                numitems = items.match(/[0-9]+/g).pop();
            }

        


        var products = $('div.pagination .top').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();
        }
        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.pagination h1').text();
        name = cleanName(name);
        return name;
    }
}
