// sm-supplier-html/suppliers/www.bareescentuals.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)

        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div#productname').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('span#stockStateMssg').text();
        if (stockMessage) {
            if (stockMessage.toLowerCase().indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('p.be_price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div#hero_pics img').attr('src');
        return image;
    },
    description: function($) {
        var description = $('div#tab1 p.double_space').text();
        if (description) {
            description = description.replace(/(\r|\n|\r\n)/gm, ' ').replace(/\s+/gm, ' ').trim();
            return description;
        }
    }
}

var Search = {
    container: function($) {
        return $('li[data-productid]');
    },
    name: function($, product) {
        var name = $(product).find('h2.prod_name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('p.be_price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        return $(product).find('h2 a').attr('href');
    },
    image: function($, product) {
        var image = $(product).find('div.prod_image img').attr('src');
        return image;
    }
}

// var Pagination = {

//      items: function($) {

//      },
//      pages: function($) {

//      }
// }

var Category = {
    name: function($) {
        var name = $('title').text();
        if(name.indexOf('|') > -1){
            name = name.split('|').shift();
        }

        name = cleanName(name);
        return name;
    }
}
