// sm-supplier-html/suppliers/www.nisbets.co.uk/logic.js
var PageInfo = require('../../lib/pageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = {
    "useJavascript": true,

    'product': function ($, next) {
        var info = new PageInfo()
        info.inStock = isStock($)
        info.name = getName($)
        info.price = price($)
        info.image = image($)
        info.country = "uk"
        return info
    },
    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var isStock = function inStock($) {
    var isInStock = true;
    var availability = $('.prod-availability-msg').text()
    isInStock=instockCheck(availability,isInStock)
    return isInStock
}


var price = function thePrice($) {
    var price =$('.price--main').text()
    price=getPrice(price)
    return price
}

var getName = function theName($) {
    var name = $('.product-details-summary').text()
    name =cleanName(name)
    return name
}

var image = function image($) {
    return 'http://www.nisbets.co.uk' + $('img.primary-image').attr('src');
}


var Search = {
    container: function ($) {
        var searchProducts = $('li.product-item');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('div.product-name a.name').html();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.product-listing-price').html();
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.nisbets.co.uk' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('a.thumb img').attr('data-src');
        return image;

    }
}


var Pagination = {

    items: function ($) {
        var items = $('div.product-list-row-new');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div.product-list-row-new');
        items = items.length;

        var products = $('p.fs14 span').eq("-1").text();
        products = products.match(/[0-9]+/g).pop();

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.centre-content h1').text();
        name = cleanName(name);
        return name;
    }
}
