// sm-supplier-html/suppliers/www.gamestop.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = gamestop =  {
    "useJavascript": true,
    "usePhantom": "gamestopUK.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';

            if(info.price !== null) {
            result.push(info)
        }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }


}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        isInStock = true;
        var availability = $('.singleVariantText p.variantName').html();

        if (availability !== null) {
            availability = availability.toLowerCase() 
            if(availability.indexOf('pre-order') > -1) {
                isInStock = false;
            }
        }
        return isInStock;
    },
    price: function($) {
        var price = $('span.prodPriceCont').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("a.prodImg img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.singleProduct');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var discounted = $(product).find('p.buyNew em.discounted').html();
        var preowned = $(product).find('p.buyNew').html();

        if (discounted === null) {
            discounted = $(product).find('p.buyNew a span').text();
        }

        if (preowned.indexOf('buyDisabled') > -1) {
            price = null;
        }

       price = getPrice(discounted);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('h3 a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://www.gamestop.co.uk' + url;

    },

    image: function($, product) {
        var image = $(product).find('a.prodImg img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}

var Category = {
    name: function($) {
        return "test";
    }
}
