// sm-supplier-html/suppliers/www.toysrus.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = toysRUs =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(""))
    },


    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("productId=")+1,Search.url($, product).lastIndexOf("&"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('div#lTitle h1').html();
        name = cleanName(name);
        return name;
        
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div#price li.retail span').html();
        price = getPrice(price);
        return price;
        
    },
    image: function($) {
        var image = $('div#marker img').attr('src');
        image = 'http://www.toysrus.com' + image;
        return image;
        

    },
    barcode: function($) {
        var barcode = $('p.upc span.value').text();
        return barcode;
    }
}

var Search = {
    container: function($) {
        // var searchProducts = $('.prodloop_cont');

        // if(searchProducts.length === 0) {
        //     searchProducts = $('div.topSellersView');  // This is for when toysrus have a different layout on the website, they use 2 diferent style bullshit // 
        // }
        // return searchProducts;
        
        var searchProducts = $('a.product-item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.product-item__product-title').text();
        name = cleanName(name);
        return name;
    
    },
    price: function($, product) {
        var price = null;
        var salePrice = $(product).find('span.tru-price.product-price__label.product-price__label__sale.tru-price__red').html();
        var standardPrice = $(product).find('span.tru-price').html()
        salePrice = getPrice(salePrice);
        standardPrice = getPrice(standardPrice);

        if(salePrice == null) {
            price = standardPrice;
        } else {
            price = salePrice;
        }
    return price;
   },

    url: function($, product) {
        var url = null;
        var anchor = $(product);
        if(anchor) {
            url = 'https://www.toysrus.com' + $(anchor).attr('href');
        }
        return url;
    },

    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.prodloop_float'); 

        if(items.length) {
          items = items.length;
          return parseInt(items);  
      } else {
        return null;
      }

        
     
     },
     pages: function($) {
        var items = $('div.prodloop_float'); 
        items = items.length;

        var result = null
        var products = $('div.showingText').text();
        
        if(products !== "") {
            products = products.match(/[0-9,]+/g).pop();
            result = Math.ceil(products / items);
        }else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
        return result
        
     }
}

var Category = {

 name: function($) {
        var name = $('div#contentbody h1#TRUFamilyBrandTitle').text();
        name = cleanName(name);
        return name;
    }

    // category: function($, categories) {
    //     var maincat = $('h1.familyBrand').text();
    //     var name = $(categories).find('h2 a').attr('title');

    //     var url = null;
    //     var anchor = $(categories).find('h2 a')
        
    //     if(anchor) {
    //         url = $(anchor).attr('href');
    //     }
    //     url = 'www.toysrus.com' + url + '&ppg=96&page={{pageNo}}';

    //     var category = '"' + maincat + ' - ' + name + '": ' + '"' + url + '",';

    //     if(category.indexOf('/category/') > -1) {
    //         category = null;
    //     } else {
    //         (category)
    //         return category
    //     }
    // }
}
