// sm-supplier-html/suppliers/www.rutlandcycling.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.productName').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div#fancybox-content h2').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('this product is currently unavailable') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = $('span.nowPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = null
        var style = $('.imageContainer.facetItemImg').attr('style')
        if(style.indexOf('background:url("') > -1){
            image = style.split('"')[1]
            return 'https://www.rutlandcycling.com'+image
        }else{
            return null
        }
        // if (style) {
        //     image =  + style
        // }
        // return image
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.item.col-facetItem');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.frItemName').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('[itemprop="offers"]').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a.frItemName');
        if (anchor) {
            url = 'https://www.rutlandcycling.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = null
        var style = $(product)

        if (style) {
            image = 'https://www.rutlandcycling.com' + style.attr('data-imgsrc')
        }
        return image
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div#FAFTitle h2').text()
        name = cleanName(name);
        return name;
    }
}
