// sm-supplier-html/suppliers/www.mscdirect.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div#pdp-prod-title h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        var status = $('div.pdp-stock p strong').text();

        if (status.indexOf('In Stock') > -1) {
            status = true;
        } else {
            status = false;
        }

        return status;
    },
    price: function ($) {
        var price = $('span#listPriceDiv').text();
        var status = $('div.pdp-stock p strong').text();

        if (status.indexOf('In Stock') > -1) {
            price = getPrice(price);
        } else {
            price = null;
        }

        return price;
    },
    image: function ($) {
        var image = $('div#zoom-box a img').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('div#pdp-specs').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.ui.divided.items div.item');

        return product;
    },
    name: function ($, product) {
        //  $(product).find('div.termnode-prodtitle a span').remove();
        var name = $(product).find('h4.ui.header.v4-tn-title').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.v4-tn-your-price').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'http://www.mscdirect.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('a.productimage img').attr('src');
        var image = 'http:' + image;

        return image;
    }
}

var Pagination = {

    items: function ($) {
        //
        // there is no element with items count on page
        //
        var items = $('div.termnode-item').length;

        return items;
    },
    pages: function ($) {
        var items = $('div.termnode-item').length;

        var products = $('div.termnode-nav-returned').first().text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / items;

        return Math.ceil(sum);
    }
}
var Category = {
    name: function ($) {
        var name = $('div#termnode-breadcrumb a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
