// x4-q/lib/sortKeys.js
module.exports = (keys, next) => {
    let qs = {
        rsmq: [],
        cat: [],
        aggsCache: [],
        scrollId: [],
        scrollIdOld: [],
        other: [],
        key: []
    }

    keys.map((q) => {
        switch(true) {
            case q.indexOf('rsmq:us') === 0 || q.indexOf('rsmq:uk') === 0:
                qs.rsmq.push(q);
                break;
            case q.indexOf('catsearch') === 0:
                qs.cat.push(q);
                break;
            case q.indexOf('AggsCacheData') === 0:
                qs.aggsCache.push(q);
                break;
            case q.indexOf('scrollid') === 0:
                qs.scrollId.push(q);
                break;
            case q.indexOf('us-') === 0 || q.indexOf('uk-') === 0:
                qs.scrollIdOld.push(q);
                break;
            case q.indexOf('key') === 0:
                qs.key.push(q);
                break;
            default:
                qs.other.push(q);
                break;
        }
    })

    next(qs)
}
