// sm-supplier-html/suppliers/www.cotswoldoutdoor.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');


module.exports = cotswoldoutdoor = {
    "useJavascript": true,
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('span.a-product-name').text();
        var brand=$('span.a-brand-name ').text()
        name=brand+' '+name
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var availbility=$('button.a-button--add-to-basket.button__pdp.button__add.at-element-click-tracking')
        var inStock=false
        if(availbility){
            inStock=true
        }
        return inStock;
    },
    price: function ($) {
        var price = $('div.a-price--sell').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("div.m-product-image__zoom img").attr('src');
        return 'https:'+ image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('a.m-product-tile__link');
        return searchProducts;
    },
    name: function ($, product) {
        var brand = $(product).find('span.a-brand-name').text();
        var title = $(product).find('span.a-product-name').text();

        name = brand + ' ' + title;
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.a-price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var anchor = $(product).attr('href');
        
        return 'https://www.cotswoldoutdoor.com' + anchor;

    },
    image: function ($, product) {
        var image = $(product).find("div.m-product-image__placeholder img").attr('src');
        return 'https:'+ image;
    }
}
var Category = {
    name: function ($) {
        var name = cleanName($("div.heading.parbase h1.text--center").text());
        

        return name;
    }
}
var Pagination = {

    items: function ($) {
        return parseInt($("div.a-text.a-text--clickable.a-text--underline").text());
    },
    pages: function ($) {
        return Math.ceil($("div.o-lister-control-bar__item").text().match(/[0-9]+/g) / this.items($));

    }
}
