// sm-supplier-html/suppliers/www.kennethcole.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').text();
        name = cleanName(name);
        return name.toUpperCase();
    },
    inStock: function($) {
        var inStock = true;
        var checkStock = $('p.in-stock-msg').text();

        if(checkStock) {
            checkStock = checkStock.toLowerCase();
            if(checkStock.indexOf('in stock') == -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.product-price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        return $('ul.product-image-list img:nth-child(1)').attr('src');
    },
    description: function($) {
        return '';
    }
}

var Search = {
    container: function($) {
        return $('li.grid-tile');
    },
    name: function($, product) {
        var name = $(product).find('a.name-link').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.product-sales-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        return $(product).find('a.name-link').attr('href');
    },
    image: function($, product) {
        return $(product).find('img.product-thumbnail-image').attr('src');
    }
}

// var Pagination = {
// 
//      items: function($) {
// 
//      },
//      pages: function($) {
// 
//      }
// }

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('-') > -1) {
            name = name.split('-').shift();
        }

        name = cleanName(name);
        return name;
    }
}
