// sm-supplier-html/suppliers/www.qvc.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.col-tn-12 h1').html();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        return "UNCHECKED"

    },
    price: function ($) {
        var price2 = $('div.qvcPrice span').html();
        var price = $('div.featuredPrice span.productDetailPrice').html();
        (price2);

        if (price === null) {
            price = price2
        }

        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.defaultPhoto img').attr('src');
        image = 'http:' + image;
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.galleryItem');
        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('p.productDesc').text()
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('span.priceSell').text();
        //var price2 = $(product).find('span.productQvcPrice').html();

        // if (price === null) {
        //     price = price2;
        // }
        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        var image = $(product).find('.productImg img').attr('src');
        //var images = $(product).find('.productImg').html()

        return 'https:'+image

    }
}
var Pagination = {

    items: function ($) {
        // var items = $('').html();
        // items = items.match(/[0-9]+/g).pop();
        // return parseInt(items);

    },
    pages: function ($) {
        // var items = $('').html(); 
        // numitems = items.match(/[0-9]+/g).pop();

        // var products = $('').text();
        // products = products.match(/[0-9]+/g).pop();

        // sum = products / numitems
        // return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div#divPageTitle h1').text();
        name = cleanName(name);
        return name;
    }
}
