// sm-supplier-html/suppliers/www.littlewoods.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.priceNow').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.amp-zoom-overflow img').attr('src');
        return image;
    }
}

var Search = {
    container: function($, product) {
        var searchProducts = $('li.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.productTitle h3').text();
        name = cleanName(name);
        return name;
    },  
    price: function($, product) {
        var price = $(product).find('.productPrice').text();
        price = getPrice(price);
        if (price === null) {
            price = $(product).find('a.productPrice .productNowPrice').text().replace('£', '');
        }
        
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a.productTitle');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.productMainImage img').attr('src');
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("h1.galleryHeading span").text());
    }
}
var Pagination = {

     items: function($) {
        return parseInt(cleanName($("div.productsPerPage span")
            .text()).split("Results:")[1].match(/[0-9]+/g)[1]);
     },
     pages: function($) {
        return parseInt(cleanName($("div.pagination ul li").text())
            .split("▸ ◂")[0].match(/[0-9]+/g).pop());

     }
}
