// sm-supplier-html/suppliers/www.target.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var generateSku = require('../../lib/testPattern.js')
let regex = /A-?[0-9]+/


module.exports = target = {

    'sku': (url) => {
        generateSku(url, regex)[0].replace('A-', '')
        return url
    },

    "useJavascript": true,
    "usePhantom": "target.phantom.js",
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = generateSku(Search.url($, product), regex)[0].replace('A-', '')
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "us"
            // result.push(info)

            if (info.price > 0 && info.name != "") {
                result.push(info)
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $("h1.h-margin-b-none.h-margin-b-tiny").text()
        if (name != null) {
            name = cleanName(name)
        }
        return name;
    },
    inStock: function ($) {


        var isInStock = true;
        var preorder = $('.add-to-cart-wrap').html();
        var checkstock = $('.add-to-cart-wrap .btn').html();
        var outOfStock = $('span.h-text-orangeDark').text()
        let noLongerAvailable = $('div.h-text-orangeDark').text()
        //  ('NO LONGER AVAILABLE: ' + noLongerAvailable)

        if (preorder !== null) {
            preorder = preorder.toLowerCase();
            if (preorder.indexOf('preorder now') > -1) {
                isInStock = false;
            }
        }

        if (checkstock !== null) {
            checkstock = checkstock.toLowerCase();
            if (checkstock.indexOf('only in stores') > -1) {
                isInStock = false;
            }
        }

        if (noLongerAvailable) {
            noLongerAvailable = noLongerAvailable.toLowerCase();
            if (noLongerAvailable.indexOf('no longer available') > -1 || noLongerAvailable.indexOf('not available') > -1 || noLongerAvailable.indexOf('out of stock') > -1) {
                isInStock = false;
            }

        }


        return isInStock;
    },
    price: function ($) {
        var price = $('span.h-text-xl').text();
        price = getPrice(price);
        if (!price) {
            price = $('span.h-text-bs').text();
            price = getPrice(price)
        }
        return price;
    },
    image: function ($) {
        var image = $('div.Col-lvtw7q-0.dCwWmF img').attr('src');
        return 'https://www.target.com' + image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('li.h-padding-a-none');
        //(searchProducts)
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('a.h-display-block').text();
        if (name != null) {
            name = cleanName(name)
        }
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.h-text-bs').text();


        //    if (price.indexOf('Reg') > -1) {
        //        price = $(product).find('.h-text-red').text();
        //    }

        //     if (price.indexOf('-') > -1) {
        //         price = null
        //     }

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = 'https://www.target.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function ($, product) {
        //   var image = $(product).find('img').attr('src');
        var image = $(product).find('picture').children(); 
        return $(image[0]).attr('srcset');
    }
}

var Pagination = {
    items: function ($) {
        var items = $('ul.products--list li.product')
        return items.length

    },
    pages: function ($) {


        var pages = $('div.dropdown button.dropdown--button').text();
        pages = pages.match(/[0-9,]+/g);


        var result = null;

        if (pages !== null) {
            pages = pages.pop();
            result = parseInt(pages);

        } else {
            if (Search.container($).length > 0) {
                result = 1;
            }
        }

        return result

    }
}

var Category = {
    name: function ($) {
        var name = $('a#js-plp-page-title').text();
        name = cleanName(name);
        return name.replace(/\w\S*/g, function (txt) {
            return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
        });;
    }
}
