// sm-supplier-html/suppliers/www.petplanet.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    // 'sku': (url) => {
    //     return url.substring(url.lastIndexOf("/")+1,url.lastIndexOf(".prd"))
    // },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/")+1,Search.url($, product).lastIndexOf(".prd"));
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {

    },
    inStock: function ($) {

    },
    price: function ($) {

    },
    image: function ($) {

    },
    url: function ($) {

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.category-grid-prod');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('td.small_text h3').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.display-price-category').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = $(product).find('td.small_text a').attr('href');
        return 'https://www.petplanet.co.uk/' + url
    },

    image: function ($, product) {
        var image = $(product).find('a img').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.category-grid-prod');
        items = items.length;
        return parseInt(items);
    },
    pages: function ($) {
        var pages = $('td.ps_scrollnav').html();
        pages = pages.match(/[0-9]+/g)
        pages = pages.split('')
        (pages)
        return 1;
    }
}

var Category = {
    name: function ($) {
    }
}
