// sm-supplier-html/suppliers/www.lordandtaylor.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');


module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if(info.price !== null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
      'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h2.detial').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return "UNCHECKED"

    },
    price: function($) {
        var price = $('span.ora').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('ul#thumb1 li img#main').attr('src');
        image = 'https:' + image;
        return image;

    },
    description: function($) {
        var description = $('div.detial_main_content ul li').text();
        return description;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('ul.pro_list li');
        return searchProducts;

    },
    name: function($, product) {
        var brand = $(product).find('a.tit').html();
        var name = $(product).find('div.info a').html();
        name = cleanName(name);
        brand = cleanName(brand);
        return brand + " " + name;

    },
    price: function($, product) {
        var price = $(product).find('div.pro_price_black').html();

        if(price.indexOf('-') > -1) {
            price = null;
        }

        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('script.catEntryDisplayUrlScript').text();
        url = url.match(/\(([^)]+)\)/)[1];
        url = url.replace('"','').replace('"','');
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.img img.lazy').attr('src');
        image = 'https:' + image;
        return image;
        
    }
}

var Pagination = {

     items: function($) {
        var items = $('ul.pro_list li');
        items = items.length;
        return items;

     },
     pages: function($) {
        var items = $('ul.pro_list li');
        items = items.length;

        var products = $('div.list_page').text();
        products = products.match(/[0-9]+/g);
        products = products[0]
        
        sum = products / items;
        return Math.ceil(sum);
     }
}
var Category = {
    name: function($) {
        var name = $('div#location div a').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
