// sm-supplier-html/suppliers/www.jdsports.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = jdsports =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }



}

var Product = {
    name: function($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.itemPrices').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img.imgMed").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.productListItem');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('span.itemTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.itemPrice').text();
        var sale = $(product).find('span.now').text();

        if(sale !== "") {
            price = sale
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.jdsports.co.uk' + url;

    },
    image: function($, product) {
        //var image = $(product).find('a.itemImage>picture').attr('src');
        var image = $(product).find('a.itemImage>picture>source').attr('data-srcset')
        return image;
    }
}
var Category = {
    name: function($) {
        return cleanName($("div#productListTitle h1").text());
    }
}
var Pagination = {

     items: function($) {
        return Search.container($).length;

     },
     pages: function($) {
        return Math.ceil(cleanName($("span.count").text()).match(/[0-9]+/g)[0] / this.items($));

     }
}
