// sm-supplier-html/suppliers/www.cvs.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = cvs =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
             //result.push(info)

            if(info.price !== null) {
                result.push(info)
            }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div div.header h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div.price span.ng-binding').text();
        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div.large-image img').attr('src');
        image = 'http://www.cvs.com' + image;

        return image;
    },
    description: function($) {
        var description = $('div.description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.plp-productGridItemContainer');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.plp-productName span').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('div[ng-show="!group.hasRange"] span').text();
        price = getPrice(price);

        if (price == '') {
            price = null;
        }

        return price;

    },
    url: function($, product) {
        var url = $(product).find('div.plp-productName a').attr('ng-href');
        url = 'http://www.cvs.com' + url;

        return url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('ng-src');
        image = 'http://www.cvs.com' + image;

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('#sortByItems > option').last().text();
        
         return parseInt(items);
     },
     pages: function($) {
         var items = $('#sortByItems > option').last().text();

         var products = $('#resultsTabs a').text();
         if(products === ""){
            return null
         }else {
            products = products.match(/[0-9]+/g).shift();  
         }
         var sum = products / items;

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1#pageHeading').text();
        name = cleanName(name);
        return name;
    }
}
