// sm-supplier-html/suppliers/store.approvedfood.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var instockCheck = require('../../lib/inStockCheck.js')


module.exports = approvedfood =  {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("=")+1,url.lastIndexOf(""))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("=")+1,Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('.pdp_title').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        let stock= true
        let check = $('h1').last().text()
        stock = instockCheck(check,stock)
	    return stock
},
    price: function($) {
        var price = $('span.price_large').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("a.zoom_img img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('span[itemprop="identifier"]').text();
        return barcode;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.ra_item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.description a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.pp_price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('div.description a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.prod_shelfimg a').attr('style').match(/\(([^)]+)\)/)[1];
        return image;
    }
}

var Category = {
    name: function($) {
        return cleanName($("span.srch_res_text").text()).replace(/"/g, "");
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.ra_item');

        var products = $('div#total_result').text();
        products = products.match(/[0-9]+/g)

        if(items && products) {
            return parseInt(items.length);
        } else {
            return null;
        }
     },
     pages: function($) {
        var numitems = $('div.ra_item');

        if(numitems) {
            numitems = numitems.length
        } else {
            return null;
        }

        var products = $('div#total_result').text();
        products = products.match(/[0-9]+/g)

        if(products) {
            products = products.pop();
        } else {
            return null;
        }

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}
