// sm-supplier-html/suppliers/www.fye.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = fye =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('meta[property="og:title"]').attr('content')
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('div[itemprop="price"]').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img[itemprop="image"]').attr('src');
        return 'www.fye.com' + image;
    },
    barcode: function($) {
        var barcode = $('div.purchase-options').text();
        
        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC:') + 5).substring(0, 12);
            return barcode;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('div.title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.prices').find('span.sale-price').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'http://www.fye.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('div.image-container img').attr('src');
        return 'www.fye.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.product');
        return parseInt(items.length);

     },
     pages: function($) {
        var pages = $('select.pagination-select').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('div.breadcrumb h2').text();
        name = cleanName(name);
        return name;
    }
}
