// sm-supplier-html/suppliers/www.koleimports.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
// Pagination not needed
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')
module.exports = koleimports =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.phone-display h1.product-name').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        let stock = true
        let check = $('button.out-of-stock').text()
        stock = instockCheck(check,stock)
	    return stock;
},
    price: function($) {
        var price = $('div.content').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("div#primary-image img").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('div.tab-content .column .attribute:nth-child(3) span').html();
        
        if(barcode !== null) {
        return barcode
         }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.product-name ').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('.new-lowest-price').text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price=$(product).find('.lowest-price').text();
            price = getPrice(price)
            return price
        }
        
        
    },
    url: function($, product) {
        let urlContainer = $(product).find('.product-name').text()
            urlContainer = 'https://www.koleimports.com/' + cleanName(urlContainer).replace(/ /g,'-').replace(/,/g,'').replace(/"/g,'').trim().toLowerCase()
            return urlContainer

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Category = {
    name: function($) {
        var name = $('title').text();
        name = cleanName(name);
        return name;
    }
}
