// sm-supplier-html/suppliers/www.musiciansfriend.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.titleWrap').text()
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        let price =$('.productPrice').first().text()
        price = getPrice(price)
        // $('#itemizedPrice > li > dl > dd > span sup').first().remove();
        
        // var coins = $('#itemizedPrice > li > dl > dd > span sup').text().trim();
        // $('#itemizedPrice > li > dl > dd > span sup').remove();

        // var pound = $('#itemizedPrice > li > dl > dd > span').text().trim();
        // var price = pound + '.' + coins;

        // price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('a.skuLink>img').attr('src');

        return image;
    },
    description: function ($) {
        var des = $('#detailsContainer > div > p.description').text().replace(/\s+/g, " ").trim();

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('li.product-container');
        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div strong a').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.productPrice').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        url = 'http://www.musiciansfriend.com' + url;

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        (image)

        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('#displayNumber > li.most > a').first().text();
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('#displayNumber > li.most > a').first().text();
        var products = $('#compareForm > div:nth-child(3) > p > var').text();

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('#main > div.clearShadow > h1').text();
        name = cleanName(name)

        return name;
    }
}
