// sm-supplier-html/suppliers/www.bloomingdales.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
let generateSku = require('../../lib/testPattern.js')
let regex = /([^?]+)\/?$/


module.exports = replaceWithwebAddressChangeToCamelCase = {
    "useJavascript": "true",
    "usePhantom": "bloomingdales.phantom.js",
    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'sku': (url) => {
        let sku = generateSku(url, regex)[0].match('[0-9]+')[0]
        return sku
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = generateSku(Search.url($, product), regex)[0].match('[0-9]+')[0]
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        let brand = cleanName($('.brand-name-link').text())
        let title = cleanName($('h1').text());

        if (!brand || brand == "") {
            return title
        } else {
            return brand + ' ' + title
        }
    },
    inStock: function ($) {
        let oos = $('div.upc-availability').text()
        if(oos.toLowerCase().indexOf('in stock') > -1) {
            return true
        } else {
            return false
        }
    },
    price: function ($) {
        let salePrice = getPrice($('span.c-bold').html());
        let stdPrice = getPrice($('div.price').html());

        if (salePrice) {
            return salePrice
        } else {
            return stdPrice
        }
    },
    image: function ($) {
        var image = $('img#productImage').attr('src');
        return image;

    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.productThumbnail');
        return searchProducts;


    },
    name: function ($, product) {
        var name = $(product).find('div.productDescription').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = null
        var salePrice = $(product).find('span.discount').html();
        var regularPrice = $(product).find('span.regular').html()
        salePrice = getPrice(salePrice);
        regularPrice = getPrice(regularPrice);

        if (salePrice != null) {
            price = salePrice
        } else {
            price = regularPrice
        }
        return price;

    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;


    },
    image: function ($, product) {
        var image = $(product).find('img.thumbnailImage').attr('src');
        return image;

    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.thumbnailItem').length;
        return parseInt(items);
    },
    pages: function ($) {
        var numitems = $('li.thumbnailItem').length;

        var products = $('div#productCount').text();
        if (products === "") {
            return null
        } else {
            products = products.match(/[0-9]+/g).pop();
        }
        var sum = products / numitems;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('h1#pageHeader').text();
        name = cleanName(name);
        return name;
    }
}
