// sm-supplier-html/suppliers/store.hmv.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = hmv = {
    "useJavascript": true,

    'sku': (url) => {
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf(""))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "uk"

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("/") + 1, Search.url($, product).lastIndexOf(""));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = "uk"

            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}
var Product = {
    name: function ($) {
        var name = $('.prod-info__inner .h1').text();
        name = cleanName(name);
        return name;

    },
    inStock: function ($) {
        var isInStock = true;
        var availability = $('#p_lt_Body_lt_WebPartZone1_wpzProductDetail_ProductDetail_ucVariantsAndOptions_ctlStock').html();
        var soldout = $('div#p_lt_Body_lt_WebPartZone1_wpzProductDetail_ProductDetail_ucVariantsAndOptions_ctlStock').html();


        if (availability !== null) {
            if (availability.indexOf('Back in stock soon') > -1) {
                isInStock = false
            }
        }
        availability = $('#p_lt_Body_lt_ctl00_MessageBox_ucMessageBox_pnlMessageBoxSettings').html()
        if (availability !== null) {
            if (availability.indexOf('has now been discontinued') > -1) {
                isInStock = false
            }
        }
        availability = $('div.free-content p').html()
        if (availability !== null) {
            if (availability.indexOf('Unfortunately the page you were looking for could not be found.') > -1) {
                isInStock = false
            }
        }

        if (soldout !== null) {
            soldout = soldout.toLowerCase();
            if (soldout.indexOf('sold out') > -1) {
                isInStock = false;
            }
        }

        return isInStock

    },
    price: function ($) {
        var price = $('.prod-info__price').text();
        price = getPrice(price);
        return price;

    },
    image: function ($) {
        var image = $('div.prod-detail__img__inner img').attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.prod-listing__prod');

        return searchProducts;

    },
    name: function ($, product) {
        var name = $(product).find('h3.prod__heading a').text();
        name = cleanName(name);

        return name;

    },
    price: function ($, product) {
        var price = $(product).find('.prod__content__inner p.prod__meta').html();

        price = getPrice(price);
        return price;

    },
    url: function ($, product) {
        var url = $(product).find('h3.prod__heading a').attr("href");
        var breadCrumb = $("a#p_lt_ctl01_Breadcrumb_ucBreadcrumbs_rptBreadcrumbs_ctl01_lnkParentItem").text();
        url = "https://store.hmv.com/" + breadCrumb + "/" + url;
        return url;


    },

    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }

}
var Pagination = {

    items: function ($) {
        var items = $('div.prod-listing__opt .select').text();
        items = items.match(/[0-9]+/g)

        if (items) {
            items = items.pop();

            return parseInt(items);
        } else {
            return null;
        }


    },
    pages: function ($) {
        var pages = $('ul.pagination__pages').text();
        pages = pages.match(/\d/g);

        if (pages) {
            return parseInt(pages.length);
        } else {
            return null;
        }

    }
}

var Category = {
    name: function ($) {
        var name = $('.page-header h1').text();
        name = cleanName(name);
        return name;
    }
}
