// sm-supplier-html/suppliers/www.aeropostale.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'sku': (url) => {
        url = url.split('/')[0]
        return url.substring(url.lastIndexOf("/") + 1, url.lastIndexOf("."))
    },

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            var sku = Search.url($, product).split('?')[0]
            info.name = Search.name($, product);
            info.sku = sku.substring(sku.lastIndexOf("/") + 1, sku.lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('#pdpContainer > div.right > h2').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return "UNCHECKED";
    },
    price: function ($) {
        var price = $('#pdpContainer > div.right > ul.price > li.now').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div#zoomIn > img').attr('src');
        image = 'http://www.aeropostale.com' + image;

        return image;
    },
    description: function ($) {
        var description = $('div.product-description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.product-tile');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('a.name-link').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var salePrice = $(product).find('span.price-sale').text();
        salePrice = getPrice(salePrice);

        var price = $(product).find('div.product-price').text();
        price = getPrice(price);

        if (salePrice == '') {
            price = price
        } else {
            price = salePrice
        }

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');

        return image;
    }
}

var Pagination = {
    //
    // there is no items count element on page
    //
    items: function ($) {
        var items = $('div.item').length;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = $('div.item').length;

        var products = $('div.pagination ul li.count').first().text()

        if (products === "") {
            return null

        } else {
            products = products.match(/[0-9]+/g).pop();
        }


        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}

var Category = {
    name: function ($) {
        var name = $('ul#breadcrumbs li.active').text();
        name = cleanName(name);

        return name;
    }
}
