// sm-supplier-html/suppliers/www.kohls.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = kohls =  {
    "useJavaScript": true,
    "usePhantom": "kohls.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("prd-")+4,Search.url($, product).lastIndexOf("/"));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.pdp-product-title').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var isInStock = true;
        var producttop = $('.addtobag-availability').html()
        var check= $('.pdp-quan-bag-links-badges h3').text()
        var checkh3=$('h3').text()
        
        isInStock = instockCheck(checkh3,isInStock)
        isInStock=instockCheck(check,isInStock)
        var outofstock = $('#content').html();

       if(producttop !== null) {
            if (((producttop.indexOf('Out of Stock') > -1) || (producttop.indexOf('Not in Stock') > -1)
            || (producttop.indexOf('Unavailable') > -1))) {  
                isInStock = false
            }
        }

        if(outofstock !== null) {
            outofstock = outofstock.toLowerCase();
            if (outofstock.indexOf('is out of stock') > -1) {
                  isInStock = false  
            }
        }

        return isInStock
    },
    price: function($) {
        var price = $('.main-price').html();
        price = getPrice(price)
        return price;

    },
    image: function($) {
        var image = $('.pdp-easyzoom img').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.products_grid');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('.prod_nameBlock').text().replace("...", "");
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('.prod_price_amount').html();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.prod_nameBlock p');
        if(anchor) {
            url = 'http://www.kohls.com' + $(anchor).attr('rel');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('.prod_img_block img').attr('src');
        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('li.products_grid');

        if(items.length) {
            items = items.length;
            return parseInt(items);
        } else {
            return null;
        }

     },
     pages: function($) {
       var items = $('li.products_grid'); 
       items = items.length;

       
       var result = null;
       var products = $('span.result_count').text();

        if(products !== "") {
            products = products.match(/[0-9,]+/g).pop();
            result = Math.ceil(products / items);
        } else {
            if(Search.container($).length > 0) {
                result = 1;
            }
        }
       
        return result;
     }
}

var Category = {
    name: function($) {
        var name = $('h1.categoryName').text();
        name = cleanName(name);
        return name;
    }
}
