// sm-supplier-html/suppliers/www.chicos.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('h1.product-name').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('div.product-no-inventory').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = null
        var container = $('div.active-product-form-wrapper span.product-price')
        if (container.length) {
            price = getPrice(container.text())
        }
        return price
    },
    image: function ($) {
        var image = $('a.pdp-img-selected img').data('src');
        image = 'https://www.chicos.com' + image
        return image;
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.collection-container div.product-block');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.product-price').text();
        price = getPrice(price);
        if(price){
            return price;
        }else{
            price =$(product).find('.product-list-price').text()
            price=getPrice(price)
            return price
        }
        
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('h2.product-name a');
        if (anchor) {
            url = 'https://www.chicos.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('img.product-default-img').attr('src')
        image = 'https://www.chicos.com' + image
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('h1').text()
        name = cleanName(name);
        return name;
    }
}
