// sm-supplier-html/suppliers/www.entirelypets.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1#item-title').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var salePrice = $('span#precio-final span').text();
        var regularPrice = $('span#precio-regular span').text();
        var price;

        if (salePrice) {
            price = salePrice;
        } else {
            price = regularPrice;
        }

        price = getPrice(price);

        return price;
    },
    image: function($) {
        var image = $('div#producto-foto img').attr('src');

        return image;
    },
    description: function($) {
        var description = $('div.description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('div.product.grid');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('div.bf-title a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var regularPrice = $(product).find('span.bf-regular-price').text();
        var salePrice = $(product).find('span.bf-blackfriday-price').text();
        var price;

        if (salePrice) {
            price = salePrice;
        } else {
            price = regularPrice;
        }

        price = getPrice(price);

        return price;
    },
    url: function($, product) {
        var url = $(product).find('div.bf-title a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('div.img-container img').attr('src');

        return image;
    }
}

var Pagination = {

     items: function($) {
         var items = $('span.sli_bct_num_results').text();
         items = items.match(/[0-9]+/g).pop();
         items = JSON.parse(items);
         
         return items;
     },
     pages: function($) {
         var items = $('span.sli_bct_num_results').text();
         items = items.match(/[0-9]+/g).pop();

         var products = $('span.sli_bct_total_records').text();

         var sum = products / items;

         return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('h1#primary-headline').text();
        name = cleanName(name);
        return name;
    }
}
