// sm-supplier-html/suppliers/www.evitamins.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replace =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.h1productHead').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        return 'unchecked'
},
    price: function($) {
        var price = $('div.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img[itemprop='image']").attr('src');
        return 'http:' + image;
    },
    barcode: function($) {
        var barcode = $('td.newsText').text();

        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC:') + 5).substring(0, 12);
            return barcode;
        }
        return barcode;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('td[style="width:33%;padding:0px 10px"]');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('td[style="padding:5px 0;font:600 1.4em gladiator, arial,Helvetica,sans-serif;color:#FC511D"]').not('span.strike').text();

        if(price.indexOf('$') > 0) {
            price = price.substring(price.indexOf('$') + 4)
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('td[style="width:33%;padding:0px 10px"]');
        return parseInt(items.length);
     },
     pages: function($) {
        var pages = $('div.pagging-sec1 ul').text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages.match(/\d/g).pop());
     }
}

var Category = {
    name: function($) {
        var name = $('li.active').text();
        name = cleanName(name);
        return name;
    }
}
