// sm-supplier-html/suppliers/www.firebox.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck')
module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-page__content-title').text();
        var name1 = $('h2.product-page__content-sub-title').text()
        name = cleanName(name+' '+name1);
        return name;
    },
    inStock: function($) {
        var inStock=true
        var check= $('.product-page__content-cta>div>div').first().text()
        inStock=instockCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('.product-page__content-price>h3>span>span>span').first().text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('div.product-full__main-image-blurried-container img').attr('src');
        return 'http:' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a.info-title').text();
        name = cleanName(name);
        return name;
    },  
    price: function($, product) {
        var price = $(product).find('span.text-currency-0').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.firebox.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return 'http:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.gridFire__block');
        var products = $('span.category-page__search-results-value').text();

        if(items && products) {
            return parseInt(items.length);
        } else {
            return null;

        }
     },
     pages: function($) {
        var numitems = $('div.gridFire__block');
        var products = $('span.category-page__search-results-value').text();

        if(numitems && products) {
            numitems = numitems.length; 

            var sum = products / numitems;
            return Math.ceil(sum);
        } else {
            return null;
        }



     }
}

var Category = {
    name: function($) {
        var name = $('div.category-page__header h1').text();
        name = cleanName(name);
        return name;
    }
}
