// sm-supplier-html/suppliers/www.saucony.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('div.persistent-details > div.product-v2-name > h1').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'unchecked';
    },
    price: function ($) {
        var price = $('div.product-price > span.price-sales.promo-price').text().trim();

        if (price == '') {
            price = $('div.product-price > span.price-standard').text().trim();
        }
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div.product-hero-image.product-col-1 > div > div > div.main-product-slider > a').attr('href');

        return image;
    },
    description: function ($) {
        var des = $('div.pdp-drawer.active.description > div.pdp-drawer-content').text().replace(/\s+/g, " ").trim();;

        return des;
    }
}

var Search = {
    container: function ($) {
        var product = $('li.grid-tile');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.product-name > h3 > a').text();
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('div.product-pricing > div > span.product-sales-price').text();

        if (price == '') {
            price = $(product).find('div.product-pricing > span').text();
        }
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('div.product-name > h3 > a').attr('href');

        return url;
    },
    image: function ($, product) {
        var image = $(product).find('div.product-image > a > img').attr('src');

        return image;
    }
}
//
// No pagination
//
/*var Pagination = {

    items: function ($) {
        var items = ;
        items = JSON.parse(items);

        return items;
    },
    pages: function ($) {
        var items = ;
        var products = ;

        var pages = products / items;
        pages = Math.ceil(pages);

        return pages;
    }
}*/

var Category = {
    name: function ($) {
        var name = $('div.wrap-top-search.clearfix > hgroup > h1').text();
        name = cleanName(name)

        return name;
    }
}
