// sm-supplier-html/suppliers/www.familychristian.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = familychristian =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.barcode = Product.barcode($)
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('div.product-name h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
return "UNCHECKED"
},
    price: function($) {
        var price = $('span.price').text();
        var saleprice = $('span.special-price .price').text();

        if(saleprice !== null) {
            price = saleprice;
        }

        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("img#featured-product-image").attr('src');
        return image;
    },
    barcode: function($) {
        var barcode = $('ul#product-attributes li.fcs_upc').text();
        
        if(barcode.indexOf('UPC') > -1) {
            barcode = barcode.substring(barcode.indexOf('UPC:') + 5).substring(0, 12);
            return barcode;
        }
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.item');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.regular-price').text();
        var saleprice = $(product).find('span.special-price .price').text();

        if(saleprice !== "") {
            price = saleprice
        }
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a.product-image img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('select[aria-labelledby="toolbar-pager"]').text();
        items = items.match(/[0-9]+/g).pop();
        return parseInt(items);
     },
     pages: function($) {
        var items = $('select[aria-labelledby="toolbar-pager"]').text();
        numitems = items.match(/[0-9]+/g).pop();

        var products = $('p.amount').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-title h1').text();
        name = cleanName(name);
        return name;
    }
}
