// sm-supplier-html/suppliers/www.vminnovations.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');

module.exports = vmInnovations =  {
    "useJavascript":"true",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            //(info)
            result.push(info)
        });

        return result
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('#product-title h1').html();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
    return 'UNCHECKED'
},
    price: function($) {
        var price = $('#priceVal .price').html();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img#prodMainImageDisplay').attr('src');
        return 'https://www.vminnovations.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('[itemprop="itemListElement"]');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('.title a').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('#priceVal .price').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.vminnovations.com' + $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return 'https:' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('.col-md-10 ul.browse-item-list li');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var pages = $('ul.bPager li').eq("-2").text();
        pages = pages.match(/[0-9]+/g).pop();
        return parseInt(pages);

       
     }
}

var Category = {
    name: function($) {
        var name = $('div.category-head h1').text();
        name = cleanName(name);
        return name;
    }
}
