// sm-supplier-html/suppliers/www.gameseek.co.uk-OLD/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');


module.exports = gameseek =  {
    "useJavascript": true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            if(info.name !== null) {
            result.push(info)
        }
        });

        return result
    }, 
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

}

var Product = {
    name: function($) {
        
        var name = $('h1.p_title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var isInStock = true;
        var checkStock = $('span.ss_tos_lrg').html();
        var lmtStock = $('div.pad span.psm').text().substring(0,2,10);
        var notforsale = $('b.warn').html();

        if(checkStock !== null) {
            checkStock = checkStock.toLowerCase();
            if (checkStock.indexOf('temporarily out of stock') > -1) {
                isInStock = false;
            }
        }

        if(lmtStock !== null) {
            lmtStock = lmtStock.toLowerCase();
            if (lmtStock < 10) {
                isInStock = false;
            }
        }

        if(notforsale !== null) {
            notforsale = notforsale.toLowerCase();
            if (notforsale.indexOf('this item is not for sale at this time!') > -1) {
                isInStock = false;
            }
        }
        return isInStock
    },
    price: function($) {
        var price = $('div.mt5 b.cost_p').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $("#product_img a img").attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('tbody tr');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('b a').html();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('b.cost').text();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('a img').attr('src');
        return image;
    }
}

var Pagination = {

     items: function($) {
        return parseInt($('tbody tr').length);
     },
     pages: function($) {
        var numitems = $('tbody tr').length;

        var products = $('td').text();
       
        if(products.indexOf('of') > -1){
            products = products.split('of').pop();
            products = products.match(/[0-9]+/g).shift();
        }

        var sum = products / numitems;
        return Math.ceil(sum);
     }
}

var Category = {
    name: function($) {
        var hyph = cleanName($("h1.rl_title").text()).indexOf("-");

        return cleanName($("h1.rl_title").text()).slice(hyph + 2);
    }
}
