// sm-supplier-html/suppliers/www.academy.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = academy =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            if(info.price && info.name) {
            result.push(info)
            }
        });

        return result
    },
    
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.css-1u7748l.e10no0t94').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;
        var checkstock = $('div.online-message').html();
        var check =$('.row.mb-0').text()
        inStock=inStockCheck(check,inStock)
        var price=$('div.no-gutters.row div:last-child div:nth-child(3)').text()
        if(price !== null){
            price=price.toLowerCase()
            if(price.indexOf('our price in cart') > -1){
                inStock=false
            }
        }

        if(checkstock !== null) {
            //checkstock = checkstock.toLoweCase();
            if(checkstock.indexOf('Currently Not Available Online') > -1) {
                inStock = false;
            }
        }
    return inStock;
    },
    price: function($) {
        let price=''
        //var price = $('div.no-gutters.row div:last-child div:nth-child(3)').text();
        var bignumber=$('div[data-auid="PDP_ProductPrice"] span span').text()
        var smallnumber= $('div[data-auid="PDP_ProductPrice"] span small:last-child').text()
        price=bignumber+'.'+smallnumber
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.no-gutters.row div:first-child div div:first-child div:last-child img').attr('src');

        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $(".col-12")
        return searchProducts;
    },
    name: function($, product) {
        var brand = $(product).find('.c-product__title').html();
        let name = $(product).find('.c-product__description').html()
        brand = cleanName(brand);
        name = cleanName(name);
        return brand + ' ' + name;

    },
    price: function($, product) {
        let price = $(product).find ('section.mt-half')
        let dollar = price.find('span:nth-child(2)').html()
        let cent = price.find('sup:nth-child(3)').html()
        price = getPrice(dollar + '.' + cent)
        return price;

    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
        return 'https://www.academy.com' + url;

    },
    image: function($, product) {
        var image = $(product).find('img').attr('src');
        image = 'http:' + image;
        return image;

    }
}


var Pagination = {

     items: function($) {
        var items = $('div.productListingWidget .show-view-all-tool').text();
        if(items === "") {
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }
        return parseInt(items);

     },
     pages: function($) {
        var items = $('div.productListingWidget .show-view-all-tool').text();
        if(items === ""){
            return null
        }else{
            numitems = items.match(/[0-9]+/g).pop();
        }
        
     
        var products = $('#ayShowMoreButton #total_count_bottom').attr('value');
        if(products === null || ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();    
        }
        

        sum = products / numitems
        return Math.ceil(sum);

     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
