// sm-supplier-html/suppliers/www.kiddicare.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = kiddiCare = {
    "useJavascript": true,

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1').text()
        name = cleanName(name);
        return name;
    },

    inStock: function ($) {
        var isInStock = true;
        var availability = $('div.ws-availability').text();

        if (availability.indexOf('Weeks') > -1) {
            isInStock = false;
        }

        return isInStock;
    },
    price: function ($) {
        var price = $('meta[itemprop="price"]').attr('content');
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $("div#item_image a img").attr('src');
        return image;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.plp-grid-view-block');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span.our-price').text();
        price = getPrice(price);
        return price;
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },

    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('div.products-list-item-container');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        var items = $('div.products-list-item-container');
        items = items.length;

        var products = $('div.showing').text();
        products = products.match(/[0-9]+/g).pop();
        (products);

        sum = products / items
        return Math.ceil(sum)
    }
}

var Category = {
    name: function ($) {
        var name = $('div.breadcrumb-container span').eq("-1").text();
        name = cleanName(name);
        return name;
    }
}
