// sm-supplier-html/suppliers/www.softball.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = replaceWithSupplierName = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },


    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('div.pdisp_product_title').text()
        name=cleanName(name)
        return name;
    },
    inStock: function ($) {
        var availabilty = $('div.pdisp_addtocart')
        var isInstock= false
        if (availabilty){
                isInstock=true
        }
	return isInstock;
    },
    price: function ($) {
        var price = $('div.pdisp_price').text();
    //    if(price.match(/\\./gi).length!=1){
    //        price=null
        if(price.indexOf('-') > -1) {
        price = null
        }
        else{
            price = getPrice(price);
        }
        return price;
    },
    image: function ($) {
        var image = $("img#pdisp_media_image").attr('src');
        return image;
    },
}

var Search = {
    container: function ($) {
        var searchProducts = $('div.proddiv');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('div.proddiv_title').text();
        name = cleanName(name);
        return name;

    },
    price: function ($, product) {
        var price = $(product).find('div.proddiv_price').text();
        if(price.indexOf('-') > -1) {
            price = null
            }
            else{
                price = getPrice(price);
            }
            return price;
    },
    url: function ($, product) {
        var url = null
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        //('this is my url             '+ url);
        return 'http://www.softball.com' + url;
    },

    image: function ($, product) {
        var image = $(product).find('img.proddiv_thumbnail_image').attr('src');
        return image;
    }
}

var Category = {
    name: function ($) {
        return cleanName($("div.search_results_top_caption").text())
    }
}
