// sm-supplier-html/suppliers/www.bhcosmetics.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        // info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-name').text();
        name = cleanName(name);

        return name;
    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('p.availability.out-of-stock').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('sold out') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var status = $('span.value').text();
        var price;

        if (status.indexOf('Out of stock') > -1) {
            price = null;
        } else {
            price = $('span.price').last().text();
            price = getPrice(price);
        }

        return price;
    },
    image: function($) {
        var image = $('picture#image-0 source').attr('srcset');

        return image;
    },
    description: function($) {
        var description = $('div.short-description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function($) {
        var product = $('li.item');

        return product;
    },
    name: function($, product) {
        var name = $(product).find('h2.product-name a').text();
        name = cleanName(name);

        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.list-price').text();
        price = price.trim().replace(/[^0-9.]/g, '')
        if (!price) {
            price = null
        }
        return price;
    },
    url: function($, product) {
        var url = $(product).find('h2.product-name a').attr('href');

        return url;
    },
    image: function($, product) {
        var image = $(product).find('a.product-image picture source').attr('data-srcset');

        return image;
    }
}

//
// no pagination
//
/*var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}*/

var Category = {
    name: function($) {
        var name = $('div.breadcrumbs ul li').last().text();
        name = cleanName(name);

        return name;
    }
}
