// sm-supplier-html/suppliers/www.cyclegear.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js')
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = cycleGear =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.product-header-name').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        return "UNCHECKED";

    },
    price: function($) {
        var price = $('div.product-price-retail span.money').text().replace("$","");

        //price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('div.gallery-active-wrapper img.gallery-active').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.product-index-listing__product-wrapper');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.product-tile__name').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.product-tile__price').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
         var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = 'https://www.cyclegear.com' + $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.product-tile__image-wrapper').html();
        image = image.substring(image.lastIndexOf("//"),image.lastIndexOf("jpg")) 
        return 'https:' + image + 'jpg';
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.product-list-item-wrapper');
        return parseInt(items.length);
     },
     pages: function($) {
        var pages = $('select.jump').text();
        if(pages === ""){
            return null
        }else {
            pages = pages.match(/[0-9]+/g).pop();
        }
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    }
}
