// sm-supplier-html/suppliers/www.thewatchhut.co.uk/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = thewatchhut =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'
        info.description = Product.description($)

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.title').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock = true;

        var stockMessage = $('div.stock-status');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function($) {
        var price = $('div.priceNow.current').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var backgroundImage = $('div.zoomWindow').css('background-image');
        if (backgroundImage) {
            var image = backgroundImage.match(/url\([\"\'](.+?)[\"\']\)/);
            return 'http:' + image[1];
        }
    },
    description: function($) {
        var description = $('div#desc-box');
        if (description) {
            return description.text().replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function($) {
        return $('div.col-md-4.col-sm-6.col-xs-6');
    },
    name: function($, product) {
        var name = $(product).find('div.productTitle').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('div.priceNowLarge').text();
        price = getPrice(price);
        return price; 
    },
    url: function($, product) {
    //    var onclick = $(product).attr('onclick');
    //    if (onclick) {
    //        var match = onclick.match(/window.location.href='(.+?)'/);
        var url = $(product).find('a').attr('href');
            return 'http://www.thewatchhut.co.uk' + url;
        },
    image: function($, product) {
  //      var backgroundImage = $(product).find('div.details').css('background-image');    
  //      if (backgroundImage) {
  //          var match = backgroundImage.match(/url\('(.+?)'\)/);
       var image = $(product).find('div.productImage').attr('style') 
           image = image.replace('background-image:url(\'/','')
           image = image.replace('\');','')
            return 'http:/' + image
        }
    }

var Pagination = {

     items: function($) {
        var products = $('div.pagination div.perpage').text();
        products = products.match(/\d+/gm).pop();
        return parseInt(products);
     },
     pages: function($) {
        var products = $('div.pagination div.perpage').text();
        products = products.match(/\d+/gm).pop();

        var total = $('p.productsfound').text();
        total = total.match(/\d+/gm).pop();

        var pages = total / products;
        return Math.ceil(pages);
     }
}
