// sm-supplier-html/suppliers/www.brandsmartusa.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instoCheck =require('../../lib/inStockCheck')
module.exports = brandSmartUsa =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = "us"

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($,product);
            info.country = "us"
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').last().text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
        var inStock =true
        var check=$('.inventory').text()
        inStock= instoCheck(check,inStock)
        return inStock
},
    price: function($) {
        var price = $('.price-final').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.image-responsive').attr('src');
        return image
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.col-xs-12.col-xs-6.col-sm-4.col-md-3.text-center');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('a strong').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        var price = $(product).find('span.p-price span').html();
        price = getPrice(price);
        return price;
    },
    url: function($, product) {
        var url = $(product).find('a').attr('href');
      //  var anchor = $(product).find('a').html();
       
      //  if(anchor) {
      //      endUrl = $(anchor).attr('onclick');
        //    endUrl = endUrl.split("'");
       //     endUrl = endUrl[1];

        //    url = 'http://www.brandsmartusa.com' + endUrl;
       // }
        return 'http://www.brandsmartusa.com' + url;
    },
    image: function($, product) {
        var image = $(product).find('img.img-responsive').attr('src');
        return 'https://www.brandsmartusa.com' + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.sort-count').text();
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }
        return parseInt(items);

     },
     pages: function($) {
        var pages = $('li.sort-paging ul').text();
        if(pages === ""){
            return null
        }else{
            pages = pages.match(/[0-9]+/g).pop();
        }
        return parseInt(pages);
     }
}

var Category = {
    name: function($) {
        var name = $('h1').text();

        if(name.indexOf('Search Results') > -1) {
            name = name.split(":").pop();
        }
        name = cleanName(name);
        return name;
    }
}
