// sm-supplier-html/suppliers/www.winfieldsoutdoors.co.uk/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');

module.exports = Habitat = {
    'useJavaScript': true,
    'usePhantom': '',

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function ($) {
        var name = $('.page-title').text()
        name = cleanName(name)
        return name;
    },
    inStock: function ($) {
        var inStock = true;
        var stockcheck = $('p.availability').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf('out of stock') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function ($) {
        var price = null 
        var regularPrice = $('[itemprop="price]').attr('content')
        //var salePrice = $('div.price-box p.special-price').text()
        // if (salePrice) {
        //     price = getPrice(salePrice)
        // } else if (regularPrice) {
            price = getPrice(regularPrice)
        // }
        return price
    },
    image: function ($) {
        var image = null 
        var style = $('div.zoomContainer div.zoomWindow').attr('style')
        if (style) {
            var parts = style.match(/background-image: url\("(.+?)"\)/)
            if (parts) {
                image = parts[1]
            }
        }
        return image
    },
    url: function ($) {
        return $;
    }
}

var Search = {
    container: function ($) {
        var searchProducts = $('.product-item-info');
        return searchProducts;
    },
    name: function ($, product) {
        var name = $(product).find('.product-item-link').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = null 
        var regularPrice = $(product).find('.normal-price').text()
        //var salePrice = $(product).find('div.price-box p.special-price').text()
        // if (salePrice) {
        //     price = getPrice(salePrice)
        // } else if (regularPrice) {
            price = getPrice(regularPrice)
        // }
        return price
    },
    url: function ($, product) {
        var url = null;
        var anchor = $(product).find('.product-item-link');
        if (anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function ($, product) {
        var image = $(product).find('.product-image-photo').attr('src')
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var items = $('li.product-tile');
        items = items.length;
        return parseInt(items);

    },
    pages: function ($) {
        return 1;
    }
}

var Category = {
    name: function ($) {
        var name = $('div.category-title h1').text()
        name = cleanName(name);
        return name;
    }
}
