// sm-supplier-html/suppliers/www.abesofmaine.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var inStockCheck =require('../../lib/inStockCheck')
module.exports = abesofmaine =  {

    'sku': (url) => {
        return url.substring(url.lastIndexOf("_")+1,url.lastIndexOf("."))
    },

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("_")+1,Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us'
            result.push(info)
        });

        return result
    },
    
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1.text-center-xs').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true
        var check =$('p.stock').text()
        inStock=inStockCheck(check,inStock)
        return inStock

    },
    price: function($) {
        var price = $('span.price').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('a.2zoom img').attr('src');
        image = 'http://www.abesofmaine.com' + image;
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.col-sm-6 div.item');
        return searchProducts;

    },
    name: function($, product) {
        var name = cleanName($(product).find("div.item-title").text());

/*        var name = $(product).find('div.prodtitle a').html();
        name = cleanName(name);*/
        return name;

    },
    price: function($, product) {
        var price = getPrice($(product).find("span.price").text());
/*        var price = $(product).find('div.prodprice').html();
        price = getPrice(price);*/
        return price;

    },
    url: function($, product) {
        var url = "http://www.abesofmaine.com" + $(product).find("div.item-title a").attr("href");
/*        var url = null;
        var anchor = $(product).find('a');*/

        /*url = 'http://www.abesofmaine.com' + $(anchor).attr('href');*/
        
        return url;

    },
    image: function($, product) {
        var image = "http://www.abesofmaine.com" + $(product).find("div.item-img img").attr("src");
/*        var image = $(product).find('div.prodimg a img').attr('src');
        image = 'http://www.abesofmaine.com' + image;*/
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = Search.container($).length;

        if(items) {
            return items;

        } else {
            return null;
        }

     },
     pages: function($) {
        var pages = $('ul.pagination').html();

        if(pages && this.items($)) {
            pages.match(/[0-9]+/g).map(function(prod) {
                if(pages < prod) {
                    pages = prod;
                }
            });

            return parseInt(pages);
        } else {
            return null;
        }

        
/*        if(!pages) {
            return null
        }else {
            pages = pages.match(/[0-9]+/g).pop();    
        }*/
        

     }
}

var Category = {
    name: function($) {

        var name = $('div#breadcrumbsload').text();
        
        if(name.indexOf('•') > -1) {
            name = name.split('•').pop();
        }
        name = cleanName(name);

        var name = $('div.resultsCrumb').text();
       
        var splitName = name.split('>');
        name = cleanName(splitName.pop());
        return name;
    }
}
