// sm-supplier-html/suppliers/www.shopyourway.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase = {

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
            
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('h1[itemprop="name"]').text();
        name = cleanName(name);

        return name;
    },
    inStock: function ($) {
        return 'UNCHECKED';
    },
    price: function ($) {
        var price = $('div > div.product-price-section.section > div > span > div.sale-price > span.price-value').text();
        price = getPrice(price);

        return price;
    },
    image: function ($) {
        var image = $('div.product-main-image-container img').attr('src');

        return image;
    },
    description: function ($) {
        var description = $('div.short-description').text().replace(/\s+/g, " ").trim();

        return description;
    }
}

var Search = {
    container: function ($) {
        var product = $('div.product-subject-box');

        return product;
    },
    name: function ($, product) {
        var name = $(product).find('div.subject-content-title').attr('title');
        name = cleanName(name);

        return name;
    },
    price: function ($, product) {
        var price = $(product).find('span[itemprop="price"]').text();
        price = getPrice(price);

        return price;
    },
    url: function ($, product) {
        var url = $(product).find('a.link').attr('href');
        return 'https://www.shopyourway.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('a.link>img').attr('data-orig-src');
        image = 'http:' + image;
        //var images =$(product).find('a.link').html();

        return image;
    }
}

//
// there is no pagination, one page
//
/*var Pagination = {

     items: function($) {

     },
     pages: function($) {

     }
}*/

var Category = {
    name: function ($) {
        var name = $('div.title-line').text();
        name = cleanName(name);

        return name;
    }
}
