// sm-supplier-html/suppliers/www.homebargains.co.uk/logic.js
var SearchPageInfo = require ('../../lib/searchPageInfo.js')
var ProductPageInfo = require ('../../lib/productPageInfo.js')
var logger = require('../../lib/logger.js')
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var routine = require ("../../lib/routine.js"); 
var inStockCheck = require('../../lib/inStockCheck.js')

module.exports = {
     'search' : function($, next) {
        var result = []

        $('.product-grid').find('.item-box').map(function(index, product) {

            var info = new SearchPageInfo()
            info.name = searchName($,product)
            info.price = searchPrice($,product)
            info.url = url($,product)
            info.image = searchImage($,product)
            info.country = "uk"
            result.push(info)
        });

        // var pages = getNumberOfPages($)
        // for (var i = 0; i < result.length; i++)
        // {
        //     var info = result[i]
        //     info.pages = pages
        //     result[i] = info
        // }

        return result
    } ,

    'product' : function($, next) {
        var info  = new ProductPageInfo()
        info.inStock = isStock($)
        info.price = getPrice($, info.inStock)
        info.name = getName($, info.inStock)
        info.image = image($, info.inStock)
        info.country = "uk"
        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }
}


function url($, product) {
    return $(product).find('.picture').find('a').attr('href')
}


function searchName($, product) {
    return $(product).find('.producttitle').text().trim()
}


function searchPrice($, product) {
    var price = $(product).find('.productprice').find('b').text().replace("£","").trim()
    price = /[0-9]+.[0-9][0-9]/.exec(price)
    if (price !== null){
        price = price[0].replace("£","").trim()
    }else{
        price = ""
    }
    return price
}


// function getNumberOfPages($) {
//     var matching = $('#pagination').find('div a').toArray()
//     var pages = (($(matching[matching.length-1]).text()))

//     if (pages == "Last") // when there are 5+ pages
//     {
//         pages = ($(matching[matching.length-1]).attr('href'))
//         pages = /pageindex=[0-9]+/.exec(pages)
//         pages = /[0-9]+/.exec(pages[0])
//         pages = pages[0]
//     }
//     else if (pages == ">") // when there are 2-5 pages
//     {
//         pages = (($(matching[matching.length-2]).text()))
//     }
//     else
//     {
//         pages = 1
//     }
//     return pages
// }

function searchImage($, product) {
    return $(product).find('.picture img').attr('src');
}





var isStock = function inStock($) {
    var isInStock = true;
    var stockImage = $('#extra-buts').find('.imageStoreCollect').attr('src')
    var stock = $('.outofstock').text()
    isInStock = inStockCheck(stock,isInStock)
    if ((stockImage == "http://www.homebargains.co.uk/images/ccyeshdno.png")
    || (stockImage == "http://www.homebargains.co.uk/images/ccnohdno.png"))
    {
        isInStock = false
    }
    return isInStock
}

var getPrice = function thePrice($, instock) {
    var price = $('#productlt').find('.product-price').text().trim()
    price = /[0-9]+.[0-9][0-9]/.exec(price)
    if (price !== null){
        price = price[0].replace("£","").trim()
    }else{
        price = ""
    }
    return price
}

var getName = function theName($, instock) {
    return $('#rightheader3').text().trim()

}

var image = function image($, instock) {
    return $('a img#ctl00_ctl00_cph1_cph1_ctl00_ctrlProductInfo_defaultImage').attr('src');
}
var Category = {
    name: function($) {
        return cleanName($("a#ctl00_ctl00_cph1_cph1_ctl00_hlkName").text());
    }
}
var Pagination = {

     items: function($) {
        
        var match = cleanName($("select#ctl00_ctl00_" + 
            "cph1_cph1_ctl00_ddlPaging option").text()).match(/[0-9]+/g);

        if(match) {
           return parseInt(match[0]);
        } else {
            return null;
        }

     },
     pages: function($) {
        //return parseInt($("div#pagination").text().match(/[0-9]+/g).pop());

        return routine.matchNumbersAndPop($, "div#pagination");

     }
}
