// sm-supplier-html/suppliers/www.kotulas.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = replaceWithwebAddressChangeToCamelCase =  {

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1 span').text();
        name = cleanName(name);
        return name;
    },
    inStock: function($) {
return 'unchecked'
},
    price: function($) {
        var price = $('.only-price').html();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('.zoomWrapperImage img').attr('src');
        image = 'http://www.kotulas.com' + image;
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.nested div.prod-record');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('div.title a').html();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('div.only-price').text().replace(/ /g,"");
        var price2 = $(product).find('div.sale-price').text().replace(/ /g,"");

        if(price === "") {
            price = price2
        }
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.title a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;
    },
    image: function($, product) {
        var image = $(product).find('.img img').attr('src');
        image = 'http://www.kotulas.com' + image;
        return image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('div.nested div.prod-record').length;
        return parseInt(items);
     },
     pages: function($) {
        var numitems = $('div.nested div.prod-record').length;

        var products = $('p.total-count').text();
        products = products.match(/[0-9]+/g).pop();

        var sum = products / numitems;
        return Math.ceil(sum);        
     }
}

var Category = {
    name: function($) {
        var name = $('div#left-nav h1').text();
        name = cleanName(name);
        return name;
    }
}
