// sm-supplier-html/suppliers/www.buydig.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = buyDig =  {
    "useJavascript" : true,
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            result.push(info)
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('span#lblProdName').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var lmtstock = $('span.instock').text();

        if(lmtstock !== null) {
            lmtstock = lmtstock.toLowerCase();
            if(lmtstock.indexOf('only') > -1) {
                inStock = false;
            }
        }
        return inStock;
    },
    price: function($) {
        var price = $('span#lblPrice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img#imgProduct').attr('src');
        return image;

    }
}

var Search = {
    container: function($) {
        var searchProducts = $('div.row div.col-xs-6');

        return searchProducts;

    },
    name: function($, product) {
        /*Category name somehow gets into the name,
        when name is queried. Needed to replace
        category name with an empty string*/
        var category = $("div.search h1").text();

        var name = $(product).find('a').text();
        name = cleanName(name);

        name = name.replace(category, "");
  
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.price').text();
        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return url;

    },
    image: function($, product) {
        var image = $(product).find('div.col-xs-6 img').attr('src');
        return image;

    }
}

var Pagination = {

     items: function($) {
        var items = $('select#drpViewLimit').text();
        if(items === ""){
            return null
        }else {
            items = items.match(/[0-9]+/g).pop();
        }
        return parseInt(items);

     },
     pages: function($) {
        var items = $('select#drpViewLimit').text(); 
        if(items === ""){
            return null
        }else {
           numitems = items.match(/[0-9]+/g).pop();
        }
        

        var products = $('span#lblResultCount').text();
        if(products === ""){
            return null
        }else {
            products = products.match(/[0-9]+/g).pop();

        }

        sum = products / numitems
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('title').text();

        if(name.indexOf('|') > -1) {
            name = name.split('|').shift();
        }
        name = cleanName(name);
        return name;
    }
}
