// sm-supplier-html/suppliers/forbiddenplanet.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');

module.exports = forbiddenplanet =  {
    "useJavaScript": true,
    "usePhantom": "forbiddenplanet.phantom.js",
    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'uk'

        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product)
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'uk';
            result.push(info)
        });

        return result
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    },
    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    }

}

var Product = {
    name: function($) {
        var name = $('h1.fn span').html();
        name = cleanName(name);
        return name;

    },
    inStock: function($) {
        var inStock = true;
        var stockcheck = $('span.OOSNotify').text();

        if (stockcheck !== null) {
            stockcheck = stockcheck.toLowerCase();
            if (stockcheck.indexOf("notify me when it's back in stock") > -1) {
                inStock = false;
            }
        }

        return inStock;

    },
    price: function($) {
        var price = $('span.final-price.ourprice').html();
        price = getPrice(price);
        return price;

    },
    image: function($) {
        var image = $('img.w-product-main-image').attr('src');
        return image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('li.single');
        return searchProducts;

    },
    name: function($, product) {
        var name = $(product).find('h3').text();
        name = cleanName(name);
        return name;

    },
    price: function($, product) {
        var price = $(product).find('span.final-price').text();

        price = getPrice(price);
        return price;

    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('a');
        if(anchor) {
            url = $(anchor).attr('href');
        }
        return 'https://forbiddenplanet.com'+url;

    },
    image: function($, product) {
        var image = $(product).find('img.list-image').attr('src');
        return image;

    }
}
var Category = {
    name: function($) {
        return cleanName($("h1#productListHeading").text());
    }
}
var Pagination = {

     items: function($) {
        return parseInt(cleanName($("div.navSplitPagesResult").text())
            .split("products")[0].match(/[0-9]+/g)[1]);

     },
     pages: function($) {
        return Math.ceil(cleanName($("div.navSplitPagesResult").text())
            .split("products")[0].match(/[0-9]+/g)[2] / this.items($));

     }
}
