// sm-supplier-html/suppliers/www.ltdcommodities.com/logic.js
var ProductPageInfo = require ('../../lib/productPageInfo.js');
var SearchPageInfo = require ('../../lib/searchPageInfo.js');
var CategoryInfo = require ('../../lib/CategoryInfo.js');
var PaginationNumberInfo = require  ('../../lib/paginationNumberInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');
var instockCheck = require('../../lib/inStockCheck.js')

module.exports = ltdcommodities =  {
    "useJavascript": true,
    "usePhantom": 'ltdcommodities.phantom.js',

    'product' : function($, next) {
        // product - Returns product information for provided url
        var info  = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        return info
    },

    'search' : function($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function(index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';

            if (info.price !== null) {
            result.push(info)
        }
        });

        return result
    },

    'pagination' : function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },
    'category' : function($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function($) {
        var name = $('h1').text();
        name = cleanName(name);
        return name;
    },

    inStock: function($) {
        var inStock = true
        var message = $('.sku-panel').text()
        inStock = instockCheck(message,inStock)
	return inStock;
    },
    price: function($) {
        var price = $('.sku-desc__caption').text();
        price = getPrice(price);
        return price;
    },
    image: function($) {
        var image = $('img.infoImg').attr('src');
        return 'http://www.ltdcommodities.com' + image;
    }
}

var Search = {
    container: function($) {
        var searchProducts = $('.prod-artc').not('.prod-artc__vm');
        return searchProducts;
    },
    name: function($, product) {
        var name = $(product).find('h2').text();
        name = cleanName(name);
        return name;
    },
    price: function($, product) {
        let price = $(product).find ('span.prod-price').text();
        
        if(price.indexOf('-') > -1) {
            price = null
        }
        price = getPrice(price);
    //  (price)
        return price;
    },
    url: function($, product) {
        var url = null;
        var anchor = $(product).find('.prod-href__a');
        if(anchor) {
            url = "http://www.ltdcommodities.com" + $(anchor).attr('href');
        }
        return url;

    },

    image: function($, product) {
        var image = $(product).find('img').attr('src');
        return "http://www.ltdcommodities.com" + image;
    }
}

var Pagination = {

     items: function($) {
        var items = $('li.prod-item');
        items = items.length;
        return parseInt(items);

     },
     pages: function($) {
        var items = $('li.prod-item');
        numitems = items.length;

        var products = $('p#numItems').text();
        products = products.match(/[0-9]+/g).pop();
        
        sum = products / numitems 
        return Math.ceil(sum)
     }
}

var Category = {
    name: function($) {
        var name = $('div.categoryHero h1').text();
        name = cleanName(name);
        return name;
    }
}
