// sm-supplier-html/suppliers/www.allegromedical.com/logic.js
var ProductPageInfo = require('../../lib/productPageInfo.js');
var SearchPageInfo = require('../../lib/searchPageInfo.js');
var PaginationNumberInfo = require('../../lib/paginationNumberInfo.js');
var CategoryInfo = require('../../lib/CategoryInfo.js');
var cleanName = require('../../lib/cleanName.js');
var getPrice = require('../../lib/getPrice.js');

module.exports = allegromedical = {
/*
    'sku': (url) => {
        return url.substring(url.lastIndexOf("-") + 1, url.lastIndexOf("."))
    },
*/

    'product': function ($, next) {
        // product - Returns product information for provided url
        var info = new ProductPageInfo()
        info.name = Product.name($)
        info.inStock = Product.inStock($)
        info.price = Product.price($)
        info.image = Product.image($)
        info.country = 'us'
        info.description = Product.description($)

        return info
    },

    'search': function ($, next) {
        // search -  Returns an array of products from the suppliers search page
        var result = []

        Search.container($).map(function (index, product) {
            var info = new SearchPageInfo()
            info.name = Search.name($, product);
           // info.sku = Search.url($, product).substring(Search.url($, product).lastIndexOf("-") + 1, Search.url($, product).lastIndexOf("."));
            info.price = Search.price($, product);
            info.url = Search.url($, product);
            info.image = Search.image($, product);
            info.country = 'us';
            //result.push(info)

            if (info.price != null) {
                result.push(info);
            }
        });

        return result
    },

    'pagination': function ($, next) {

        var info = new PaginationNumberInfo()
        info.items = Pagination.items($);
        info.pages = Pagination.pages($);

        return info
    },

    'category': function ($, next) {
        var info = new CategoryInfo()
        info.name = Category.name($);

        return info
    }
}

var Product = {
    name: function ($) {
        var name = $('span#mainProductName').text();
        name = cleanName(name);
        return name;
    },
    inStock: function ($) {
        var inStock = true;

        var stockMessage = $('div.description h3');
        if (stockMessage) {
            stockMessage = stockMessage.text().toLowerCase();
            if (stockMessage.indexOf('temporarily unavailable') > -1) {
                inStock = false;
            }
        }

        return inStock;
    },
    price: function ($) {
        var price = $('dd#mainCurrentSkuPrice').text();
        price = getPrice(price);
        return price;
    },
    image: function ($) {
        var image = $('img#prodImage').attr('src');
        return image;
    },
    description: function ($) {
        var description = $('div.description').text();
        if (description) {
            return description.replace(/\s+/gm, ' ').trim();
        }
    }
}

var Search = {
    container: function ($) {
        return $('div.category');
    },
    name: function ($, product) {
        var name = $(product).find('div.cat-inside').text();
        name = cleanName(name);
        return name;
    },
    price: function ($, product) {
        var price = $(product).find('a.hoverOn div:nth-child(1), a.hoverOn span').text();
        price = price.replace(/\s+/gm, ' ');
        if (price.indexOf(' - ') == -1) {
            price = $(product).find('span.our-price').text();
            price = getPrice(price.replace(/\s+/gm, ''));
            return price;
        }
    },
    url: function ($, product) {
        var url = $(product).find('a').attr('href');
        return 'http://www.allegromedical.com' + url;
    },
    image: function ($, product) {
        var image = $(product).find('img').attr('src');
        return image;
    }
}

var Pagination = {

    items: function ($) {
        var products = $('select[name=productsPerPage]').text();
        if (products === "") {
            return null
        } else {
            products = products.match(/\d+/gm).pop();
        }

        return parseInt(products);
    },
    pages: function ($) {
        var products = $('select[name=productsPerPage]').text();

        if (products === "") {
            return null
        } else {
            products = products.match(/\d+/gm).pop();
        }



        var total = $('div#centerContent table td:contains(items)').text();

        if (total === "") {
            return null
        } else {
            total = total.match(/\d+/gm).pop();
        }

        var sum = total / products;
        return Math.ceil(sum);
    }
}

var Category = {
    name: function ($) {
        var name = $('div#browse-navtab span.crumb.final-crumb').text();
        name = cleanName(name);
        return name;
    }
}
